{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "d4b09cde-ecb6-480b-8e98-b535c630986d",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "import scipy.stats as st\n",
    "import seaborn as sns\n",
    "import sys\n",
    "import os\n",
    "import gseapy as gp\n",
    "import matplotlib.pyplot as plt\n",
    "import swan_vis as swan\n",
    "from pandarallel import pandarallel\n",
    "\n",
    "p = os.path.dirname(os.path.dirname(os.getcwd()))\n",
    "sys.path.append(p)\n",
    "\n",
    "from scripts.utils import *\n",
    "from scripts.plotting import *"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "6a639cf2-1849-4a02-aded-aca39a2e5bc1",
   "metadata": {},
   "outputs": [],
   "source": [
    "ab = '../../lr_bulk/talon/human_talon_abundance.tsv'\n",
    "filt_ab = '../../lr_bulk/cerberus/cerberus_filtered_abundance.tsv'\n",
    "read_annot = '../../lr_bulk/talon/human_talon_read_annot.tsv'\n",
    "t_metadata = '../../refs/cerberus/v40_transcript_metadata.tsv'\n",
    "t_orig_metadata = '../../refs/v40_transcript_metadata.tsv'\n",
    "swan_file = '../../lr_bulk/cerberus/swan/swan_meta.p'\n",
    "cerberus_h5 = '../../lr_bulk/cerberus/triplets/cerberus_annot_triplets.h5'\n",
    "major_isos = '../../lr_bulk/cerberus/swan/major_isos.tsv'\n",
    "mane_file = '../../refs/v40_gene_metadata.tsv'\n",
    "\n",
    "ver = 'v40_cerberus'\n",
    "\n",
    "min_tpm = 1\n",
    "gene_subset = 'protein_coding'\n",
    "obs_col = 'sample'\n",
    "\n",
    "mouse_ab = '../../mouse/lr_bulk/talon/mouse_talon_abundance.tsv'\n",
    "mouse_filt_ab = '../../mouse/lr_bulk/cerberus/cerberus_filtered_abundance.tsv'\n",
    "mouse_read_annot = '../../mouse/lr_bulk/talon/mouse_talon_read_annot.tsv'\n",
    "mouse_swan_file = '../../mouse/lr_bulk/swan/swan.p'\n",
    "mouse_meta = '../../mouse/lr_bulk/lr_mouse_library_data_summary.tsv'\n",
    "mouse_t_metadata = '../../mouse/refs/cerberus/vM25_transcript_metadata.tsv'\n",
    "mouse_ver = 'vM25_cerberus'\n",
    "mouse_h5 = '../../mouse/lr_bulk/triplets/cerberus_annot_triplets.h5'\n",
    "\n",
    "orth_table = '../../refs/biomart_human_to_mouse.tsv'"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "821a963b-7899-4692-8fcb-c35987b9362a",
   "metadata": {},
   "source": [
    "## Human vs mouse sector sankey"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "994c14e8-9c3b-4952-826a-252adda22f7a",
   "metadata": {},
   "outputs": [],
   "source": [
    "# major isoform triplets mouse match \n",
    "h_source = 'obs_mm_major'\n",
    "m_source = 'obs_major'\n",
    "\n",
    "df = human_v_mouse_sectors(cerberus_h5,\n",
    "                      mouse_h5,\n",
    "                      h_source,\n",
    "                      m_source,\n",
    "                      gene_subset=[gene_subset],\n",
    "                      ofile='figures/obs_major_mouse_v_human_triplets_sankey.pdf')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4cd2475a-43ed-4b7d-9916-9e6f5d55504c",
   "metadata": {},
   "source": [
    "## Simplex plots for human and mouse ARF4 / Arf4"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "e985bf01-a017-4ed6-9ecd-2b2041d49641",
   "metadata": {},
   "outputs": [],
   "source": [
    "# restrict sample triplets to mm ones\n",
    "ca = cerberus.read(cerberus_h5)\n",
    "samples = get_mouse_match_samples()\n",
    "inds = ca.triplets.loc[(ca.triplets.source == 'sample_det')&\\\n",
    "                       ~(ca.triplets['sample'].isin(samples))].index\n",
    "ca.triplets.drop(inds, axis=0, inplace=True)\n",
    "ca.triplets.loc[ca.triplets.source == 'sample_det', 'source'] = 'sample_mm_det'\n",
    "ca = get_centroids(ca, \n",
    "                   source='sample_mm_det',\n",
    "                   gene_subset=gene_subset,\n",
    "                   ver=ver)\n",
    "\n",
    "# add suffix to mouse sources\n",
    "m_ca = cerberus.read(mouse_h5)\n",
    "m_ca.triplets.loc[m_ca.triplets.source == 'sample_det', 'source'] ='sample_det_mouse'\n",
    "m_ca = get_centroids(m_ca,\n",
    "                     source='sample_det_mouse',\n",
    "                     gene_subset=gene_subset,\n",
    "                     ver=mouse_ver)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "0133aff9-027e-4540-9a46-a6c8586ad8bb",
   "metadata": {},
   "outputs": [],
   "source": [
    "# merge in gids for orthologs\n",
    "def get_human_mouse_gid_table(fname):\n",
    "    # get matching gids from human and mouse\n",
    "    df = pd.read_csv(fname, sep='\\t')\n",
    "\n",
    "    # drop nans in either human or mouse\n",
    "    df = df[['Gene stable ID', 'Mouse gene stable ID']]\n",
    "    df = df.loc[~df['Gene stable ID'].isnull()]\n",
    "    df = df.loc[~df['Mouse gene stable ID'].isnull()]    \n",
    "    \n",
    "    df = df.drop_duplicates()\n",
    "    return df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "785d2a8c-b84b-48fd-b102-a6a2e252aea7",
   "metadata": {},
   "outputs": [],
   "source": [
    "hm_gids = get_human_mouse_gid_table(orth_table)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "2134da8d-bdd5-44c6-a56d-185a2cd6dbe2",
   "metadata": {},
   "outputs": [],
   "source": [
    "# merge in mouse gids that we need\n",
    "ca.triplets = ca.triplets.merge(hm_gids, how='inner', left_on='gid', right_on='Gene stable ID')\n",
    "\n",
    "# merge in mouse data\n",
    "m_ca.triplets = m_ca.triplets.merge(hm_gids, how='inner', left_on='gid', right_on='Mouse gene stable ID')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "a58e5ee0-1cbd-46e1-aeca-c951395d9465",
   "metadata": {},
   "outputs": [],
   "source": [
    "# concatenate the dfs\n",
    "ca.triplets = pd.concat([ca.triplets, m_ca.triplets], axis=0)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "dd218a12-12cd-49c3-a02e-56ab6a229dc7",
   "metadata": {},
   "outputs": [],
   "source": [
    "# ca.triplets['gid'] = ca.triplets['Gene stable ID'] \n",
    "# ca.triplets['gname'] = 'temp'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "4bdd73e2-4e37-4439-9e5d-469645495890",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array(['lapa', 'gtex', 'sample_mm_det', 'sample_major',\n",
       "       'sample_mm_det_centroid', 'v29', 'obs_mm_det', 'obs_mm_major',\n",
       "       'v40', 'obs_det', 'obs_major', 'all', 'vM25', 'vM21',\n",
       "       'sample_det_mouse', 'tissue_det', 'tissue_major',\n",
       "       'tissue_adult_det', 'tissue_adult_major',\n",
       "       'sample_det_mouse_centroid'], dtype=object)"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ca.triplets.source.unique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "486568b5-fac4-4870-ba52-799159d8fca0",
   "metadata": {},
   "outputs": [],
   "source": [
    "df = compute_dists(ca, 'sample_mm_det_centroid', 'sample_det_mouse_centroid', gene_merge=['Gene stable ID'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "0aff03a8-5b94-4ac7-b9fd-4f54784f7a8e",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>source_sample_mm_det_centroid</th>\n",
       "      <th>gid_sample_mm_det_centroid</th>\n",
       "      <th>n_tss_sample_mm_det_centroid</th>\n",
       "      <th>n_tes_sample_mm_det_centroid</th>\n",
       "      <th>n_ic_sample_mm_det_centroid</th>\n",
       "      <th>n_iso_sample_mm_det_centroid</th>\n",
       "      <th>splicing_ratio_sample_mm_det_centroid</th>\n",
       "      <th>tss_ratio_sample_mm_det_centroid</th>\n",
       "      <th>tes_ratio_sample_mm_det_centroid</th>\n",
       "      <th>spl_ratio_sample_mm_det_centroid</th>\n",
       "      <th>...</th>\n",
       "      <th>spl_ratio_sample_det_mouse_centroid</th>\n",
       "      <th>sector_sample_det_mouse_centroid</th>\n",
       "      <th>gname_sample_det_mouse_centroid</th>\n",
       "      <th>sample_sample_det_mouse_centroid</th>\n",
       "      <th>gene_tpm_sample_det_mouse_centroid</th>\n",
       "      <th>Mouse gene stable ID_sample_det_mouse_centroid</th>\n",
       "      <th>tissue_sample_det_mouse_centroid</th>\n",
       "      <th>tissue_adult_sample_det_mouse_centroid</th>\n",
       "      <th>dist</th>\n",
       "      <th>z_score</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>sample_mm_det_centroid</td>\n",
       "      <td>ENSG00000109667</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.333333</td>\n",
       "      <td>0.333333</td>\n",
       "      <td>0.333333</td>\n",
       "      <td>...</td>\n",
       "      <td>0.321476</td>\n",
       "      <td>mixed</td>\n",
       "      <td>Slc2a9</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENSMUSG00000005107</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.009765</td>\n",
       "      <td>-1.272098</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>sample_mm_det_centroid</td>\n",
       "      <td>ENSG00000109805</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>2.428571</td>\n",
       "      <td>2.285714</td>\n",
       "      <td>3.571429</td>\n",
       "      <td>1.261905</td>\n",
       "      <td>0.242615</td>\n",
       "      <td>0.470387</td>\n",
       "      <td>0.286999</td>\n",
       "      <td>...</td>\n",
       "      <td>0.293995</td>\n",
       "      <td>mixed</td>\n",
       "      <td>Ncapg</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENSMUSG00000015880</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.053128</td>\n",
       "      <td>-0.611812</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>sample_mm_det_centroid</td>\n",
       "      <td>ENSG00000143340</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.333333</td>\n",
       "      <td>0.333333</td>\n",
       "      <td>0.333333</td>\n",
       "      <td>...</td>\n",
       "      <td>0.568421</td>\n",
       "      <td>splicing</td>\n",
       "      <td>Fam163a</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENSMUSG00000015484</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.167865</td>\n",
       "      <td>1.135293</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>sample_mm_det_centroid</td>\n",
       "      <td>ENSG00000149633</td>\n",
       "      <td>2.333333</td>\n",
       "      <td>2.000000</td>\n",
       "      <td>2.333333</td>\n",
       "      <td>3.000000</td>\n",
       "      <td>1.066667</td>\n",
       "      <td>0.415259</td>\n",
       "      <td>0.361495</td>\n",
       "      <td>0.223246</td>\n",
       "      <td>...</td>\n",
       "      <td>0.333333</td>\n",
       "      <td>simple</td>\n",
       "      <td>D630003M21Rik</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENSMUSG00000037813</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.089504</td>\n",
       "      <td>-0.057906</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>sample_mm_det_centroid</td>\n",
       "      <td>ENSG00000149636</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>5.666667</td>\n",
       "      <td>5.666667</td>\n",
       "      <td>5.666667</td>\n",
       "      <td>0.145326</td>\n",
       "      <td>0.145326</td>\n",
       "      <td>0.709347</td>\n",
       "      <td>...</td>\n",
       "      <td>0.350068</td>\n",
       "      <td>mixed</td>\n",
       "      <td>Dsn1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENSMUSG00000027635</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.263908</td>\n",
       "      <td>2.597731</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 37 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "  source_sample_mm_det_centroid gid_sample_mm_det_centroid  \\\n",
       "0        sample_mm_det_centroid            ENSG00000109667   \n",
       "1        sample_mm_det_centroid            ENSG00000109805   \n",
       "2        sample_mm_det_centroid            ENSG00000143340   \n",
       "3        sample_mm_det_centroid            ENSG00000149633   \n",
       "4        sample_mm_det_centroid            ENSG00000149636   \n",
       "\n",
       "   n_tss_sample_mm_det_centroid  n_tes_sample_mm_det_centroid  \\\n",
       "0                      1.000000                      1.000000   \n",
       "1                      1.000000                      2.428571   \n",
       "2                      1.000000                      1.000000   \n",
       "3                      2.333333                      2.000000   \n",
       "4                      1.000000                      1.000000   \n",
       "\n",
       "   n_ic_sample_mm_det_centroid  n_iso_sample_mm_det_centroid  \\\n",
       "0                     1.000000                      1.000000   \n",
       "1                     2.285714                      3.571429   \n",
       "2                     1.000000                      1.000000   \n",
       "3                     2.333333                      3.000000   \n",
       "4                     5.666667                      5.666667   \n",
       "\n",
       "   splicing_ratio_sample_mm_det_centroid  tss_ratio_sample_mm_det_centroid  \\\n",
       "0                               1.000000                          0.333333   \n",
       "1                               1.261905                          0.242615   \n",
       "2                               1.000000                          0.333333   \n",
       "3                               1.066667                          0.415259   \n",
       "4                               5.666667                          0.145326   \n",
       "\n",
       "   tes_ratio_sample_mm_det_centroid  spl_ratio_sample_mm_det_centroid  ...  \\\n",
       "0                          0.333333                          0.333333  ...   \n",
       "1                          0.470387                          0.286999  ...   \n",
       "2                          0.333333                          0.333333  ...   \n",
       "3                          0.361495                          0.223246  ...   \n",
       "4                          0.145326                          0.709347  ...   \n",
       "\n",
       "  spl_ratio_sample_det_mouse_centroid sector_sample_det_mouse_centroid  \\\n",
       "0                            0.321476                            mixed   \n",
       "1                            0.293995                            mixed   \n",
       "2                            0.568421                         splicing   \n",
       "3                            0.333333                           simple   \n",
       "4                            0.350068                            mixed   \n",
       "\n",
       "  gname_sample_det_mouse_centroid  sample_sample_det_mouse_centroid  \\\n",
       "0                          Slc2a9                               NaN   \n",
       "1                           Ncapg                               NaN   \n",
       "2                         Fam163a                               NaN   \n",
       "3                   D630003M21Rik                               NaN   \n",
       "4                            Dsn1                               NaN   \n",
       "\n",
       "  gene_tpm_sample_det_mouse_centroid  \\\n",
       "0                                NaN   \n",
       "1                                NaN   \n",
       "2                                NaN   \n",
       "3                                NaN   \n",
       "4                                NaN   \n",
       "\n",
       "  Mouse gene stable ID_sample_det_mouse_centroid  \\\n",
       "0                             ENSMUSG00000005107   \n",
       "1                             ENSMUSG00000015880   \n",
       "2                             ENSMUSG00000015484   \n",
       "3                             ENSMUSG00000037813   \n",
       "4                             ENSMUSG00000027635   \n",
       "\n",
       "  tissue_sample_det_mouse_centroid tissue_adult_sample_det_mouse_centroid  \\\n",
       "0                              NaN                                    NaN   \n",
       "1                              NaN                                    NaN   \n",
       "2                              NaN                                    NaN   \n",
       "3                              NaN                                    NaN   \n",
       "4                              NaN                                    NaN   \n",
       "\n",
       "       dist   z_score  \n",
       "0  0.009765 -1.272098  \n",
       "1  0.053128 -0.611812  \n",
       "2  0.167865  1.135293  \n",
       "3  0.089504 -0.057906  \n",
       "4  0.263908  2.597731  \n",
       "\n",
       "[5 rows x 37 columns]"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5de75b8c-d570-41b3-8edd-c2e94094e680",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c881b1d7-07b9-4ba8-91de-32f993d76488",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e17a6f2f-7e5b-4a3f-aba3-974fede6571d",
   "metadata": {},
   "outputs": [],
   "source": [
    "# mouse match stuff\n",
    "c_dict, order = get_biosample_colors()\n",
    "c_dict[np.nan] = 'k'\n",
    "\n",
    "mm_tissues = get_mouse_match_samples()\n",
    "print(mm_tissues)\n",
    "mm_tissues += [np.nan]\n",
    "mmap = {'v40': '*', 'v29': 'x', 'obs_mm_det': '^', 'cerberus': '2'}\n",
    "subset = {'source': ['v40', 'obs_mm_det', 'sample_det']}\n",
    "\n",
    "# can't subset the way that I want because I want to use and and or logic...\n",
    "ca.triplets = ca.triplets.loc[ca.triplets.source.isin(subset['source'])]\n",
    "ca.triplets = ca.triplets.loc[ca.triplets['sample'].isin(mm_tissues)]"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.4"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
