{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "c0ca800f-893c-4184-b84a-5259c939de7a",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import pyranges as pr\n",
    "import upsetplot\n",
    "import seaborn as sns\n",
    "import matplotlib.pyplot as plt\n",
    "import os\n",
    "import sys\n",
    "\n",
    "p = os.path.dirname(os.path.dirname(os.getcwd()))\n",
    "sys.path.append(p)\n",
    "\n",
    "from scripts.utils import *\n",
    "from scripts.plotting import *"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "id": "a514f3f9-ab4a-4100-8bca-3c850f3199f3",
   "metadata": {},
   "outputs": [],
   "source": [
    "# # download data\n",
    "# !meta='https://storage.googleapis.com/isoform.io/iso_all_v1.2.csv'\n",
    "# !gtf='https://storage.googleapis.com/isoform.io/chess_structure_v1.2.1_sorted.gtf'\n",
    "\n",
    "# !wget $meta\n",
    "# !wget $gtf\n",
    "# !wget https://github.com/chess-genome/chess/releases/download/v.3.0/chess3.0.gtf.gz\n",
    "# !gunzip chess3.0.gtf.gz"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7620601e-f03a-4a4d-ad25-af6f1946848f",
   "metadata": {},
   "source": [
    "## 1/3/2023 -- chess 3"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 55,
   "id": "17da0632-5382-4125-b731-60d5b9572980",
   "metadata": {},
   "outputs": [],
   "source": [
    "gtf = 'chess3.0.gtf'\n",
    "df = pr.read_gtf(gtf, duplicate_attr=True).as_df()\n",
    "df = df.loc[df.Feature == 'transcript']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 56,
   "id": "7478a150-36a9-499e-a17c-330b1dceb7b0",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 57,
   "id": "e1c8657f-fb2a-4512-b06c-a46b39b20d98",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Source</th>\n",
       "      <th>Feature</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Score</th>\n",
       "      <th>Strand</th>\n",
       "      <th>Frame</th>\n",
       "      <th>transcript_id</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>gene_type</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>db_xref</th>\n",
       "      <th>num_samples</th>\n",
       "      <th>max_tpm</th>\n",
       "      <th>assembly_id</th>\n",
       "      <th>tag</th>\n",
       "      <th>comment</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>BestRefSeq</td>\n",
       "      <td>transcript</td>\n",
       "      <td>11873</td>\n",
       "      <td>14409</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>CHS.1.1</td>\n",
       "      <td>CHS.1</td>\n",
       "      <td>transcribed_pseudogene</td>\n",
       "      <td>DDX11L1</td>\n",
       "      <td>RefSeq:NR_046018.2,GENCODE:ENST00000456328.2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>transcript</td>\n",
       "      <td>29553</td>\n",
       "      <td>31097</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>CHS.3.1</td>\n",
       "      <td>CHS.3</td>\n",
       "      <td>lncRNA</td>\n",
       "      <td>NaN</td>\n",
       "      <td>GENCODE:ENST00000473358.1</td>\n",
       "      <td>903</td>\n",
       "      <td>9.496638</td>\n",
       "      <td>ALL_00000008</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>chr1</td>\n",
       "      <td>BestRefSeq</td>\n",
       "      <td>transcript</td>\n",
       "      <td>30365</td>\n",
       "      <td>30503</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>CHS.166735.1</td>\n",
       "      <td>CHS.166735</td>\n",
       "      <td>miRNA</td>\n",
       "      <td>MIR1302-2</td>\n",
       "      <td>RefSeq:NR_036051.1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>chr1</td>\n",
       "      <td>BestRefSeq</td>\n",
       "      <td>transcript</td>\n",
       "      <td>30437</td>\n",
       "      <td>30458</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>CHS.166735.2</td>\n",
       "      <td>CHS.166735</td>\n",
       "      <td>miRNA</td>\n",
       "      <td>MIR1302-2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>chr1</td>\n",
       "      <td>Curated Genomic</td>\n",
       "      <td>transcript</td>\n",
       "      <td>52452</td>\n",
       "      <td>53396</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>CHS.144069.1</td>\n",
       "      <td>CHS.144069</td>\n",
       "      <td>pseudogene</td>\n",
       "      <td>OR4G4P</td>\n",
       "      <td>RefSeq:OR4G4P,GENCODE:ENST00000606857.1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   Chromosome           Source     Feature  Start    End Score Strand Frame  \\\n",
       "0        chr1       BestRefSeq  transcript  11873  14409     .      +     .   \n",
       "4        chr1           HAVANA  transcript  29553  31097     .      +     .   \n",
       "8        chr1       BestRefSeq  transcript  30365  30503     .      +     .   \n",
       "10       chr1       BestRefSeq  transcript  30437  30458     .      +     .   \n",
       "12       chr1  Curated Genomic  transcript  52452  53396     .      +     .   \n",
       "\n",
       "   transcript_id     gene_id               gene_type  gene_name  \\\n",
       "0        CHS.1.1       CHS.1  transcribed_pseudogene    DDX11L1   \n",
       "4        CHS.3.1       CHS.3                  lncRNA        NaN   \n",
       "8   CHS.166735.1  CHS.166735                   miRNA  MIR1302-2   \n",
       "10  CHS.166735.2  CHS.166735                   miRNA  MIR1302-2   \n",
       "12  CHS.144069.1  CHS.144069              pseudogene     OR4G4P   \n",
       "\n",
       "                                         db_xref num_samples   max_tpm  \\\n",
       "0   RefSeq:NR_046018.2,GENCODE:ENST00000456328.2         NaN       NaN   \n",
       "4                      GENCODE:ENST00000473358.1         903  9.496638   \n",
       "8                             RefSeq:NR_036051.1         NaN       NaN   \n",
       "10                                           NaN         NaN       NaN   \n",
       "12       RefSeq:OR4G4P,GENCODE:ENST00000606857.1         NaN       NaN   \n",
       "\n",
       "     assembly_id  tag comment  \n",
       "0            NaN  NaN     NaN  \n",
       "4   ALL_00000008  NaN     NaN  \n",
       "8            NaN  NaN     NaN  \n",
       "10           NaN  NaN     NaN  \n",
       "12           NaN  NaN     NaN  "
      ]
     },
     "execution_count": 57,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 63,
   "id": "09493340-9b5f-425e-91ab-e0f2d21a6f2c",
   "metadata": {},
   "outputs": [],
   "source": [
    "# does each chess gene id have a gencode transcript id associated with it\n",
    "df['gencode_tid'] = df.db_xref.str.split('GENCODE:', expand=True)[1]\n",
    "df['gencode_tid'] = df.gencode_tid.str.split(',', expand=True)[0]\n",
    "df['has_gencode_tid'] = ~df.gencode_tid.isnull()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 64,
   "id": "ecd15c0b-3cf3-4eb6-bb70-45a9ff9495c1",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Source</th>\n",
       "      <th>Feature</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Score</th>\n",
       "      <th>Strand</th>\n",
       "      <th>Frame</th>\n",
       "      <th>transcript_id</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>gene_type</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>db_xref</th>\n",
       "      <th>num_samples</th>\n",
       "      <th>max_tpm</th>\n",
       "      <th>assembly_id</th>\n",
       "      <th>tag</th>\n",
       "      <th>comment</th>\n",
       "      <th>gencode_tid</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>BestRefSeq</td>\n",
       "      <td>transcript</td>\n",
       "      <td>11873</td>\n",
       "      <td>14409</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>CHS.1.1</td>\n",
       "      <td>CHS.1</td>\n",
       "      <td>transcribed_pseudogene</td>\n",
       "      <td>DDX11L1</td>\n",
       "      <td>RefSeq:NR_046018.2,GENCODE:ENST00000456328.2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENST00000456328.2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>transcript</td>\n",
       "      <td>29553</td>\n",
       "      <td>31097</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>CHS.3.1</td>\n",
       "      <td>CHS.3</td>\n",
       "      <td>lncRNA</td>\n",
       "      <td>NaN</td>\n",
       "      <td>GENCODE:ENST00000473358.1</td>\n",
       "      <td>903</td>\n",
       "      <td>9.496638</td>\n",
       "      <td>ALL_00000008</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENST00000473358.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>chr1</td>\n",
       "      <td>BestRefSeq</td>\n",
       "      <td>transcript</td>\n",
       "      <td>30365</td>\n",
       "      <td>30503</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>CHS.166735.1</td>\n",
       "      <td>CHS.166735</td>\n",
       "      <td>miRNA</td>\n",
       "      <td>MIR1302-2</td>\n",
       "      <td>RefSeq:NR_036051.1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>chr1</td>\n",
       "      <td>BestRefSeq</td>\n",
       "      <td>transcript</td>\n",
       "      <td>30437</td>\n",
       "      <td>30458</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>CHS.166735.2</td>\n",
       "      <td>CHS.166735</td>\n",
       "      <td>miRNA</td>\n",
       "      <td>MIR1302-2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>chr1</td>\n",
       "      <td>Curated Genomic</td>\n",
       "      <td>transcript</td>\n",
       "      <td>52452</td>\n",
       "      <td>53396</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>CHS.144069.1</td>\n",
       "      <td>CHS.144069</td>\n",
       "      <td>pseudogene</td>\n",
       "      <td>OR4G4P</td>\n",
       "      <td>RefSeq:OR4G4P,GENCODE:ENST00000606857.1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENST00000606857.1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   Chromosome           Source     Feature  Start    End Score Strand Frame  \\\n",
       "0        chr1       BestRefSeq  transcript  11873  14409     .      +     .   \n",
       "4        chr1           HAVANA  transcript  29553  31097     .      +     .   \n",
       "8        chr1       BestRefSeq  transcript  30365  30503     .      +     .   \n",
       "10       chr1       BestRefSeq  transcript  30437  30458     .      +     .   \n",
       "12       chr1  Curated Genomic  transcript  52452  53396     .      +     .   \n",
       "\n",
       "   transcript_id     gene_id               gene_type  gene_name  \\\n",
       "0        CHS.1.1       CHS.1  transcribed_pseudogene    DDX11L1   \n",
       "4        CHS.3.1       CHS.3                  lncRNA        NaN   \n",
       "8   CHS.166735.1  CHS.166735                   miRNA  MIR1302-2   \n",
       "10  CHS.166735.2  CHS.166735                   miRNA  MIR1302-2   \n",
       "12  CHS.144069.1  CHS.144069              pseudogene     OR4G4P   \n",
       "\n",
       "                                         db_xref num_samples   max_tpm  \\\n",
       "0   RefSeq:NR_046018.2,GENCODE:ENST00000456328.2         NaN       NaN   \n",
       "4                      GENCODE:ENST00000473358.1         903  9.496638   \n",
       "8                             RefSeq:NR_036051.1         NaN       NaN   \n",
       "10                                           NaN         NaN       NaN   \n",
       "12       RefSeq:OR4G4P,GENCODE:ENST00000606857.1         NaN       NaN   \n",
       "\n",
       "     assembly_id  tag comment        gencode_tid  \n",
       "0            NaN  NaN     NaN  ENST00000456328.2  \n",
       "4   ALL_00000008  NaN     NaN  ENST00000473358.1  \n",
       "8            NaN  NaN     NaN               None  \n",
       "10           NaN  NaN     NaN                NaN  \n",
       "12           NaN  NaN     NaN  ENST00000606857.1  "
      ]
     },
     "execution_count": 64,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 74,
   "id": "dccd02ce-d6b0-43a4-bda3-de07f8233efd",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>has_gencode_tid</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>n_no_gencode</th>\n",
       "      <th>n_gencode</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>CHS.1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>CHS.100</td>\n",
       "      <td>1.0</td>\n",
       "      <td>3.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>CHS.10000</td>\n",
       "      <td>31.0</td>\n",
       "      <td>7.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>CHS.10001</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>CHS.10002</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "has_gencode_tid    gene_id  n_no_gencode  n_gencode\n",
       "0                    CHS.1           NaN        1.0\n",
       "1                  CHS.100           1.0        3.0\n",
       "2                CHS.10000          31.0        7.0\n",
       "3                CHS.10001           NaN        1.0\n",
       "4                CHS.10002           NaN        2.0"
      ]
     },
     "execution_count": 74,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "temp = df[['gene_id', 'has_gencode_tid', 'transcript_id']].groupby(['gene_id', 'has_gencode_tid']).count().reset_index()\n",
    "temp.head()\n",
    "temp = temp.pivot(index='gene_id', columns='has_gencode_tid', values='transcript_id').reset_index()\n",
    "temp.rename({False: 'n_no_gencode', True: 'n_gencode'}, axis=1, inplace=True)\n",
    "temp.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 99,
   "id": "0e597c84-a3c8-460f-8137-09be99fea15a",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "63774\n",
      "23648\n"
     ]
    }
   ],
   "source": [
    "print(len(temp.index))\n",
    "print(len(temp.loc[temp.n_gencode.isnull()].index))\n",
    "# temp.loc[temp.gene_id=='CHS.166735']\n",
    "# all genes have at least one transcript with a matching tid JUST KIDDING"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "bad418f3-4bf4-453b-8d64-fc3a9b2918ce",
   "metadata": {},
   "source": [
    "## 1/5/2023 - chess3"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 91,
   "id": "a932b2c6-0a66-4d03-89df-ba24f3380323",
   "metadata": {},
   "outputs": [],
   "source": [
    "# gtf = 'chess3.0.gtf'\n",
    "# df = pr.read_gtf(gtf, duplicate_attr=True).as_df()\n",
    "# df = df.loc[df.Feature == 'transcript']\n",
    "df['gencode_tid'] = df.db_xref.str.split('GENCODE:', expand=True)[1]\n",
    "df['gencode_tid'] = df.gencode_tid.str.split(',', expand=True)[0]\n",
    "df['gencode_stable_tid'] = cerberus.get_stable_gid(df, col='gencode_tid')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 86,
   "id": "3d13aaa3-8ba1-4375-b740-5797d55a6d8b",
   "metadata": {},
   "outputs": [],
   "source": [
    "# merge with gencode\n",
    "gtf_df, _, _ = get_gtf_info(how='iso', ver='v29')\n",
    "gtf_df['gid_stable'] = cerberus.get_stable_gid(gtf_df, col='gid')\n",
    "gtf_df['tid_stable'] = cerberus.get_stable_gid(gtf_df, col='tid')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 94,
   "id": "91ee9fab-bab3-485b-b4c5-654c76a3a102",
   "metadata": {},
   "outputs": [],
   "source": [
    "# limit to tid / gid pairings \n",
    "df = df[['gene_id', 'gencode_stable_tid']]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 93,
   "id": "17dec108-bad5-4094-bad0-d8e67c7a91e8",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>tid</th>\n",
       "      <th>t_len</th>\n",
       "      <th>gid</th>\n",
       "      <th>gname</th>\n",
       "      <th>biotype</th>\n",
       "      <th>biotype_category</th>\n",
       "      <th>tf</th>\n",
       "      <th>gid_stable</th>\n",
       "      <th>tid_stable</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENST00000000233.9</td>\n",
       "      <td>1103</td>\n",
       "      <td>ENSG00000004059.10</td>\n",
       "      <td>ARF5</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>False</td>\n",
       "      <td>ENSG00000004059</td>\n",
       "      <td>ENST00000000233</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENST00000000412.7</td>\n",
       "      <td>2756</td>\n",
       "      <td>ENSG00000003056.7</td>\n",
       "      <td>M6PR</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>False</td>\n",
       "      <td>ENSG00000003056</td>\n",
       "      <td>ENST00000000412</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENST00000000442.10</td>\n",
       "      <td>2215</td>\n",
       "      <td>ENSG00000173153.13</td>\n",
       "      <td>ESRRA</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>True</td>\n",
       "      <td>ENSG00000173153</td>\n",
       "      <td>ENST00000000442</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENST00000001008.5</td>\n",
       "      <td>3732</td>\n",
       "      <td>ENSG00000004478.7</td>\n",
       "      <td>FKBP4</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>False</td>\n",
       "      <td>ENSG00000004478</td>\n",
       "      <td>ENST00000001008</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENST00000001146.6</td>\n",
       "      <td>4732</td>\n",
       "      <td>ENSG00000003137.8</td>\n",
       "      <td>CYP26B1</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>False</td>\n",
       "      <td>ENSG00000003137</td>\n",
       "      <td>ENST00000001146</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                  tid  t_len                 gid    gname         biotype  \\\n",
       "0   ENST00000000233.9   1103  ENSG00000004059.10     ARF5  protein_coding   \n",
       "1   ENST00000000412.7   2756   ENSG00000003056.7     M6PR  protein_coding   \n",
       "2  ENST00000000442.10   2215  ENSG00000173153.13    ESRRA  protein_coding   \n",
       "3   ENST00000001008.5   3732   ENSG00000004478.7    FKBP4  protein_coding   \n",
       "4   ENST00000001146.6   4732   ENSG00000003137.8  CYP26B1  protein_coding   \n",
       "\n",
       "  biotype_category     tf       gid_stable       tid_stable  \n",
       "0   protein_coding  False  ENSG00000004059  ENST00000000233  \n",
       "1   protein_coding  False  ENSG00000003056  ENST00000000412  \n",
       "2   protein_coding   True  ENSG00000173153  ENST00000000442  \n",
       "3   protein_coding  False  ENSG00000004478  ENST00000001008  \n",
       "4   protein_coding  False  ENSG00000003137  ENST00000001146  "
      ]
     },
     "execution_count": 93,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gtf_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 96,
   "id": "f9135d0e-b27f-489b-995f-0139359217ea",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_id</th>\n",
       "      <th>gencode_stable_tid</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>CHS.166735</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>CHS.166735</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       gene_id gencode_stable_tid\n",
       "8   CHS.166735                NaN\n",
       "10  CHS.166735                NaN"
      ]
     },
     "execution_count": 96,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.loc[df.gene_id=='CHS.166735']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1f1f999b-effb-4ad4-8a5d-c60ba137bc6e",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d97e4e93-570a-40c9-ac09-71ae5be8ae7a",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "abab1b4d-b543-444b-b918-837105f69091",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c5297b4e-873a-4ee8-ba46-c1fee56607fb",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "4fcc932e-aabd-4653-bfa1-160836c46eb1",
   "metadata": {},
   "source": [
    "## earlier"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "id": "f98dba62-5932-4ad0-866b-4effb5a12534",
   "metadata": {},
   "outputs": [],
   "source": [
    "gtf = 'chess_structure_v1.2.1_sorted.gtf'\n",
    "meta = 'iso_all_v1.2.csv'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "41ec1971-83aa-47a1-8089-17bfd6678ac1",
   "metadata": {},
   "outputs": [],
   "source": [
    "gtf_df = pr.read_gtf(gtf, duplicate_attr=True)\n",
    "meta_df = pd.read_csv(meta)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "ce6c65f7-0d8d-43cc-85d8-424b3f00e144",
   "metadata": {},
   "outputs": [],
   "source": [
    "gtf_df = gtf_df.as_df()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "49f61d27-2f75-4561-9c1b-71424798d311",
   "metadata": {},
   "outputs": [],
   "source": [
    "t_df = gtf_df.loc[gtf_df.Feature == 'transcript']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "0503c129-3d28-498d-8629-39164de01a1c",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0      ENSG00000186092.7\n",
       "3                    NaN\n",
       "12                   NaN\n",
       "13    ENSG00000187634.13\n",
       "16                   NaN\n",
       "Name: old_gene_id, dtype: object"
      ]
     },
     "execution_count": 23,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "t_df.head()\n",
    "t_df.old_gene_id.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "id": "f422acb2-2000-4397-9b7e-ca1b9d113084",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "219915\n",
      "17360\n"
     ]
    }
   ],
   "source": [
    "print(len(t_df.loc[t_df.old_gene_id.isnull()]))\n",
    "print(len(t_df.loc[~t_df.old_gene_id.isnull()]))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "9477923c-209c-4a54-8d4c-fde556488c74",
   "metadata": {},
   "outputs": [],
   "source": [
    "# fix the transcript id section and make the gene id correct aswell\n",
    "t_df['gene_id'] = t_df['old_gene_id']\n",
    "t_df['transcript_id'] = t_df.transcript_id.str.split(';', expand=True)[0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "id": "f631f0f2-4b5a-420d-a18d-0aac359780cb",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "219915\n",
      "17360\n"
     ]
    }
   ],
   "source": [
    "print(len(t_df.loc[t_df.gene_id.isnull()]))\n",
    "print(len(t_df.loc[~t_df.gene_id.isnull()]))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "id": "bc0b24c7-4dae-4dd6-8321-5f77a6567e89",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "128323\n",
      "108952\n"
     ]
    }
   ],
   "source": [
    "print(len(t_df.loc[t_df.gene_name.isnull()]))\n",
    "print(len(t_df.loc[~t_df.gene_name.isnull()]))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d59e2d59-76a0-44b2-b8ec-e2eeac90d7e6",
   "metadata": {},
   "outputs": [],
   "source": [
    "t_df.db_xref.head()\n",
    "print(len(t_df.loc[~(t_df.db_xref.isnull())&(t_df.db_xref.str.contains('GENCODE'))]))\n",
    "print(len(t_df.loc[~(t_df.db_xref.isnull())&~(t_df.db_xref.str.contains('GENCODE'))]))\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "id": "851fee0c-b432-4870-8bf6-f3efe626790b",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array([], dtype=object)"
      ]
     },
     "execution_count": 43,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# this column does not contain gencode gene IDs\n",
    "# t_df.loc[~t_df.Dbxref.isnull(), 'Dbxref'].head().values\n",
    "# t_df.loc[~(t_df.Dbxref.isnull())&(t_df.Dbxref.str.contains('ENSG')), 'Dbxref'].head().values\n",
    "\n",
    "# print(len(t_df.loc[~(t_df.Dbxref.isnull())&(t_df.Dbxref.str.contains('GENCODE'))]))\n",
    "# print(len(t_df.loc[~(t_df.Dbxref.isnull())&~(t_df.Dbxref.str.contains('GENCODE'))]))\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "id": "b5491eb1-b353-4aa7-b7ab-9d310bf68d3a",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "194663\n",
      "42612\n",
      "    Chromosome     Source     Feature   Start     End Score Strand Frame  \\\n",
      "48        chr1  StringTie  transcript  925941  944153     .      +     .   \n",
      "70        chr1  StringTie  transcript  930311  944575     .      +     .   \n",
      "167       chr1  StringTie  transcript  941075  942994     .      +     .   \n",
      "214       chr1  StringTie  transcript  942165  942892     .      +     .   \n",
      "394       chr1  StringTie  transcript  961448  962478     .      +     .   \n",
      "\n",
      "     transcript_id old_transcript_id  ... CDS_inference end_range start_range  \\\n",
      "48   CHS.39.alt104               NaN  ...           NaN       NaN         NaN   \n",
      "70   CHS.39.alt105               NaN  ...           NaN       NaN         NaN   \n",
      "167  CHS.39.alt109               NaN  ...           NaN       NaN         NaN   \n",
      "214  CHS.39.alt110               NaN  ...           NaN       NaN         NaN   \n",
      "394  CHS.42.alt110               NaN  ...           NaN       NaN         NaN   \n",
      "\n",
      "    MANE_gid MANE_tid model_evidence standard_name pseudo transl_except  \\\n",
      "48       NaN      NaN            NaN           NaN    NaN           NaN   \n",
      "70       NaN      NaN            NaN           NaN    NaN           NaN   \n",
      "167      NaN      NaN            NaN           NaN    NaN           NaN   \n",
      "214      NaN      NaN            NaN           NaN    NaN           NaN   \n",
      "394      NaN      NaN            NaN           NaN    NaN           NaN   \n",
      "\n",
      "    CDS_exception  \n",
      "48            NaN  \n",
      "70            NaN  \n",
      "167           NaN  \n",
      "214           NaN  \n",
      "394           NaN  \n",
      "\n",
      "[5 rows x 81 columns]\n"
     ]
    }
   ],
   "source": [
    "print(len(t_df.loc[t_df.GENCODE_ID.isnull()]))\n",
    "print(len(t_df.loc[~t_df.GENCODE_ID.isnull()]))\n",
    "print(t_df.loc[~t_df.GENCODE_ID.isnull()].head())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "10a077a9-694e-4f11-8848-9c0adcbbe7c7",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>CHESS ID</th>\n",
       "      <th>GENE</th>\n",
       "      <th>RefSeq ID</th>\n",
       "      <th>GENCODE ID</th>\n",
       "      <th>pLDDT</th>\n",
       "      <th>introns in mouse</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>137610</th>\n",
       "      <td>CHS.39.alt101</td>\n",
       "      <td>SAMD11</td>\n",
       "      <td>NM_001385641.1</td>\n",
       "      <td>ENST00000616016.5</td>\n",
       "      <td>46.4</td>\n",
       "      <td>FALSE</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "             CHESS ID    GENE       RefSeq ID         GENCODE ID  pLDDT  \\\n",
       "137610  CHS.39.alt101  SAMD11  NM_001385641.1  ENST00000616016.5   46.4   \n",
       "\n",
       "       introns in mouse  \n",
       "137610            FALSE  "
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "meta_df.head()\n",
    "# meta_df.loc[meta_df['CHESS ID'].str.contains('gene_id')]\n",
    "meta_df.loc[meta_df['CHESS ID'] == 'CHS.39.alt101']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "70a3ffc4-d423-44ea-b10e-0ca10555f326",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>transcript_id</th>\n",
       "      <th>gene_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>CHS.131169.alt100;gene_id CHS.131169</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>CHS.131169.0</td>\n",
       "      <td>CHS.131169</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>CHS.39.7</td>\n",
       "      <td>CHS.39</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>CHS.39.alt101;gene_id CHS.39</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>CHS.39.5</td>\n",
       "      <td>CHS.39</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                           transcript_id     gene_id\n",
       "0   CHS.131169.alt100;gene_id CHS.131169         NaN\n",
       "3                           CHS.131169.0  CHS.131169\n",
       "12                              CHS.39.7      CHS.39\n",
       "13          CHS.39.alt101;gene_id CHS.39         NaN\n",
       "16                              CHS.39.5      CHS.39"
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "t_df.loc[t_df.transcript_id.str.contains(';'), ['transcript_id', 'gene_id']]\n",
    "t_df[['transcript_id', 'gene_id']].head()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "80c7f384-af82-4d78-93ee-59a343c07765",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "b43b95ca-d5b0-4564-8068-315cac350c13",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>transcript_id</th>\n",
       "      <th>gene_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>CHS.131169.alt100;gene_id CHS.131169</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>CHS.39.alt101;gene_id CHS.39</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>CHS.39.alt100;gene_id CHS.39</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>CHS.39.alt102;gene_id CHS.39</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>44</th>\n",
       "      <td>CHS.39.alt103;gene_id CHS.39</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4357738</th>\n",
       "      <td>CHS.59348.alt101;gene_id CHS.59348</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4357747</th>\n",
       "      <td>CHS.59356.alt100;gene_id CHS.59356</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4357749</th>\n",
       "      <td>CHS.59356.alt101;gene_id CHS.59356</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4357753</th>\n",
       "      <td>CHS.59356.alt102;gene_id CHS.59356</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4357875</th>\n",
       "      <td>CHS.59359.alt100;gene_id CHS.59359</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>145686 rows Ã— 2 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                transcript_id gene_id\n",
       "0        CHS.131169.alt100;gene_id CHS.131169     NaN\n",
       "13               CHS.39.alt101;gene_id CHS.39     NaN\n",
       "21               CHS.39.alt100;gene_id CHS.39     NaN\n",
       "23               CHS.39.alt102;gene_id CHS.39     NaN\n",
       "44               CHS.39.alt103;gene_id CHS.39     NaN\n",
       "...                                       ...     ...\n",
       "4357738    CHS.59348.alt101;gene_id CHS.59348     NaN\n",
       "4357747    CHS.59356.alt100;gene_id CHS.59356     NaN\n",
       "4357749    CHS.59356.alt101;gene_id CHS.59356     NaN\n",
       "4357753    CHS.59356.alt102;gene_id CHS.59356     NaN\n",
       "4357875    CHS.59359.alt100;gene_id CHS.59359     NaN\n",
       "\n",
       "[145686 rows x 2 columns]"
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "\n",
    "# fix entries with the gene id concatenated\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "19fe7ac7-5453-4bb5-ac96-2109ec17345b",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.4"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
