{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "fd2b45b7-5600-4257-8c22-89422a2c1627",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pyranges as pr\n",
    "import cerberus\n",
    "import pandas as pd"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "625779b0-8b8f-44b1-a132-b97439a684a8",
   "metadata": {},
   "source": [
    "## Some known transcripts aren't from a v29 / v41 source 10/21/22"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "9c1d3786-8102-49a2-9d0e-d04bb50bc88f",
   "metadata": {},
   "outputs": [],
   "source": [
    "h5 = 'gtex_cerberus.h5'\n",
    "ca = cerberus.read(h5)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "40aa73c8-39cc-415d-b726-b05a35ce7af1",
   "metadata": {},
   "outputs": [],
   "source": [
    "df = pd.read_csv('cerberus_transcript_novelty.tsv', sep='\\t')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "a4b898fc-e2f4-45e9-a0c4-89b5b8595c20",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>original_transcript_id</th>\n",
       "      <th>ic</th>\n",
       "      <th>ic_id</th>\n",
       "      <th>tss_id</th>\n",
       "      <th>tss</th>\n",
       "      <th>tes_id</th>\n",
       "      <th>tes</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>original_transcript_name</th>\n",
       "      <th>transcript_triplet</th>\n",
       "      <th>transcript_id</th>\n",
       "      <th>transcript_name</th>\n",
       "      <th>tss_first_sd_issue</th>\n",
       "      <th>tes_last_sa_issue</th>\n",
       "      <th>source</th>\n",
       "      <th>ic_novelty</th>\n",
       "      <th>tss_novelty</th>\n",
       "      <th>tes_novelty</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>462643</th>\n",
       "      <td>ENST00000368347.8#4</td>\n",
       "      <td>6</td>\n",
       "      <td>ENSG00000160753_6</td>\n",
       "      <td>ENSG00000160753_2</td>\n",
       "      <td>2</td>\n",
       "      <td>ENSG00000160753_1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>ENSG00000160753</td>\n",
       "      <td>RUSC1</td>\n",
       "      <td>RUSC1-202</td>\n",
       "      <td>[2,6,1]</td>\n",
       "      <td>ENSG00000160753[2,6,1]</td>\n",
       "      <td>RUSC1[2,6,1]</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>lapa</td>\n",
       "      <td>Known</td>\n",
       "      <td>Known</td>\n",
       "      <td>Known</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>778395</th>\n",
       "      <td>ENST00000368347.8</td>\n",
       "      <td>6</td>\n",
       "      <td>ENSG00000160753_6</td>\n",
       "      <td>ENSG00000160753_2</td>\n",
       "      <td>2</td>\n",
       "      <td>ENSG00000160753_1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>ENSG00000160753</td>\n",
       "      <td>RUSC1</td>\n",
       "      <td>ENST00000368347.8</td>\n",
       "      <td>[2,6,1]</td>\n",
       "      <td>ENSG00000160753[2,6,1]</td>\n",
       "      <td>RUSC1[2,6,1]</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>gtex</td>\n",
       "      <td>Known</td>\n",
       "      <td>Known</td>\n",
       "      <td>Known</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       original_transcript_id  ic              ic_id             tss_id  tss  \\\n",
       "462643    ENST00000368347.8#4   6  ENSG00000160753_6  ENSG00000160753_2    2   \n",
       "778395      ENST00000368347.8   6  ENSG00000160753_6  ENSG00000160753_2    2   \n",
       "\n",
       "                   tes_id  tes          gene_id gene_name  \\\n",
       "462643  ENSG00000160753_1  1.0  ENSG00000160753     RUSC1   \n",
       "778395  ENSG00000160753_1  1.0  ENSG00000160753     RUSC1   \n",
       "\n",
       "       original_transcript_name transcript_triplet           transcript_id  \\\n",
       "462643                RUSC1-202            [2,6,1]  ENSG00000160753[2,6,1]   \n",
       "778395        ENST00000368347.8            [2,6,1]  ENSG00000160753[2,6,1]   \n",
       "\n",
       "       transcript_name  tss_first_sd_issue  tes_last_sa_issue source  \\\n",
       "462643    RUSC1[2,6,1]               False              False   lapa   \n",
       "778395    RUSC1[2,6,1]               False              False   gtex   \n",
       "\n",
       "       ic_novelty tss_novelty tes_novelty  \n",
       "462643      Known       Known       Known  \n",
       "778395      Known       Known       Known  "
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.loc[df.transcript_id == 'ENSG00000160753[2,6,1]']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "e6791d8c-f010-45c9-bb30-16177c7a5511",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>original_transcript_id</th>\n",
       "      <th>ic</th>\n",
       "      <th>ic_id</th>\n",
       "      <th>tss_id</th>\n",
       "      <th>tss</th>\n",
       "      <th>tes_id</th>\n",
       "      <th>tes</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>original_transcript_name</th>\n",
       "      <th>transcript_triplet</th>\n",
       "      <th>transcript_id</th>\n",
       "      <th>transcript_name</th>\n",
       "      <th>tss_first_sd_issue</th>\n",
       "      <th>tes_last_sa_issue</th>\n",
       "      <th>source</th>\n",
       "      <th>ic_novelty</th>\n",
       "      <th>tss_novelty</th>\n",
       "      <th>tes_novelty</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>180</th>\n",
       "      <td>ENST00000292254.8</td>\n",
       "      <td>3</td>\n",
       "      <td>ENSG00000160753_3</td>\n",
       "      <td>ENSG00000160753_2</td>\n",
       "      <td>2</td>\n",
       "      <td>ENSG00000160753_1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>ENSG00000160753</td>\n",
       "      <td>RUSC1</td>\n",
       "      <td>RUSC1-201</td>\n",
       "      <td>[2,3,1]</td>\n",
       "      <td>ENSG00000160753[2,3,1]</td>\n",
       "      <td>RUSC1[2,3,1]</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>v40</td>\n",
       "      <td>Known</td>\n",
       "      <td>Known</td>\n",
       "      <td>Known</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1250</th>\n",
       "      <td>ENST00000368347.8</td>\n",
       "      <td>6</td>\n",
       "      <td>ENSG00000160753_6</td>\n",
       "      <td>ENSG00000160753_3</td>\n",
       "      <td>3</td>\n",
       "      <td>ENSG00000160753_1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>ENSG00000160753</td>\n",
       "      <td>RUSC1</td>\n",
       "      <td>RUSC1-202</td>\n",
       "      <td>[3,6,1]</td>\n",
       "      <td>ENSG00000160753[3,6,1]</td>\n",
       "      <td>RUSC1[3,6,1]</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>v40</td>\n",
       "      <td>Known</td>\n",
       "      <td>Known</td>\n",
       "      <td>Known</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1251</th>\n",
       "      <td>ENST00000368349.8</td>\n",
       "      <td>2</td>\n",
       "      <td>ENSG00000160753_2</td>\n",
       "      <td>ENSG00000160753_2</td>\n",
       "      <td>2</td>\n",
       "      <td>ENSG00000160753_1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>ENSG00000160753</td>\n",
       "      <td>RUSC1</td>\n",
       "      <td>RUSC1-203</td>\n",
       "      <td>[2,2,1]</td>\n",
       "      <td>ENSG00000160753[2,2,1]</td>\n",
       "      <td>RUSC1[2,2,1]</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>v40</td>\n",
       "      <td>Known</td>\n",
       "      <td>Known</td>\n",
       "      <td>Known</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1252</th>\n",
       "      <td>ENST00000368352.10</td>\n",
       "      <td>1</td>\n",
       "      <td>ENSG00000160753_1</td>\n",
       "      <td>ENSG00000160753_1</td>\n",
       "      <td>1</td>\n",
       "      <td>ENSG00000160753_1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>ENSG00000160753</td>\n",
       "      <td>RUSC1</td>\n",
       "      <td>RUSC1-204</td>\n",
       "      <td>[1,1,1]</td>\n",
       "      <td>ENSG00000160753[1,1,1]</td>\n",
       "      <td>RUSC1[1,1,1]</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>v40</td>\n",
       "      <td>Known</td>\n",
       "      <td>Known</td>\n",
       "      <td>Known</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1253</th>\n",
       "      <td>ENST00000368354.7</td>\n",
       "      <td>5</td>\n",
       "      <td>ENSG00000160753_5</td>\n",
       "      <td>ENSG00000160753_1</td>\n",
       "      <td>1</td>\n",
       "      <td>ENSG00000160753_1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>ENSG00000160753</td>\n",
       "      <td>RUSC1</td>\n",
       "      <td>RUSC1-205</td>\n",
       "      <td>[1,5,1]</td>\n",
       "      <td>ENSG00000160753[1,5,1]</td>\n",
       "      <td>RUSC1[1,5,1]</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>v40</td>\n",
       "      <td>Known</td>\n",
       "      <td>Known</td>\n",
       "      <td>Known</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4315</th>\n",
       "      <td>ENST00000462780.5</td>\n",
       "      <td>12</td>\n",
       "      <td>ENSG00000160753_12</td>\n",
       "      <td>ENSG00000160753_2</td>\n",
       "      <td>2</td>\n",
       "      <td>ENSG00000160753_2</td>\n",
       "      <td>2.0</td>\n",
       "      <td>ENSG00000160753</td>\n",
       "      <td>RUSC1</td>\n",
       "      <td>RUSC1-206</td>\n",
       "      <td>[2,12,2]</td>\n",
       "      <td>ENSG00000160753[2,12,2]</td>\n",
       "      <td>RUSC1[2,12,2]</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>v40</td>\n",
       "      <td>Known</td>\n",
       "      <td>Known</td>\n",
       "      <td>Known</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4703</th>\n",
       "      <td>ENST00000468764.5</td>\n",
       "      <td>14</td>\n",
       "      <td>ENSG00000160753_14</td>\n",
       "      <td>ENSG00000160753_2</td>\n",
       "      <td>2</td>\n",
       "      <td>ENSG00000160753_3</td>\n",
       "      <td>3.0</td>\n",
       "      <td>ENSG00000160753</td>\n",
       "      <td>RUSC1</td>\n",
       "      <td>RUSC1-208</td>\n",
       "      <td>[2,14,3]</td>\n",
       "      <td>ENSG00000160753[2,14,3]</td>\n",
       "      <td>RUSC1[2,14,3]</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>v40</td>\n",
       "      <td>Known</td>\n",
       "      <td>Known</td>\n",
       "      <td>Known</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4953</th>\n",
       "      <td>ENST00000471876.5</td>\n",
       "      <td>8</td>\n",
       "      <td>ENSG00000160753_8</td>\n",
       "      <td>ENSG00000160753_2</td>\n",
       "      <td>2</td>\n",
       "      <td>ENSG00000160753_1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>ENSG00000160753</td>\n",
       "      <td>RUSC1</td>\n",
       "      <td>RUSC1-209</td>\n",
       "      <td>[2,8,1]</td>\n",
       "      <td>ENSG00000160753[2,8,1]</td>\n",
       "      <td>RUSC1[2,8,1]</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>v40</td>\n",
       "      <td>Known</td>\n",
       "      <td>Known</td>\n",
       "      <td>Known</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5516</th>\n",
       "      <td>ENST00000479924.5</td>\n",
       "      <td>13</td>\n",
       "      <td>ENSG00000160753_13</td>\n",
       "      <td>ENSG00000160753_2</td>\n",
       "      <td>2</td>\n",
       "      <td>ENSG00000160753_5</td>\n",
       "      <td>5.0</td>\n",
       "      <td>ENSG00000160753</td>\n",
       "      <td>RUSC1</td>\n",
       "      <td>RUSC1-211</td>\n",
       "      <td>[2,13,5]</td>\n",
       "      <td>ENSG00000160753[2,13,5]</td>\n",
       "      <td>RUSC1[2,13,5]</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>v40</td>\n",
       "      <td>Known</td>\n",
       "      <td>Known</td>\n",
       "      <td>Known</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6173</th>\n",
       "      <td>ENST00000490373.5</td>\n",
       "      <td>7</td>\n",
       "      <td>ENSG00000160753_7</td>\n",
       "      <td>ENSG00000160753_2</td>\n",
       "      <td>2</td>\n",
       "      <td>ENSG00000160753_2</td>\n",
       "      <td>2.0</td>\n",
       "      <td>ENSG00000160753</td>\n",
       "      <td>RUSC1</td>\n",
       "      <td>RUSC1-214</td>\n",
       "      <td>[2,7,2]</td>\n",
       "      <td>ENSG00000160753[2,7,2]</td>\n",
       "      <td>RUSC1[2,7,2]</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>v40</td>\n",
       "      <td>Known</td>\n",
       "      <td>Known</td>\n",
       "      <td>Known</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6310</th>\n",
       "      <td>ENST00000492536.5</td>\n",
       "      <td>15</td>\n",
       "      <td>ENSG00000160753_15</td>\n",
       "      <td>ENSG00000160753_2</td>\n",
       "      <td>2</td>\n",
       "      <td>ENSG00000160753_3</td>\n",
       "      <td>3.0</td>\n",
       "      <td>ENSG00000160753</td>\n",
       "      <td>RUSC1</td>\n",
       "      <td>RUSC1-215</td>\n",
       "      <td>[2,15,3]</td>\n",
       "      <td>ENSG00000160753[2,15,3]</td>\n",
       "      <td>RUSC1[2,15,3]</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>v40</td>\n",
       "      <td>Known</td>\n",
       "      <td>Known</td>\n",
       "      <td>Known</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6649</th>\n",
       "      <td>ENST00000497930.6</td>\n",
       "      <td>9</td>\n",
       "      <td>ENSG00000160753_9</td>\n",
       "      <td>ENSG00000160753_1</td>\n",
       "      <td>1</td>\n",
       "      <td>ENSG00000160753_1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>ENSG00000160753</td>\n",
       "      <td>RUSC1</td>\n",
       "      <td>RUSC1-216</td>\n",
       "      <td>[1,9,1]</td>\n",
       "      <td>ENSG00000160753[1,9,1]</td>\n",
       "      <td>RUSC1[1,9,1]</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>v40</td>\n",
       "      <td>Known</td>\n",
       "      <td>Known</td>\n",
       "      <td>Known</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11311</th>\n",
       "      <td>ENST00000696599.1</td>\n",
       "      <td>4</td>\n",
       "      <td>ENSG00000160753_4</td>\n",
       "      <td>ENSG00000160753_1</td>\n",
       "      <td>1</td>\n",
       "      <td>ENSG00000160753_1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>ENSG00000160753</td>\n",
       "      <td>RUSC1</td>\n",
       "      <td>RUSC1-217</td>\n",
       "      <td>[1,4,1]</td>\n",
       "      <td>ENSG00000160753[1,4,1]</td>\n",
       "      <td>RUSC1[1,4,1]</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>v40</td>\n",
       "      <td>Known</td>\n",
       "      <td>Known</td>\n",
       "      <td>Known</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>246807</th>\n",
       "      <td>ENST00000292254.8</td>\n",
       "      <td>3</td>\n",
       "      <td>ENSG00000160753_3</td>\n",
       "      <td>ENSG00000160753_2</td>\n",
       "      <td>2</td>\n",
       "      <td>ENSG00000160753_1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>ENSG00000160753</td>\n",
       "      <td>RUSC1</td>\n",
       "      <td>RUSC1-201</td>\n",
       "      <td>[2,3,1]</td>\n",
       "      <td>ENSG00000160753[2,3,1]</td>\n",
       "      <td>RUSC1[2,3,1]</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>v29</td>\n",
       "      <td>Known</td>\n",
       "      <td>Known</td>\n",
       "      <td>Known</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>247912</th>\n",
       "      <td>ENST00000368347.8</td>\n",
       "      <td>6</td>\n",
       "      <td>ENSG00000160753_6</td>\n",
       "      <td>ENSG00000160753_3</td>\n",
       "      <td>3</td>\n",
       "      <td>ENSG00000160753_1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>ENSG00000160753</td>\n",
       "      <td>RUSC1</td>\n",
       "      <td>RUSC1-202</td>\n",
       "      <td>[3,6,1]</td>\n",
       "      <td>ENSG00000160753[3,6,1]</td>\n",
       "      <td>RUSC1[3,6,1]</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>v29</td>\n",
       "      <td>Known</td>\n",
       "      <td>Known</td>\n",
       "      <td>Known</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>247913</th>\n",
       "      <td>ENST00000368349.8</td>\n",
       "      <td>2</td>\n",
       "      <td>ENSG00000160753_2</td>\n",
       "      <td>ENSG00000160753_2</td>\n",
       "      <td>2</td>\n",
       "      <td>ENSG00000160753_1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>ENSG00000160753</td>\n",
       "      <td>RUSC1</td>\n",
       "      <td>RUSC1-203</td>\n",
       "      <td>[2,2,1]</td>\n",
       "      <td>ENSG00000160753[2,2,1]</td>\n",
       "      <td>RUSC1[2,2,1]</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>v29</td>\n",
       "      <td>Known</td>\n",
       "      <td>Known</td>\n",
       "      <td>Known</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>247914</th>\n",
       "      <td>ENST00000368352.9</td>\n",
       "      <td>1</td>\n",
       "      <td>ENSG00000160753_1</td>\n",
       "      <td>ENSG00000160753_1</td>\n",
       "      <td>1</td>\n",
       "      <td>ENSG00000160753_1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>ENSG00000160753</td>\n",
       "      <td>RUSC1</td>\n",
       "      <td>RUSC1-204</td>\n",
       "      <td>[1,1,1]</td>\n",
       "      <td>ENSG00000160753[1,1,1]</td>\n",
       "      <td>RUSC1[1,1,1]</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>v29</td>\n",
       "      <td>Known</td>\n",
       "      <td>Known</td>\n",
       "      <td>Known</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>247915</th>\n",
       "      <td>ENST00000368354.7</td>\n",
       "      <td>5</td>\n",
       "      <td>ENSG00000160753_5</td>\n",
       "      <td>ENSG00000160753_1</td>\n",
       "      <td>1</td>\n",
       "      <td>ENSG00000160753_1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>ENSG00000160753</td>\n",
       "      <td>RUSC1</td>\n",
       "      <td>RUSC1-205</td>\n",
       "      <td>[1,5,1]</td>\n",
       "      <td>ENSG00000160753[1,5,1]</td>\n",
       "      <td>RUSC1[1,5,1]</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>v29</td>\n",
       "      <td>Known</td>\n",
       "      <td>Known</td>\n",
       "      <td>Known</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>251011</th>\n",
       "      <td>ENST00000462780.5</td>\n",
       "      <td>12</td>\n",
       "      <td>ENSG00000160753_12</td>\n",
       "      <td>ENSG00000160753_2</td>\n",
       "      <td>2</td>\n",
       "      <td>ENSG00000160753_2</td>\n",
       "      <td>2.0</td>\n",
       "      <td>ENSG00000160753</td>\n",
       "      <td>RUSC1</td>\n",
       "      <td>RUSC1-206</td>\n",
       "      <td>[2,12,2]</td>\n",
       "      <td>ENSG00000160753[2,12,2]</td>\n",
       "      <td>RUSC1[2,12,2]</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>v29</td>\n",
       "      <td>Known</td>\n",
       "      <td>Known</td>\n",
       "      <td>Known</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>251400</th>\n",
       "      <td>ENST00000468764.5</td>\n",
       "      <td>14</td>\n",
       "      <td>ENSG00000160753_14</td>\n",
       "      <td>ENSG00000160753_2</td>\n",
       "      <td>2</td>\n",
       "      <td>ENSG00000160753_3</td>\n",
       "      <td>3.0</td>\n",
       "      <td>ENSG00000160753</td>\n",
       "      <td>RUSC1</td>\n",
       "      <td>RUSC1-208</td>\n",
       "      <td>[2,14,3]</td>\n",
       "      <td>ENSG00000160753[2,14,3]</td>\n",
       "      <td>RUSC1[2,14,3]</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>v29</td>\n",
       "      <td>Known</td>\n",
       "      <td>Known</td>\n",
       "      <td>Known</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>251650</th>\n",
       "      <td>ENST00000471876.5</td>\n",
       "      <td>8</td>\n",
       "      <td>ENSG00000160753_8</td>\n",
       "      <td>ENSG00000160753_2</td>\n",
       "      <td>2</td>\n",
       "      <td>ENSG00000160753_1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>ENSG00000160753</td>\n",
       "      <td>RUSC1</td>\n",
       "      <td>RUSC1-209</td>\n",
       "      <td>[2,8,1]</td>\n",
       "      <td>ENSG00000160753[2,8,1]</td>\n",
       "      <td>RUSC1[2,8,1]</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>v29</td>\n",
       "      <td>Known</td>\n",
       "      <td>Known</td>\n",
       "      <td>Known</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>252218</th>\n",
       "      <td>ENST00000479924.5</td>\n",
       "      <td>13</td>\n",
       "      <td>ENSG00000160753_13</td>\n",
       "      <td>ENSG00000160753_2</td>\n",
       "      <td>2</td>\n",
       "      <td>ENSG00000160753_5</td>\n",
       "      <td>5.0</td>\n",
       "      <td>ENSG00000160753</td>\n",
       "      <td>RUSC1</td>\n",
       "      <td>RUSC1-211</td>\n",
       "      <td>[2,13,5]</td>\n",
       "      <td>ENSG00000160753[2,13,5]</td>\n",
       "      <td>RUSC1[2,13,5]</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>v29</td>\n",
       "      <td>Known</td>\n",
       "      <td>Known</td>\n",
       "      <td>Known</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>252877</th>\n",
       "      <td>ENST00000490373.5</td>\n",
       "      <td>7</td>\n",
       "      <td>ENSG00000160753_7</td>\n",
       "      <td>ENSG00000160753_2</td>\n",
       "      <td>2</td>\n",
       "      <td>ENSG00000160753_2</td>\n",
       "      <td>2.0</td>\n",
       "      <td>ENSG00000160753</td>\n",
       "      <td>RUSC1</td>\n",
       "      <td>RUSC1-214</td>\n",
       "      <td>[2,7,2]</td>\n",
       "      <td>ENSG00000160753[2,7,2]</td>\n",
       "      <td>RUSC1[2,7,2]</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>v29</td>\n",
       "      <td>Known</td>\n",
       "      <td>Known</td>\n",
       "      <td>Known</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>253014</th>\n",
       "      <td>ENST00000492536.5</td>\n",
       "      <td>15</td>\n",
       "      <td>ENSG00000160753_15</td>\n",
       "      <td>ENSG00000160753_2</td>\n",
       "      <td>2</td>\n",
       "      <td>ENSG00000160753_3</td>\n",
       "      <td>3.0</td>\n",
       "      <td>ENSG00000160753</td>\n",
       "      <td>RUSC1</td>\n",
       "      <td>RUSC1-215</td>\n",
       "      <td>[2,15,3]</td>\n",
       "      <td>ENSG00000160753[2,15,3]</td>\n",
       "      <td>RUSC1[2,15,3]</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>v29</td>\n",
       "      <td>Known</td>\n",
       "      <td>Known</td>\n",
       "      <td>Known</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       original_transcript_id  ic               ic_id             tss_id  tss  \\\n",
       "180         ENST00000292254.8   3   ENSG00000160753_3  ENSG00000160753_2    2   \n",
       "1250        ENST00000368347.8   6   ENSG00000160753_6  ENSG00000160753_3    3   \n",
       "1251        ENST00000368349.8   2   ENSG00000160753_2  ENSG00000160753_2    2   \n",
       "1252       ENST00000368352.10   1   ENSG00000160753_1  ENSG00000160753_1    1   \n",
       "1253        ENST00000368354.7   5   ENSG00000160753_5  ENSG00000160753_1    1   \n",
       "4315        ENST00000462780.5  12  ENSG00000160753_12  ENSG00000160753_2    2   \n",
       "4703        ENST00000468764.5  14  ENSG00000160753_14  ENSG00000160753_2    2   \n",
       "4953        ENST00000471876.5   8   ENSG00000160753_8  ENSG00000160753_2    2   \n",
       "5516        ENST00000479924.5  13  ENSG00000160753_13  ENSG00000160753_2    2   \n",
       "6173        ENST00000490373.5   7   ENSG00000160753_7  ENSG00000160753_2    2   \n",
       "6310        ENST00000492536.5  15  ENSG00000160753_15  ENSG00000160753_2    2   \n",
       "6649        ENST00000497930.6   9   ENSG00000160753_9  ENSG00000160753_1    1   \n",
       "11311       ENST00000696599.1   4   ENSG00000160753_4  ENSG00000160753_1    1   \n",
       "246807      ENST00000292254.8   3   ENSG00000160753_3  ENSG00000160753_2    2   \n",
       "247912      ENST00000368347.8   6   ENSG00000160753_6  ENSG00000160753_3    3   \n",
       "247913      ENST00000368349.8   2   ENSG00000160753_2  ENSG00000160753_2    2   \n",
       "247914      ENST00000368352.9   1   ENSG00000160753_1  ENSG00000160753_1    1   \n",
       "247915      ENST00000368354.7   5   ENSG00000160753_5  ENSG00000160753_1    1   \n",
       "251011      ENST00000462780.5  12  ENSG00000160753_12  ENSG00000160753_2    2   \n",
       "251400      ENST00000468764.5  14  ENSG00000160753_14  ENSG00000160753_2    2   \n",
       "251650      ENST00000471876.5   8   ENSG00000160753_8  ENSG00000160753_2    2   \n",
       "252218      ENST00000479924.5  13  ENSG00000160753_13  ENSG00000160753_2    2   \n",
       "252877      ENST00000490373.5   7   ENSG00000160753_7  ENSG00000160753_2    2   \n",
       "253014      ENST00000492536.5  15  ENSG00000160753_15  ENSG00000160753_2    2   \n",
       "\n",
       "                   tes_id  tes          gene_id gene_name  \\\n",
       "180     ENSG00000160753_1  1.0  ENSG00000160753     RUSC1   \n",
       "1250    ENSG00000160753_1  1.0  ENSG00000160753     RUSC1   \n",
       "1251    ENSG00000160753_1  1.0  ENSG00000160753     RUSC1   \n",
       "1252    ENSG00000160753_1  1.0  ENSG00000160753     RUSC1   \n",
       "1253    ENSG00000160753_1  1.0  ENSG00000160753     RUSC1   \n",
       "4315    ENSG00000160753_2  2.0  ENSG00000160753     RUSC1   \n",
       "4703    ENSG00000160753_3  3.0  ENSG00000160753     RUSC1   \n",
       "4953    ENSG00000160753_1  1.0  ENSG00000160753     RUSC1   \n",
       "5516    ENSG00000160753_5  5.0  ENSG00000160753     RUSC1   \n",
       "6173    ENSG00000160753_2  2.0  ENSG00000160753     RUSC1   \n",
       "6310    ENSG00000160753_3  3.0  ENSG00000160753     RUSC1   \n",
       "6649    ENSG00000160753_1  1.0  ENSG00000160753     RUSC1   \n",
       "11311   ENSG00000160753_1  1.0  ENSG00000160753     RUSC1   \n",
       "246807  ENSG00000160753_1  1.0  ENSG00000160753     RUSC1   \n",
       "247912  ENSG00000160753_1  1.0  ENSG00000160753     RUSC1   \n",
       "247913  ENSG00000160753_1  1.0  ENSG00000160753     RUSC1   \n",
       "247914  ENSG00000160753_1  1.0  ENSG00000160753     RUSC1   \n",
       "247915  ENSG00000160753_1  1.0  ENSG00000160753     RUSC1   \n",
       "251011  ENSG00000160753_2  2.0  ENSG00000160753     RUSC1   \n",
       "251400  ENSG00000160753_3  3.0  ENSG00000160753     RUSC1   \n",
       "251650  ENSG00000160753_1  1.0  ENSG00000160753     RUSC1   \n",
       "252218  ENSG00000160753_5  5.0  ENSG00000160753     RUSC1   \n",
       "252877  ENSG00000160753_2  2.0  ENSG00000160753     RUSC1   \n",
       "253014  ENSG00000160753_3  3.0  ENSG00000160753     RUSC1   \n",
       "\n",
       "       original_transcript_name transcript_triplet            transcript_id  \\\n",
       "180                   RUSC1-201            [2,3,1]   ENSG00000160753[2,3,1]   \n",
       "1250                  RUSC1-202            [3,6,1]   ENSG00000160753[3,6,1]   \n",
       "1251                  RUSC1-203            [2,2,1]   ENSG00000160753[2,2,1]   \n",
       "1252                  RUSC1-204            [1,1,1]   ENSG00000160753[1,1,1]   \n",
       "1253                  RUSC1-205            [1,5,1]   ENSG00000160753[1,5,1]   \n",
       "4315                  RUSC1-206           [2,12,2]  ENSG00000160753[2,12,2]   \n",
       "4703                  RUSC1-208           [2,14,3]  ENSG00000160753[2,14,3]   \n",
       "4953                  RUSC1-209            [2,8,1]   ENSG00000160753[2,8,1]   \n",
       "5516                  RUSC1-211           [2,13,5]  ENSG00000160753[2,13,5]   \n",
       "6173                  RUSC1-214            [2,7,2]   ENSG00000160753[2,7,2]   \n",
       "6310                  RUSC1-215           [2,15,3]  ENSG00000160753[2,15,3]   \n",
       "6649                  RUSC1-216            [1,9,1]   ENSG00000160753[1,9,1]   \n",
       "11311                 RUSC1-217            [1,4,1]   ENSG00000160753[1,4,1]   \n",
       "246807                RUSC1-201            [2,3,1]   ENSG00000160753[2,3,1]   \n",
       "247912                RUSC1-202            [3,6,1]   ENSG00000160753[3,6,1]   \n",
       "247913                RUSC1-203            [2,2,1]   ENSG00000160753[2,2,1]   \n",
       "247914                RUSC1-204            [1,1,1]   ENSG00000160753[1,1,1]   \n",
       "247915                RUSC1-205            [1,5,1]   ENSG00000160753[1,5,1]   \n",
       "251011                RUSC1-206           [2,12,2]  ENSG00000160753[2,12,2]   \n",
       "251400                RUSC1-208           [2,14,3]  ENSG00000160753[2,14,3]   \n",
       "251650                RUSC1-209            [2,8,1]   ENSG00000160753[2,8,1]   \n",
       "252218                RUSC1-211           [2,13,5]  ENSG00000160753[2,13,5]   \n",
       "252877                RUSC1-214            [2,7,2]   ENSG00000160753[2,7,2]   \n",
       "253014                RUSC1-215           [2,15,3]  ENSG00000160753[2,15,3]   \n",
       "\n",
       "       transcript_name  tss_first_sd_issue  tes_last_sa_issue source  \\\n",
       "180       RUSC1[2,3,1]               False              False    v40   \n",
       "1250      RUSC1[3,6,1]               False              False    v40   \n",
       "1251      RUSC1[2,2,1]               False              False    v40   \n",
       "1252      RUSC1[1,1,1]               False              False    v40   \n",
       "1253      RUSC1[1,5,1]               False              False    v40   \n",
       "4315     RUSC1[2,12,2]               False              False    v40   \n",
       "4703     RUSC1[2,14,3]               False              False    v40   \n",
       "4953      RUSC1[2,8,1]               False              False    v40   \n",
       "5516     RUSC1[2,13,5]               False              False    v40   \n",
       "6173      RUSC1[2,7,2]               False              False    v40   \n",
       "6310     RUSC1[2,15,3]               False              False    v40   \n",
       "6649      RUSC1[1,9,1]               False              False    v40   \n",
       "11311     RUSC1[1,4,1]               False              False    v40   \n",
       "246807    RUSC1[2,3,1]               False              False    v29   \n",
       "247912    RUSC1[3,6,1]               False              False    v29   \n",
       "247913    RUSC1[2,2,1]               False              False    v29   \n",
       "247914    RUSC1[1,1,1]               False              False    v29   \n",
       "247915    RUSC1[1,5,1]               False              False    v29   \n",
       "251011   RUSC1[2,12,2]               False              False    v29   \n",
       "251400   RUSC1[2,14,3]               False              False    v29   \n",
       "251650    RUSC1[2,8,1]               False              False    v29   \n",
       "252218   RUSC1[2,13,5]               False              False    v29   \n",
       "252877    RUSC1[2,7,2]               False              False    v29   \n",
       "253014   RUSC1[2,15,3]               False              False    v29   \n",
       "\n",
       "       ic_novelty tss_novelty tes_novelty  \n",
       "180         Known       Known       Known  \n",
       "1250        Known       Known       Known  \n",
       "1251        Known       Known       Known  \n",
       "1252        Known       Known       Known  \n",
       "1253        Known       Known       Known  \n",
       "4315        Known       Known       Known  \n",
       "4703        Known       Known       Known  \n",
       "4953        Known       Known       Known  \n",
       "5516        Known       Known       Known  \n",
       "6173        Known       Known       Known  \n",
       "6310        Known       Known       Known  \n",
       "6649        Known       Known       Known  \n",
       "11311       Known       Known       Known  \n",
       "246807      Known       Known       Known  \n",
       "247912      Known       Known       Known  \n",
       "247913      Known       Known       Known  \n",
       "247914      Known       Known       Known  \n",
       "247915      Known       Known       Known  \n",
       "251011      Known       Known       Known  \n",
       "251400      Known       Known       Known  \n",
       "251650      Known       Known       Known  \n",
       "252218      Known       Known       Known  \n",
       "252877      Known       Known       Known  \n",
       "253014      Known       Known       Known  "
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "temp = df.loc[(df.ic_id=='ENSG00000160753_6')|(df.tss_id=='ENSG00000160753_2')|(df.tes_id=='ENSG00000160753_1')]\n",
    "ref_sources = ['v29', 'v40']\n",
    "temp = temp.loc[temp.source.isin(ref_sources)]\n",
    "temp\n",
    "                "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4a03550c-0480-437c-bf79-8489db75ee0a",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "caaf8f42-2a25-40e0-b0c7-cf40a7e61e05",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6d9ef63f-1983-42c8-b68f-838693af789a",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9cc8eff8-1ce3-4f09-9aae-d3ddf51db82c",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "701b724f-4cee-40e1-bae7-e86296e4644a",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "cda43cbd-4e09-4052-95ec-c0b0ce168d1a",
   "metadata": {},
   "source": [
    "## Mapping file has more entries than unique cerberus IDs\n",
    "\n",
    "Regarding the transcripts that are assigned to the same cerberus ID, these mostly look like they have the same original_transcript_id but differ in the suffix. (See first screenshot).\n",
    "I looked at a pair of these transcripts with (ENST00000044462.11 and ENST00000044462.11-1) in the original GTF, and found that they differ only in the 3' UTR (See second screenshot).\n",
    "These regions essentially were collapsed in the cerberus pipeline, which caused the transcripts with differing 3' UTRs in gtex to be consolidated in to one transcript from cerberus"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "id": "7c8ff379-a193-464f-b424-1e31ef9464e5",
   "metadata": {},
   "outputs": [],
   "source": [
    "f1 = 'gtex_to_cerberus_id.tsv'\n",
    "\n",
    "df = pd.read_csv(f1, sep='\\t')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "id": "88db4146-0cd5-46d1-9e7d-24d430443cf6",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "88696\n",
      "88666\n",
      "88696\n"
     ]
    }
   ],
   "source": [
    "print(len(df.index))\n",
    "print(len(df.transcript_id.unique()))\n",
    "print(len(df.original_transcript_id.unique()))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "id": "6fcfe531-be84-4d63-bb18-9262db4325db",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>transcript_id</th>\n",
       "      <th>original_transcript_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>28637</th>\n",
       "      <td>ENSG00000041357[1,1,2]</td>\n",
       "      <td>ENST00000044462.11-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28636</th>\n",
       "      <td>ENSG00000041357[1,1,2]</td>\n",
       "      <td>ENST00000044462.11</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>49250</th>\n",
       "      <td>ENSG00000079156[1,11,2]</td>\n",
       "      <td>694903e4-a452-41de-a9ea-15c71909a3bc_ENSG00000...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>49249</th>\n",
       "      <td>ENSG00000079156[1,11,2]</td>\n",
       "      <td>694903e4-a452-41de-a9ea-15c71909a3bc-1_ENSG000...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>44255</th>\n",
       "      <td>ENSG00000083845[2,2,1]</td>\n",
       "      <td>ENST00000601521.5-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>44254</th>\n",
       "      <td>ENSG00000083845[2,2,1]</td>\n",
       "      <td>ENST00000601521.5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>65889</th>\n",
       "      <td>ENSG00000085871[1,2,6]</td>\n",
       "      <td>ENST00000616265.4-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>65888</th>\n",
       "      <td>ENSG00000085871[1,2,6]</td>\n",
       "      <td>ENST00000616265.4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>56425</th>\n",
       "      <td>ENSG00000101182[1,1,1]</td>\n",
       "      <td>ENST00000370873.8-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>56424</th>\n",
       "      <td>ENSG00000101182[1,1,1]</td>\n",
       "      <td>ENST00000370873.8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>82608</th>\n",
       "      <td>ENSG00000104763[1,8,4]</td>\n",
       "      <td>ENST00000636435.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>82609</th>\n",
       "      <td>ENSG00000104763[1,8,4]</td>\n",
       "      <td>ENST00000636435.1-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>70829</th>\n",
       "      <td>ENSG00000113558[1,11,2]</td>\n",
       "      <td>ENST00000328392.10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>70830</th>\n",
       "      <td>ENSG00000113558[1,11,2]</td>\n",
       "      <td>ENST00000328392.10-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>53249</th>\n",
       "      <td>ENSG00000115053[1,1,2]</td>\n",
       "      <td>ENST00000322723.8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>53250</th>\n",
       "      <td>ENSG00000115053[1,1,2]</td>\n",
       "      <td>ENST00000322723.8-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7732</th>\n",
       "      <td>ENSG00000116251[2,2,2]</td>\n",
       "      <td>ENST00000497965.5-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7731</th>\n",
       "      <td>ENSG00000116251[2,2,2]</td>\n",
       "      <td>ENST00000497965.5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>52109</th>\n",
       "      <td>ENSG00000119865[1,9,1]</td>\n",
       "      <td>38cab2e3-f7be-43a8-ad3d-6bd685755dc7-1_ENSG000...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>52110</th>\n",
       "      <td>ENSG00000119865[1,9,1]</td>\n",
       "      <td>38cab2e3-f7be-43a8-ad3d-6bd685755dc7_ENSG00000...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>55158</th>\n",
       "      <td>ENSG00000124228[1,8,1]</td>\n",
       "      <td>9c35a31c-b633-4fb1-bf27-c7eb7b977961_ENSG00000...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>55157</th>\n",
       "      <td>ENSG00000124228[1,8,1]</td>\n",
       "      <td>9c35a31c-b633-4fb1-bf27-c7eb7b977961-1_ENSG000...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>68799</th>\n",
       "      <td>ENSG00000134056[1,1,5]</td>\n",
       "      <td>ENST00000256441.4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>68800</th>\n",
       "      <td>ENSG00000134056[1,1,5]</td>\n",
       "      <td>ENST00000256441.4-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23727</th>\n",
       "      <td>ENSG00000134897[1,1,3]</td>\n",
       "      <td>ENST00000257336.5-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23726</th>\n",
       "      <td>ENSG00000134897[1,1,3]</td>\n",
       "      <td>ENST00000257336.5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10284</th>\n",
       "      <td>ENSG00000138326[1,7,10]</td>\n",
       "      <td>ENST00000440692.5-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10283</th>\n",
       "      <td>ENSG00000138326[1,7,10]</td>\n",
       "      <td>ENST00000440692.5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>81431</th>\n",
       "      <td>ENSG00000147586[2,10,1]</td>\n",
       "      <td>09eff090-a04f-4960-8ab1-90f428736e8c-1_ENSG000...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>81432</th>\n",
       "      <td>ENSG00000147586[2,10,1]</td>\n",
       "      <td>09eff090-a04f-4960-8ab1-90f428736e8c_ENSG00000...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>82437</th>\n",
       "      <td>ENSG00000147586[2,6,1]</td>\n",
       "      <td>ENST00000519120.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>82438</th>\n",
       "      <td>ENSG00000147586[2,6,1]</td>\n",
       "      <td>ENST00000519120.1-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15203</th>\n",
       "      <td>ENSG00000149196[1,20,1]</td>\n",
       "      <td>e83e2408-1f45-434b-854c-00590f9282ea-1_ENSG000...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15204</th>\n",
       "      <td>ENSG00000149196[1,20,1]</td>\n",
       "      <td>e83e2408-1f45-434b-854c-00590f9282ea_ENSG00000...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>82459</th>\n",
       "      <td>ENSG00000154582[1,1,2]</td>\n",
       "      <td>ENST00000520242.5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>82460</th>\n",
       "      <td>ENSG00000154582[1,1,2]</td>\n",
       "      <td>ENST00000520242.5-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>82435</th>\n",
       "      <td>ENSG00000154582[1,15,2]</td>\n",
       "      <td>ENST00000519082.5-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>82434</th>\n",
       "      <td>ENSG00000154582[1,15,2]</td>\n",
       "      <td>ENST00000519082.5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>82423</th>\n",
       "      <td>ENSG00000154582[1,6,2]</td>\n",
       "      <td>ENST00000518127.5-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>82422</th>\n",
       "      <td>ENSG00000154582[1,6,2]</td>\n",
       "      <td>ENST00000518127.5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>81468</th>\n",
       "      <td>ENSG00000154582[1,7,2]</td>\n",
       "      <td>11a0e2ea-f876-4f1b-bc3a-3a663251644e_ENSG00000...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>81467</th>\n",
       "      <td>ENSG00000154582[1,7,2]</td>\n",
       "      <td>11a0e2ea-f876-4f1b-bc3a-3a663251644e-1_ENSG000...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4305</th>\n",
       "      <td>ENSG00000196890[1,1,5]</td>\n",
       "      <td>ed124237-136f-4bd5-81ba-199d210ed077</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>192</th>\n",
       "      <td>ENSG00000196890[1,1,5]</td>\n",
       "      <td>0e715173-e3c3-4616-921e-ae193fdc93c7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>59557</th>\n",
       "      <td>ENSG00000198125[2,23,1]</td>\n",
       "      <td>c94c828f-e38a-42e3-b45d-b6d0a10cbf1b-1_ENSG000...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>59558</th>\n",
       "      <td>ENSG00000198125[2,23,1]</td>\n",
       "      <td>c94c828f-e38a-42e3-b45d-b6d0a10cbf1b_ENSG00000...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>79260</th>\n",
       "      <td>ENSG00000236753[1,11,5]</td>\n",
       "      <td>ENST00000454515.5-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>79259</th>\n",
       "      <td>ENSG00000236753[1,11,5]</td>\n",
       "      <td>ENST00000454515.5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>77967</th>\n",
       "      <td>ENSG00000236753[1,17,5]</td>\n",
       "      <td>0c1778a0-7fda-43d7-a558-df28207e84dd_ENSG00000...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>77966</th>\n",
       "      <td>ENSG00000236753[1,17,5]</td>\n",
       "      <td>0c1778a0-7fda-43d7-a558-df28207e84dd-1_ENSG000...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>79217</th>\n",
       "      <td>ENSG00000236753[1,7,5]</td>\n",
       "      <td>ENST00000444245.5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>79218</th>\n",
       "      <td>ENSG00000236753[1,7,5]</td>\n",
       "      <td>ENST00000444245.5-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>79148</th>\n",
       "      <td>ENSG00000236753[1,8,5]</td>\n",
       "      <td>ENST00000416220.5-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>79147</th>\n",
       "      <td>ENSG00000236753[1,8,5]</td>\n",
       "      <td>ENST00000416220.5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>75036</th>\n",
       "      <td>ENSG00000272325[1,1,6]</td>\n",
       "      <td>ENST00000607016.1-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>75035</th>\n",
       "      <td>ENSG00000272325[1,1,6]</td>\n",
       "      <td>ENST00000607016.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>71051</th>\n",
       "      <td>ENSG00000273841[1,5,1]</td>\n",
       "      <td>ENST00000508954.3-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>71050</th>\n",
       "      <td>ENSG00000273841[1,5,1]</td>\n",
       "      <td>ENST00000508954.3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>81217</th>\n",
       "      <td>ENSG00000280064[1,1,5]</td>\n",
       "      <td>db8b00b0-87a9-4fd4-a33d-550be7331c10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>80262</th>\n",
       "      <td>ENSG00000280064[1,1,5]</td>\n",
       "      <td>506d0114-3eb0-43fa-a43a-71d8aa9b42a1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                 transcript_id  \\\n",
       "28637   ENSG00000041357[1,1,2]   \n",
       "28636   ENSG00000041357[1,1,2]   \n",
       "49250  ENSG00000079156[1,11,2]   \n",
       "49249  ENSG00000079156[1,11,2]   \n",
       "44255   ENSG00000083845[2,2,1]   \n",
       "44254   ENSG00000083845[2,2,1]   \n",
       "65889   ENSG00000085871[1,2,6]   \n",
       "65888   ENSG00000085871[1,2,6]   \n",
       "56425   ENSG00000101182[1,1,1]   \n",
       "56424   ENSG00000101182[1,1,1]   \n",
       "82608   ENSG00000104763[1,8,4]   \n",
       "82609   ENSG00000104763[1,8,4]   \n",
       "70829  ENSG00000113558[1,11,2]   \n",
       "70830  ENSG00000113558[1,11,2]   \n",
       "53249   ENSG00000115053[1,1,2]   \n",
       "53250   ENSG00000115053[1,1,2]   \n",
       "7732    ENSG00000116251[2,2,2]   \n",
       "7731    ENSG00000116251[2,2,2]   \n",
       "52109   ENSG00000119865[1,9,1]   \n",
       "52110   ENSG00000119865[1,9,1]   \n",
       "55158   ENSG00000124228[1,8,1]   \n",
       "55157   ENSG00000124228[1,8,1]   \n",
       "68799   ENSG00000134056[1,1,5]   \n",
       "68800   ENSG00000134056[1,1,5]   \n",
       "23727   ENSG00000134897[1,1,3]   \n",
       "23726   ENSG00000134897[1,1,3]   \n",
       "10284  ENSG00000138326[1,7,10]   \n",
       "10283  ENSG00000138326[1,7,10]   \n",
       "81431  ENSG00000147586[2,10,1]   \n",
       "81432  ENSG00000147586[2,10,1]   \n",
       "82437   ENSG00000147586[2,6,1]   \n",
       "82438   ENSG00000147586[2,6,1]   \n",
       "15203  ENSG00000149196[1,20,1]   \n",
       "15204  ENSG00000149196[1,20,1]   \n",
       "82459   ENSG00000154582[1,1,2]   \n",
       "82460   ENSG00000154582[1,1,2]   \n",
       "82435  ENSG00000154582[1,15,2]   \n",
       "82434  ENSG00000154582[1,15,2]   \n",
       "82423   ENSG00000154582[1,6,2]   \n",
       "82422   ENSG00000154582[1,6,2]   \n",
       "81468   ENSG00000154582[1,7,2]   \n",
       "81467   ENSG00000154582[1,7,2]   \n",
       "4305    ENSG00000196890[1,1,5]   \n",
       "192     ENSG00000196890[1,1,5]   \n",
       "59557  ENSG00000198125[2,23,1]   \n",
       "59558  ENSG00000198125[2,23,1]   \n",
       "79260  ENSG00000236753[1,11,5]   \n",
       "79259  ENSG00000236753[1,11,5]   \n",
       "77967  ENSG00000236753[1,17,5]   \n",
       "77966  ENSG00000236753[1,17,5]   \n",
       "79217   ENSG00000236753[1,7,5]   \n",
       "79218   ENSG00000236753[1,7,5]   \n",
       "79148   ENSG00000236753[1,8,5]   \n",
       "79147   ENSG00000236753[1,8,5]   \n",
       "75036   ENSG00000272325[1,1,6]   \n",
       "75035   ENSG00000272325[1,1,6]   \n",
       "71051   ENSG00000273841[1,5,1]   \n",
       "71050   ENSG00000273841[1,5,1]   \n",
       "81217   ENSG00000280064[1,1,5]   \n",
       "80262   ENSG00000280064[1,1,5]   \n",
       "\n",
       "                                  original_transcript_id  \n",
       "28637                               ENST00000044462.11-1  \n",
       "28636                                 ENST00000044462.11  \n",
       "49250  694903e4-a452-41de-a9ea-15c71909a3bc_ENSG00000...  \n",
       "49249  694903e4-a452-41de-a9ea-15c71909a3bc-1_ENSG000...  \n",
       "44255                                ENST00000601521.5-1  \n",
       "44254                                  ENST00000601521.5  \n",
       "65889                                ENST00000616265.4-1  \n",
       "65888                                  ENST00000616265.4  \n",
       "56425                                ENST00000370873.8-1  \n",
       "56424                                  ENST00000370873.8  \n",
       "82608                                  ENST00000636435.1  \n",
       "82609                                ENST00000636435.1-1  \n",
       "70829                                 ENST00000328392.10  \n",
       "70830                               ENST00000328392.10-1  \n",
       "53249                                  ENST00000322723.8  \n",
       "53250                                ENST00000322723.8-1  \n",
       "7732                                 ENST00000497965.5-1  \n",
       "7731                                   ENST00000497965.5  \n",
       "52109  38cab2e3-f7be-43a8-ad3d-6bd685755dc7-1_ENSG000...  \n",
       "52110  38cab2e3-f7be-43a8-ad3d-6bd685755dc7_ENSG00000...  \n",
       "55158  9c35a31c-b633-4fb1-bf27-c7eb7b977961_ENSG00000...  \n",
       "55157  9c35a31c-b633-4fb1-bf27-c7eb7b977961-1_ENSG000...  \n",
       "68799                                  ENST00000256441.4  \n",
       "68800                                ENST00000256441.4-1  \n",
       "23727                                ENST00000257336.5-1  \n",
       "23726                                  ENST00000257336.5  \n",
       "10284                                ENST00000440692.5-1  \n",
       "10283                                  ENST00000440692.5  \n",
       "81431  09eff090-a04f-4960-8ab1-90f428736e8c-1_ENSG000...  \n",
       "81432  09eff090-a04f-4960-8ab1-90f428736e8c_ENSG00000...  \n",
       "82437                                  ENST00000519120.1  \n",
       "82438                                ENST00000519120.1-1  \n",
       "15203  e83e2408-1f45-434b-854c-00590f9282ea-1_ENSG000...  \n",
       "15204  e83e2408-1f45-434b-854c-00590f9282ea_ENSG00000...  \n",
       "82459                                  ENST00000520242.5  \n",
       "82460                                ENST00000520242.5-1  \n",
       "82435                                ENST00000519082.5-1  \n",
       "82434                                  ENST00000519082.5  \n",
       "82423                                ENST00000518127.5-1  \n",
       "82422                                  ENST00000518127.5  \n",
       "81468  11a0e2ea-f876-4f1b-bc3a-3a663251644e_ENSG00000...  \n",
       "81467  11a0e2ea-f876-4f1b-bc3a-3a663251644e-1_ENSG000...  \n",
       "4305                ed124237-136f-4bd5-81ba-199d210ed077  \n",
       "192                 0e715173-e3c3-4616-921e-ae193fdc93c7  \n",
       "59557  c94c828f-e38a-42e3-b45d-b6d0a10cbf1b-1_ENSG000...  \n",
       "59558  c94c828f-e38a-42e3-b45d-b6d0a10cbf1b_ENSG00000...  \n",
       "79260                                ENST00000454515.5-1  \n",
       "79259                                  ENST00000454515.5  \n",
       "77967  0c1778a0-7fda-43d7-a558-df28207e84dd_ENSG00000...  \n",
       "77966  0c1778a0-7fda-43d7-a558-df28207e84dd-1_ENSG000...  \n",
       "79217                                  ENST00000444245.5  \n",
       "79218                                ENST00000444245.5-1  \n",
       "79148                                ENST00000416220.5-1  \n",
       "79147                                  ENST00000416220.5  \n",
       "75036                                ENST00000607016.1-1  \n",
       "75035                                  ENST00000607016.1  \n",
       "71051                                ENST00000508954.3-1  \n",
       "71050                                  ENST00000508954.3  \n",
       "81217               db8b00b0-87a9-4fd4-a33d-550be7331c10  \n",
       "80262               506d0114-3eb0-43fa-a43a-71d8aa9b42a1  "
      ]
     },
     "execution_count": 31,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.loc[df.transcript_id.duplicated(keep=False)].sort_values(by='transcript_id')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "626ff7c4-e7f2-41d3-9fbf-cd105530d32f",
   "metadata": {},
   "outputs": [],
   "source": [
    "h5_annot = '../cerberus/cerberus_annot.h5'\n",
    "ca = cerberus.read(h5_annot)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "id": "e62fca2e-2aee-49a2-89bb-813527f6d69d",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Strand</th>\n",
       "      <th>source</th>\n",
       "      <th>Name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>786449</th>\n",
       "      <td>chr15</td>\n",
       "      <td>78548945</td>\n",
       "      <td>78549046</td>\n",
       "      <td>+</td>\n",
       "      <td>gtex</td>\n",
       "      <td>ENSG00000041357_2</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       Chromosome     Start       End Strand source               Name\n",
       "786449      chr15  78548945  78549046      +   gtex  ENSG00000041357_2"
      ]
     },
     "execution_count": 40,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ca.tes_map.loc[(ca.tes_map.source == 'gtex')&(ca.tes_map.Chromosome=='chr15')&(ca.tes_map.Start <= 78548996)&(ca.tes_map.Name.str.contains('ENSG00000041357'))].sort_values(by='Start', ascending=False).head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "id": "9327b1cb-c44d-42b3-883a-c1f2b4b0d571",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Strand</th>\n",
       "      <th>source</th>\n",
       "      <th>Name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>778953</th>\n",
       "      <td>chr15</td>\n",
       "      <td>78548769</td>\n",
       "      <td>78549254</td>\n",
       "      <td>+</td>\n",
       "      <td>v40</td>\n",
       "      <td>ENSG00000041357_2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>783815</th>\n",
       "      <td>chr15</td>\n",
       "      <td>78548769</td>\n",
       "      <td>78549254</td>\n",
       "      <td>+</td>\n",
       "      <td>v29</td>\n",
       "      <td>ENSG00000041357_2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>785589</th>\n",
       "      <td>chr15</td>\n",
       "      <td>78548769</td>\n",
       "      <td>78549254</td>\n",
       "      <td>+</td>\n",
       "      <td>lapa</td>\n",
       "      <td>ENSG00000041357_2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>786449</th>\n",
       "      <td>chr15</td>\n",
       "      <td>78548945</td>\n",
       "      <td>78549046</td>\n",
       "      <td>+</td>\n",
       "      <td>gtex</td>\n",
       "      <td>ENSG00000041357_2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>786450</th>\n",
       "      <td>chr15</td>\n",
       "      <td>78549153</td>\n",
       "      <td>78549254</td>\n",
       "      <td>+</td>\n",
       "      <td>gtex</td>\n",
       "      <td>ENSG00000041357_2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>787255</th>\n",
       "      <td>chr15</td>\n",
       "      <td>78548868</td>\n",
       "      <td>78548869</td>\n",
       "      <td>+</td>\n",
       "      <td>pas</td>\n",
       "      <td>ENSG00000041357_2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>787256</th>\n",
       "      <td>chr15</td>\n",
       "      <td>78548907</td>\n",
       "      <td>78548930</td>\n",
       "      <td>+</td>\n",
       "      <td>pas</td>\n",
       "      <td>ENSG00000041357_2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>787257</th>\n",
       "      <td>chr15</td>\n",
       "      <td>78549188</td>\n",
       "      <td>78549261</td>\n",
       "      <td>+</td>\n",
       "      <td>pas</td>\n",
       "      <td>ENSG00000041357_2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>789942</th>\n",
       "      <td>chr15</td>\n",
       "      <td>78548791</td>\n",
       "      <td>78548792</td>\n",
       "      <td>+</td>\n",
       "      <td>polya_atlas</td>\n",
       "      <td>ENSG00000041357_2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>789943</th>\n",
       "      <td>chr15</td>\n",
       "      <td>78548798</td>\n",
       "      <td>78548833</td>\n",
       "      <td>+</td>\n",
       "      <td>polya_atlas</td>\n",
       "      <td>ENSG00000041357_2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>789944</th>\n",
       "      <td>chr15</td>\n",
       "      <td>78548851</td>\n",
       "      <td>78548879</td>\n",
       "      <td>+</td>\n",
       "      <td>polya_atlas</td>\n",
       "      <td>ENSG00000041357_2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>789945</th>\n",
       "      <td>chr15</td>\n",
       "      <td>78549041</td>\n",
       "      <td>78549126</td>\n",
       "      <td>+</td>\n",
       "      <td>polya_atlas</td>\n",
       "      <td>ENSG00000041357_2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>789946</th>\n",
       "      <td>chr15</td>\n",
       "      <td>78549126</td>\n",
       "      <td>78549160</td>\n",
       "      <td>+</td>\n",
       "      <td>polya_atlas</td>\n",
       "      <td>ENSG00000041357_2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>789947</th>\n",
       "      <td>chr15</td>\n",
       "      <td>78549210</td>\n",
       "      <td>78549263</td>\n",
       "      <td>+</td>\n",
       "      <td>polya_atlas</td>\n",
       "      <td>ENSG00000041357_2</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       Chromosome     Start       End Strand       source               Name\n",
       "778953      chr15  78548769  78549254      +          v40  ENSG00000041357_2\n",
       "783815      chr15  78548769  78549254      +          v29  ENSG00000041357_2\n",
       "785589      chr15  78548769  78549254      +         lapa  ENSG00000041357_2\n",
       "786449      chr15  78548945  78549046      +         gtex  ENSG00000041357_2\n",
       "786450      chr15  78549153  78549254      +         gtex  ENSG00000041357_2\n",
       "787255      chr15  78548868  78548869      +          pas  ENSG00000041357_2\n",
       "787256      chr15  78548907  78548930      +          pas  ENSG00000041357_2\n",
       "787257      chr15  78549188  78549261      +          pas  ENSG00000041357_2\n",
       "789942      chr15  78548791  78548792      +  polya_atlas  ENSG00000041357_2\n",
       "789943      chr15  78548798  78548833      +  polya_atlas  ENSG00000041357_2\n",
       "789944      chr15  78548851  78548879      +  polya_atlas  ENSG00000041357_2\n",
       "789945      chr15  78549041  78549126      +  polya_atlas  ENSG00000041357_2\n",
       "789946      chr15  78549126  78549160      +  polya_atlas  ENSG00000041357_2\n",
       "789947      chr15  78549210  78549263      +  polya_atlas  ENSG00000041357_2"
      ]
     },
     "execution_count": 41,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ca.tes_map.loc[ca.tes_map.Name == 'ENSG00000041357_2']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "id": "6306b833-2250-49d2-af2b-196e1dc330bb",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Strand</th>\n",
       "      <th>Name</th>\n",
       "      <th>source</th>\n",
       "      <th>novelty</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>tes</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>164051</th>\n",
       "      <td>chr15</td>\n",
       "      <td>78548769</td>\n",
       "      <td>78549254</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000041357_2</td>\n",
       "      <td>v40,v29,lapa,gtex,pas,polya_atlas</td>\n",
       "      <td>Known</td>\n",
       "      <td>ENSG00000041357</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       Chromosome     Start       End Strand               Name  \\\n",
       "164051      chr15  78548769  78549254      +  ENSG00000041357_2   \n",
       "\n",
       "                                   source novelty          gene_id  tes  \n",
       "164051  v40,v29,lapa,gtex,pas,polya_atlas   Known  ENSG00000041357    2  "
      ]
     },
     "execution_count": 42,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ca.tes.loc[ca.tes.Name == 'ENSG00000041357_2']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "260c78a2-683d-4998-8d4d-1ab44a422e67",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "878bdfd8-1068-4f57-a551-c5555061cfff",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c4e6f119-abfc-441b-b298-8a32371edf0a",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "e10852ff-f0a3-4503-ac01-44f593849bc5",
   "metadata": {},
   "source": [
    "## Weird stuff going on w/ gtex stuff? \n",
    "* Yunzhe says that some transcripts from the transcript novelty file (which is output directly from the gtex gtf) don't match to the cerberus ids\n",
    "* This is because all of the ones that are removed b/w the two files are from novel genes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "ea1e9d20-046c-41f4-a8d7-c53bfd02355d",
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_gtex_gene_nov(df):\n",
    "    df['gene_novelty'] = 'Known'\n",
    "    df.loc[df.gene_id.str.contains('chr'), 'gene_novelty'] = 'Novel'\n",
    "    return df\n",
    "\n",
    "def get_gtex_transcript_nov(df):\n",
    "    df['transcript_novelty'] = 'Novel'\n",
    "    df.loc[df.transcript_id.str.contains('ENST'), 'transcript_novelty'] = 'Known'\n",
    "    return df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "161828b2-b440-4246-bf7e-b542ebc2646f",
   "metadata": {},
   "outputs": [],
   "source": [
    "f1 = 'gtex_to_cerberus_id.tsv'\n",
    "f2 = 'gtex_transcript_novelty.tsv'\n",
    "\n",
    "df1 = pd.read_csv(f1, sep='\\t')\n",
    "df2 = pd.read_csv(f2, sep='\\t')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "025820fc-a195-47d2-a0c7-f92e1d5311e6",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Unnamed: 0</th>\n",
       "      <th>transcript_id</th>\n",
       "      <th>transcript_novelty</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0</td>\n",
       "      <td>dfe30c5c-ddd7-4ee8-a9ba-74556bd9dd87_chr10:628000</td>\n",
       "      <td>Novel</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>2</td>\n",
       "      <td>01252fde-2aa4-4b0d-99f6-c95a4efcc758_chr10:650000</td>\n",
       "      <td>Novel</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>4</td>\n",
       "      <td>5d705740-efaf-4cd9-9b68-71d6b775a3be_chr10:306...</td>\n",
       "      <td>Novel</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>6</td>\n",
       "      <td>115765b8-4233-49bb-a2bf-93683ac4e094_chr10:313...</td>\n",
       "      <td>Novel</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>8</td>\n",
       "      <td>6d54e9f3-0815-4ea6-a878-1ef2947c1cd0_chr10:482...</td>\n",
       "      <td>Novel</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   Unnamed: 0                                      transcript_id  \\\n",
       "0           0  dfe30c5c-ddd7-4ee8-a9ba-74556bd9dd87_chr10:628000   \n",
       "1           2  01252fde-2aa4-4b0d-99f6-c95a4efcc758_chr10:650000   \n",
       "2           4  5d705740-efaf-4cd9-9b68-71d6b775a3be_chr10:306...   \n",
       "3           6  115765b8-4233-49bb-a2bf-93683ac4e094_chr10:313...   \n",
       "4           8  6d54e9f3-0815-4ea6-a878-1ef2947c1cd0_chr10:482...   \n",
       "\n",
       "  transcript_novelty  \n",
       "0              Novel  \n",
       "1              Novel  \n",
       "2              Novel  \n",
       "3              Novel  \n",
       "4              Novel  "
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df2.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "6eb0175d-05ee-4d19-bdb2-08f2ff85e938",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>transcript_id</th>\n",
       "      <th>original_transcript_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENSG00000086015[5,22,1]</td>\n",
       "      <td>0016cd97-0137-4194-9821-910931c6e972_ENSG00000...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENSG00000162736[1,28,1]</td>\n",
       "      <td>001eea3f-368c-444f-96c0-539cb2204403_ENSG00000...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENSG00000084072[1,23,3]</td>\n",
       "      <td>0039fbf5-7334-40e1-a1a9-2aa34b6e009c_ENSG00000...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENSG00000137965[1,13,1]</td>\n",
       "      <td>003cbbd1-0cb2-43d0-ba48-daabd4d7d518_ENSG00000...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENSG00000127603[39,77,36]</td>\n",
       "      <td>004a6304-0523-477e-ad32-90215c977dca_ENSG00000...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "               transcript_id  \\\n",
       "0    ENSG00000086015[5,22,1]   \n",
       "1    ENSG00000162736[1,28,1]   \n",
       "2    ENSG00000084072[1,23,3]   \n",
       "3    ENSG00000137965[1,13,1]   \n",
       "4  ENSG00000127603[39,77,36]   \n",
       "\n",
       "                              original_transcript_id  \n",
       "0  0016cd97-0137-4194-9821-910931c6e972_ENSG00000...  \n",
       "1  001eea3f-368c-444f-96c0-539cb2204403_ENSG00000...  \n",
       "2  0039fbf5-7334-40e1-a1a9-2aa34b6e009c_ENSG00000...  \n",
       "3  003cbbd1-0cb2-43d0-ba48-daabd4d7d518_ENSG00000...  \n",
       "4  004a6304-0523-477e-ad32-90215c977dca_ENSG00000...  "
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df1.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "26819d36-46e7-4ee0-9f8c-0c960adbdf66",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array(['dfe30c5c-ddd7-4ee8-a9ba-74556bd9dd87_chr10:628000',\n",
       "       '01252fde-2aa4-4b0d-99f6-c95a4efcc758_chr10:650000',\n",
       "       '5d705740-efaf-4cd9-9b68-71d6b775a3be_chr10:3069000',\n",
       "       '115765b8-4233-49bb-a2bf-93683ac4e094_chr10:3132000',\n",
       "       '6d54e9f3-0815-4ea6-a878-1ef2947c1cd0_chr10:4826000'], dtype=object)"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df2.transcript_id.head().values"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "ba3738fc-1d3c-447c-bd2b-795e437d4bd5",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "5022"
      ]
     },
     "execution_count": 23,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# get list of transcripts that don't have a cerberus mapping\n",
    "tids = list(set(df2.transcript_id.tolist())-set(df1.original_transcript_id.tolist()))\n",
    "len(tids)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "0cca4653-e9bf-4d10-b10e-db09bdfdf4a5",
   "metadata": {},
   "outputs": [],
   "source": [
    "gtf = 'flair_filter_transcripts.gtf'\n",
    "df = pr.read_gtf(gtf, as_df=True)\n",
    "df = get_gtex_gene_nov(df)\n",
    "df = get_gtex_transcript_nov(df)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "id": "1e3cc86f-63ba-4e46-8140-a36506a6ca2d",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>transcript_id</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gene_novelty</th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>Novel</th>\n",
       "      <td>5022</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "              transcript_id\n",
       "gene_novelty               \n",
       "Novel                  5022"
      ]
     },
     "execution_count": 26,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "temp = df.loc[df.transcript_id.isin(tids)]\n",
    "temp = temp[['transcript_id', 'gene_novelty']].drop_duplicates()\n",
    "temp.groupby('gene_novelty').count()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0e7419f5-b7e9-4482-8fa4-ffa4341652ef",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a47910b8-1810-49f8-a65e-d61320795f01",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "cc180081-406d-46ed-b0dd-7349f05f341c",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8126a25f-666a-46d2-9f6f-05e2cdf4755f",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "b43958ce-f1d1-4c6d-957c-e6ae3c378e17",
   "metadata": {},
   "source": [
    "## Get cerberus IDs for each transcript"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "22da7ebf-a287-447c-922b-c85ab2a8edf3",
   "metadata": {},
   "outputs": [],
   "source": [
    "ca = cerberus.read('gtex_cerberus.h5')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "2fdb9a5c-f5fc-4691-806a-1e6a9fb2f6bb",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array(['v40', 'v29', 'lapa', 'gtex'], dtype=object)"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ca.t_map.source.unique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "3160a9e4-679c-4406-b341-0b031d9fa8b6",
   "metadata": {},
   "outputs": [],
   "source": [
    "df = ca.t_map.loc[ca.t_map.source == 'gtex']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "96b56d28-d078-4e69-b03f-2467e510ecc9",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>original_transcript_id</th>\n",
       "      <th>ic</th>\n",
       "      <th>ic_id</th>\n",
       "      <th>tss_id</th>\n",
       "      <th>tss</th>\n",
       "      <th>tes_id</th>\n",
       "      <th>tes</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>original_transcript_name</th>\n",
       "      <th>transcript_triplet</th>\n",
       "      <th>transcript_id</th>\n",
       "      <th>transcript_name</th>\n",
       "      <th>tss_first_sd_issue</th>\n",
       "      <th>tes_last_sa_issue</th>\n",
       "      <th>source</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0016cd97-0137-4194-9821-910931c6e972_ENSG00000...</td>\n",
       "      <td>22</td>\n",
       "      <td>ENSG00000086015_22</td>\n",
       "      <td>ENSG00000086015_5</td>\n",
       "      <td>5</td>\n",
       "      <td>ENSG00000086015_1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>ENSG00000086015</td>\n",
       "      <td>MAST2</td>\n",
       "      <td>0016cd97-0137-4194-9821-910931c6e972_ENSG00000...</td>\n",
       "      <td>[5,22,1]</td>\n",
       "      <td>ENSG00000086015[5,22,1]</td>\n",
       "      <td>MAST2[5,22,1]</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>gtex</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>001eea3f-368c-444f-96c0-539cb2204403_ENSG00000...</td>\n",
       "      <td>28</td>\n",
       "      <td>ENSG00000162736_28</td>\n",
       "      <td>ENSG00000162736_1</td>\n",
       "      <td>1</td>\n",
       "      <td>ENSG00000162736_1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>ENSG00000162736</td>\n",
       "      <td>NCSTN</td>\n",
       "      <td>001eea3f-368c-444f-96c0-539cb2204403_ENSG00000...</td>\n",
       "      <td>[1,28,1]</td>\n",
       "      <td>ENSG00000162736[1,28,1]</td>\n",
       "      <td>NCSTN[1,28,1]</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>gtex</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>0039fbf5-7334-40e1-a1a9-2aa34b6e009c_ENSG00000...</td>\n",
       "      <td>23</td>\n",
       "      <td>ENSG00000084072_23</td>\n",
       "      <td>ENSG00000084072_1</td>\n",
       "      <td>1</td>\n",
       "      <td>ENSG00000084072_3</td>\n",
       "      <td>3.0</td>\n",
       "      <td>ENSG00000084072</td>\n",
       "      <td>PPIE</td>\n",
       "      <td>0039fbf5-7334-40e1-a1a9-2aa34b6e009c_ENSG00000...</td>\n",
       "      <td>[1,23,3]</td>\n",
       "      <td>ENSG00000084072[1,23,3]</td>\n",
       "      <td>PPIE[1,23,3]</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>gtex</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>003cbbd1-0cb2-43d0-ba48-daabd4d7d518_ENSG00000...</td>\n",
       "      <td>13</td>\n",
       "      <td>ENSG00000137965_13</td>\n",
       "      <td>ENSG00000137965_1</td>\n",
       "      <td>1</td>\n",
       "      <td>ENSG00000137965_1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>ENSG00000137965</td>\n",
       "      <td>IFI44</td>\n",
       "      <td>003cbbd1-0cb2-43d0-ba48-daabd4d7d518_ENSG00000...</td>\n",
       "      <td>[1,13,1]</td>\n",
       "      <td>ENSG00000137965[1,13,1]</td>\n",
       "      <td>IFI44[1,13,1]</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>gtex</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>004a6304-0523-477e-ad32-90215c977dca_ENSG00000...</td>\n",
       "      <td>77</td>\n",
       "      <td>ENSG00000127603_77</td>\n",
       "      <td>ENSG00000127603_39</td>\n",
       "      <td>39</td>\n",
       "      <td>ENSG00000127603_36</td>\n",
       "      <td>36.0</td>\n",
       "      <td>ENSG00000127603</td>\n",
       "      <td>MACF1</td>\n",
       "      <td>004a6304-0523-477e-ad32-90215c977dca_ENSG00000...</td>\n",
       "      <td>[39,77,36]</td>\n",
       "      <td>ENSG00000127603[39,77,36]</td>\n",
       "      <td>MACF1[39,77,36]</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>gtex</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                              original_transcript_id  ic               ic_id  \\\n",
       "0  0016cd97-0137-4194-9821-910931c6e972_ENSG00000...  22  ENSG00000086015_22   \n",
       "1  001eea3f-368c-444f-96c0-539cb2204403_ENSG00000...  28  ENSG00000162736_28   \n",
       "2  0039fbf5-7334-40e1-a1a9-2aa34b6e009c_ENSG00000...  23  ENSG00000084072_23   \n",
       "3  003cbbd1-0cb2-43d0-ba48-daabd4d7d518_ENSG00000...  13  ENSG00000137965_13   \n",
       "4  004a6304-0523-477e-ad32-90215c977dca_ENSG00000...  77  ENSG00000127603_77   \n",
       "\n",
       "               tss_id  tss              tes_id   tes          gene_id  \\\n",
       "0   ENSG00000086015_5    5   ENSG00000086015_1   1.0  ENSG00000086015   \n",
       "1   ENSG00000162736_1    1   ENSG00000162736_1   1.0  ENSG00000162736   \n",
       "2   ENSG00000084072_1    1   ENSG00000084072_3   3.0  ENSG00000084072   \n",
       "3   ENSG00000137965_1    1   ENSG00000137965_1   1.0  ENSG00000137965   \n",
       "4  ENSG00000127603_39   39  ENSG00000127603_36  36.0  ENSG00000127603   \n",
       "\n",
       "  gene_name                           original_transcript_name  \\\n",
       "0     MAST2  0016cd97-0137-4194-9821-910931c6e972_ENSG00000...   \n",
       "1     NCSTN  001eea3f-368c-444f-96c0-539cb2204403_ENSG00000...   \n",
       "2      PPIE  0039fbf5-7334-40e1-a1a9-2aa34b6e009c_ENSG00000...   \n",
       "3     IFI44  003cbbd1-0cb2-43d0-ba48-daabd4d7d518_ENSG00000...   \n",
       "4     MACF1  004a6304-0523-477e-ad32-90215c977dca_ENSG00000...   \n",
       "\n",
       "  transcript_triplet              transcript_id  transcript_name  \\\n",
       "0           [5,22,1]    ENSG00000086015[5,22,1]    MAST2[5,22,1]   \n",
       "1           [1,28,1]    ENSG00000162736[1,28,1]    NCSTN[1,28,1]   \n",
       "2           [1,23,3]    ENSG00000084072[1,23,3]     PPIE[1,23,3]   \n",
       "3           [1,13,1]    ENSG00000137965[1,13,1]    IFI44[1,13,1]   \n",
       "4         [39,77,36]  ENSG00000127603[39,77,36]  MACF1[39,77,36]   \n",
       "\n",
       "  tss_first_sd_issue tes_last_sa_issue source  \n",
       "0              False             False   gtex  \n",
       "1              False             False   gtex  \n",
       "2              False             False   gtex  \n",
       "3              False             False   gtex  \n",
       "4              False             False   gtex  "
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "d92f784e-9382-49d2-a2dc-58a379cb88f2",
   "metadata": {},
   "outputs": [],
   "source": [
    "df[['transcript_id', 'original_transcript_id']].to_csv('gtex_to_cerberus_id.tsv', sep='\\t', index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7c526ed7-a89d-40c8-bf23-df2986b4f64c",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "befff62d-36cb-449f-8e54-0b86f1974813",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "79559fa8-ce6d-485e-a006-45dc45f00349",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "08538a39-8219-4eae-b24c-38a2b103e81c",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8661ea24-cfcb-43eb-9846-86c50c241b40",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0fc8692e-3349-4510-ad90-da675f944178",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f401345a-7f8a-4764-a469-5b638c6b9b6e",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6d2716ba-e954-4aeb-844a-f38c98428838",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "65cb8eaf-646a-42b5-8785-48d1b5d99e5d",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "4e0c0afc-b7e5-4555-9fcd-894b08177c9e",
   "metadata": {},
   "source": [
    "## Get lists of known / novel genes / transcripts "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "ded2ce67-a730-45b2-a6a3-eae079b1e8ec",
   "metadata": {},
   "outputs": [],
   "source": [
    "gtf = 'flair_filter_transcripts.gtf'\n",
    "df = pr.read_gtf(gtf, as_df=True)\n",
    "df = get_gtex_gene_nov(df)\n",
    "df = get_gtex_transcript_nov(df)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "d89c2777-6853-47dc-b98d-5f710d443c81",
   "metadata": {},
   "outputs": [],
   "source": [
    "df = pr.read_gtf(gtf, as_df=True)\n",
    "df = df[['gene_id']].drop_duplicates()\n",
    "df['gene_novelty'] = 'Known'\n",
    "df.loc[df.gene_id.str.contains('chr'), 'gene_novelty'] = 'Novel'\n",
    "df.head()\n",
    "df.to_csv('gtex_gene_novelty.tsv', sep='\\t')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "36915a50-c0da-4d93-9952-90e979b4788d",
   "metadata": {},
   "outputs": [],
   "source": [
    "df = pr.read_gtf(gtf, as_df=True)\n",
    "ts = df[['transcript_id']].drop_duplicates()\n",
    "ts['transcript_novelty'] = 'Novel'\n",
    "ts.loc[ts.transcript_id.str.contains('ENST'), 'transcript_novelty'] = 'Known'\n",
    "ts.groupby('transcript_novelty').count()\n",
    "ts.head()\n",
    "# ts.to_csv('gtex_transcript_novelty.tsv', sep='\\t')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4819c622-7c1c-4823-80a6-058e53599bfb",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "cc1c54b6-46f6-42f7-a635-3ea7a5921a7e",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "957677ee-9e94-4ca1-8a80-cc913a8e2e59",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a8a1e67a-505f-47fd-b3ed-acdd35306cb6",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "4ab85af9-23b3-4883-95ca-2c6a29b322a0",
   "metadata": {},
   "source": [
    "## Get only known genes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "id": "7bae0bf6-cced-4384-9e45-14a36cccaffb",
   "metadata": {},
   "outputs": [],
   "source": [
    "df = pr.read_gtf('flair_filter_transcripts.gtf', as_df=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "id": "e3259e1c-5954-4ea2-b569-05f0dc9e2985",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "962547\n",
      "951905\n"
     ]
    }
   ],
   "source": [
    "print(len(df.index))\n",
    "df = df.loc[df.gene_id.str.contains('ENSG')]\n",
    "print(len(df.index))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "id": "9d769bae-dd3c-49a1-b786-12a110baffc1",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Source</th>\n",
       "      <th>Feature</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Score</th>\n",
       "      <th>Strand</th>\n",
       "      <th>Frame</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>transcript_id</th>\n",
       "      <th>exon_number</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>chr10</td>\n",
       "      <td>FLAIR</td>\n",
       "      <td>transcript</td>\n",
       "      <td>5498696</td>\n",
       "      <td>5499570</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000178372.7</td>\n",
       "      <td>6168c158-18b2-4590-92f6-c1c968c4b3cb_ENSG00000...</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>chr10</td>\n",
       "      <td>FLAIR</td>\n",
       "      <td>exon</td>\n",
       "      <td>5498696</td>\n",
       "      <td>5499570</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>0</td>\n",
       "      <td>ENSG00000178372.7</td>\n",
       "      <td>6168c158-18b2-4590-92f6-c1c968c4b3cb_ENSG00000...</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>chr10</td>\n",
       "      <td>FLAIR</td>\n",
       "      <td>transcript</td>\n",
       "      <td>14877700</td>\n",
       "      <td>14878678</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>chr10:14873000_ENSG00000272853.1</td>\n",
       "      <td>c81d16e1-be22-46bf-b4a3-27d127f5c5b9</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>chr10</td>\n",
       "      <td>FLAIR</td>\n",
       "      <td>exon</td>\n",
       "      <td>14877700</td>\n",
       "      <td>14878678</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>0</td>\n",
       "      <td>chr10:14873000_ENSG00000272853.1</td>\n",
       "      <td>c81d16e1-be22-46bf-b4a3-27d127f5c5b9</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>34</th>\n",
       "      <td>chr10</td>\n",
       "      <td>FLAIR</td>\n",
       "      <td>transcript</td>\n",
       "      <td>22252065</td>\n",
       "      <td>22255955</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>chr10:22252000_ENSG00000279631.1</td>\n",
       "      <td>7be43b56-a050-448a-8d71-324d0c33b516</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   Chromosome Source     Feature     Start       End Score Strand Frame  \\\n",
       "12      chr10  FLAIR  transcript   5498696   5499570     .      -     .   \n",
       "13      chr10  FLAIR        exon   5498696   5499570     .      -     0   \n",
       "28      chr10  FLAIR  transcript  14877700  14878678     .      -     .   \n",
       "29      chr10  FLAIR        exon  14877700  14878678     .      -     0   \n",
       "34      chr10  FLAIR  transcript  22252065  22255955     .      +     .   \n",
       "\n",
       "                             gene_id  \\\n",
       "12                 ENSG00000178372.7   \n",
       "13                 ENSG00000178372.7   \n",
       "28  chr10:14873000_ENSG00000272853.1   \n",
       "29  chr10:14873000_ENSG00000272853.1   \n",
       "34  chr10:22252000_ENSG00000279631.1   \n",
       "\n",
       "                                        transcript_id exon_number  \n",
       "12  6168c158-18b2-4590-92f6-c1c968c4b3cb_ENSG00000...         NaN  \n",
       "13  6168c158-18b2-4590-92f6-c1c968c4b3cb_ENSG00000...           0  \n",
       "28               c81d16e1-be22-46bf-b4a3-27d127f5c5b9         NaN  \n",
       "29               c81d16e1-be22-46bf-b4a3-27d127f5c5b9           0  \n",
       "34               7be43b56-a050-448a-8d71-324d0c33b516         NaN  "
      ]
     },
     "execution_count": 34,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "id": "64ea7bf1-fcd5-434c-8f8c-31427af28be5",
   "metadata": {},
   "outputs": [],
   "source": [
    "inds = df.loc[df.gene_id.str.contains('chr')].index.tolist()\n",
    "df.loc[inds, 'gene_id'] = df.loc[inds, 'gene_id'].str.split('_', n=1, expand=True)[1]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "id": "c716da06-a00c-4372-abdf-d69a9e3992b8",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Source</th>\n",
       "      <th>Feature</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Score</th>\n",
       "      <th>Strand</th>\n",
       "      <th>Frame</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>transcript_id</th>\n",
       "      <th>exon_number</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>chr10</td>\n",
       "      <td>FLAIR</td>\n",
       "      <td>transcript</td>\n",
       "      <td>5498696</td>\n",
       "      <td>5499570</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000178372.7</td>\n",
       "      <td>6168c158-18b2-4590-92f6-c1c968c4b3cb_ENSG00000...</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>chr10</td>\n",
       "      <td>FLAIR</td>\n",
       "      <td>exon</td>\n",
       "      <td>5498696</td>\n",
       "      <td>5499570</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>0</td>\n",
       "      <td>ENSG00000178372.7</td>\n",
       "      <td>6168c158-18b2-4590-92f6-c1c968c4b3cb_ENSG00000...</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>chr10</td>\n",
       "      <td>FLAIR</td>\n",
       "      <td>transcript</td>\n",
       "      <td>14877700</td>\n",
       "      <td>14878678</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000272853.1</td>\n",
       "      <td>c81d16e1-be22-46bf-b4a3-27d127f5c5b9</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>chr10</td>\n",
       "      <td>FLAIR</td>\n",
       "      <td>exon</td>\n",
       "      <td>14877700</td>\n",
       "      <td>14878678</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>0</td>\n",
       "      <td>ENSG00000272853.1</td>\n",
       "      <td>c81d16e1-be22-46bf-b4a3-27d127f5c5b9</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>34</th>\n",
       "      <td>chr10</td>\n",
       "      <td>FLAIR</td>\n",
       "      <td>transcript</td>\n",
       "      <td>22252065</td>\n",
       "      <td>22255955</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000279631.1</td>\n",
       "      <td>7be43b56-a050-448a-8d71-324d0c33b516</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   Chromosome Source     Feature     Start       End Score Strand Frame  \\\n",
       "12      chr10  FLAIR  transcript   5498696   5499570     .      -     .   \n",
       "13      chr10  FLAIR        exon   5498696   5499570     .      -     0   \n",
       "28      chr10  FLAIR  transcript  14877700  14878678     .      -     .   \n",
       "29      chr10  FLAIR        exon  14877700  14878678     .      -     0   \n",
       "34      chr10  FLAIR  transcript  22252065  22255955     .      +     .   \n",
       "\n",
       "              gene_id                                      transcript_id  \\\n",
       "12  ENSG00000178372.7  6168c158-18b2-4590-92f6-c1c968c4b3cb_ENSG00000...   \n",
       "13  ENSG00000178372.7  6168c158-18b2-4590-92f6-c1c968c4b3cb_ENSG00000...   \n",
       "28  ENSG00000272853.1               c81d16e1-be22-46bf-b4a3-27d127f5c5b9   \n",
       "29  ENSG00000272853.1               c81d16e1-be22-46bf-b4a3-27d127f5c5b9   \n",
       "34  ENSG00000279631.1               7be43b56-a050-448a-8d71-324d0c33b516   \n",
       "\n",
       "   exon_number  \n",
       "12         NaN  \n",
       "13           0  \n",
       "28         NaN  \n",
       "29           0  \n",
       "34         NaN  "
      ]
     },
     "execution_count": 36,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "id": "d4bf81f9-0bce-4d88-ba70-eed1cf146ad0",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "770"
      ]
     },
     "execution_count": 31,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(df.loc[df.gene_id.str.contains('chr')].index)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "id": "586f7a54-06bb-48bd-b647-9f0604f1f05d",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "738165"
      ]
     },
     "execution_count": 28,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(df.loc[~df.transcript_id.str.contains('ENST')].index)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "id": "ab57edef-7a73-4075-9090-a6fb649d6711",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Source</th>\n",
       "      <th>Feature</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Score</th>\n",
       "      <th>Strand</th>\n",
       "      <th>Frame</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>transcript_id</th>\n",
       "      <th>exon_number</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>FLAIR</td>\n",
       "      <td>transcript</td>\n",
       "      <td>1232253</td>\n",
       "      <td>1235041</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>chr1:1232000_ENSG00000176022.4</td>\n",
       "      <td>a9ca82a1-c1d2-4d0a-bcff-672c3e94ba81</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>FLAIR</td>\n",
       "      <td>exon</td>\n",
       "      <td>1232253</td>\n",
       "      <td>1235041</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>chr1:1232000_ENSG00000176022.4</td>\n",
       "      <td>a9ca82a1-c1d2-4d0a-bcff-672c3e94ba81</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>FLAIR</td>\n",
       "      <td>transcript</td>\n",
       "      <td>6393509</td>\n",
       "      <td>6394391</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>chr1:6393000_ENSG00000271746.1</td>\n",
       "      <td>a1359d21-b10f-4022-8db8-9bc0e9407d12</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>chr1</td>\n",
       "      <td>FLAIR</td>\n",
       "      <td>exon</td>\n",
       "      <td>6393509</td>\n",
       "      <td>6394391</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>chr1:6393000_ENSG00000271746.1</td>\n",
       "      <td>a1359d21-b10f-4022-8db8-9bc0e9407d12</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>chr1</td>\n",
       "      <td>FLAIR</td>\n",
       "      <td>transcript</td>\n",
       "      <td>23080845</td>\n",
       "      <td>23084727</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>chr1:23080000_ENSG00000004487.15</td>\n",
       "      <td>13906ae4-0e7e-4e90-88b8-c8fe4b07f94c</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   Chromosome Source     Feature     Start       End Score Strand Frame  \\\n",
       "2        chr1  FLAIR  transcript   1232253   1235041     .      +     .   \n",
       "3        chr1  FLAIR        exon   1232253   1235041     .      +     0   \n",
       "4        chr1  FLAIR  transcript   6393509   6394391     .      +     .   \n",
       "5        chr1  FLAIR        exon   6393509   6394391     .      +     0   \n",
       "10       chr1  FLAIR  transcript  23080845  23084727     .      +     .   \n",
       "\n",
       "                             gene_id                         transcript_id  \\\n",
       "2     chr1:1232000_ENSG00000176022.4  a9ca82a1-c1d2-4d0a-bcff-672c3e94ba81   \n",
       "3     chr1:1232000_ENSG00000176022.4  a9ca82a1-c1d2-4d0a-bcff-672c3e94ba81   \n",
       "4     chr1:6393000_ENSG00000271746.1  a1359d21-b10f-4022-8db8-9bc0e9407d12   \n",
       "5     chr1:6393000_ENSG00000271746.1  a1359d21-b10f-4022-8db8-9bc0e9407d12   \n",
       "10  chr1:23080000_ENSG00000004487.15  13906ae4-0e7e-4e90-88b8-c8fe4b07f94c   \n",
       "\n",
       "   exon_number  \n",
       "2          NaN  \n",
       "3            0  \n",
       "4          NaN  \n",
       "5            0  \n",
       "10         NaN  "
      ]
     },
     "execution_count": 26,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.loc[(df.gene_id.str.contains('chr'))&~(df.transcript_id.str.contains('ENST'))].head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "d3f0ef76-ca32-4997-92f3-3d86d3219c6a",
   "metadata": {},
   "outputs": [],
   "source": [
    "df = pr.PyRanges(df)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "58e117c4-1d69-474f-8393-eeed1007e52e",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "be513cbb-5a45-4f77-a669-5d19f7c1ca8f",
   "metadata": {},
   "outputs": [],
   "source": [
    "## oTher stuff"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "6c8e69cb-5f70-4b57-b984-333c437ee6ff",
   "metadata": {},
   "outputs": [],
   "source": [
    "df = pr.read_gtf('flair_filter_transcripts.gtf')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "e46ace0f-3613-4fdc-a840-f0ad1481a1ce",
   "metadata": {},
   "outputs": [],
   "source": [
    "df = df.as_df()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "d9269478-8a64-46cd-8619-08188d7a71ee",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Source</th>\n",
       "      <th>Feature</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Score</th>\n",
       "      <th>Strand</th>\n",
       "      <th>Frame</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>transcript_id</th>\n",
       "      <th>exon_number</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>FLAIR</td>\n",
       "      <td>transcript</td>\n",
       "      <td>785781</td>\n",
       "      <td>787672</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000237491.8</td>\n",
       "      <td>c0e70edb-ee9e-4863-b844-af03092e0afe</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>FLAIR</td>\n",
       "      <td>exon</td>\n",
       "      <td>785781</td>\n",
       "      <td>787672</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>1862</td>\n",
       "      <td>ENSG00000237491.8</td>\n",
       "      <td>c0e70edb-ee9e-4863-b844-af03092e0afe</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>FLAIR</td>\n",
       "      <td>transcript</td>\n",
       "      <td>1013422</td>\n",
       "      <td>1014541</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>chr1:1013000</td>\n",
       "      <td>f940fb0a-3c9b-4989-b060-7adc659c4467_chr1:1013000</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>FLAIR</td>\n",
       "      <td>exon</td>\n",
       "      <td>1013422</td>\n",
       "      <td>1014541</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>chr1:1013000</td>\n",
       "      <td>f940fb0a-3c9b-4989-b060-7adc659c4467_chr1:1013000</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>FLAIR</td>\n",
       "      <td>transcript</td>\n",
       "      <td>1232253</td>\n",
       "      <td>1235041</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>chr1:1232000_ENSG00000176022.4</td>\n",
       "      <td>a9ca82a1-c1d2-4d0a-bcff-672c3e94ba81</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  Chromosome Source     Feature    Start      End Score Strand Frame  \\\n",
       "0       chr1  FLAIR  transcript   785781   787672     .      +     .   \n",
       "1       chr1  FLAIR        exon   785781   787672     .      +  1862   \n",
       "2       chr1  FLAIR  transcript  1013422  1014541     .      +     .   \n",
       "3       chr1  FLAIR        exon  1013422  1014541     .      +     0   \n",
       "4       chr1  FLAIR  transcript  1232253  1235041     .      +     .   \n",
       "\n",
       "                          gene_id  \\\n",
       "0               ENSG00000237491.8   \n",
       "1               ENSG00000237491.8   \n",
       "2                    chr1:1013000   \n",
       "3                    chr1:1013000   \n",
       "4  chr1:1232000_ENSG00000176022.4   \n",
       "\n",
       "                                       transcript_id exon_number  \n",
       "0               c0e70edb-ee9e-4863-b844-af03092e0afe         NaN  \n",
       "1               c0e70edb-ee9e-4863-b844-af03092e0afe           0  \n",
       "2  f940fb0a-3c9b-4989-b060-7adc659c4467_chr1:1013000         NaN  \n",
       "3  f940fb0a-3c9b-4989-b060-7adc659c4467_chr1:1013000           0  \n",
       "4               a9ca82a1-c1d2-4d0a-bcff-672c3e94ba81         NaN  "
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "e460ee08-4f25-488a-880e-f34d045f2c7a",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_id</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gene_novelty</th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>Known</th>\n",
       "      <td>16274</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Novel</th>\n",
       "      <td>4793</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "              gene_id\n",
       "gene_novelty         \n",
       "Known           16274\n",
       "Novel            4793"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "genes = df[['gene_id']].drop_duplicates()\n",
    "genes['gene_novelty'] = 'Novel'\n",
    "genes.loc[genes.gene_id.str.contains('ENSG'), 'gene_novelty'] = 'Known'\n",
    "genes.groupby('gene_novelty').count()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "c2219bc8-8640-4a19-9fb4-1f00f4434e6c",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>transcript_id</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>transcript_novelty</th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>Known</th>\n",
       "      <td>21393</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Novel</th>\n",
       "      <td>72325</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                    transcript_id\n",
       "transcript_novelty               \n",
       "Known                       21393\n",
       "Novel                       72325"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ts = df[['transcript_id']].drop_duplicates()\n",
    "ts['transcript_novelty'] = 'Novel'\n",
    "ts.loc[ts.transcript_id.str.contains('ENST'), 'transcript_novelty'] = 'Known'\n",
    "ts.groupby('transcript_novelty').count()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "36853452-56af-406e-9081-84c195578ecc",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>transcript_id</th>\n",
       "      <th>transcript_novelty</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>c0e70edb-ee9e-4863-b844-af03092e0afe</td>\n",
       "      <td>Novel</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>f940fb0a-3c9b-4989-b060-7adc659c4467_chr1:1013000</td>\n",
       "      <td>Novel</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>a9ca82a1-c1d2-4d0a-bcff-672c3e94ba81</td>\n",
       "      <td>Novel</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>ffcf6444-2d57-4390-a915-c658140698f5_chr1:1318000</td>\n",
       "      <td>Novel</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>8cd8c06a-ed11-405f-9955-efe2ea3df6bf</td>\n",
       "      <td>Novel</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>962433</th>\n",
       "      <td>6c0c928d-51d1-4083-9382-37fdd1dc7124_ENSG00000...</td>\n",
       "      <td>Novel</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>962460</th>\n",
       "      <td>022b0ce1-a05a-43f4-a43b-64b84d7c5b49_ENSG00000...</td>\n",
       "      <td>Novel</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>962466</th>\n",
       "      <td>ec1fae0e-8321-49b9-846b-ac185a689b42_ENSG00000...</td>\n",
       "      <td>Novel</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>962472</th>\n",
       "      <td>924b8dec-fc8b-49be-a5c9-299e8e31443e_ENSG00000...</td>\n",
       "      <td>Novel</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>962521</th>\n",
       "      <td>02eb4919-694b-4bf3-96db-621db54048c6</td>\n",
       "      <td>Novel</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>72325 rows  2 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                            transcript_id transcript_novelty\n",
       "0                    c0e70edb-ee9e-4863-b844-af03092e0afe              Novel\n",
       "2       f940fb0a-3c9b-4989-b060-7adc659c4467_chr1:1013000              Novel\n",
       "4                    a9ca82a1-c1d2-4d0a-bcff-672c3e94ba81              Novel\n",
       "6       ffcf6444-2d57-4390-a915-c658140698f5_chr1:1318000              Novel\n",
       "8                    8cd8c06a-ed11-405f-9955-efe2ea3df6bf              Novel\n",
       "...                                                   ...                ...\n",
       "962433  6c0c928d-51d1-4083-9382-37fdd1dc7124_ENSG00000...              Novel\n",
       "962460  022b0ce1-a05a-43f4-a43b-64b84d7c5b49_ENSG00000...              Novel\n",
       "962466  ec1fae0e-8321-49b9-846b-ac185a689b42_ENSG00000...              Novel\n",
       "962472  924b8dec-fc8b-49be-a5c9-299e8e31443e_ENSG00000...              Novel\n",
       "962521               02eb4919-694b-4bf3-96db-621db54048c6              Novel\n",
       "\n",
       "[72325 rows x 2 columns]"
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ts.loc[ts.transcript_novelty == 'Novel']"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.4"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
