{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "fd2b45b7-5600-4257-8c22-89422a2c1627",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pyranges as pr"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "6c8e69cb-5f70-4b57-b984-333c437ee6ff",
   "metadata": {},
   "outputs": [],
   "source": [
    "df = pr.read_gtf('flair_filter_transcripts.gtf')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "e46ace0f-3613-4fdc-a840-f0ad1481a1ce",
   "metadata": {},
   "outputs": [],
   "source": [
    "df = df.as_df()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "d9269478-8a64-46cd-8619-08188d7a71ee",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Source</th>\n",
       "      <th>Feature</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Score</th>\n",
       "      <th>Strand</th>\n",
       "      <th>Frame</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>transcript_id</th>\n",
       "      <th>exon_number</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>FLAIR</td>\n",
       "      <td>transcript</td>\n",
       "      <td>785781</td>\n",
       "      <td>787672</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000237491.8</td>\n",
       "      <td>c0e70edb-ee9e-4863-b844-af03092e0afe</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>FLAIR</td>\n",
       "      <td>exon</td>\n",
       "      <td>785781</td>\n",
       "      <td>787672</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>1862</td>\n",
       "      <td>ENSG00000237491.8</td>\n",
       "      <td>c0e70edb-ee9e-4863-b844-af03092e0afe</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>FLAIR</td>\n",
       "      <td>transcript</td>\n",
       "      <td>1013422</td>\n",
       "      <td>1014541</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>chr1:1013000</td>\n",
       "      <td>f940fb0a-3c9b-4989-b060-7adc659c4467_chr1:1013000</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>FLAIR</td>\n",
       "      <td>exon</td>\n",
       "      <td>1013422</td>\n",
       "      <td>1014541</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>chr1:1013000</td>\n",
       "      <td>f940fb0a-3c9b-4989-b060-7adc659c4467_chr1:1013000</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>FLAIR</td>\n",
       "      <td>transcript</td>\n",
       "      <td>1232253</td>\n",
       "      <td>1235041</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>chr1:1232000_ENSG00000176022.4</td>\n",
       "      <td>a9ca82a1-c1d2-4d0a-bcff-672c3e94ba81</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  Chromosome Source     Feature    Start      End Score Strand Frame  \\\n",
       "0       chr1  FLAIR  transcript   785781   787672     .      +     .   \n",
       "1       chr1  FLAIR        exon   785781   787672     .      +  1862   \n",
       "2       chr1  FLAIR  transcript  1013422  1014541     .      +     .   \n",
       "3       chr1  FLAIR        exon  1013422  1014541     .      +     0   \n",
       "4       chr1  FLAIR  transcript  1232253  1235041     .      +     .   \n",
       "\n",
       "                          gene_id  \\\n",
       "0               ENSG00000237491.8   \n",
       "1               ENSG00000237491.8   \n",
       "2                    chr1:1013000   \n",
       "3                    chr1:1013000   \n",
       "4  chr1:1232000_ENSG00000176022.4   \n",
       "\n",
       "                                       transcript_id exon_number  \n",
       "0               c0e70edb-ee9e-4863-b844-af03092e0afe         NaN  \n",
       "1               c0e70edb-ee9e-4863-b844-af03092e0afe           0  \n",
       "2  f940fb0a-3c9b-4989-b060-7adc659c4467_chr1:1013000         NaN  \n",
       "3  f940fb0a-3c9b-4989-b060-7adc659c4467_chr1:1013000           0  \n",
       "4               a9ca82a1-c1d2-4d0a-bcff-672c3e94ba81         NaN  "
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "e460ee08-4f25-488a-880e-f34d045f2c7a",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_id</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gene_novelty</th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>Known</th>\n",
       "      <td>16274</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Novel</th>\n",
       "      <td>4793</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "              gene_id\n",
       "gene_novelty         \n",
       "Known           16274\n",
       "Novel            4793"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "genes = df[['gene_id']].drop_duplicates()\n",
    "genes['gene_novelty'] = 'Novel'\n",
    "genes.loc[genes.gene_id.str.contains('ENSG'), 'gene_novelty'] = 'Known'\n",
    "genes.groupby('gene_novelty').count()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "c2219bc8-8640-4a19-9fb4-1f00f4434e6c",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>transcript_id</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>transcript_novelty</th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>Known</th>\n",
       "      <td>21393</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Novel</th>\n",
       "      <td>72325</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                    transcript_id\n",
       "transcript_novelty               \n",
       "Known                       21393\n",
       "Novel                       72325"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ts = df[['transcript_id']].drop_duplicates()\n",
    "ts['transcript_novelty'] = 'Novel'\n",
    "ts.loc[ts.transcript_id.str.contains('ENST'), 'transcript_novelty'] = 'Known'\n",
    "ts.groupby('transcript_novelty').count()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "36853452-56af-406e-9081-84c195578ecc",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>transcript_id</th>\n",
       "      <th>transcript_novelty</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>c0e70edb-ee9e-4863-b844-af03092e0afe</td>\n",
       "      <td>Novel</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>f940fb0a-3c9b-4989-b060-7adc659c4467_chr1:1013000</td>\n",
       "      <td>Novel</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>a9ca82a1-c1d2-4d0a-bcff-672c3e94ba81</td>\n",
       "      <td>Novel</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>ffcf6444-2d57-4390-a915-c658140698f5_chr1:1318000</td>\n",
       "      <td>Novel</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>8cd8c06a-ed11-405f-9955-efe2ea3df6bf</td>\n",
       "      <td>Novel</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>962433</th>\n",
       "      <td>6c0c928d-51d1-4083-9382-37fdd1dc7124_ENSG00000...</td>\n",
       "      <td>Novel</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>962460</th>\n",
       "      <td>022b0ce1-a05a-43f4-a43b-64b84d7c5b49_ENSG00000...</td>\n",
       "      <td>Novel</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>962466</th>\n",
       "      <td>ec1fae0e-8321-49b9-846b-ac185a689b42_ENSG00000...</td>\n",
       "      <td>Novel</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>962472</th>\n",
       "      <td>924b8dec-fc8b-49be-a5c9-299e8e31443e_ENSG00000...</td>\n",
       "      <td>Novel</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>962521</th>\n",
       "      <td>02eb4919-694b-4bf3-96db-621db54048c6</td>\n",
       "      <td>Novel</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>72325 rows × 2 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                            transcript_id transcript_novelty\n",
       "0                    c0e70edb-ee9e-4863-b844-af03092e0afe              Novel\n",
       "2       f940fb0a-3c9b-4989-b060-7adc659c4467_chr1:1013000              Novel\n",
       "4                    a9ca82a1-c1d2-4d0a-bcff-672c3e94ba81              Novel\n",
       "6       ffcf6444-2d57-4390-a915-c658140698f5_chr1:1318000              Novel\n",
       "8                    8cd8c06a-ed11-405f-9955-efe2ea3df6bf              Novel\n",
       "...                                                   ...                ...\n",
       "962433  6c0c928d-51d1-4083-9382-37fdd1dc7124_ENSG00000...              Novel\n",
       "962460  022b0ce1-a05a-43f4-a43b-64b84d7c5b49_ENSG00000...              Novel\n",
       "962466  ec1fae0e-8321-49b9-846b-ac185a689b42_ENSG00000...              Novel\n",
       "962472  924b8dec-fc8b-49be-a5c9-299e8e31443e_ENSG00000...              Novel\n",
       "962521               02eb4919-694b-4bf3-96db-621db54048c6              Novel\n",
       "\n",
       "[72325 rows x 2 columns]"
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ts.loc[ts.transcript_novelty == 'Novel']"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.4"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
