{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "6e146929-84bd-47ac-b816-f51ef3bf2d4c",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import pyranges as pr"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "775726d6-576c-42ea-9d7f-a7687fb13970",
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_ic(gtf_pr):\n",
    "    \"\"\"\n",
    "    Get a hyphen-separated representation of each transcript's intron chain\n",
    "    from a PyRanges GTF\n",
    "\n",
    "    Parameters:\n",
    "        gtf_pr (pyranges PyRanges): GTF PyRanges object\n",
    "\n",
    "    Returns:\n",
    "        df (pandas DataFrame): DataFrame detailing intron chain, gene, strand,\n",
    "            chromosome, and transcript that intron chain was seen in\n",
    "    \"\"\"\n",
    "    df = gtf_pr.df.copy(deep=True)\n",
    "\n",
    "    # restrict to exon entries\n",
    "    df = df.loc[df.Feature == 'exon']\n",
    "    cols = ['Chromosome', 'Strand', 'Start', 'End', 'transcript_id', 'gene_id']\n",
    "    df = df[cols]\n",
    "\n",
    "    # melt to isolate individual coordinates\n",
    "    df = pd.melt(df, id_vars=['Chromosome', 'Strand', 'transcript_id', 'gene_id'],\n",
    "                value_vars=['Start', 'End'],\n",
    "                value_name='Coord')\n",
    "    df.drop('variable', axis=1, inplace=True)\n",
    "\n",
    "    # sort to order coordinates correctly\n",
    "    df.Coord = df.Coord.astype(int)\n",
    "    fwd = df.loc[df.Strand == '+'].copy(deep=True)\n",
    "    rev = df.loc[df.Strand == '-'].copy(deep=True)\n",
    "\n",
    "    fwd.sort_values(by=['Chromosome', 'transcript_id', 'Coord'],\n",
    "                    ascending=[True, True, True], inplace=True)\n",
    "    rev.sort_values(by=['Chromosome', 'transcript_id', 'Coord'],\n",
    "                    ascending=[True, True, False], inplace=True)\n",
    "    df = pd.concat([fwd, rev])\n",
    "\n",
    "    # create intron chain strings\n",
    "    df.Coord = df.Coord.astype(str)\n",
    "    df = df.groupby(['Chromosome', 'Strand',\n",
    "                     'transcript_id', 'gene_id'], observed=True)['Coord'].apply('-'.join).reset_index()\n",
    "\n",
    "    # remove tss and tes from intron chain\n",
    "    df['temp'] = df.Coord.str.split('-', n=1, expand=True)[1]\n",
    "    df['ic'] = df.temp.str.rsplit('-', n=1, expand=True)[0]\n",
    "\n",
    "    return df"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b2aa16ab-4fa8-4ade-a69f-2f0833e2d422",
   "metadata": {},
   "source": [
    "## How many filtered novel transcripts' intron chains are supported in v39?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 94,
   "id": "d938b4a6-51e9-486a-a3e2-214ab8110b73",
   "metadata": {},
   "outputs": [],
   "source": [
    "# TALON transcripts\n",
    "fname = '/Users/fairliereese/mortazavi_lab/data/rnawg/lr_bulk/talon/human_known_nic_nnc_talon.gtf'\n",
    "talon = pr.read_gtf(fname, duplicate_attr=True)\n",
    "\n",
    "nov_df = get_ic(talon)\n",
    "\n",
    "# limit to just NIC and NNC transcripts\n",
    "talon = talon.df\n",
    "nov_tids = talon.loc[(talon.NNC_transcript=='TRUE')|(talon.NNC_transcript=='TRUE')].transcript_id.tolist()\n",
    "nov_df = nov_df.loc[nov_df.transcript_id.isin(nov_tids)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 95,
   "id": "4732c95d-892d-41af-a8cd-10a8b5cfd1c8",
   "metadata": {},
   "outputs": [],
   "source": [
    "# GENCODE transcripts \n",
    "fname = '/Users/fairliereese/mortazavi_lab/ref/gencode.v39/gencode.v39.annotation.gtf'\n",
    "genc = pr.read_gtf(fname)\n",
    "genc_df = get_ic(genc)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 96,
   "id": "3d1f6062-cdc8-4865-a5de-c56bb8781a8a",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Strand</th>\n",
       "      <th>transcript_id</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>Coord</th>\n",
       "      <th>temp</th>\n",
       "      <th>ic</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>SIRV1</td>\n",
       "      <td>+</td>\n",
       "      <td>ENCODEHT000206942</td>\n",
       "      <td>ENCODEHG000058846</td>\n",
       "      <td>10647-10791-10882-11057-11434-11606</td>\n",
       "      <td>10791-10882-11057-11434-11606</td>\n",
       "      <td>10791-10882-11057-11434</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>SIRV1</td>\n",
       "      <td>-</td>\n",
       "      <td>ENCODEHT000206867</td>\n",
       "      <td>ENCODEHG000058837</td>\n",
       "      <td>10790-10647-10366-10282-7814-7552-6813-6560-64...</td>\n",
       "      <td>10647-10366-10282-7814-7552-6813-6560-6473-633...</td>\n",
       "      <td>10647-10366-10282-7814-7552-6813-6560-6473-633...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>SIRV1</td>\n",
       "      <td>-</td>\n",
       "      <td>ENCODEHT000206868</td>\n",
       "      <td>ENCODEHG000058837</td>\n",
       "      <td>10790-10444-10366-10282-7814-7552-6813-6560-64...</td>\n",
       "      <td>10444-10366-10282-7814-7552-6813-6560-6473-633...</td>\n",
       "      <td>10444-10366-10282-7814-7552-6813-6560-6473-633...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>SIRV1</td>\n",
       "      <td>-</td>\n",
       "      <td>ENCODEHT000206870</td>\n",
       "      <td>ENCODEHG000058837</td>\n",
       "      <td>10790-10553-7808-7552-1484-999</td>\n",
       "      <td>10553-7808-7552-1484-999</td>\n",
       "      <td>10553-7808-7552-1484</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>SIRV3</td>\n",
       "      <td>+</td>\n",
       "      <td>ENCODEHT000206896</td>\n",
       "      <td>ENCODEHG000058844</td>\n",
       "      <td>1980-2005-4568-4779-6057-7988-8127-8207-8755-8937</td>\n",
       "      <td>2005-4568-4779-6057-7988-8127-8207-8755-8937</td>\n",
       "      <td>2005-4568-4779-6057-7988-8127-8207-8755</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  Chromosome Strand      transcript_id            gene_id  \\\n",
       "0      SIRV1      +  ENCODEHT000206942  ENCODEHG000058846   \n",
       "1      SIRV1      -  ENCODEHT000206867  ENCODEHG000058837   \n",
       "2      SIRV1      -  ENCODEHT000206868  ENCODEHG000058837   \n",
       "3      SIRV1      -  ENCODEHT000206870  ENCODEHG000058837   \n",
       "5      SIRV3      +  ENCODEHT000206896  ENCODEHG000058844   \n",
       "\n",
       "                                               Coord  \\\n",
       "0                10647-10791-10882-11057-11434-11606   \n",
       "1  10790-10647-10366-10282-7814-7552-6813-6560-64...   \n",
       "2  10790-10444-10366-10282-7814-7552-6813-6560-64...   \n",
       "3                     10790-10553-7808-7552-1484-999   \n",
       "5  1980-2005-4568-4779-6057-7988-8127-8207-8755-8937   \n",
       "\n",
       "                                                temp  \\\n",
       "0                      10791-10882-11057-11434-11606   \n",
       "1  10647-10366-10282-7814-7552-6813-6560-6473-633...   \n",
       "2  10444-10366-10282-7814-7552-6813-6560-6473-633...   \n",
       "3                           10553-7808-7552-1484-999   \n",
       "5       2005-4568-4779-6057-7988-8127-8207-8755-8937   \n",
       "\n",
       "                                                  ic  \n",
       "0                            10791-10882-11057-11434  \n",
       "1  10647-10366-10282-7814-7552-6813-6560-6473-633...  \n",
       "2  10444-10366-10282-7814-7552-6813-6560-6473-633...  \n",
       "3                               10553-7808-7552-1484  \n",
       "5            2005-4568-4779-6057-7988-8127-8207-8755  "
      ]
     },
     "execution_count": 96,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "nov_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 97,
   "id": "9fea17bf-5d34-4869-8def-fa481a5d6ab7",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Strand</th>\n",
       "      <th>transcript_id</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>Coord</th>\n",
       "      <th>temp</th>\n",
       "      <th>ic</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>+</td>\n",
       "      <td>ENST00000003912.7</td>\n",
       "      <td>ENSG00000001461.17</td>\n",
       "      <td>24415802-24415904-24419290-24419640-24433103-2...</td>\n",
       "      <td>24415904-24419290-24419640-24433103-24433213-2...</td>\n",
       "      <td>24415904-24419290-24419640-24433103-24433213-2...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>+</td>\n",
       "      <td>ENST00000008440.9</td>\n",
       "      <td>ENSG00000010072.16</td>\n",
       "      <td>231338255-231338604-231339768-231339868-231351...</td>\n",
       "      <td>231338604-231339768-231339868-231351303-231352521</td>\n",
       "      <td>231338604-231339768-231339868-231351303</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>+</td>\n",
       "      <td>ENST00000009105.5</td>\n",
       "      <td>ENSG00000008118.10</td>\n",
       "      <td>209583716-209583932-209594954-209595075-209599...</td>\n",
       "      <td>209583932-209594954-209595075-209599982-209600...</td>\n",
       "      <td>209583932-209594954-209595075-209599982-209600...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>+</td>\n",
       "      <td>ENST00000010299.10</td>\n",
       "      <td>ENSG00000009780.16</td>\n",
       "      <td>27726056-27726161-27727471-27727536-27730231-2...</td>\n",
       "      <td>27726161-27727471-27727536-27730231-27730333-2...</td>\n",
       "      <td>27726161-27727471-27727536-27730231-27730333-2...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>+</td>\n",
       "      <td>ENST00000011700.10</td>\n",
       "      <td>ENSG00000048707.15</td>\n",
       "      <td>12277120-12278038-12279498-12279650-12282704-1...</td>\n",
       "      <td>12278038-12279498-12279650-12282704-12283736-1...</td>\n",
       "      <td>12278038-12279498-12279650-12282704-12283736-1...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  Chromosome Strand       transcript_id             gene_id  \\\n",
       "0       chr1      +   ENST00000003912.7  ENSG00000001461.17   \n",
       "1       chr1      +   ENST00000008440.9  ENSG00000010072.16   \n",
       "2       chr1      +   ENST00000009105.5  ENSG00000008118.10   \n",
       "3       chr1      +  ENST00000010299.10  ENSG00000009780.16   \n",
       "4       chr1      +  ENST00000011700.10  ENSG00000048707.15   \n",
       "\n",
       "                                               Coord  \\\n",
       "0  24415802-24415904-24419290-24419640-24433103-2...   \n",
       "1  231338255-231338604-231339768-231339868-231351...   \n",
       "2  209583716-209583932-209594954-209595075-209599...   \n",
       "3  27726056-27726161-27727471-27727536-27730231-2...   \n",
       "4  12277120-12278038-12279498-12279650-12282704-1...   \n",
       "\n",
       "                                                temp  \\\n",
       "0  24415904-24419290-24419640-24433103-24433213-2...   \n",
       "1  231338604-231339768-231339868-231351303-231352521   \n",
       "2  209583932-209594954-209595075-209599982-209600...   \n",
       "3  27726161-27727471-27727536-27730231-27730333-2...   \n",
       "4  12278038-12279498-12279650-12282704-12283736-1...   \n",
       "\n",
       "                                                  ic  \n",
       "0  24415904-24419290-24419640-24433103-24433213-2...  \n",
       "1            231338604-231339768-231339868-231351303  \n",
       "2  209583932-209594954-209595075-209599982-209600...  \n",
       "3  27726161-27727471-27727536-27730231-27730333-2...  \n",
       "4  12278038-12279498-12279650-12282704-12283736-1...  "
      ]
     },
     "execution_count": 97,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "genc_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 98,
   "id": "ba62e2bf-01da-4752-b936-46fcecf88120",
   "metadata": {},
   "outputs": [],
   "source": [
    "# which intron chains from the novel data exist in the new gencode annotation?\n",
    "nov_df.drop(['temp', 'Coord'], axis=1, inplace=True)\n",
    "genc_df.drop(['temp', 'Coord'], axis=1, inplace=True)\n",
    "\n",
    "nov_df = nov_df.merge(genc_df, how='left', on=['Chromosome', 'Strand', 'ic'], suffixes=('_talon', '_genc'))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 99,
   "id": "5cb44080-6565-4e53-a0fe-2c67bbd6fc16",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Strand</th>\n",
       "      <th>transcript_id_talon</th>\n",
       "      <th>gene_id_talon</th>\n",
       "      <th>ic</th>\n",
       "      <th>transcript_id_genc</th>\n",
       "      <th>gene_id_genc</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>SIRV1</td>\n",
       "      <td>+</td>\n",
       "      <td>ENCODEHT000206942</td>\n",
       "      <td>ENCODEHG000058846</td>\n",
       "      <td>10791-10882-11057-11434</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>SIRV1</td>\n",
       "      <td>-</td>\n",
       "      <td>ENCODEHT000206867</td>\n",
       "      <td>ENCODEHG000058837</td>\n",
       "      <td>10647-10366-10282-7814-7552-6813-6560-6473-633...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>SIRV1</td>\n",
       "      <td>-</td>\n",
       "      <td>ENCODEHT000206868</td>\n",
       "      <td>ENCODEHG000058837</td>\n",
       "      <td>10444-10366-10282-7814-7552-6813-6560-6473-633...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>SIRV1</td>\n",
       "      <td>-</td>\n",
       "      <td>ENCODEHT000206870</td>\n",
       "      <td>ENCODEHG000058837</td>\n",
       "      <td>10553-7808-7552-1484</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>SIRV3</td>\n",
       "      <td>+</td>\n",
       "      <td>ENCODEHT000206896</td>\n",
       "      <td>ENCODEHG000058844</td>\n",
       "      <td>2005-4568-4779-6057-7988-8127-8207-8755</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  Chromosome Strand transcript_id_talon      gene_id_talon  \\\n",
       "0      SIRV1      +   ENCODEHT000206942  ENCODEHG000058846   \n",
       "1      SIRV1      -   ENCODEHT000206867  ENCODEHG000058837   \n",
       "2      SIRV1      -   ENCODEHT000206868  ENCODEHG000058837   \n",
       "3      SIRV1      -   ENCODEHT000206870  ENCODEHG000058837   \n",
       "4      SIRV3      +   ENCODEHT000206896  ENCODEHG000058844   \n",
       "\n",
       "                                                  ic transcript_id_genc  \\\n",
       "0                            10791-10882-11057-11434                NaN   \n",
       "1  10647-10366-10282-7814-7552-6813-6560-6473-633...                NaN   \n",
       "2  10444-10366-10282-7814-7552-6813-6560-6473-633...                NaN   \n",
       "3                               10553-7808-7552-1484                NaN   \n",
       "4            2005-4568-4779-6057-7988-8127-8207-8755                NaN   \n",
       "\n",
       "  gene_id_genc  \n",
       "0          NaN  \n",
       "1          NaN  \n",
       "2          NaN  \n",
       "3          NaN  \n",
       "4          NaN  "
      ]
     },
     "execution_count": 99,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "nov_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 100,
   "id": "7ae5c49a-4cc9-411f-a9b5-9b4d76cca75a",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Found 389 novel transcripts w/ their intron chains supported by v39\n"
     ]
    }
   ],
   "source": [
    "temp = nov_df.loc[~nov_df.transcript_id_genc.isnull()]\n",
    "n = len(temp.index)\n",
    "supp_tids = temp.transcript_id_genc.str.split('.', n=1, expand=True)[0].tolist()\n",
    "print('Found {} novel transcripts w/ their intron chains supported by v39'.format(n))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 101,
   "id": "ae99264f-d578-4e80-8511-6629f9762433",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "8285 novel transcripts (w/o sirv / ercc) total\n"
     ]
    }
   ],
   "source": [
    "n = len(nov_df.loc[~(nov_df.Chromosome.str.contains('SIRV'))&~(nov_df.Chromosome.str.contains('ERCC'))].index)\n",
    "print('{} novel transcripts (w/o sirv / ercc) total'.format(n))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 117,
   "id": "a01335d1-588c-47a1-a5c3-b0d990ec8473",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>transcript_id_talon</th>\n",
       "      <th>transcript_id_genc</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>ENCODEHT000222627</td>\n",
       "      <td>ENST00000394517.7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>ENCODEHT000238573</td>\n",
       "      <td>ENST00000685968.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>51</th>\n",
       "      <td>ENCODEHT000494112</td>\n",
       "      <td>ENST00000690421.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>100</th>\n",
       "      <td>ENCODEHT000670446</td>\n",
       "      <td>ENST00000677378.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>101</th>\n",
       "      <td>ENCODEHT000670479</td>\n",
       "      <td>ENST00000678689.1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    transcript_id_talon transcript_id_genc\n",
       "23    ENCODEHT000222627  ENST00000394517.7\n",
       "25    ENCODEHT000238573  ENST00000685968.1\n",
       "51    ENCODEHT000494112  ENST00000690421.1\n",
       "100   ENCODEHT000670446  ENST00000677378.1\n",
       "101   ENCODEHT000670479  ENST00000678689.1"
      ]
     },
     "execution_count": 117,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "temp[['transcript_id_talon', 'transcript_id_genc']].head()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ae6290a7-6676-4ae8-bca0-9c20ed6c9715",
   "metadata": {},
   "source": [
    "## How many of these stable transcript ids exist in v29 and in v39? "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 79,
   "id": "30979b10-4888-46e3-8699-2a0017019380",
   "metadata": {},
   "outputs": [],
   "source": [
    "# were these transcripts designated at novel because these transcripts just didn't exist in prior versions of the annotation?\n",
    "fname = '/Users/fairliereese/mortazavi_lab/data/rnawg/refs/gencode_v29_sirv4_ercc.gtf'\n",
    "v29 = pr.read_gtf(fname)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 80,
   "id": "5997ef88-b487-4241-b1f1-f1b099938130",
   "metadata": {},
   "outputs": [],
   "source": [
    "fname = '/Users/fairliereese/mortazavi_lab/ref/gencode.v39/gencode.v39.annotation.gtf'\n",
    "v39 = pr.read_gtf(fname)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 81,
   "id": "ecadab33-fb6a-4050-bdd3-011a5d44928a",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['ENST00000394517',\n",
       " 'ENST00000685968',\n",
       " 'ENST00000690421',\n",
       " 'ENST00000677378',\n",
       " 'ENST00000678689']"
      ]
     },
     "execution_count": 81,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "supp_tids[:5]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 85,
   "id": "0228386c-ee58-42f7-af7f-12a91ba9aff6",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "59 transcripts whose intron chains are supported in v39 also are annotated in v29\n"
     ]
    }
   ],
   "source": [
    "v29 = v29.df\n",
    "v29['tid_stable'] = v29.transcript_id.str.split('.', n=1, expand=True)[0]\n",
    "\n",
    "temp = v29.loc[v29.tid_stable.isin(supp_tids)]\n",
    "temp = temp.loc[temp.Feature == 'transcript']\n",
    "n = len(temp.index)\n",
    "print('{} transcripts whose intron chains are supported in v39 also are annotated in v29'.format(n))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 86,
   "id": "cbc052ba-7b20-438f-9b3f-6f2c29948fbd",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Source</th>\n",
       "      <th>Feature</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Score</th>\n",
       "      <th>Strand</th>\n",
       "      <th>Frame</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>gene_type</th>\n",
       "      <th>...</th>\n",
       "      <th>tag</th>\n",
       "      <th>havana_transcript</th>\n",
       "      <th>exon_number</th>\n",
       "      <th>exon_id</th>\n",
       "      <th>ont</th>\n",
       "      <th>protein_id</th>\n",
       "      <th>ccdsid</th>\n",
       "      <th>gene_biotype</th>\n",
       "      <th>exon_assignment</th>\n",
       "      <th>tid_stable</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ERCC-00002</td>\n",
       "      <td>ERCC</td>\n",
       "      <td>exon</td>\n",
       "      <td>0</td>\n",
       "      <td>1045</td>\n",
       "      <td>0</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ERCC-00002A</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ERCC_spike_in</td>\n",
       "      <td>NaN</td>\n",
       "      <td>DQ459430</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ERCC-00003</td>\n",
       "      <td>ERCC</td>\n",
       "      <td>exon</td>\n",
       "      <td>0</td>\n",
       "      <td>1007</td>\n",
       "      <td>0</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ERCC-00003A</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ERCC_spike_in</td>\n",
       "      <td>NaN</td>\n",
       "      <td>DQ516784</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ERCC-00004</td>\n",
       "      <td>ERCC</td>\n",
       "      <td>exon</td>\n",
       "      <td>0</td>\n",
       "      <td>507</td>\n",
       "      <td>0</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ERCC-00004A</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ERCC_spike_in</td>\n",
       "      <td>NaN</td>\n",
       "      <td>DQ516752</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ERCC-00009</td>\n",
       "      <td>ERCC</td>\n",
       "      <td>exon</td>\n",
       "      <td>0</td>\n",
       "      <td>968</td>\n",
       "      <td>0</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ERCC-00009A</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ERCC_spike_in</td>\n",
       "      <td>NaN</td>\n",
       "      <td>DQ668364</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ERCC-00012</td>\n",
       "      <td>ERCC</td>\n",
       "      <td>exon</td>\n",
       "      <td>0</td>\n",
       "      <td>978</td>\n",
       "      <td>0</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ERCC-00012A</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ERCC_spike_in</td>\n",
       "      <td>NaN</td>\n",
       "      <td>DQ883670</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 27 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "   Chromosome Source Feature  Start   End Score Strand Frame      gene_id  \\\n",
       "0  ERCC-00002   ERCC    exon      0  1045     0      +     .  ERCC-00002A   \n",
       "1  ERCC-00003   ERCC    exon      0  1007     0      +     .  ERCC-00003A   \n",
       "2  ERCC-00004   ERCC    exon      0   507     0      +     .  ERCC-00004A   \n",
       "3  ERCC-00009   ERCC    exon      0   968     0      +     .  ERCC-00009A   \n",
       "4  ERCC-00012   ERCC    exon      0   978     0      +     .  ERCC-00012A   \n",
       "\n",
       "  gene_type  ...  tag havana_transcript exon_number exon_id  ont protein_id  \\\n",
       "0       NaN  ...  NaN               NaN         NaN     NaN  NaN        NaN   \n",
       "1       NaN  ...  NaN               NaN         NaN     NaN  NaN        NaN   \n",
       "2       NaN  ...  NaN               NaN         NaN     NaN  NaN        NaN   \n",
       "3       NaN  ...  NaN               NaN         NaN     NaN  NaN        NaN   \n",
       "4       NaN  ...  NaN               NaN         NaN     NaN  NaN        NaN   \n",
       "\n",
       "  ccdsid   gene_biotype exon_assignment tid_stable  \n",
       "0    NaN  ERCC_spike_in             NaN   DQ459430  \n",
       "1    NaN  ERCC_spike_in             NaN   DQ516784  \n",
       "2    NaN  ERCC_spike_in             NaN   DQ516752  \n",
       "3    NaN  ERCC_spike_in             NaN   DQ668364  \n",
       "4    NaN  ERCC_spike_in             NaN   DQ883670  \n",
       "\n",
       "[5 rows x 27 columns]"
      ]
     },
     "execution_count": 86,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "v29.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 87,
   "id": "e1929c4e-38bf-4453-8c41-3eb687a74368",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array(['exon', 'gene', 'transcript', 'CDS', 'start_codon', 'stop_codon',\n",
       "       'UTR', 'Selenocysteine'], dtype=object)"
      ]
     },
     "execution_count": 87,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "v29.Feature.unique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 88,
   "id": "dfd6f063-0a09-4abb-b871-aa3d6b84335c",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Source</th>\n",
       "      <th>Feature</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Score</th>\n",
       "      <th>Strand</th>\n",
       "      <th>Frame</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>gene_type</th>\n",
       "      <th>...</th>\n",
       "      <th>tag</th>\n",
       "      <th>havana_transcript</th>\n",
       "      <th>exon_number</th>\n",
       "      <th>exon_id</th>\n",
       "      <th>ont</th>\n",
       "      <th>protein_id</th>\n",
       "      <th>ccdsid</th>\n",
       "      <th>gene_biotype</th>\n",
       "      <th>exon_assignment</th>\n",
       "      <th>tid_stable</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>20380</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>Selenocysteine</td>\n",
       "      <td>25802092</td>\n",
       "      <td>25802095</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000162430.16</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>...</td>\n",
       "      <td>seleno</td>\n",
       "      <td>OTTHUMT00000019314.2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENSP00000355141.2</td>\n",
       "      <td>CCDS41282.1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENST00000361547</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20381</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>Selenocysteine</td>\n",
       "      <td>25812788</td>\n",
       "      <td>25812791</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000162430.16</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>...</td>\n",
       "      <td>seleno</td>\n",
       "      <td>OTTHUMT00000019314.2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENSP00000355141.2</td>\n",
       "      <td>CCDS41282.1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENST00000361547</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20413</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>Selenocysteine</td>\n",
       "      <td>25812788</td>\n",
       "      <td>25812791</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000162430.16</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>...</td>\n",
       "      <td>seleno</td>\n",
       "      <td>OTTHUMT00000019315.2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENSP00000363434.1</td>\n",
       "      <td>CCDS41283.1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENST00000374315</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>46423</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>Selenocysteine</td>\n",
       "      <td>53904703</td>\n",
       "      <td>53904706</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000211452.10</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>...</td>\n",
       "      <td>seleno</td>\n",
       "      <td>OTTHUMT00000388166.2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENSP00000432797.1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENST00000529589</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>46435</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>Selenocysteine</td>\n",
       "      <td>53904703</td>\n",
       "      <td>53904706</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000211452.10</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>...</td>\n",
       "      <td>seleno</td>\n",
       "      <td>OTTHUMT00000023247.3</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENSP00000354643.3</td>\n",
       "      <td>CCDS41339.1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENST00000361921</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2620238</th>\n",
       "      <td>chr22</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>Selenocysteine</td>\n",
       "      <td>19877110</td>\n",
       "      <td>19877113</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000184470.20</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>...</td>\n",
       "      <td>seleno</td>\n",
       "      <td>OTTHUMT00000314903.4</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENSP00000383365.1</td>\n",
       "      <td>CCDS42981.1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENST00000400521</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2620286</th>\n",
       "      <td>chr22</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>Selenocysteine</td>\n",
       "      <td>19877110</td>\n",
       "      <td>19877113</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000184470.20</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>...</td>\n",
       "      <td>seleno</td>\n",
       "      <td>OTTHUMT00000314927.3</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENSP00000485499.2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENST00000485358</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2620306</th>\n",
       "      <td>chr22</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>Selenocysteine</td>\n",
       "      <td>19877110</td>\n",
       "      <td>19877113</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000184470.20</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>...</td>\n",
       "      <td>seleno</td>\n",
       "      <td>OTTHUMT00000314928.3</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENSP00000485466.2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENST00000462843</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2629658</th>\n",
       "      <td>chr22</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>Selenocysteine</td>\n",
       "      <td>31105950</td>\n",
       "      <td>31105953</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000198832.10</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>...</td>\n",
       "      <td>seleno</td>\n",
       "      <td>OTTHUMT00000321789.1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENSP00000384564.1</td>\n",
       "      <td>CCDS43003.1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENST00000402395</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2629683</th>\n",
       "      <td>chr22</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>Selenocysteine</td>\n",
       "      <td>31105950</td>\n",
       "      <td>31105953</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000198832.10</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>...</td>\n",
       "      <td>seleno</td>\n",
       "      <td>OTTHUMT00000321788.1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENSP00000383155.2</td>\n",
       "      <td>CCDS43003.1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENST00000400299</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>120 rows × 27 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        Chromosome  Source         Feature     Start       End Score Strand  \\\n",
       "20380         chr1  HAVANA  Selenocysteine  25802092  25802095     .      +   \n",
       "20381         chr1  HAVANA  Selenocysteine  25812788  25812791     .      +   \n",
       "20413         chr1  HAVANA  Selenocysteine  25812788  25812791     .      +   \n",
       "46423         chr1  HAVANA  Selenocysteine  53904703  53904706     .      +   \n",
       "46435         chr1  HAVANA  Selenocysteine  53904703  53904706     .      +   \n",
       "...            ...     ...             ...       ...       ...   ...    ...   \n",
       "2620238      chr22  HAVANA  Selenocysteine  19877110  19877113     .      -   \n",
       "2620286      chr22  HAVANA  Selenocysteine  19877110  19877113     .      -   \n",
       "2620306      chr22  HAVANA  Selenocysteine  19877110  19877113     .      -   \n",
       "2629658      chr22  HAVANA  Selenocysteine  31105950  31105953     .      -   \n",
       "2629683      chr22  HAVANA  Selenocysteine  31105950  31105953     .      -   \n",
       "\n",
       "        Frame             gene_id       gene_type  ...     tag  \\\n",
       "20380       .  ENSG00000162430.16  protein_coding  ...  seleno   \n",
       "20381       .  ENSG00000162430.16  protein_coding  ...  seleno   \n",
       "20413       .  ENSG00000162430.16  protein_coding  ...  seleno   \n",
       "46423       .  ENSG00000211452.10  protein_coding  ...  seleno   \n",
       "46435       .  ENSG00000211452.10  protein_coding  ...  seleno   \n",
       "...       ...                 ...             ...  ...     ...   \n",
       "2620238     .  ENSG00000184470.20  protein_coding  ...  seleno   \n",
       "2620286     .  ENSG00000184470.20  protein_coding  ...  seleno   \n",
       "2620306     .  ENSG00000184470.20  protein_coding  ...  seleno   \n",
       "2629658     .  ENSG00000198832.10  protein_coding  ...  seleno   \n",
       "2629683     .  ENSG00000198832.10  protein_coding  ...  seleno   \n",
       "\n",
       "            havana_transcript exon_number exon_id  ont         protein_id  \\\n",
       "20380    OTTHUMT00000019314.2         NaN     NaN  NaN  ENSP00000355141.2   \n",
       "20381    OTTHUMT00000019314.2         NaN     NaN  NaN  ENSP00000355141.2   \n",
       "20413    OTTHUMT00000019315.2         NaN     NaN  NaN  ENSP00000363434.1   \n",
       "46423    OTTHUMT00000388166.2         NaN     NaN  NaN  ENSP00000432797.1   \n",
       "46435    OTTHUMT00000023247.3         NaN     NaN  NaN  ENSP00000354643.3   \n",
       "...                       ...         ...     ...  ...                ...   \n",
       "2620238  OTTHUMT00000314903.4         NaN     NaN  NaN  ENSP00000383365.1   \n",
       "2620286  OTTHUMT00000314927.3         NaN     NaN  NaN  ENSP00000485499.2   \n",
       "2620306  OTTHUMT00000314928.3         NaN     NaN  NaN  ENSP00000485466.2   \n",
       "2629658  OTTHUMT00000321789.1         NaN     NaN  NaN  ENSP00000384564.1   \n",
       "2629683  OTTHUMT00000321788.1         NaN     NaN  NaN  ENSP00000383155.2   \n",
       "\n",
       "              ccdsid gene_biotype exon_assignment       tid_stable  \n",
       "20380    CCDS41282.1          NaN             NaN  ENST00000361547  \n",
       "20381    CCDS41282.1          NaN             NaN  ENST00000361547  \n",
       "20413    CCDS41283.1          NaN             NaN  ENST00000374315  \n",
       "46423            NaN          NaN             NaN  ENST00000529589  \n",
       "46435    CCDS41339.1          NaN             NaN  ENST00000361921  \n",
       "...              ...          ...             ...              ...  \n",
       "2620238  CCDS42981.1          NaN             NaN  ENST00000400521  \n",
       "2620286          NaN          NaN             NaN  ENST00000485358  \n",
       "2620306          NaN          NaN             NaN  ENST00000462843  \n",
       "2629658  CCDS43003.1          NaN             NaN  ENST00000402395  \n",
       "2629683  CCDS43003.1          NaN             NaN  ENST00000400299  \n",
       "\n",
       "[120 rows x 27 columns]"
      ]
     },
     "execution_count": 88,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "v29.loc[v29.Feature == 'Selenocysteine']"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "99d1af69-7efe-42a0-b8a3-b92341ca69d3",
   "metadata": {},
   "source": [
    "## How many more reads would be known if we were using v39?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "e7eed1a5-4ced-4b49-bb96-487a434daf10",
   "metadata": {},
   "outputs": [],
   "source": [
    "import sqlite3\n",
    "import pandas as pd\n",
    "import pyranges as pr"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "af72844e-f0c1-4848-ab06-4ffc65904268",
   "metadata": {},
   "outputs": [],
   "source": [
    "db = '../human.db'\n",
    "annot = '../human_talon_read_annot.tsv'\n",
    "genome_build = 'hg38'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "b03c21f5-ea95-40bb-964d-23e325208515",
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_ic_from_talon_db(db, annot, genome_build):\n",
    "    conn = sqlite3.connect(db)\n",
    "\n",
    "    # get edge annotations\n",
    "    query = \"\"\"\n",
    "            SELECT \n",
    "                e.edge_ID,\n",
    "                loc1.chromosome,\n",
    "                MIN(loc1.position,loc2.position),\n",
    "                MAX(loc1.position,loc2.position),\n",
    "                e.strand\n",
    "             FROM edge e\n",
    "             LEFT JOIN location loc1 ON e.v1 = loc1.location_ID\n",
    "             LEFT JOIN location loc2 ON e.v2 = loc2.location_ID\n",
    "             WHERE loc1.genome_build = '\"\"\" + genome_build + \"\"\"' AND\n",
    "             loc2.genome_build = '\"\"\" + genome_build + \\\n",
    "             \"\"\"' AND e.edge_type = 'exon';\"\"\"\n",
    "\n",
    "    e_df = pd.read_sql_query(query, conn)\n",
    "    e_df.columns = ['eid', 'Chromosome', 'Start', 'End', 'Strand']\n",
    "    e_df.eid = e_df.eid.astype(int)\n",
    "\n",
    "    # get transcript annotations \n",
    "    query = \"\"\"\n",
    "            SELECT \n",
    "                   start_exon,\n",
    "                   end_exon,\n",
    "                   jn_path,\n",
    "                   transcript_ID,\n",
    "                   gene_ID\n",
    "            FROM transcripts\n",
    "            \"\"\"\n",
    "    t_df = pd.read_sql_query(query, conn)\n",
    "    t_df['temp'] = t_df.jn_path.str.split(',')\n",
    "    t_df.drop('jn_path', axis=1, inplace=True)\n",
    "    t_df = t_df.explode(column='temp')\n",
    "    t_df = t_df.melt(id_vars=['transcript_ID', 'gene_ID'],\n",
    "              value_vars=['temp', 'start_exon', 'end_exon'])\n",
    "    t_df.drop('variable', axis=1, inplace=True)\n",
    "    t_df.rename({'value': 'temp'}, axis=1, inplace=True)\n",
    "    t_df.drop_duplicates(inplace=True)\n",
    "    t_df.rename({'transcript_ID': 'transcript_id',\n",
    "                 'gene_ID': 'gene_id'},\n",
    "                axis=1, inplace=True)\n",
    "\n",
    "    # drop things w/o intron chains (ie monoexonic transcripts)\n",
    "    t_df = t_df.loc[~t_df.temp.isnull()]\n",
    "\n",
    "    t_df.temp = t_df.temp.astype(int)\n",
    "\n",
    "    # add information about each coordinate\n",
    "    t_df = t_df.merge(e_df, how='left', left_on='temp', right_on='eid')\n",
    "\n",
    "    # drop all introns\n",
    "    t_df = t_df.loc[~t_df.eid.isnull()]\n",
    "\n",
    "    # format as pyranges\n",
    "    t_df.drop(['temp', 'eid'], axis=1, inplace=True)\n",
    "    t_df['Feature'] = 'exon'\n",
    "    t_df.Start = t_df.Start - 1\n",
    "    t_df = pr.PyRanges(t_df)\n",
    "\n",
    "    df = get_ic(t_df)\n",
    "    \n",
    "    # get novelty type of each transcript\n",
    "    query = \"\"\"\n",
    "            SELECT \n",
    "                ID,\n",
    "                value\n",
    "            FROM transcript_annotations\n",
    "            WHERE transcript_annotations.attribute = 'transcript_status'\n",
    "                AND transcript_annotations.value = 'KNOWN'\n",
    "            \"\"\"\n",
    "    known_df = pd.read_sql_query(query, conn)\n",
    "    \n",
    "    known_df.rename({'ID': 'transcript_id'}, axis=1, inplace=True)\n",
    "    known_df.drop('value', axis=1, inplace=True)\n",
    "    known_df['v29'] = True\n",
    "    \n",
    "    df = df.merge(known_df, how='left', on='transcript_id')    \n",
    "    \n",
    "    # everything that wasn't known gets \"false\"\n",
    "    df.v29.fillna(value=False, inplace=True)\n",
    "\n",
    "    return df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "da9bf392-a469-4a98-b2ae-44ed921d4424",
   "metadata": {},
   "outputs": [],
   "source": [
    "df = get_ic_from_talon_db(db, annot, genome_build)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "834d98d8-7994-49b7-86b9-6f168e943a2d",
   "metadata": {},
   "outputs": [],
   "source": [
    "df.to_csv('human_talon_ics.tsv', sep='\\t', index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "a0503ea0-52bb-4c14-b230-f08d02eed61a",
   "metadata": {},
   "outputs": [],
   "source": [
    "df = pd.read_csv('human_talon_ics.tsv', sep='\\t')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "60c01c55-c730-4732-aee0-4709b39d661f",
   "metadata": {},
   "outputs": [],
   "source": [
    "# restrict to only known transcripts\n",
    "df = df.loc[df.v29 == False]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "fccce3c7-05fe-4604-9179-6fa551b3392a",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['1059782-1060276-1060393-1061019-1061117-1063078-1063201-1065829']"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.loc[df.transcript_id == 222627, 'ic'].tolist()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "7a557d47-b7d3-4dae-b011-a79254a18277",
   "metadata": {},
   "outputs": [],
   "source": [
    "# get intron chains from v39 transcripts\n",
    "fname = '/Users/fairliereese/mortazavi_lab/ref/gencode.v39/gencode.v39.annotation.gtf'\n",
    "genc = pr.read_gtf(fname)\n",
    "genc_df = get_ic(genc)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "82cd3cd1-496d-42e6-8c3e-c59043491fbf",
   "metadata": {},
   "outputs": [],
   "source": [
    "genc_df.head()\n",
    "genc_df.to_csv('gencode_v39_ics.tsv', sep='\\t', index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "65fead6b-9a39-45c3-81c7-eedff79f7179",
   "metadata": {},
   "outputs": [],
   "source": [
    "df = df.merge(genc_df, how='left', on=['Chromosome', 'Strand', 'ic'], suffixes=('_talon', '_genc'))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "acf7f6d2-5e5a-480e-a943-1698776bc4a3",
   "metadata": {},
   "outputs": [],
   "source": [
    "print('from talon gtf')\n",
    "print(temp.loc[temp.transcript_id_talon == 'ENCODEHT000222627', ['Chromosome', 'Strand']])\n",
    "temp.loc[temp.transcript_id_talon == 'ENCODEHT000222627', 'ic'].tolist()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "09b0c1c4-2efb-4d24-9d70-c0ca12e0144b",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "from gencode v39\n",
      "     Chromosome Strand\n",
      "2141       chr1      +\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "['1059782-1060276-1060393-1061019-1061117-1063078-1063201-1065829']"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "print('from gencode v39')\n",
    "print(genc_df.loc[genc_df.transcript_id == 'ENST00000394517.7', ['Chromosome', 'Strand']])\n",
    "genc_df.loc[genc_df.transcript_id == 'ENST00000394517.7', 'ic'].tolist()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "0fbd1efb-5ca3-4c5f-bafa-4041f30d295e",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "from talon db\n",
      "     Chromosome Strand\n",
      "1287       chr1      +\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "['1059782-1060276-1060393-1061019-1061117-1063078-1063201-1065829']"
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "print('from talon db')\n",
    "print(df.loc[df.transcript_id_talon == 222627, ['Chromosome', 'Strand']])\n",
    "df.loc[df.transcript_id_talon == 222627, 'ic'].tolist()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "15d914c3-e025-4f3c-8814-3a3501e81710",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Strand</th>\n",
       "      <th>transcript_id_talon</th>\n",
       "      <th>gene_id_talon</th>\n",
       "      <th>Coord_talon</th>\n",
       "      <th>temp_talon</th>\n",
       "      <th>ic</th>\n",
       "      <th>v29</th>\n",
       "      <th>transcript_id_genc</th>\n",
       "      <th>gene_id_genc</th>\n",
       "      <th>Coord_genc</th>\n",
       "      <th>temp_genc</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ERCC-00002</td>\n",
       "      <td>+</td>\n",
       "      <td>206802</td>\n",
       "      <td>58798</td>\n",
       "      <td>0-1035</td>\n",
       "      <td>1035</td>\n",
       "      <td>1035</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ERCC-00002</td>\n",
       "      <td>+</td>\n",
       "      <td>206838</td>\n",
       "      <td>58798</td>\n",
       "      <td>0-61-141-1035</td>\n",
       "      <td>61-141-1035</td>\n",
       "      <td>61-141</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ERCC-00002</td>\n",
       "      <td>+</td>\n",
       "      <td>206839</td>\n",
       "      <td>58798</td>\n",
       "      <td>0-344-399-1035</td>\n",
       "      <td>344-399-1035</td>\n",
       "      <td>344-399</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ERCC-00002</td>\n",
       "      <td>+</td>\n",
       "      <td>1149583</td>\n",
       "      <td>58798</td>\n",
       "      <td>0-887-942-1035</td>\n",
       "      <td>887-942-1035</td>\n",
       "      <td>887-942</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ERCC-00002</td>\n",
       "      <td>+</td>\n",
       "      <td>1149585</td>\n",
       "      <td>58798</td>\n",
       "      <td>0-195-510-1035</td>\n",
       "      <td>195-510-1035</td>\n",
       "      <td>195-510</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   Chromosome Strand  transcript_id_talon  gene_id_talon     Coord_talon  \\\n",
       "0  ERCC-00002      +               206802          58798          0-1035   \n",
       "1  ERCC-00002      +               206838          58798   0-61-141-1035   \n",
       "2  ERCC-00002      +               206839          58798  0-344-399-1035   \n",
       "3  ERCC-00002      +              1149583          58798  0-887-942-1035   \n",
       "4  ERCC-00002      +              1149585          58798  0-195-510-1035   \n",
       "\n",
       "     temp_talon       ic    v29 transcript_id_genc gene_id_genc Coord_genc  \\\n",
       "0          1035     1035  False                NaN          NaN        NaN   \n",
       "1   61-141-1035   61-141  False                NaN          NaN        NaN   \n",
       "2  344-399-1035  344-399  False                NaN          NaN        NaN   \n",
       "3  887-942-1035  887-942  False                NaN          NaN        NaN   \n",
       "4  195-510-1035  195-510  False                NaN          NaN        NaN   \n",
       "\n",
       "  temp_genc  \n",
       "0       NaN  \n",
       "1       NaN  \n",
       "2       NaN  \n",
       "3       NaN  \n",
       "4       NaN  "
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "d60755fa-5326-4848-acfa-85026ebbdb40",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Strand</th>\n",
       "      <th>transcript_id_talon</th>\n",
       "      <th>gene_id_talon</th>\n",
       "      <th>Coord_talon</th>\n",
       "      <th>temp_talon</th>\n",
       "      <th>ic</th>\n",
       "      <th>v29</th>\n",
       "      <th>transcript_id_genc</th>\n",
       "      <th>gene_id_genc</th>\n",
       "      <th>Coord_genc</th>\n",
       "      <th>temp_genc</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1287</th>\n",
       "      <td>chr1</td>\n",
       "      <td>+</td>\n",
       "      <td>222627</td>\n",
       "      <td>72</td>\n",
       "      <td>1059733-1059782-1060276-1060393-1061019-106111...</td>\n",
       "      <td>1059782-1060276-1060393-1061019-1061117-106307...</td>\n",
       "      <td>1059782-1060276-1060393-1061019-1061117-106307...</td>\n",
       "      <td>False</td>\n",
       "      <td>ENST00000394517.7</td>\n",
       "      <td>ENSG00000217801.11</td>\n",
       "      <td>1059707-1059782-1060276-1060393-1061019-106111...</td>\n",
       "      <td>1059782-1060276-1060393-1061019-1061117-106307...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     Chromosome Strand  transcript_id_talon  gene_id_talon  \\\n",
       "1287       chr1      +               222627             72   \n",
       "\n",
       "                                            Coord_talon  \\\n",
       "1287  1059733-1059782-1060276-1060393-1061019-106111...   \n",
       "\n",
       "                                             temp_talon  \\\n",
       "1287  1059782-1060276-1060393-1061019-1061117-106307...   \n",
       "\n",
       "                                                     ic    v29  \\\n",
       "1287  1059782-1060276-1060393-1061019-1061117-106307...  False   \n",
       "\n",
       "     transcript_id_genc        gene_id_genc  \\\n",
       "1287  ENST00000394517.7  ENSG00000217801.11   \n",
       "\n",
       "                                             Coord_genc  \\\n",
       "1287  1059707-1059782-1060276-1060393-1061019-106111...   \n",
       "\n",
       "                                              temp_genc  \n",
       "1287  1059782-1060276-1060393-1061019-1061117-106307...  "
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.loc[df.transcript_id_talon == 222627]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "280164f5-ab64-4068-897e-f7e1d14b6e1c",
   "metadata": {},
   "outputs": [],
   "source": [
    "df['v39'] = False\n",
    "df.loc[~df.transcript_id_genc.isnull(), 'v39'] = True"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "c243cded-179c-4f3a-ac75-4dd6c98501f3",
   "metadata": {},
   "outputs": [],
   "source": [
    "df.loc[df.v39 == True]\n",
    "v39_tids = df.loc[df.v39 == True, 'transcript_id_talon'].tolist()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "df872a4d-0190-461c-96e9-e46e7c642b91",
   "metadata": {},
   "outputs": [],
   "source": [
    "# how many reads do these transcripts correspond to in the read_annot file?\n",
    "read_df = pd.read_csv('../human_talon_read_annot.tsv', sep='\\t', usecols=[10])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "id": "60867d6a-17d8-408e-b7dc-f9e11316294b",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "181558331\n"
     ]
    }
   ],
   "source": [
    "total_reads = len(read_df.index)\n",
    "print(total_reads)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "d3582536-5658-4734-a8a7-40dbabf889b5",
   "metadata": {},
   "outputs": [],
   "source": [
    "nov_v39_reads = len(read_df.loc[read_df.transcript_ID.isin(v39_tids)].index)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "904f4fe3-8737-408e-a5ff-9ef51d45dc6b",
   "metadata": {},
   "outputs": [],
   "source": [
    "known_reads = len(read_df.index)-nov_v39_reads"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "b4231349-29a7-487d-8131-78ac73d5db12",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2058984\n"
     ]
    }
   ],
   "source": [
    "print(nov_v39_reads)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "id": "b99749fc-b027-41fd-b464-0862db301a87",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "179499347\n"
     ]
    }
   ],
   "source": [
    "print(known_reads)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "id": "08483bff-6077-408d-907b-a9f231aae608",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "1.134061978130874\n"
     ]
    }
   ],
   "source": [
    "print((nov_v39_reads/total_reads)*100)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "id": "f48e9961-d8a2-48c8-8bf2-218f9388ec6e",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "16324"
      ]
     },
     "execution_count": 26,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(v39_tids)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "id": "92a0cd7b-1136-4450-a2d7-0c9d3a5d7013",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[208393, 208410, 208584, 208597, 208599]"
      ]
     },
     "execution_count": 27,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "v39_tids[:5]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "id": "89c31c0e-bfba-42e1-beb5-3951b0873848",
   "metadata": {},
   "outputs": [],
   "source": [
    "# and then how many of these table transcript IDs are in v29 but these reads were\n",
    "# not assigned to those models (presumably because of differences at the ends)\n",
    "temp = df.loc[~df.transcript_id_genc.isnull()]\n",
    "supp_tids = temp.transcript_id_genc.str.split('.', n=1, expand=True)[0].tolist()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "id": "418c8265-4fad-4e98-8a69-63f8ebf84435",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "16324"
      ]
     },
     "execution_count": 31,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(supp_tids)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "id": "5337492f-e1a3-4f26-861a-f1d36e301542",
   "metadata": {},
   "outputs": [],
   "source": [
    "# were these transcripts designated at novel because these transcripts just didn't exist in prior versions of the annotation?\n",
    "fname = '/Users/fairliereese/mortazavi_lab/data/rnawg/refs/gencode_v29_sirv4_ercc.gtf'\n",
    "v29 = pr.read_gtf(fname)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "id": "b6dc9e30-0aff-4ce8-859c-44496bf94cf7",
   "metadata": {},
   "outputs": [],
   "source": [
    "fname = '/Users/fairliereese/mortazavi_lab/ref/gencode.v39/gencode.v39.annotation.gtf'\n",
    "v39 = pr.read_gtf(fname)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "id": "3e1e5f43-7b8e-423b-8055-e7689000b0ed",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2318 transcripts whose intron chains are supported in v39 also are annotated in v29\n"
     ]
    }
   ],
   "source": [
    "v29 = v29.df\n",
    "v29['tid_stable'] = v29.transcript_id.str.split('.', n=1, expand=True)[0]\n",
    "\n",
    "temp = v29.loc[v29.tid_stable.isin(supp_tids)]\n",
    "temp = temp.loc[temp.Feature == 'transcript']\n",
    "n = len(temp.index)\n",
    "print('{} transcripts whose intron chains are supported in v39 also are annotated in v29'.format(n))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "aed557b8-56a8-4ed1-a73e-a73479b7286d",
   "metadata": {},
   "source": [
    "## Does pyranges lengthen intervals when it reads GTFs in? \n",
    "Yes [gh issue](https://github.com/biocore-ntnu/pyranges/pull/260)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "5d83d8cc-fae1-441f-996e-46643f2d2ff1",
   "metadata": {},
   "outputs": [],
   "source": [
    "beep = pr.read_gtf('confusing_transcript_talon.gtf')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "a32c14f1-34cf-4952-9c34-1419975b91d1",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Source</th>\n",
       "      <th>Feature</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Score</th>\n",
       "      <th>Strand</th>\n",
       "      <th>Frame</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>...</th>\n",
       "      <th>transcript_name</th>\n",
       "      <th>talon_transcript</th>\n",
       "      <th>NNC_transcript</th>\n",
       "      <th>exon_number</th>\n",
       "      <th>exon_id</th>\n",
       "      <th>talon_exon</th>\n",
       "      <th>exon_status</th>\n",
       "      <th>source</th>\n",
       "      <th>ont</th>\n",
       "      <th>tag</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>gene</td>\n",
       "      <td>1059733</td>\n",
       "      <td>1066441</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000217801.9</td>\n",
       "      <td>AL390719.1</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>TALON</td>\n",
       "      <td>transcript</td>\n",
       "      <td>1059733</td>\n",
       "      <td>1066441</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000217801.9</td>\n",
       "      <td>AL390719.1</td>\n",
       "      <td>...</td>\n",
       "      <td>ENCODEHT000222627</td>\n",
       "      <td>222627</td>\n",
       "      <td>TRUE</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>TALON</td>\n",
       "      <td>exon</td>\n",
       "      <td>1059733</td>\n",
       "      <td>1059782</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000217801.9</td>\n",
       "      <td>AL390719.1</td>\n",
       "      <td>...</td>\n",
       "      <td>ENCODEHT000222627</td>\n",
       "      <td>222627</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1</td>\n",
       "      <td>971902</td>\n",
       "      <td>971902</td>\n",
       "      <td>NOVEL</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>TALON</td>\n",
       "      <td>exon</td>\n",
       "      <td>1060276</td>\n",
       "      <td>1060393</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000217801.9</td>\n",
       "      <td>AL390719.1</td>\n",
       "      <td>...</td>\n",
       "      <td>ENCODEHT000222627</td>\n",
       "      <td>222627</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2</td>\n",
       "      <td>971874</td>\n",
       "      <td>971874</td>\n",
       "      <td>NOVEL</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>1061019</td>\n",
       "      <td>1061117</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000217801.9</td>\n",
       "      <td>AL390719.1</td>\n",
       "      <td>...</td>\n",
       "      <td>ENCODEHT000222627</td>\n",
       "      <td>222627</td>\n",
       "      <td>NaN</td>\n",
       "      <td>3</td>\n",
       "      <td>ENSE00001630784.1</td>\n",
       "      <td>936</td>\n",
       "      <td>KNOWN</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>1063078</td>\n",
       "      <td>1063201</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000217801.9</td>\n",
       "      <td>AL390719.1</td>\n",
       "      <td>...</td>\n",
       "      <td>ENCODEHT000222627</td>\n",
       "      <td>222627</td>\n",
       "      <td>NaN</td>\n",
       "      <td>4</td>\n",
       "      <td>ENSE00001613074.1</td>\n",
       "      <td>933</td>\n",
       "      <td>KNOWN</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>PGO:0000019</td>\n",
       "      <td>basic</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>chr1</td>\n",
       "      <td>TALON</td>\n",
       "      <td>exon</td>\n",
       "      <td>1065829</td>\n",
       "      <td>1066441</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000217801.9</td>\n",
       "      <td>AL390719.1</td>\n",
       "      <td>...</td>\n",
       "      <td>ENCODEHT000222627</td>\n",
       "      <td>222627</td>\n",
       "      <td>NaN</td>\n",
       "      <td>5</td>\n",
       "      <td>971887</td>\n",
       "      <td>971887</td>\n",
       "      <td>NOVEL</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>7 rows × 27 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "+--------------+------------+--------------+-----------+-----------+-------+\n",
       "| Chromosome   | Source     | Feature      |     Start |       End | +22   |\n",
       "| (category)   | (object)   | (category)   |   (int32) |   (int32) | ...   |\n",
       "|--------------+------------+--------------+-----------+-----------+-------|\n",
       "| chr1         | HAVANA     | gene         |   1059733 |   1066441 | ...   |\n",
       "| chr1         | TALON      | transcript   |   1059733 |   1066441 | ...   |\n",
       "| chr1         | TALON      | exon         |   1059733 |   1059782 | ...   |\n",
       "| chr1         | TALON      | exon         |   1060276 |   1060393 | ...   |\n",
       "| chr1         | HAVANA     | exon         |   1061019 |   1061117 | ...   |\n",
       "| chr1         | HAVANA     | exon         |   1063078 |   1063201 | ...   |\n",
       "| chr1         | TALON      | exon         |   1065829 |   1066441 | ...   |\n",
       "+--------------+------------+--------------+-----------+-----------+-------+\n",
       "Stranded PyRanges object has 7 rows and 27 columns from 1 chromosomes.\n",
       "For printing, the PyRanges was sorted on Chromosome and Strand.\n",
       "22 hidden columns: Score, Strand, Frame, gene_id, gene_name, gene_status, gene_type, ... (+ 15 more.)"
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "beep"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "44f35787-66e2-41f1-b678-22b7b0471440",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.4"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
