{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "c32584f6-ad73-4934-b944-7234c14b798c",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "import scipy.stats as st\n",
    "import seaborn as sns\n",
    "import sys\n",
    "import os\n",
    "import gseapy as gp\n",
    "import matplotlib.pyplot as plt\n",
    "import swan_vis as swan\n",
    "from pandarallel import pandarallel\n",
    "\n",
    "p = os.path.dirname(os.path.dirname(os.path.dirname(os.getcwd())))\n",
    "sys.path.append(p)\n",
    "\n",
    "from scripts.utils import *\n",
    "from scripts.plotting import *"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "4f76672a-1759-4bdc-95d8-71e99b12a3b3",
   "metadata": {},
   "outputs": [],
   "source": [
    "ab = '../../../lr_bulk/talon/human_talon_abundance.tsv'\n",
    "cerberus_h5 = '../../../lr_bulk/cerberus/triplets/cerberus_annot_triplets.h5'\n",
    "swan_file = '../../..//lr_bulk/cerberus/swan/swan_meta.p'\n",
    "\n",
    "ver = 'v40_cerberus'\n",
    "\n",
    "min_tpm = 1\n",
    "gene_subset = 'protein_coding'\n",
    "obs_col = 'dataset'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "df2076f0-142c-4322-bc9b-1d5b41002646",
   "metadata": {},
   "outputs": [],
   "source": [
    "ca = cerberus.read(cerberus_h5)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "d1c8a6b3-8714-44d5-9822-84bfed99723a",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Read in graph from ../../..//lr_bulk/cerberus/swan/swan_meta.p\n"
     ]
    }
   ],
   "source": [
    "sg = swan.read(swan_file)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "11d8c6c9-70d0-4fa9-a5ca-15775881b461",
   "metadata": {},
   "outputs": [],
   "source": [
    "# add library-level expressed triplets\n",
    "df = ca.get_expressed_triplets(sg,\n",
    "                               obs_col=obs_col,\n",
    "                               min_tpm=min_tpm,\n",
    "                               source='library_det')\n",
    "ca.add_triplets(df)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "437b216e-80ac-47f0-8381-f8f92eb32677",
   "metadata": {},
   "outputs": [],
   "source": [
    "# remove non-polya genes\n",
    "df, _, _ = get_gtf_info(how='gene',\n",
    "                        ver=ver,\n",
    "                        subset=gene_subset)\n",
    "df['gid_stable'] = cerberus.get_stable_gid(df, 'gid')\n",
    "polya_gids = df.gid_stable.tolist()\n",
    "ca.triplets = ca.triplets.loc[ca.triplets.gid.isin(polya_gids)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "9c453e3b-414c-43a7-8290-4532ac2ebf4b",
   "metadata": {},
   "outputs": [],
   "source": [
    "ca.write('cerberus_ad.h5')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b8853a8c-8700-40de-a62e-f76e2ef810db",
   "metadata": {},
   "source": [
    "## Analysis"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "f62f3003-0625-43e3-afa7-185c7558fd5a",
   "metadata": {},
   "outputs": [
    {
     "ename": "NameError",
     "evalue": "name 'swan_file' is not defined",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mNameError\u001b[0m                                 Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-6-cae230154c9b>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[1;32m      1\u001b[0m \u001b[0mca\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mcerberus\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mread\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m'cerberus_ad.h5'\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m----> 2\u001b[0;31m \u001b[0msg\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mswan\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mread\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mswan_file\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m",
      "\u001b[0;31mNameError\u001b[0m: name 'swan_file' is not defined"
     ]
    }
   ],
   "source": [
    "ca = cerberus.read('cerberus_ad.h5')\n",
    "sg = swan.read(swan_file)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b7e3dc14-0d90-40bb-a228-71a03fd78b04",
   "metadata": {},
   "outputs": [],
   "source": [
    "# limit triplets to just the brain ones\n",
    "ca.triplets = ca.triplets.loc[~(ca.triplets.dataset.isnull())&\\\n",
    "                              (ca.triplets.dataset.str.contains('brodmann'))&\\\n",
    "                              (ca.triplets.source=='library_det')]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1eff1c73-382e-40f4-86f6-95177b8f97e6",
   "metadata": {},
   "outputs": [],
   "source": [
    "# merge in disease status\n",
    "sg.adata.obs.health_status.unique()\n",
    "sg.adata.obs.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "647d4e82-a9e1-4746-9647-01d7bf4ec56d",
   "metadata": {},
   "outputs": [],
   "source": [
    "ca.triplets = ca.triplets.merge(sg.adata.obs[['health_status']],\n",
    "                                how='left',\n",
    "                                left_on='dataset',\n",
    "                                right_index=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "24a67b48-8a5e-43c5-a510-dda75af64481",
   "metadata": {},
   "outputs": [],
   "source": [
    "ca.triplets.health_status = ca.triplets.health_status.astype(str)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "48fe8532-89c0-4502-bb41-2a117a0aa57b",
   "metadata": {},
   "outputs": [],
   "source": [
    "# get centroids for each health status\n",
    "inds = ca.triplets.loc[ca.triplets.health_status == 'AD'].index\n",
    "ca.triplets.loc[inds, 'source'] = 'library_ad_det'\n",
    "\n",
    "inds = ca.triplets.loc[ca.triplets.health_status == 'healthy'].index\n",
    "ca.triplets.loc[inds, 'source'] = 'library_healthy_det'\n",
    "\n",
    "ca = get_centroids(ca,\n",
    "                   source='library_ad_det',\n",
    "                   gene_subset=gene_subset,\n",
    "                   ver=ver)\n",
    "ca = get_centroids(ca,\n",
    "                   source='library_healthy_det',\n",
    "                   gene_subset=gene_subset,\n",
    "                   ver=ver)\n",
    "\n",
    "\n",
    "df = compute_dists([ca,ca],\n",
    "                   ['library_ad_det_centroid',\n",
    "                    'library_healthy_det_centroid'],\n",
    "                   rm_1_isos=[False,False],\n",
    "                   ver=[ver,ver])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7a163714-48b3-4d4d-9132-6a17b272d86d",
   "metadata": {},
   "outputs": [],
   "source": [
    "temp = df.loc[df.z_score>2, ['gname', 'z_score']].drop_duplicates()\n",
    "temp.to_csv('ad_vs_healthy_centroid_z_score_2_genes.txt',index=False)\n",
    "len(temp.index)\n",
    "temp.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "84e2e898-0fbe-4a43-99ab-6eb879f749b4",
   "metadata": {},
   "outputs": [],
   "source": [
    "sns.displot(df, x='dist', kind='kde')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1eab9b21-130f-40e9-bd23-6ad2452a2127",
   "metadata": {},
   "outputs": [],
   "source": [
    "sns.displot(df, x='z_score', kind='kde')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b6fdab03-7262-4b75-9409-0db7d3484640",
   "metadata": {},
   "outputs": [],
   "source": [
    "# which genes are the farthest apart\n",
    "df.sort_values(by='dist', ascending=False)[['gname', 'dist', 'z_score']].head(20)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1724e43f-084b-417f-8222-da684fc9dcca",
   "metadata": {},
   "outputs": [],
   "source": [
    "c_dict, order = get_ad_colors()\n",
    "mmap = {'library_ad_det_centroid': 'x',\n",
    "        'library_healthy_det_centroid': 'x'}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "72da20d8-4a11-4c02-a1bc-548831bfce60",
   "metadata": {},
   "outputs": [],
   "source": [
    "df = ca.plot_simplex(top='splicing_ratio', \n",
    "            gene='SPIDR',\n",
    "            hue='health_status',\n",
    "            cmap=c_dict,\n",
    "            size='gene_tpm',\n",
    "            mmap=mmap,\n",
    "            marker_style='source',\n",
    "            log_size=True,\n",
    "            sectors=True,\n",
    "            legend=False,\n",
    "            jitter=True,\n",
    "            size_scale=1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a965ac38-298f-4560-b274-f0e3805ae929",
   "metadata": {},
   "outputs": [],
   "source": [
    "df = ca.plot_simplex(top='splicing_ratio', \n",
    "            gene='FLII',\n",
    "            hue='health_status',\n",
    "            cmap=c_dict,\n",
    "            size='gene_tpm',\n",
    "            mmap=mmap,\n",
    "            marker_style='source',\n",
    "            log_size=True,\n",
    "            sectors=True,\n",
    "            legend=False,\n",
    "            jitter=True,\n",
    "            size_scale=1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5f44036c-6fb1-4b5b-bb06-c73dce1fb985",
   "metadata": {},
   "outputs": [],
   "source": [
    "df = ca.plot_simplex(top='splicing_ratio', \n",
    "            gene='ACY1',\n",
    "            hue='health_status',\n",
    "            cmap=c_dict,\n",
    "            size='gene_tpm',\n",
    "            mmap=mmap,\n",
    "            marker_style='source',\n",
    "            log_size=True,\n",
    "            sectors=True,\n",
    "            legend=False,\n",
    "            jitter=True,\n",
    "            size_scale=1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8d4547ea-7aa0-4c4c-84eb-b4d265c1882f",
   "metadata": {},
   "outputs": [],
   "source": [
    "df = ca.plot_simplex(top='splicing_ratio', \n",
    "            gene='CDCA7L',\n",
    "            hue='health_status',\n",
    "            cmap=c_dict,\n",
    "            size='gene_tpm',\n",
    "            mmap=mmap,\n",
    "            marker_style='source',\n",
    "            log_size=True,\n",
    "            sectors=True,\n",
    "            legend=False,\n",
    "            jitter=True,\n",
    "            size_scale=1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "08e541c8-0c50-4e77-a506-21b8bf00dc74",
   "metadata": {},
   "outputs": [],
   "source": [
    "c_dict, order = get_ad_colors()\n",
    "df = ca.plot_simplex(top='splicing_ratio', \n",
    "            gene='MAPT',\n",
    "            hue='health_status',\n",
    "            cmap=c_dict,\n",
    "            size='gene_tpm', \n",
    "            log_size=True,\n",
    "            sectors=True,\n",
    "            legend=False,\n",
    "            jitter=True,\n",
    "            size_scale=1)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8f29e447-de08-48d3-89a7-15e4a2ffadc0",
   "metadata": {},
   "source": [
    "## What about predominant isoform switches? "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "e135a43b-6b6c-4900-8aed-01b7ac8a5179",
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_exp_gene_subset(ab, min_tpm, obs_col):\n",
    "    df = pd.read_csv(ab, sep='\\t')\n",
    "    if obs_col == 'dataset':\n",
    "        gb = 'library'\n",
    "    else:\n",
    "        gb = obs_col\n",
    "    df = get_det_table(df,\n",
    "                  groupby=gb,\n",
    "                  how='gene',\n",
    "                  min_tpm=min_tpm,\n",
    "                  gene_subset='polya')\n",
    "\n",
    "    df = df.melt(ignore_index=False, var_name='gid', value_name='detected').reset_index()\n",
    "    df.rename({'library': 'dataset', 'biosample':'sample'}, axis=1, inplace=True)\n",
    "    \n",
    "    # limit to detected genes\n",
    "    df = df.loc[df.detected==True]\n",
    "    df.drop('detected', axis=1, inplace=True)\n",
    "    \n",
    "    fname = 'gids_{}_1_tpm.tsv'.format(obs_col)\n",
    "    df.to_csv(fname, sep='\\t', index=False)\n",
    "    \n",
    "def get_exp_mane_genes(obs_col):\n",
    "    \"\"\"\n",
    "    Get list of genes that are expressed and have MANE isoforms\n",
    "    \"\"\"\n",
    "    # get list of gene inds that are >=1 tpm and \n",
    "    # had a mane iso\n",
    "    fname = 'gids_{}_1_tpm.tsv'.format(obs_col)\n",
    "    df = pd.read_csv(fname, sep='\\t')\n",
    "    \n",
    "    # genes w/ mane isoforms\n",
    "    mane = pd.read_csv(mane_file, sep='\\t')\n",
    "    mane = mane.loc[mane.MANE_Select == True]\n",
    "    mane['gid'] = cerberus.get_stable_gid(mane, col='gid')\n",
    "    mane = mane.gid.tolist()    \n",
    "    df = df.loc[df.gid.isin(mane)]\n",
    "    \n",
    "\n",
    "    # inds = list(set(inds)&(set(mane)))\n",
    "    # print(len(inds))\n",
    "    \n",
    "    return df\n",
    "\n",
    "def compute_feat_tpm(adata, obs_col, feat, how, id_col, min_tpm=None):\n",
    "    \"\"\"\n",
    "    Get TPM from SwanGraph for a given gene. Melt to long format.\n",
    "    \"\"\"\n",
    "    tpm_df = swan.calc_tpm(adata, obs_col=obs_col, how=how)\n",
    "    tpm_df = tpm_df.sparse.to_dense()\n",
    "    tpm_df = tpm_df.T\n",
    "    tpm_df = tpm_df.melt(var_name=obs_col, value_name='tpm', ignore_index=False)\n",
    "    tpm_df.reset_index(inplace=True)\n",
    "    tpm_df.rename({'index': id_col}, axis=1, inplace=True)\n",
    "    \n",
    "    # only expressed samples\n",
    "    tpm_df = tpm_df.loc[tpm_df.tpm > 0]\n",
    "    tpm_df = tpm_df.loc[~tpm_df.tpm.isnull()]\n",
    "    \n",
    "    # enforce additional expression thresh\n",
    "    if min_tpm:\n",
    "        tpm_df = tpm_df.loc[tpm_df.tpm >= min_tpm]\n",
    "        \n",
    "    tpm_df.rename({'tpm': '{}_tpm'.format(feat)}, axis=1, inplace=True)\n",
    "    return tpm_df\n",
    "\n",
    "def compute_feat_pi(adata, obs_col, t_df, feat):\n",
    "    \"\"\"\n",
    "    Get Pi from SwanGraph for a given gene. Melt to long format.\n",
    "    \n",
    "        adata: adata, tss_adata, ic_adata, tes_adata\n",
    "        t_df: t_df, tss_adata.var, ic_adata.var, tes_adata.var\n",
    "        feat: {'triplet', 'tss', 'ic', 'tes'}\n",
    "    \"\"\"\n",
    "    # get pi value for each feature in the specified gb category\n",
    "    df, _ = swan.calc_pi(adata, t_df, obs_col=obs_col)\n",
    "    df = df.sparse.to_dense()\n",
    "    df = df.transpose()\n",
    "\n",
    "    df = df.melt(var_name=obs_col, value_name='pi', ignore_index=False)\n",
    "    df.reset_index(inplace=True)\n",
    "\n",
    "    # only expressed samples\n",
    "    df = df.loc[df.pi > 0]\n",
    "    df = df.loc[~df.pi.isnull()]\n",
    "\n",
    "    df.rename({'pi': '{}_pi'.format(feat)}, axis=1, inplace=True)\n",
    "\n",
    "    return df\n",
    "\n",
    "def get_principal_feats(df, feat, obs_col, id_col):\n",
    "    \"\"\"\n",
    "    Get the most highly-expressed feature per gene / sample \n",
    "    \"\"\"\n",
    "    df = get_ranked_feats(df, feat, obs_col, id_col, rank=1)\n",
    "\n",
    "    return df\n",
    "\n",
    "def rank_feats(df, feat, obs_col, id_col):\n",
    "    \"\"\"\n",
    "    Rank features from high -> low expression and number them\n",
    "    \"\"\"\n",
    "    pi_col = '{}_tpm'.format(feat)\n",
    "    tpm_col = '{}_pi'.format(feat)\n",
    "    rank_col = '{}_rank'.format(feat)\n",
    "\n",
    "    df = df.copy(deep=True)\n",
    "    \n",
    "    # sort by gene, sample, and pi value\n",
    "    # dedupe across the gene and sample cols; take the top-expressed isoform\n",
    "    df = df.sort_values(by=['gid', obs_col, tpm_col, pi_col],\n",
    "                        ascending=[False, False, False, False])\n",
    "    df[rank_col] = df.groupby(['gid', obs_col]).cumcount()+1\n",
    "    \n",
    "    return df\n",
    "\n",
    "def get_ranked_feats(df, feat, obs_col, id_col, rank):\n",
    "    \"\"\"\n",
    "    Get entries that correspond to transcript w/ rank=rank\n",
    "    \"\"\"\n",
    "    df = rank_feats(df, feat, obs_col, id_col)\n",
    "    rank_col = '{}_rank'.format(feat)\n",
    "    df = df.loc[df[rank_col] == rank]\n",
    "    return df\n",
    "\n",
    "def get_mane_feats(ca, df, feat, obs_col, id_col, id_col_2):\n",
    "    \"\"\"\n",
    "    Get the mane feature expressed per gene / sample\n",
    "    \"\"\"\n",
    "    # get mane feature ids by looking at v40 metadata\n",
    "    # then and cross referencing with how they were mapped w/ cerberus\n",
    "    meta_df = pd.read_csv(t_orig_metadata, sep='\\t')\n",
    "    mane_tids = meta_df.loc[meta_df.MANE_Select].tid.tolist()\n",
    "    mane_df = ca.t_map.copy(deep=True)\n",
    "    mane_df = mane_df.loc[mane_df.source=='v40']\n",
    "    mane_df = mane_df.loc[mane_df.original_transcript_id.isin(mane_tids)]\n",
    "    mane_feats = mane_df[id_col_2].unique().tolist()\n",
    "    \n",
    "    df = df.copy(deep=True)\n",
    "    df = df.loc[df[id_col].isin(mane_feats)]\n",
    "    \n",
    "    return df\n",
    "\n",
    "def plot_tpms(temp, feat, how, x='mane', y='princ'):\n",
    "    \n",
    "    fig_dir = 'figures/'\n",
    "        \n",
    "    princ_col = f'{feat}_{how}_{y}'\n",
    "    mane_col = f'{feat}_{how}_{x}'\n",
    "    if how == 'tpm':\n",
    "        princ_col_2 = f'log_{feat}_{how}_{y}'\n",
    "        mane_col_2 = f'log_{feat}_{how}_{x}'\n",
    "        temp[princ_col_2] = np.log2(temp[princ_col]+1)\n",
    "        temp[mane_col_2] = np.log2(temp[mane_col]+1)\n",
    "        princ_col = princ_col_2\n",
    "        mane_col = mane_col_2\n",
    "        \n",
    "    lim_max = max(temp[princ_col].max(), temp[mane_col].max())\n",
    "    xlim = ylim = (-2, lim_max)\n",
    "\n",
    "    c_dict, order = get_feat_triplet_colors()  \n",
    "    color = c_dict[feat]\n",
    "    \n",
    "    ax = sns.jointplot(data=temp, x=mane_col, y=princ_col, color=color, \n",
    "                       xlim=xlim, ylim=ylim, kind='hex',\n",
    "                       marginal_kws={'linewidth': 0}, \n",
    "                       joint_kws={'edgecolors': 'none'})\n",
    "\n",
    "    if feat == 'triplet':\n",
    "        label = 'isoform'\n",
    "    else:\n",
    "        label = feat.upper()\n",
    "        \n",
    "    def get_hr_label(a):\n",
    "        if a == 'mane':\n",
    "            return 'MANE'\n",
    "        elif a == 'princ':\n",
    "            return 'Predominant'\n",
    "        elif a == 'sec':\n",
    "            return 'Secondary'\n",
    "        \n",
    "    x_hr = get_hr_label(x)\n",
    "    y_hr = get_hr_label(y)\n",
    "    \n",
    "    if how == 'tpm':\n",
    "        # TODO change labels based on x and y\n",
    "        ylabel = f'log2({y_hr} {label} TPM+1)'\n",
    "        xlabel = f'log2({x_hr} {label} TPM+1)'\n",
    "    elif how == 'pi':\n",
    "        ylabel = f'{y_hr} isoform PI'\n",
    "        xlabel = f'{x_hr} isoform PI'\n",
    "\n",
    "    tick_range = range(0, int(lim_max+1), 5)\n",
    "    ax.ax_joint.set_xticks(tick_range)\n",
    "    ax.ax_joint.set_yticks(tick_range)\n",
    "    ax.ax_joint.set(ylabel=ylabel, xlabel=xlabel)\n",
    "    \n",
    "    \n",
    "    ax = ax.ax_joint\n",
    "    if y == 'princ' and x == 'mane':\n",
    "        \n",
    "        # add sector lines\n",
    "        xlims = ax.get_xlim()\n",
    "\n",
    "        # line @ 30 tpm\n",
    "        y_coord = math.log2(30+1)\n",
    "        color = '#5c5c5c'\n",
    "        ax.hlines(y_coord, xlims[0], xlims[1],\n",
    "                      colors=color, linestyles='dashed',\n",
    "                      linewidth=2)\n",
    "    \n",
    "    fname = f'{fig_dir}/{x}_vs_{y}_{feat}_{how}.pdf'\n",
    "    plt.savefig(fname, dpi=800, bbox_inches='tight')\n",
    "\n",
    "def get_loopers(sg):\n",
    "    \"\"\"\n",
    "    Get things to loop through when processing one feature at a time\n",
    "    \"\"\"\n",
    "    feats = ['triplet', 'tss', 'ic', 'tes']\n",
    "    adatas = [sg.adata, sg.tss_adata, sg.ic_adata, sg.tes_adata]\n",
    "    t_dfs = [sg.t_df, sg.tss_adata.var, sg.ic_adata.var, sg.tes_adata.var]\n",
    "    id_cols = {'triplet': 'tid', 'tss': 'tss_id', 'ic': 'ic_id', 'tes': 'tes_id'}\n",
    "    id_cols_2 = {'triplet': 'transcript_id', 'tss': 'tss_id', 'ic': 'ic_id', 'tes': 'tes_id'}\n",
    "    return feats, adatas, t_dfs, id_cols, id_cols_2\n",
    "\n",
    "def get_pi_tpm(sg, adata, t_df, obs_col, feat, id_col, id_col_2, min_tpm_how=None, min_tpm=None):\n",
    "    \"\"\"\n",
    "    Get long-form table w/ pi and tpm values for each feature\n",
    "    \"\"\"    \n",
    "    tpm_df = compute_feat_tpm(adata, obs_col, feat, 'mean', id_col, min_tpm=min_tpm)\n",
    "    pi_df = compute_feat_pi(adata, obs_col, t_df, feat)    \n",
    "    df = tpm_df.merge(pi_df, how='left', on=[obs_col, id_col])\n",
    "    \n",
    "    # get gene id, gene name\n",
    "    if feat == 'triplet':\n",
    "        pat = '['\n",
    "    else:\n",
    "        pat = '_'\n",
    "    df['gid'] = df[id_col].str.split(pat, expand=True)[0]\n",
    "    g_df = sg.t_df[['gid', 'gname']].reset_index(drop=True).drop_duplicates()\n",
    "    df = df.merge(g_df, how='left', on='gid')\n",
    "    \n",
    "    df = rank_feats(df, feat, obs_col, id_col)\n",
    "    return df\n",
    "\n",
    "def get_pi_tpm_tables(sg, obs_col):\n",
    "    feats, adatas, t_dfs, id_cols, id_cols_2 = get_loopers(sg)\n",
    "    for feat, adata, t_df in zip(feats, adatas, t_dfs):\n",
    "\n",
    "        # var init\n",
    "        id_col = id_cols[feat]\n",
    "        id_col_2 = id_cols_2[feat]\n",
    "        t_df = t_df.copy(deep=True)\n",
    "\n",
    "        # compute tpm and pi of each isoform\n",
    "        df = get_pi_tpm(sg, adata, t_df, obs_col, feat, id_col, id_col_2)\n",
    "\n",
    "        # save file\n",
    "        fname = '{}_{}_pi_tpm.tsv'.format(feat, obs_col)\n",
    "        df.to_csv(fname, sep='\\t', index=False)\n",
    "        \n",
    "def plot_ranked_feats(df, feat, obs_col, x='mane', opref='figures/'):\n",
    "    \"\"\"\n",
    "    Plot features by their rank\n",
    "    \"\"\"\n",
    "    rank_col = f'{feat}_rank_{x}'\n",
    "    temp = df.loc[~df[f'{feat}_rank_{x}'].isnull()].copy(deep=True)\n",
    "    temp = temp[['gid', obs_col, rank_col]]\n",
    "\n",
    "    # call anything @ rank 10+ same colum\n",
    "    temp['rank'] = temp[rank_col].astype(int).astype(str)\n",
    "    temp.loc[temp[rank_col]>=10, 'rank'] = '10+'\n",
    "\n",
    "    # count gene / dataset combos @ each rank\n",
    "    gb_cols = 'rank'\n",
    "    temp = temp[['gid', 'rank']].groupby(gb_cols).count().reset_index()\n",
    "    temp.rename({'gid': 'counts'}, axis=1, inplace=True)\n",
    "\n",
    "    sns.set_context('paper', font_scale=2)\n",
    "    mpl.rcParams['font.family'] = 'Arial'\n",
    "    mpl.rcParams['pdf.fonttype'] = 42\n",
    "    plt.figure(figsize=(3,4))\n",
    "\n",
    "    order = [str(int(i)) for i in range(1,10)]\n",
    "    order += ['10+']\n",
    "\n",
    "    if feat == 'triplet':\n",
    "        feat_label = 'isoform'\n",
    "    else:\n",
    "        feat_label = feat.upper()\n",
    "\n",
    "    if x == 'mane':\n",
    "        x_hr = 'MANE'\n",
    "    elif x == 'princ':\n",
    "        x_hr = 'Predominant'\n",
    "    elif x == 'sec':\n",
    "        x_hr = 'Secondary'\n",
    "\n",
    "    c_dict, _ = get_feat_triplet_colors()  \n",
    "    color = c_dict[feat]\n",
    "\n",
    "    ax = sns.catplot(data=temp, x='rank', y='counts', order=order, kind='bar', color=color, saturation=1)\n",
    "    ylabel = 'Gene / library combinations'\n",
    "    xlabel = f'Rank of {x_hr} {feat_label}'\n",
    "    ax.set(xlabel=xlabel, ylabel=ylabel)\n",
    "\n",
    "\n",
    "    ax = plt.gca()\n",
    "    from matplotlib import ticker\n",
    "    formatter = ticker.ScalarFormatter(useMathText=True)\n",
    "    formatter.set_scientific(True) \n",
    "    formatter.set_powerlimits((-1,1)) \n",
    "    ax.yaxis.set_major_formatter(formatter) \n",
    "\n",
    "    fname = f'{opref}/{x}_{feat}s_by_rank_hist.pdf'\n",
    "    plt.savefig(fname, dpi=800, bbox_inches='tight')\n",
    "\n",
    "def plot_major_principal_feat_counts(df, feat, obs_col, opref='figures/', **kwargs):\n",
    "    \"\"\"\n",
    "    Plot a histogram of the number of datasets where the major isoform\n",
    "    is the principal isoform.\n",
    "    \n",
    "    Parameters:\n",
    "        sg (swan_vis SwanGraph): SwanGraph with transcript\n",
    "                abundance data added\n",
    "    \n",
    "    Returns:\n",
    "        temp (pandas DataFrame): Output from `count_major_principal_feats`.\n",
    "    \"\"\"\n",
    "    mpl.rcParams['font.family'] = 'Arial'\n",
    "    mpl.rcParams['pdf.fonttype'] = 42\n",
    "    sns.set_context('paper', font_scale=2.8)\n",
    "    c_dict, order = get_feat_triplet_colors()  \n",
    "    col = 'mp_perc'\n",
    "    ax = sns.displot(data=df,\n",
    "                     x=col,\n",
    "                     linewidth=0, \n",
    "                     color=c_dict[feat],\n",
    "                     binwidth=5,\n",
    "                     alpha=1)\n",
    "    plt.gcf().set_size_inches(1.3333*4,0.7639*4)\n",
    "    ylabel = '# genes'\n",
    "    if obs_col == 'dataset':\n",
    "        label_col = 'librarie'\n",
    "    else:\n",
    "        label_col = obs_col\n",
    "    if feat in ['tss', 'ic', 'tes']: \n",
    "        xlabel = '% of {}s where predominant\\n{} is from MANE'.format(label_col, feat.upper())\n",
    "    else:\n",
    "        xlabel = '% of {}s where predominant\\ntranscript is MANE'.format(label_col)\n",
    "\n",
    "    ax.set(ylabel=ylabel, xlabel=xlabel, ylim=(0,9500))\n",
    "\n",
    "    fname = '{}/mane_vs_principal_{}_{}_hist.pdf'.format(opref, feat, obs_col)\n",
    "    plt.savefig(fname, dpi=800, bbox_inches='tight')\n",
    "            \n",
    "                \n",
    "#                 mpl.rcParams['font.family'] = 'Arial'\n",
    "#     mpl.rcParams['pdf.fonttype'] = 42\n",
    "#     sns.set_context('paper', font_scale=1.8)\n",
    "#     # fig, ax = plt.subplots(figsize=(1.3333,0.7639))\n",
    "#     # sns.set(rc={'figure.figsize':(1.3333,0.7639)})\n",
    "    \n",
    "#     c_dict, order = get_feat_triplet_colors()  \n",
    "#     col = 'mp_perc'\n",
    "#     ax = sns.displot(data=df,\n",
    "#                      x=col,\n",
    "#                      linewidth=0, \n",
    "#                      color=c_dict[feat],\n",
    "#                      binwidth=5,\n",
    "#                      alpha=1)\n",
    "#     plt.gcf().set_size_inches(1.3333,0.7639)\n",
    "#     ylabel = '# genes'\n",
    "#     if obs_col == 'dataset':\n",
    "#         label_col = 'librarie'\n",
    "#     else:\n",
    "#         label_col = obs_col\n",
    "#     if feat in ['tss', 'ic', 'tes']: \n",
    "#         xlabel = '% of {}s where major {} is from MANE'.format(obs_col, feat.upper())\n",
    "#     else:\n",
    "#         xlabel = '% of {}s where major isoform is MANE'.format(obs_col)\n",
    "\n",
    "#     # ax.set(ylabel=ylabel, xlabel=xlabel, labelsize=6.71)\n",
    "#     plt.xlabel(xlabel, fontsize=6.71)\n",
    "#     plt.ylabel(ylabel, fontsize=6.71)\n",
    "#     plt.xticks(fontsize=6)\n",
    "#     plt.yticks(fontsize=6)\n",
    "\n",
    "#     fname = '{}/mane_vs_principal_{}_{}_hist.pdf'.format(opref, feat, obs_col)\n",
    "#     plt.savefig(fname, dpi=800, bbox_inches='tight')\n",
    "\n",
    "def mane_analysis(sg, obs_col, min_feat_tpm):\n",
    "    \n",
    "    mp_dfs = {}\n",
    "    feats, adatas, t_dfs, id_cols, id_cols_2 = get_loopers(sg)\n",
    "    for feat, adata, t_df in zip(feats, adatas, t_dfs):\n",
    "        print(feat)\n",
    "\n",
    "        # var init\n",
    "        id_col = id_cols[feat]\n",
    "        id_col_2 = id_cols_2[feat]\n",
    "        t_df = t_df.copy(deep=True)\n",
    "        \n",
    "        # get pi / tpm table\n",
    "        fname = '{}_{}_pi_tpm.tsv'.format(feat, obs_col)\n",
    "        df = pd.read_csv(fname, sep='\\t')\n",
    "\n",
    "        # merge mane and principal; make sure that exp. of null genes == 0\n",
    "        princ_df = get_principal_feats(df, feat, obs_col, id_col)\n",
    "        sec_df = get_ranked_feats(df, feat, obs_col, id_col, rank=2) \n",
    "        rename_cols = [id_col, '{}_tpm'.format(feat), '{}_pi'.format(feat), '{}_rank'.format(feat)]\n",
    "        rename_dict = dict([(c,'{}_sec'.format(c)) for c in rename_cols])\n",
    "        sec_df.rename(rename_dict, axis=1, inplace=True)\n",
    "        mane_df = get_mane_feats(ca, df, feat, obs_col, id_col, id_col_2)\n",
    "        mp_df = princ_df.merge(mane_df, how='left', on=['gid', 'gname', obs_col], suffixes=('_princ', '_mane'))\n",
    "        mp_df = mp_df.merge(sec_df, how='left', on=['gid', 'gname', obs_col], suffixes=('', '_sec'))\n",
    "        mp_df['{}_tpm_mane'.format(feat)].fillna(0, inplace=True)\n",
    "        mp_df['{}_tpm_sec'.format(feat)].fillna(0, inplace=True)\n",
    "        \n",
    "        # remove genes that don't have annotated, expressed MANE isoforms in this sample\n",
    "        exp_mane_genes = get_exp_mane_genes(obs_col=obs_col)\n",
    "        mp_df = mp_df.merge(exp_mane_genes, how='inner', on=['gid', obs_col])\n",
    "\n",
    "        # threshold on tpm\n",
    "        mp_df = mp_df.loc[mp_df['{}_tpm_princ'.format(feat)] >= min_feat_tpm] # exclude princ. isos < min\n",
    "        mp_df.loc[mp_df['{}_tpm_mane'.format(feat)]<min_feat_tpm, '{}_tpm_mane'.format(feat)] = 0 # 0 out mane isos < min\n",
    "        mp_df.loc[mp_df['{}_tpm_sec'.format(feat)]<min_feat_tpm, '{}_tpm_sec'.format(feat)] = 0 # 0 out sec isos < min\n",
    "\n",
    "        # is principal == mane?\n",
    "        mp_col = '{}_princ_is_mane'.format(feat)\n",
    "        mp_df[mp_col] = mp_df[id_col+'_princ'] == mp_df[id_col+'_mane']\n",
    "\n",
    "        # is secondary == mane?\n",
    "        ms_col = '{}_sec_is_mane'.format(feat)\n",
    "        mp_df[ms_col] = mp_df[id_col+'_mane'] == mp_df[id_col+'_sec']\n",
    "\n",
    "        # histograms\n",
    "        # count number of libraries / gene where mane == principal\n",
    "        temp1 = mp_df[['gid', mp_col, obs_col]].groupby(['gid', mp_col]).count().reset_index()\n",
    "        # count number of libraries / gene where the gene is actually expressed (and has an isoform >= 1 tpm)\n",
    "        temp2 = temp1[['gid', obs_col]].groupby('gid').sum().reset_index()\n",
    "        temp1.rename({obs_col: 'n_{}s'.format(obs_col)}, axis=1, inplace=True)\n",
    "        temp2.rename({obs_col: 'n_{}s'.format(obs_col)}, axis=1, inplace=True)\n",
    "        # merge total # expressed libraries w/ # libraries where mane == pricipal\n",
    "        temp1 = temp1.merge(temp2, on='gid', suffixes=('_mp', ''))\n",
    "        print('# genes: {}'.format(len(temp1.gid.unique().tolist())))\n",
    "        n = len(temp1.gid.unique())\n",
    "        # only get instances where mane == principal\n",
    "        temp = temp1.loc[temp1[mp_col] == True].copy(deep=True)\n",
    "        temp['mp_perc'] = (temp['n_{}s_mp'.format(obs_col)]/temp['n_{}s'.format(obs_col)])*100\n",
    "        n_num = len(temp.loc[temp.mp_perc < 80, 'gid'].unique())\n",
    "        # pdb.set_trace()\n",
    "        print('{:.2f}% ({}/{}) of genes have a non-MANE predominant {} in 80% of expressed {}s'.format((n_num/n)*100, n_num, n, feat, obs_col))\n",
    "        plot_major_principal_feat_counts(temp, feat, obs_col, opref='figures/')\n",
    "        \n",
    "        # plot mane features based on their ranks\n",
    "        plot_ranked_feats(mp_df, feat, obs_col, x='mane', opref='figures/')\n",
    "        \n",
    "        # plot secondary isoform tpm vs. mane when mane == principal\n",
    "        temp = mp_df.loc[mp_df[f'{feat}_princ_is_mane']==True]\n",
    "        n = len(temp.index)\n",
    "        temp = temp.loc[temp[f'{feat}_tpm_sec']>0]\n",
    "        n_num = len(temp.index)\n",
    "        print('{:.2f}% ({}/{}) of gene / {} combos where MANE is predominant {} also have a secondary expressed {}'.format((n_num/n)*100, n_num, n, obs_col, feat, feat))\n",
    "        plot_tpms(temp, feat, 'tpm', x='mane', y='sec')\n",
    "        # sns.displot(temp, x='log_triplet_tpm_sec', kind='hist')\n",
    "        # plt.xticks(np.arange(0, 11, 1.0))\n",
    "        # break\n",
    "\n",
    "        # stats for principal != mane but mane still expressed\n",
    "        temp = mp_df.loc[mp_df[mp_col] == False]\n",
    "        n = len(temp.index)\n",
    "        temp = temp.loc[temp['{}_tpm_mane'.format(feat)]>0]\n",
    "        n_num = len(temp.index)\n",
    "        print('{:.2f}% ({}/{}) of gene / {} combos  where predominant {} is not MANE have MANE expression'.format((n_num/n)*100, n_num, n, obs_col, feat))\n",
    "        plot_tpms(temp, feat, 'tpm')\n",
    "        plot_tpms(temp, feat, 'pi')\n",
    "        \n",
    "        temp['tpm_diff'] = temp['{}_tpm_princ'.format(feat)] - temp['{}_tpm_mane'.format(feat)]\n",
    "        temp['log_tpm_diff'] = temp['log_{}_tpm_princ'.format(feat)] - temp['log_{}_tpm_mane'.format(feat)]\n",
    "        sns.displot(data=temp, x='tpm_diff')\n",
    "        sns.displot(data=temp, x='log_tpm_diff')\n",
    "        \n",
    "        # stats for principal != mane and mane NOT expressed\n",
    "        temp = mp_df.loc[mp_df[mp_col] == False]\n",
    "        n = len(temp.index)\n",
    "        temp = temp.loc[temp['{}_tpm_mane'.format(feat)]==0]\n",
    "        n_num = len(temp.index)\n",
    "        print('{:.2f}% ({}/{}) of gene / {} combos  where predominant {} is not MANE have NO MANE expression'.format((n_num/n)*100, n_num, n, obs_col, feat))\n",
    "\n",
    "        # what % of genes w/ mane isoforms that are expressed have >= 1 library where principal != mane\n",
    "        n = len(mp_df.gid.unique().tolist())\n",
    "        n_num = len(mp_df.loc[mp_df[mp_col]==False, 'gid'].unique().tolist())\n",
    "        print('{:.2f}% ({}/{}) of genes have >= 1 {} where the predominant {} is not MANE'.format((n_num/n)*100, n_num, n, obs_col, feat))\n",
    "\n",
    "        # what % of genes w/ mane isoforms that are expressed have >= 1 library where principal == mane\n",
    "        n = len(mp_df.gid.unique())\n",
    "        n_num = len(mp_df.loc[mp_df[mp_col]==True, 'gid'].unique().tolist())\n",
    "        print('{:.2f}% ({}/{}) of genes have >= 1 {} where the predominant {} is MANE'.format((n_num/n)*100, n_num, n, obs_col, feat))\n",
    "        print()    \n",
    "        \n",
    "        mp_dfs[feat] = mp_df\n",
    "        \n",
    "    return mp_dfs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "7401ca4c-e2aa-4885-9ebb-04575cdf612b",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Read in graph from ../../..//lr_bulk/cerberus/swan/swan_meta.p\n"
     ]
    }
   ],
   "source": [
    "#### get tpm / pi rank for each ad vs. healthy sample\n",
    "# save a file for library, gene id for detected polya genes\n",
    "# get_exp_gene_subset(ab, min_tpm, obs_col='health_status')\n",
    "sg = swan.read(swan_file)\n",
    "sample_inds = sg.adata.obs.loc[~sg.adata.obs.health_status.isnull()].index\n",
    "sg.adata = sg.adata[sample_inds, :]\n",
    "df = get_pi_tpm(sg, sg.adata, sg.t_df, 'health_status', 'triplet', 'tid', 'transcript_id', min_tpm=1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 75,
   "id": "9172102e-6006-47e7-baf6-780b0c253389",
   "metadata": {},
   "outputs": [],
   "source": [
    "major_isos = '../../../figures/fig_4_sample_major/triplet_dataset_pi_tpm.tsv'\n",
    "df = pd.read_csv(major_isos, sep='\\t')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 78,
   "id": "bbde0baf-6894-47f0-a159-ff545019f968",
   "metadata": {},
   "outputs": [],
   "source": [
    "# only get predominant isoforms\n",
    "df = df.loc[df.triplet_rank == 1]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 84,
   "id": "497c165f-df02-47a1-8061-0078978727bf",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Read in graph from ../../..//lr_bulk/cerberus/swan/swan_meta.p\n"
     ]
    }
   ],
   "source": [
    "# add in disease status\n",
    "sg = swan.read(swan_file)\n",
    "df = df.merge(sg.adata.obs[['health_status']], how='left',\n",
    "              left_on='dataset', right_index=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 86,
   "id": "5e741f9d-3bda-4fd7-a6f7-555061e38f61",
   "metadata": {},
   "outputs": [],
   "source": [
    "# only brain data\n",
    "df = df.loc[df.dataset.str.contains('brodman')]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 87,
   "id": "29af8f3f-a460-4630-8709-fecad5c1ffb3",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>tid</th>\n",
       "      <th>dataset</th>\n",
       "      <th>triplet_tpm</th>\n",
       "      <th>triplet_pi</th>\n",
       "      <th>gid</th>\n",
       "      <th>gname</th>\n",
       "      <th>triplet_rank</th>\n",
       "      <th>health_status</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>ENSG00000285991[1,2,2]</td>\n",
       "      <td>brodmann_area_46_5_1</td>\n",
       "      <td>0.726895</td>\n",
       "      <td>100.000000</td>\n",
       "      <td>ENSG00000285991</td>\n",
       "      <td>ENSG00000285991</td>\n",
       "      <td>1</td>\n",
       "      <td>AD</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>44</th>\n",
       "      <td>ENSG00000285980[2,2,2]</td>\n",
       "      <td>brodmann_area_46_5_1</td>\n",
       "      <td>0.726895</td>\n",
       "      <td>100.000000</td>\n",
       "      <td>ENSG00000285980</td>\n",
       "      <td>ENSG00000285980</td>\n",
       "      <td>1</td>\n",
       "      <td>AD</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>45</th>\n",
       "      <td>ENSG00000285978[1,1,1]</td>\n",
       "      <td>brodmann_area_46_4_1</td>\n",
       "      <td>0.636714</td>\n",
       "      <td>100.000000</td>\n",
       "      <td>ENSG00000285978</td>\n",
       "      <td>ENSG00000285978</td>\n",
       "      <td>1</td>\n",
       "      <td>AD</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>493</th>\n",
       "      <td>ENSG00000285976[1,1,3]</td>\n",
       "      <td>brodmann_area_46_9_1</td>\n",
       "      <td>15.481815</td>\n",
       "      <td>57.500000</td>\n",
       "      <td>ENSG00000285976</td>\n",
       "      <td>ENSG00000285976</td>\n",
       "      <td>1</td>\n",
       "      <td>AD</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>496</th>\n",
       "      <td>ENSG00000285976[2,1,1]</td>\n",
       "      <td>brodmann_area_46_8_1</td>\n",
       "      <td>23.452063</td>\n",
       "      <td>96.666664</td>\n",
       "      <td>ENSG00000285976</td>\n",
       "      <td>ENSG00000285976</td>\n",
       "      <td>1</td>\n",
       "      <td>AD</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                        tid               dataset  triplet_tpm  triplet_pi  \\\n",
       "20   ENSG00000285991[1,2,2]  brodmann_area_46_5_1     0.726895  100.000000   \n",
       "44   ENSG00000285980[2,2,2]  brodmann_area_46_5_1     0.726895  100.000000   \n",
       "45   ENSG00000285978[1,1,1]  brodmann_area_46_4_1     0.636714  100.000000   \n",
       "493  ENSG00000285976[1,1,3]  brodmann_area_46_9_1    15.481815   57.500000   \n",
       "496  ENSG00000285976[2,1,1]  brodmann_area_46_8_1    23.452063   96.666664   \n",
       "\n",
       "                 gid            gname  triplet_rank health_status  \n",
       "20   ENSG00000285991  ENSG00000285991             1            AD  \n",
       "44   ENSG00000285980  ENSG00000285980             1            AD  \n",
       "45   ENSG00000285978  ENSG00000285978             1            AD  \n",
       "493  ENSG00000285976  ENSG00000285976             1            AD  \n",
       "496  ENSG00000285976  ENSG00000285976             1            AD  "
      ]
     },
     "execution_count": 87,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3743c3ee-e038-4a9d-a303-039a57e55fca",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.4"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
