{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "7b3a405c-7ae3-4fda-94d5-84c6f54bcbfd",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "import scipy.stats as st\n",
    "import seaborn as sns\n",
    "import sys\n",
    "import os\n",
    "import gseapy as gp\n",
    "import pdb\n",
    "import copy\n",
    "from scipy import sparse\n",
    "import anndata\n",
    "import cerberus\n",
    "import subprocess\n",
    "\n",
    "p = os.path.dirname(os.path.dirname(os.path.dirname(os.getcwd())))\n",
    "sys.path.append(p)\n",
    "\n",
    "from scripts.utils import *\n",
    "from scripts.plotting import *"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "e4ff697e-217d-4283-97c8-e8458730ea07",
   "metadata": {},
   "outputs": [],
   "source": [
    "major_set = '../swan/isos_sample_gene_90.tsv'\n",
    "c_annot = 'cerberus_annot_triplets.h5'\n",
    "gtf = '../cerberus.gtf'\n",
    "swan_file = '../swan/swan.p'\n",
    "filt_ab = '../cerberus_filtered_abundance.tsv'\n",
    "h5 = '../cerberus_annot.h5'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "id": "bf0c6540-acaf-4fdc-9d8f-09564f755ec6",
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_med_exons_per_gene(ca):\n",
    "    \"\"\"\n",
    "    Get the median number of exons per gene across \n",
    "    all identified transcripts in cerberus\n",
    "    \"\"\"\n",
    "    exon_df = ca.t_map[['transcript_id']]\n",
    "    exon_df = add_feat(exon_df, kind='ic', col='transcript_id')\n",
    "    exon_df.drop_duplicates(inplace=True)\n",
    "    \n",
    "    # merge w/ ic table to get # exons\n",
    "    exon_df = exon_df.merge(ca.ic[['Name', 'Coordinates']],\n",
    "                        how='left',\n",
    "                        left_on='ic',\n",
    "                        right_on='Name')\n",
    "    exon_df['n_exons'] = exon_df['Coordinates'].str.count('-')\n",
    "    exon_df.loc[exon_df.Coordinates=='-', 'exon_df'] = 0\n",
    "    \n",
    "    # get gene id and compute median\n",
    "    exon_df['gid'] = exon_df.ic.str.split('_', expand=True)[0]\n",
    "    gene_exon_df = exon_df[['gid', 'n_exons']].groupby('gid').median().reset_index()\n",
    "    gene_exon_df.rename({'n_exons': 'median_exons'}, axis=1, inplace=True)\n",
    "    \n",
    "    return gene_exon_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "3b894c4c-4c0f-4595-bd84-4b698b02530d",
   "metadata": {},
   "outputs": [],
   "source": [
    "ca = cerberus.read(c_annot)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "482ab524-c62d-4f82-beba-ed2e6f2cb821",
   "metadata": {},
   "source": [
    "## 1,1,1 isoform"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 66,
   "id": "221b2bf5-5e17-406f-8e53-1571793bcfde",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>source</th>\n",
       "      <th>gid</th>\n",
       "      <th>n_tss</th>\n",
       "      <th>n_tes</th>\n",
       "      <th>n_ic</th>\n",
       "      <th>n_iso</th>\n",
       "      <th>splicing_ratio</th>\n",
       "      <th>gname</th>\n",
       "      <th>sample</th>\n",
       "      <th>gene_tpm</th>\n",
       "      <th>triplet</th>\n",
       "      <th>median_exons</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>84888</th>\n",
       "      <td>sample_det</td>\n",
       "      <td>ENSG00000164879</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>CA3</td>\n",
       "      <td>muscle</td>\n",
       "      <td>1181.528564</td>\n",
       "      <td>1,1,1</td>\n",
       "      <td>5.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>110312</th>\n",
       "      <td>sample_det</td>\n",
       "      <td>ENSG00000142676</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>RPL11</td>\n",
       "      <td>mcf7</td>\n",
       "      <td>1067.494995</td>\n",
       "      <td>1,1,1</td>\n",
       "      <td>7.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>167491</th>\n",
       "      <td>sample_det</td>\n",
       "      <td>ENSG00000142676</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>RPL11</td>\n",
       "      <td>hl60</td>\n",
       "      <td>941.440308</td>\n",
       "      <td>1,1,1</td>\n",
       "      <td>7.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>168037</th>\n",
       "      <td>sample_det</td>\n",
       "      <td>ENSG00000171314</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>PGAM1</td>\n",
       "      <td>hl60</td>\n",
       "      <td>941.217407</td>\n",
       "      <td>1,1,1</td>\n",
       "      <td>5.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>139089</th>\n",
       "      <td>sample_det</td>\n",
       "      <td>ENSG00000171314</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>PGAM1</td>\n",
       "      <td>adipose</td>\n",
       "      <td>843.580444</td>\n",
       "      <td>1,1,1</td>\n",
       "      <td>5.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>84309</th>\n",
       "      <td>sample_det</td>\n",
       "      <td>ENSG00000143318</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>CASQ1</td>\n",
       "      <td>muscle</td>\n",
       "      <td>793.289673</td>\n",
       "      <td>1,1,1</td>\n",
       "      <td>15.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>151674</th>\n",
       "      <td>sample_det</td>\n",
       "      <td>ENSG00000142676</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>RPL11</td>\n",
       "      <td>panc1</td>\n",
       "      <td>748.990295</td>\n",
       "      <td>1,1,1</td>\n",
       "      <td>7.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>47977</th>\n",
       "      <td>sample_det</td>\n",
       "      <td>ENSG00000148965</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>SAA4</td>\n",
       "      <td>liver</td>\n",
       "      <td>743.821594</td>\n",
       "      <td>1,1,1</td>\n",
       "      <td>5.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>48175</th>\n",
       "      <td>sample_det</td>\n",
       "      <td>ENSG00000157131</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>C8A</td>\n",
       "      <td>liver</td>\n",
       "      <td>721.939087</td>\n",
       "      <td>1,1,1</td>\n",
       "      <td>13.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>126781</th>\n",
       "      <td>sample_det</td>\n",
       "      <td>ENSG00000142676</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>RPL11</td>\n",
       "      <td>pc3</td>\n",
       "      <td>709.075439</td>\n",
       "      <td>1,1,1</td>\n",
       "      <td>7.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>55254</th>\n",
       "      <td>sample_det</td>\n",
       "      <td>ENSG00000146247</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>PHIP</td>\n",
       "      <td>h9_panc_progen</td>\n",
       "      <td>705.141846</td>\n",
       "      <td>1,1,1</td>\n",
       "      <td>77.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>132952</th>\n",
       "      <td>sample_det</td>\n",
       "      <td>ENSG00000249751</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>ECSCR</td>\n",
       "      <td>pgp1_endo</td>\n",
       "      <td>673.123535</td>\n",
       "      <td>1,1,1</td>\n",
       "      <td>17.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20546</th>\n",
       "      <td>sample_det</td>\n",
       "      <td>ENSG00000064961</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>HMG20B</td>\n",
       "      <td>hl60_m0</td>\n",
       "      <td>671.065308</td>\n",
       "      <td>1,1,1</td>\n",
       "      <td>13.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23769</th>\n",
       "      <td>sample_det</td>\n",
       "      <td>ENSG00000102837</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>OLFM4</td>\n",
       "      <td>colon</td>\n",
       "      <td>660.868896</td>\n",
       "      <td>1,1,1</td>\n",
       "      <td>7.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>55037</th>\n",
       "      <td>sample_det</td>\n",
       "      <td>ENSG00000136628</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>EPRS1</td>\n",
       "      <td>h9_panc_progen</td>\n",
       "      <td>580.577881</td>\n",
       "      <td>1,1,1</td>\n",
       "      <td>37.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>87745</th>\n",
       "      <td>sample_det</td>\n",
       "      <td>ENSG00000125538</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>IL1B</td>\n",
       "      <td>hl60_m1_72hr</td>\n",
       "      <td>564.614380</td>\n",
       "      <td>1,1,1</td>\n",
       "      <td>6.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>103971</th>\n",
       "      <td>sample_det</td>\n",
       "      <td>ENSG00000158710</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>TAGLN2</td>\n",
       "      <td>hl60_m2_72hr</td>\n",
       "      <td>557.584045</td>\n",
       "      <td>1,1,1</td>\n",
       "      <td>7.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>74746</th>\n",
       "      <td>sample_det</td>\n",
       "      <td>ENSG00000108821</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>COL1A1</td>\n",
       "      <td>ovary</td>\n",
       "      <td>547.942749</td>\n",
       "      <td>1,1,1</td>\n",
       "      <td>5.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>145001</th>\n",
       "      <td>sample_det</td>\n",
       "      <td>ENSG00000141753</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>IGFBP4</td>\n",
       "      <td>hffc6</td>\n",
       "      <td>523.172363</td>\n",
       "      <td>1,1,1</td>\n",
       "      <td>5.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>139063</th>\n",
       "      <td>sample_det</td>\n",
       "      <td>ENSG00000170759</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>KIF5B</td>\n",
       "      <td>adipose</td>\n",
       "      <td>494.224457</td>\n",
       "      <td>1,1,1</td>\n",
       "      <td>18.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "            source              gid  n_tss  n_tes  n_ic  n_iso  \\\n",
       "84888   sample_det  ENSG00000164879    1.0    1.0   1.0    1.0   \n",
       "110312  sample_det  ENSG00000142676    1.0    1.0   1.0    1.0   \n",
       "167491  sample_det  ENSG00000142676    1.0    1.0   1.0    1.0   \n",
       "168037  sample_det  ENSG00000171314    1.0    1.0   1.0    1.0   \n",
       "139089  sample_det  ENSG00000171314    1.0    1.0   1.0    1.0   \n",
       "84309   sample_det  ENSG00000143318    1.0    1.0   1.0    1.0   \n",
       "151674  sample_det  ENSG00000142676    1.0    1.0   1.0    1.0   \n",
       "47977   sample_det  ENSG00000148965    1.0    1.0   1.0    1.0   \n",
       "48175   sample_det  ENSG00000157131    1.0    1.0   1.0    1.0   \n",
       "126781  sample_det  ENSG00000142676    1.0    1.0   1.0    1.0   \n",
       "55254   sample_det  ENSG00000146247    1.0    1.0   1.0    1.0   \n",
       "132952  sample_det  ENSG00000249751    1.0    1.0   1.0    1.0   \n",
       "20546   sample_det  ENSG00000064961    1.0    1.0   1.0    1.0   \n",
       "23769   sample_det  ENSG00000102837    1.0    1.0   1.0    1.0   \n",
       "55037   sample_det  ENSG00000136628    1.0    1.0   1.0    1.0   \n",
       "87745   sample_det  ENSG00000125538    1.0    1.0   1.0    1.0   \n",
       "103971  sample_det  ENSG00000158710    1.0    1.0   1.0    1.0   \n",
       "74746   sample_det  ENSG00000108821    1.0    1.0   1.0    1.0   \n",
       "145001  sample_det  ENSG00000141753    1.0    1.0   1.0    1.0   \n",
       "139063  sample_det  ENSG00000170759    1.0    1.0   1.0    1.0   \n",
       "\n",
       "        splicing_ratio   gname          sample     gene_tpm triplet  \\\n",
       "84888              1.0     CA3          muscle  1181.528564   1,1,1   \n",
       "110312             1.0   RPL11            mcf7  1067.494995   1,1,1   \n",
       "167491             1.0   RPL11            hl60   941.440308   1,1,1   \n",
       "168037             1.0   PGAM1            hl60   941.217407   1,1,1   \n",
       "139089             1.0   PGAM1         adipose   843.580444   1,1,1   \n",
       "84309              1.0   CASQ1          muscle   793.289673   1,1,1   \n",
       "151674             1.0   RPL11           panc1   748.990295   1,1,1   \n",
       "47977              1.0    SAA4           liver   743.821594   1,1,1   \n",
       "48175              1.0     C8A           liver   721.939087   1,1,1   \n",
       "126781             1.0   RPL11             pc3   709.075439   1,1,1   \n",
       "55254              1.0    PHIP  h9_panc_progen   705.141846   1,1,1   \n",
       "132952             1.0   ECSCR       pgp1_endo   673.123535   1,1,1   \n",
       "20546              1.0  HMG20B         hl60_m0   671.065308   1,1,1   \n",
       "23769              1.0   OLFM4           colon   660.868896   1,1,1   \n",
       "55037              1.0   EPRS1  h9_panc_progen   580.577881   1,1,1   \n",
       "87745              1.0    IL1B    hl60_m1_72hr   564.614380   1,1,1   \n",
       "103971             1.0  TAGLN2    hl60_m2_72hr   557.584045   1,1,1   \n",
       "74746              1.0  COL1A1           ovary   547.942749   1,1,1   \n",
       "145001             1.0  IGFBP4           hffc6   523.172363   1,1,1   \n",
       "139063             1.0   KIF5B         adipose   494.224457   1,1,1   \n",
       "\n",
       "        median_exons  \n",
       "84888            5.0  \n",
       "110312           7.0  \n",
       "167491           7.0  \n",
       "168037           5.0  \n",
       "139089           5.0  \n",
       "84309           15.0  \n",
       "151674           7.0  \n",
       "47977            5.0  \n",
       "48175           13.0  \n",
       "126781           7.0  \n",
       "55254           77.0  \n",
       "132952          17.0  \n",
       "20546           13.0  \n",
       "23769            7.0  \n",
       "55037           37.0  \n",
       "87745            6.0  \n",
       "103971           7.0  \n",
       "74746            5.0  \n",
       "145001           5.0  \n",
       "139063          18.0  "
      ]
     },
     "execution_count": 66,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# get all sample det isoforms\n",
    "source = 'sample_det'\n",
    "df = ca.triplets.loc[ca.triplets.source == source]\n",
    "\n",
    "# get triplet id and limit to 1,1,1 genes\n",
    "df['triplet'] = df.n_tss.astype(int).astype(str)+','+\\\n",
    "                df.n_ic.astype(int).astype(str)+','+\\\n",
    "                df.n_tes.astype(int).astype(str)\n",
    "df = df.loc[df.triplet == '1,1,1']\n",
    "\n",
    "# get the median number of exons per gene\n",
    "ge_df = get_med_exons_per_gene(ca)\n",
    "df = df.merge(ge_df, how='left', on='gid')\n",
    "              \n",
    "# remove genes without a lot of exons\n",
    "df = df.loc[df.median_exons > 3]\n",
    "              \n",
    "df.sort_values(by='gene_tpm', ascending=False).head(20)              "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 63,
   "id": "642f54f2-459d-4d27-8428-f64271a98040",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>source</th>\n",
       "      <th>gid</th>\n",
       "      <th>n_tss</th>\n",
       "      <th>n_tes</th>\n",
       "      <th>n_ic</th>\n",
       "      <th>n_iso</th>\n",
       "      <th>splicing_ratio</th>\n",
       "      <th>gname</th>\n",
       "      <th>sample</th>\n",
       "      <th>gene_tpm</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>525202</th>\n",
       "      <td>sample_det</td>\n",
       "      <td>ENSG00000108821</td>\n",
       "      <td>3.0</td>\n",
       "      <td>5.0</td>\n",
       "      <td>4.0</td>\n",
       "      <td>7.0</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>COL1A1</td>\n",
       "      <td>h9_chondro</td>\n",
       "      <td>63934.390625</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>140365</th>\n",
       "      <td>sample_det</td>\n",
       "      <td>ENSG00000108821</td>\n",
       "      <td>2.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>4.0</td>\n",
       "      <td>0.800000</td>\n",
       "      <td>COL1A1</td>\n",
       "      <td>pgp1_astro</td>\n",
       "      <td>17760.802734</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>128010</th>\n",
       "      <td>sample_det</td>\n",
       "      <td>ENSG00000108821</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>COL1A1</td>\n",
       "      <td>imr90</td>\n",
       "      <td>9858.122070</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>487597</th>\n",
       "      <td>sample_det</td>\n",
       "      <td>ENSG00000108821</td>\n",
       "      <td>2.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>0.800000</td>\n",
       "      <td>COL1A1</td>\n",
       "      <td>hffc6</td>\n",
       "      <td>1834.368774</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>210700</th>\n",
       "      <td>sample_det</td>\n",
       "      <td>ENSG00000108821</td>\n",
       "      <td>2.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>0.800000</td>\n",
       "      <td>COL1A1</td>\n",
       "      <td>vessels</td>\n",
       "      <td>680.690735</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>252630</th>\n",
       "      <td>sample_det</td>\n",
       "      <td>ENSG00000108821</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>COL1A1</td>\n",
       "      <td>ovary</td>\n",
       "      <td>547.942749</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>572436</th>\n",
       "      <td>sample_det</td>\n",
       "      <td>ENSG00000108821</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>COL1A1</td>\n",
       "      <td>h9_osteocyte</td>\n",
       "      <td>442.539429</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>58620</th>\n",
       "      <td>sample_det</td>\n",
       "      <td>ENSG00000108821</td>\n",
       "      <td>1.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>0.666667</td>\n",
       "      <td>COL1A1</td>\n",
       "      <td>lung</td>\n",
       "      <td>175.406296</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>466298</th>\n",
       "      <td>sample_det</td>\n",
       "      <td>ENSG00000108821</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>COL1A1</td>\n",
       "      <td>adipose</td>\n",
       "      <td>131.496017</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>87306</th>\n",
       "      <td>sample_det</td>\n",
       "      <td>ENSG00000108821</td>\n",
       "      <td>1.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>0.666667</td>\n",
       "      <td>COL1A1</td>\n",
       "      <td>colon</td>\n",
       "      <td>131.359818</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>341063</th>\n",
       "      <td>sample_det</td>\n",
       "      <td>ENSG00000108821</td>\n",
       "      <td>1.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>0.500000</td>\n",
       "      <td>COL1A1</td>\n",
       "      <td>h1_de</td>\n",
       "      <td>121.919708</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>500370</th>\n",
       "      <td>sample_det</td>\n",
       "      <td>ENSG00000108821</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>COL1A1</td>\n",
       "      <td>h9</td>\n",
       "      <td>111.554153</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>549577</th>\n",
       "      <td>sample_det</td>\n",
       "      <td>ENSG00000108821</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>COL1A1</td>\n",
       "      <td>adrenal gland</td>\n",
       "      <td>110.667442</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>620253</th>\n",
       "      <td>sample_det</td>\n",
       "      <td>ENSG00000108821</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>COL1A1</td>\n",
       "      <td>kidney</td>\n",
       "      <td>89.649818</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>226748</th>\n",
       "      <td>sample_det</td>\n",
       "      <td>ENSG00000108821</td>\n",
       "      <td>1.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>0.666667</td>\n",
       "      <td>COL1A1</td>\n",
       "      <td>h1</td>\n",
       "      <td>88.303543</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>389259</th>\n",
       "      <td>sample_det</td>\n",
       "      <td>ENSG00000108821</td>\n",
       "      <td>1.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>0.666667</td>\n",
       "      <td>COL1A1</td>\n",
       "      <td>heart</td>\n",
       "      <td>57.795326</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>304747</th>\n",
       "      <td>sample_det</td>\n",
       "      <td>ENSG00000108821</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>COL1A1</td>\n",
       "      <td>wtc11</td>\n",
       "      <td>43.026535</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>164735</th>\n",
       "      <td>sample_det</td>\n",
       "      <td>ENSG00000108821</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>COL1A1</td>\n",
       "      <td>liver</td>\n",
       "      <td>41.023026</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>153148</th>\n",
       "      <td>sample_det</td>\n",
       "      <td>ENSG00000108821</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>COL1A1</td>\n",
       "      <td>h9_de</td>\n",
       "      <td>39.249863</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>45267</th>\n",
       "      <td>sample_det</td>\n",
       "      <td>ENSG00000108821</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>COL1A1</td>\n",
       "      <td>a673</td>\n",
       "      <td>39.099773</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>511739</th>\n",
       "      <td>sample_det</td>\n",
       "      <td>ENSG00000108821</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>COL1A1</td>\n",
       "      <td>panc1</td>\n",
       "      <td>30.944794</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>430218</th>\n",
       "      <td>sample_det</td>\n",
       "      <td>ENSG00000108821</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>COL1A1</td>\n",
       "      <td>pc3</td>\n",
       "      <td>25.717762</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>282096</th>\n",
       "      <td>sample_det</td>\n",
       "      <td>ENSG00000108821</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>COL1A1</td>\n",
       "      <td>muscle</td>\n",
       "      <td>18.003805</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2600</th>\n",
       "      <td>sample_det</td>\n",
       "      <td>ENSG00000108821</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>COL1A1</td>\n",
       "      <td>caco2</td>\n",
       "      <td>7.774920</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>267651</th>\n",
       "      <td>sample_det</td>\n",
       "      <td>ENSG00000108821</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>COL1A1</td>\n",
       "      <td>gm12878</td>\n",
       "      <td>4.240708</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>406464</th>\n",
       "      <td>sample_det</td>\n",
       "      <td>ENSG00000108821</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>COL1A1</td>\n",
       "      <td>pgp1</td>\n",
       "      <td>3.797228</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16131</th>\n",
       "      <td>sample_det</td>\n",
       "      <td>ENSG00000108821</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>COL1A1</td>\n",
       "      <td>brain</td>\n",
       "      <td>2.852873</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>593631</th>\n",
       "      <td>sample_det</td>\n",
       "      <td>ENSG00000108821</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>COL1A1</td>\n",
       "      <td>hepg2</td>\n",
       "      <td>0.814126</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "            source              gid  n_tss  n_tes  n_ic  n_iso  \\\n",
       "525202  sample_det  ENSG00000108821    3.0    5.0   4.0    7.0   \n",
       "140365  sample_det  ENSG00000108821    2.0    3.0   2.0    4.0   \n",
       "128010  sample_det  ENSG00000108821    2.0    2.0   2.0    2.0   \n",
       "487597  sample_det  ENSG00000108821    2.0    3.0   2.0    3.0   \n",
       "210700  sample_det  ENSG00000108821    2.0    3.0   2.0    3.0   \n",
       "252630  sample_det  ENSG00000108821    1.0    1.0   1.0    1.0   \n",
       "572436  sample_det  ENSG00000108821    2.0    2.0   2.0    2.0   \n",
       "58620   sample_det  ENSG00000108821    1.0    2.0   1.0    2.0   \n",
       "466298  sample_det  ENSG00000108821    1.0    1.0   1.0    1.0   \n",
       "87306   sample_det  ENSG00000108821    1.0    2.0   1.0    2.0   \n",
       "341063  sample_det  ENSG00000108821    1.0    3.0   1.0    3.0   \n",
       "500370  sample_det  ENSG00000108821    1.0    1.0   1.0    1.0   \n",
       "549577  sample_det  ENSG00000108821    1.0    1.0   1.0    1.0   \n",
       "620253  sample_det  ENSG00000108821    1.0    1.0   1.0    1.0   \n",
       "226748  sample_det  ENSG00000108821    1.0    2.0   1.0    2.0   \n",
       "389259  sample_det  ENSG00000108821    1.0    2.0   1.0    2.0   \n",
       "304747  sample_det  ENSG00000108821    1.0    1.0   1.0    1.0   \n",
       "164735  sample_det  ENSG00000108821    1.0    1.0   1.0    1.0   \n",
       "153148  sample_det  ENSG00000108821    1.0    1.0   1.0    1.0   \n",
       "45267   sample_det  ENSG00000108821    1.0    1.0   1.0    1.0   \n",
       "511739  sample_det  ENSG00000108821    1.0    1.0   1.0    1.0   \n",
       "430218  sample_det  ENSG00000108821    1.0    1.0   1.0    1.0   \n",
       "282096  sample_det  ENSG00000108821    1.0    1.0   1.0    1.0   \n",
       "2600    sample_det  ENSG00000108821    1.0    1.0   1.0    1.0   \n",
       "267651  sample_det  ENSG00000108821    1.0    1.0   1.0    1.0   \n",
       "406464  sample_det  ENSG00000108821    1.0    1.0   1.0    1.0   \n",
       "16131   sample_det  ENSG00000108821    1.0    1.0   1.0    1.0   \n",
       "593631  sample_det  ENSG00000108821    1.0    1.0   1.0    1.0   \n",
       "\n",
       "        splicing_ratio   gname         sample      gene_tpm  \n",
       "525202        1.000000  COL1A1     h9_chondro  63934.390625  \n",
       "140365        0.800000  COL1A1     pgp1_astro  17760.802734  \n",
       "128010        1.000000  COL1A1          imr90   9858.122070  \n",
       "487597        0.800000  COL1A1          hffc6   1834.368774  \n",
       "210700        0.800000  COL1A1        vessels    680.690735  \n",
       "252630        1.000000  COL1A1          ovary    547.942749  \n",
       "572436        1.000000  COL1A1   h9_osteocyte    442.539429  \n",
       "58620         0.666667  COL1A1           lung    175.406296  \n",
       "466298        1.000000  COL1A1        adipose    131.496017  \n",
       "87306         0.666667  COL1A1          colon    131.359818  \n",
       "341063        0.500000  COL1A1          h1_de    121.919708  \n",
       "500370        1.000000  COL1A1             h9    111.554153  \n",
       "549577        1.000000  COL1A1  adrenal gland    110.667442  \n",
       "620253        1.000000  COL1A1         kidney     89.649818  \n",
       "226748        0.666667  COL1A1             h1     88.303543  \n",
       "389259        0.666667  COL1A1          heart     57.795326  \n",
       "304747        1.000000  COL1A1          wtc11     43.026535  \n",
       "164735        1.000000  COL1A1          liver     41.023026  \n",
       "153148        1.000000  COL1A1          h9_de     39.249863  \n",
       "45267         1.000000  COL1A1           a673     39.099773  \n",
       "511739        1.000000  COL1A1          panc1     30.944794  \n",
       "430218        1.000000  COL1A1            pc3     25.717762  \n",
       "282096        1.000000  COL1A1         muscle     18.003805  \n",
       "2600          1.000000  COL1A1          caco2      7.774920  \n",
       "267651        1.000000  COL1A1        gm12878      4.240708  \n",
       "406464        1.000000  COL1A1           pgp1      3.797228  \n",
       "16131         1.000000  COL1A1          brain      2.852873  \n",
       "593631        1.000000  COL1A1          hepg2      0.814126  "
      ]
     },
     "execution_count": 63,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ca.triplets.loc[(ca.triplets.source == 'sample_det')&(ca.triplets.gname=='COL1A1')].sort_values(by='gene_tpm', ascending=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "id": "71919ed2-45be-44e4-ae37-eaceb788b5ee",
   "metadata": {},
   "outputs": [],
   "source": [
    "ge_df = get_med_exons_per_gene(ca)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8ca22cba-0498-4fe9-85e9-9f0f034cc056",
   "metadata": {},
   "source": [
    "## Highly-expressed non 1,1,1 isoform"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 70,
   "id": "2a486f3c-fb21-41ab-8c0b-952a20f99f9c",
   "metadata": {},
   "outputs": [],
   "source": [
    "source = 'sample_det'\n",
    "df = ca.triplets.loc[ca.triplets.source == source]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 78,
   "id": "0d928482-fbd2-49fd-bad7-8c91c45f28c4",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "        source              gid  n_tss  n_tes  n_ic  n_iso  splicing_ratio  \\\n",
      "0   sample_det  ENSG00000000003    1.0    5.0   6.0   18.0             2.0   \n",
      "1   sample_det  ENSG00000000419    1.0    4.0   8.0   11.0             3.2   \n",
      "4   sample_det  ENSG00000001036    1.0    4.0   6.0   11.0             2.4   \n",
      "19  sample_det  ENSG00000002586    1.0    2.0   9.0    9.0             6.0   \n",
      "23  sample_det  ENSG00000002822    4.0    2.0  18.0   19.0             6.0   \n",
      "\n",
      "     gname sample    gene_tpm triplet  \n",
      "0   TSPAN6  caco2  347.757385   1,6,5  \n",
      "1     DPM1  caco2  126.812943   1,8,4  \n",
      "4    FUCA2  caco2  261.958191   1,6,4  \n",
      "19    CD99  caco2  408.017517   1,9,2  \n",
      "23  MAD1L1  caco2   97.261612  4,18,2  \n"
     ]
    }
   ],
   "source": [
    "# get triplet id and limit to non 1,1,1 genes\n",
    "df['triplet'] = df.n_tss.astype(int).astype(str)+','+\\\n",
    "                df.n_ic.astype(int).astype(str)+','+\\\n",
    "                df.n_tes.astype(int).astype(str)\n",
    "df = df.loc[df.triplet != '1,1,1']\n",
    "print(df.head())\n",
    "\n",
    "# let's try to get genes w/ a lot of ic variation\n",
    "df = df.loc[df.n_ic > 5]\n",
    "\n",
    "# add tf designation\n",
    "gene_df, _, _ = get_gtf_info(how='gene', ver='v40_cerberus')\n",
    "gene_df['gid_stable'] = cerberus.get_stable_gid(gene_df, 'gid')\n",
    "df = df.merge(gene_df[['tf', 'gid_stable']],\n",
    "              how='left', left_on='gid',\n",
    "              right_on='gid_stable', \n",
    "              suffixes=('', '_gtf'))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 80,
   "id": "0dcdae3f-caaa-48b5-aed7-6379d3ff83fe",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>source</th>\n",
       "      <th>gid</th>\n",
       "      <th>n_tss</th>\n",
       "      <th>n_tes</th>\n",
       "      <th>n_ic</th>\n",
       "      <th>n_iso</th>\n",
       "      <th>splicing_ratio</th>\n",
       "      <th>gname</th>\n",
       "      <th>sample</th>\n",
       "      <th>gene_tpm</th>\n",
       "      <th>triplet</th>\n",
       "      <th>tf</th>\n",
       "      <th>gid_stable</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>39390</th>\n",
       "      <td>sample_det</td>\n",
       "      <td>ENSG00000067066</td>\n",
       "      <td>2.0</td>\n",
       "      <td>5.0</td>\n",
       "      <td>13.0</td>\n",
       "      <td>15.0</td>\n",
       "      <td>3.714286</td>\n",
       "      <td>SP100</td>\n",
       "      <td>ovary</td>\n",
       "      <td>102.093597</td>\n",
       "      <td>2,13,5</td>\n",
       "      <td>True</td>\n",
       "      <td>ENSG00000067066</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>39391</th>\n",
       "      <td>sample_det</td>\n",
       "      <td>ENSG00000067082</td>\n",
       "      <td>2.0</td>\n",
       "      <td>5.0</td>\n",
       "      <td>7.0</td>\n",
       "      <td>9.0</td>\n",
       "      <td>2.000000</td>\n",
       "      <td>KLF6</td>\n",
       "      <td>ovary</td>\n",
       "      <td>197.415817</td>\n",
       "      <td>2,7,5</td>\n",
       "      <td>True</td>\n",
       "      <td>ENSG00000067082</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>39624</th>\n",
       "      <td>sample_det</td>\n",
       "      <td>ENSG00000100219</td>\n",
       "      <td>3.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>6.0</td>\n",
       "      <td>6.0</td>\n",
       "      <td>2.400000</td>\n",
       "      <td>XBP1</td>\n",
       "      <td>ovary</td>\n",
       "      <td>136.043945</td>\n",
       "      <td>3,6,2</td>\n",
       "      <td>True</td>\n",
       "      <td>ENSG00000100219</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>39762</th>\n",
       "      <td>sample_det</td>\n",
       "      <td>ENSG00000103495</td>\n",
       "      <td>5.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>9.0</td>\n",
       "      <td>9.0</td>\n",
       "      <td>2.571429</td>\n",
       "      <td>MAZ</td>\n",
       "      <td>ovary</td>\n",
       "      <td>149.227646</td>\n",
       "      <td>5,9,2</td>\n",
       "      <td>True</td>\n",
       "      <td>ENSG00000103495</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>39855</th>\n",
       "      <td>sample_det</td>\n",
       "      <td>ENSG00000106624</td>\n",
       "      <td>4.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>7.0</td>\n",
       "      <td>7.0</td>\n",
       "      <td>2.800000</td>\n",
       "      <td>AEBP1</td>\n",
       "      <td>ovary</td>\n",
       "      <td>388.119843</td>\n",
       "      <td>4,7,1</td>\n",
       "      <td>True</td>\n",
       "      <td>ENSG00000106624</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>40072</th>\n",
       "      <td>sample_det</td>\n",
       "      <td>ENSG00000116044</td>\n",
       "      <td>4.0</td>\n",
       "      <td>5.0</td>\n",
       "      <td>7.0</td>\n",
       "      <td>9.0</td>\n",
       "      <td>1.555556</td>\n",
       "      <td>NFE2L2</td>\n",
       "      <td>ovary</td>\n",
       "      <td>156.310120</td>\n",
       "      <td>4,7,5</td>\n",
       "      <td>True</td>\n",
       "      <td>ENSG00000116044</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>40205</th>\n",
       "      <td>sample_det</td>\n",
       "      <td>ENSG00000123358</td>\n",
       "      <td>5.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>11.0</td>\n",
       "      <td>11.0</td>\n",
       "      <td>2.750000</td>\n",
       "      <td>NR4A1</td>\n",
       "      <td>ovary</td>\n",
       "      <td>179.301956</td>\n",
       "      <td>5,11,3</td>\n",
       "      <td>True</td>\n",
       "      <td>ENSG00000123358</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>40352</th>\n",
       "      <td>sample_det</td>\n",
       "      <td>ENSG00000130844</td>\n",
       "      <td>7.0</td>\n",
       "      <td>7.0</td>\n",
       "      <td>10.0</td>\n",
       "      <td>23.0</td>\n",
       "      <td>1.428571</td>\n",
       "      <td>ZNF331</td>\n",
       "      <td>ovary</td>\n",
       "      <td>204.941071</td>\n",
       "      <td>7,10,7</td>\n",
       "      <td>True</td>\n",
       "      <td>ENSG00000130844</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>40495</th>\n",
       "      <td>sample_det</td>\n",
       "      <td>ENSG00000136574</td>\n",
       "      <td>6.0</td>\n",
       "      <td>5.0</td>\n",
       "      <td>16.0</td>\n",
       "      <td>18.0</td>\n",
       "      <td>2.909091</td>\n",
       "      <td>GATA4</td>\n",
       "      <td>ovary</td>\n",
       "      <td>199.088608</td>\n",
       "      <td>6,16,5</td>\n",
       "      <td>True</td>\n",
       "      <td>ENSG00000136574</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>40583</th>\n",
       "      <td>sample_det</td>\n",
       "      <td>ENSG00000140262</td>\n",
       "      <td>1.0</td>\n",
       "      <td>5.0</td>\n",
       "      <td>9.0</td>\n",
       "      <td>12.0</td>\n",
       "      <td>3.000000</td>\n",
       "      <td>TCF12</td>\n",
       "      <td>ovary</td>\n",
       "      <td>107.840759</td>\n",
       "      <td>1,9,5</td>\n",
       "      <td>True</td>\n",
       "      <td>ENSG00000140262</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>40628</th>\n",
       "      <td>sample_det</td>\n",
       "      <td>ENSG00000141448</td>\n",
       "      <td>9.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>6.0</td>\n",
       "      <td>15.0</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>GATA6</td>\n",
       "      <td>ovary</td>\n",
       "      <td>198.567398</td>\n",
       "      <td>9,6,3</td>\n",
       "      <td>True</td>\n",
       "      <td>ENSG00000141448</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>40642</th>\n",
       "      <td>sample_det</td>\n",
       "      <td>ENSG00000141905</td>\n",
       "      <td>2.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>6.0</td>\n",
       "      <td>8.0</td>\n",
       "      <td>2.400000</td>\n",
       "      <td>NFIC</td>\n",
       "      <td>ovary</td>\n",
       "      <td>327.426086</td>\n",
       "      <td>2,6,3</td>\n",
       "      <td>True</td>\n",
       "      <td>ENSG00000141905</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>40672</th>\n",
       "      <td>sample_det</td>\n",
       "      <td>ENSG00000143321</td>\n",
       "      <td>5.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>9.0</td>\n",
       "      <td>11.0</td>\n",
       "      <td>2.250000</td>\n",
       "      <td>HDGF</td>\n",
       "      <td>ovary</td>\n",
       "      <td>128.039261</td>\n",
       "      <td>5,9,3</td>\n",
       "      <td>True</td>\n",
       "      <td>ENSG00000143321</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>41149</th>\n",
       "      <td>sample_det</td>\n",
       "      <td>ENSG00000167081</td>\n",
       "      <td>5.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>7.0</td>\n",
       "      <td>13.0</td>\n",
       "      <td>1.750000</td>\n",
       "      <td>PBX3</td>\n",
       "      <td>ovary</td>\n",
       "      <td>137.806519</td>\n",
       "      <td>5,7,3</td>\n",
       "      <td>True</td>\n",
       "      <td>ENSG00000167081</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>41204</th>\n",
       "      <td>sample_det</td>\n",
       "      <td>ENSG00000168610</td>\n",
       "      <td>3.0</td>\n",
       "      <td>5.0</td>\n",
       "      <td>14.0</td>\n",
       "      <td>31.0</td>\n",
       "      <td>3.500000</td>\n",
       "      <td>STAT3</td>\n",
       "      <td>ovary</td>\n",
       "      <td>187.051453</td>\n",
       "      <td>3,14,5</td>\n",
       "      <td>True</td>\n",
       "      <td>ENSG00000168610</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>41247</th>\n",
       "      <td>sample_det</td>\n",
       "      <td>ENSG00000170345</td>\n",
       "      <td>4.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>7.0</td>\n",
       "      <td>8.0</td>\n",
       "      <td>2.000000</td>\n",
       "      <td>FOS</td>\n",
       "      <td>ovary</td>\n",
       "      <td>120.485855</td>\n",
       "      <td>4,7,3</td>\n",
       "      <td>True</td>\n",
       "      <td>ENSG00000170345</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>41439</th>\n",
       "      <td>sample_det</td>\n",
       "      <td>ENSG00000181722</td>\n",
       "      <td>5.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>7.0</td>\n",
       "      <td>9.0</td>\n",
       "      <td>1.750000</td>\n",
       "      <td>ZBTB20</td>\n",
       "      <td>ovary</td>\n",
       "      <td>105.372246</td>\n",
       "      <td>5,7,3</td>\n",
       "      <td>True</td>\n",
       "      <td>ENSG00000181722</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>41507</th>\n",
       "      <td>sample_det</td>\n",
       "      <td>ENSG00000185122</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>13.0</td>\n",
       "      <td>13.0</td>\n",
       "      <td>13.000000</td>\n",
       "      <td>HSF1</td>\n",
       "      <td>ovary</td>\n",
       "      <td>190.860550</td>\n",
       "      <td>1,13,1</td>\n",
       "      <td>True</td>\n",
       "      <td>ENSG00000185122</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>41565</th>\n",
       "      <td>sample_det</td>\n",
       "      <td>ENSG00000189067</td>\n",
       "      <td>4.0</td>\n",
       "      <td>5.0</td>\n",
       "      <td>6.0</td>\n",
       "      <td>11.0</td>\n",
       "      <td>1.333333</td>\n",
       "      <td>LITAF</td>\n",
       "      <td>ovary</td>\n",
       "      <td>227.072357</td>\n",
       "      <td>4,6,5</td>\n",
       "      <td>True</td>\n",
       "      <td>ENSG00000189067</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>41585</th>\n",
       "      <td>sample_det</td>\n",
       "      <td>ENSG00000196378</td>\n",
       "      <td>3.0</td>\n",
       "      <td>4.0</td>\n",
       "      <td>8.0</td>\n",
       "      <td>10.0</td>\n",
       "      <td>2.285714</td>\n",
       "      <td>ZNF34</td>\n",
       "      <td>ovary</td>\n",
       "      <td>229.851639</td>\n",
       "      <td>3,8,4</td>\n",
       "      <td>True</td>\n",
       "      <td>ENSG00000196378</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>41599</th>\n",
       "      <td>sample_det</td>\n",
       "      <td>ENSG00000196628</td>\n",
       "      <td>8.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>15.0</td>\n",
       "      <td>15.0</td>\n",
       "      <td>2.727273</td>\n",
       "      <td>TCF4</td>\n",
       "      <td>ovary</td>\n",
       "      <td>129.414719</td>\n",
       "      <td>8,15,3</td>\n",
       "      <td>True</td>\n",
       "      <td>ENSG00000196628</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>41652</th>\n",
       "      <td>sample_det</td>\n",
       "      <td>ENSG00000198300</td>\n",
       "      <td>1.0</td>\n",
       "      <td>6.0</td>\n",
       "      <td>7.0</td>\n",
       "      <td>14.0</td>\n",
       "      <td>2.000000</td>\n",
       "      <td>PEG3</td>\n",
       "      <td>ovary</td>\n",
       "      <td>173.894638</td>\n",
       "      <td>1,7,6</td>\n",
       "      <td>True</td>\n",
       "      <td>ENSG00000198300</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>41736</th>\n",
       "      <td>sample_det</td>\n",
       "      <td>ENSG00000213928</td>\n",
       "      <td>2.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>9.0</td>\n",
       "      <td>11.0</td>\n",
       "      <td>3.600000</td>\n",
       "      <td>IRF9</td>\n",
       "      <td>ovary</td>\n",
       "      <td>139.000351</td>\n",
       "      <td>2,9,3</td>\n",
       "      <td>True</td>\n",
       "      <td>ENSG00000213928</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "           source              gid  n_tss  n_tes  n_ic  n_iso  splicing_ratio  \\\n",
       "39390  sample_det  ENSG00000067066    2.0    5.0  13.0   15.0        3.714286   \n",
       "39391  sample_det  ENSG00000067082    2.0    5.0   7.0    9.0        2.000000   \n",
       "39624  sample_det  ENSG00000100219    3.0    2.0   6.0    6.0        2.400000   \n",
       "39762  sample_det  ENSG00000103495    5.0    2.0   9.0    9.0        2.571429   \n",
       "39855  sample_det  ENSG00000106624    4.0    1.0   7.0    7.0        2.800000   \n",
       "40072  sample_det  ENSG00000116044    4.0    5.0   7.0    9.0        1.555556   \n",
       "40205  sample_det  ENSG00000123358    5.0    3.0  11.0   11.0        2.750000   \n",
       "40352  sample_det  ENSG00000130844    7.0    7.0  10.0   23.0        1.428571   \n",
       "40495  sample_det  ENSG00000136574    6.0    5.0  16.0   18.0        2.909091   \n",
       "40583  sample_det  ENSG00000140262    1.0    5.0   9.0   12.0        3.000000   \n",
       "40628  sample_det  ENSG00000141448    9.0    3.0   6.0   15.0        1.000000   \n",
       "40642  sample_det  ENSG00000141905    2.0    3.0   6.0    8.0        2.400000   \n",
       "40672  sample_det  ENSG00000143321    5.0    3.0   9.0   11.0        2.250000   \n",
       "41149  sample_det  ENSG00000167081    5.0    3.0   7.0   13.0        1.750000   \n",
       "41204  sample_det  ENSG00000168610    3.0    5.0  14.0   31.0        3.500000   \n",
       "41247  sample_det  ENSG00000170345    4.0    3.0   7.0    8.0        2.000000   \n",
       "41439  sample_det  ENSG00000181722    5.0    3.0   7.0    9.0        1.750000   \n",
       "41507  sample_det  ENSG00000185122    1.0    1.0  13.0   13.0       13.000000   \n",
       "41565  sample_det  ENSG00000189067    4.0    5.0   6.0   11.0        1.333333   \n",
       "41585  sample_det  ENSG00000196378    3.0    4.0   8.0   10.0        2.285714   \n",
       "41599  sample_det  ENSG00000196628    8.0    3.0  15.0   15.0        2.727273   \n",
       "41652  sample_det  ENSG00000198300    1.0    6.0   7.0   14.0        2.000000   \n",
       "41736  sample_det  ENSG00000213928    2.0    3.0   9.0   11.0        3.600000   \n",
       "\n",
       "        gname sample    gene_tpm triplet    tf       gid_stable  \n",
       "39390   SP100  ovary  102.093597  2,13,5  True  ENSG00000067066  \n",
       "39391    KLF6  ovary  197.415817   2,7,5  True  ENSG00000067082  \n",
       "39624    XBP1  ovary  136.043945   3,6,2  True  ENSG00000100219  \n",
       "39762     MAZ  ovary  149.227646   5,9,2  True  ENSG00000103495  \n",
       "39855   AEBP1  ovary  388.119843   4,7,1  True  ENSG00000106624  \n",
       "40072  NFE2L2  ovary  156.310120   4,7,5  True  ENSG00000116044  \n",
       "40205   NR4A1  ovary  179.301956  5,11,3  True  ENSG00000123358  \n",
       "40352  ZNF331  ovary  204.941071  7,10,7  True  ENSG00000130844  \n",
       "40495   GATA4  ovary  199.088608  6,16,5  True  ENSG00000136574  \n",
       "40583   TCF12  ovary  107.840759   1,9,5  True  ENSG00000140262  \n",
       "40628   GATA6  ovary  198.567398   9,6,3  True  ENSG00000141448  \n",
       "40642    NFIC  ovary  327.426086   2,6,3  True  ENSG00000141905  \n",
       "40672    HDGF  ovary  128.039261   5,9,3  True  ENSG00000143321  \n",
       "41149    PBX3  ovary  137.806519   5,7,3  True  ENSG00000167081  \n",
       "41204   STAT3  ovary  187.051453  3,14,5  True  ENSG00000168610  \n",
       "41247     FOS  ovary  120.485855   4,7,3  True  ENSG00000170345  \n",
       "41439  ZBTB20  ovary  105.372246   5,7,3  True  ENSG00000181722  \n",
       "41507    HSF1  ovary  190.860550  1,13,1  True  ENSG00000185122  \n",
       "41565   LITAF  ovary  227.072357   4,6,5  True  ENSG00000189067  \n",
       "41585   ZNF34  ovary  229.851639   3,8,4  True  ENSG00000196378  \n",
       "41599    TCF4  ovary  129.414719  8,15,3  True  ENSG00000196628  \n",
       "41652    PEG3  ovary  173.894638   1,7,6  True  ENSG00000198300  \n",
       "41736    IRF9  ovary  139.000351   2,9,3  True  ENSG00000213928  "
      ]
     },
     "execution_count": 80,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.loc[df['sample']=='ovary'].sort_values(by='gene_tpm', ascending=False)\n",
    "df.loc[(df['sample']=='ovary')&(df.gene_tpm>100)&(df.gene_tpm<1000)&(df.tf==True)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e03a42a7-cc49-41a1-90c4-dcdff8c890cf",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.4"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
