{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 43,
   "id": "dba567c2-1d01-4fcd-b115-a054d8bf709a",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "import scipy.stats as st\n",
    "import seaborn as sns\n",
    "import sys\n",
    "import os\n",
    "import gseapy as gp\n",
    "import pdb\n",
    "import copy\n",
    "from scipy import sparse\n",
    "import anndata\n",
    "import cerberus\n",
    "import subprocess\n",
    "import re\n",
    "\n",
    "p = os.path.dirname(os.path.dirname(os.path.dirname(os.getcwd())))\n",
    "sys.path.append(p)\n",
    "\n",
    "from scripts.utils import *\n",
    "from scripts.plotting import *"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 132,
   "id": "7bb6552a-73db-4822-954f-330a0828b2f1",
   "metadata": {},
   "outputs": [],
   "source": [
    "major_set = '../swan/isos_sample_gene_90.tsv'\n",
    "c_annot = 'cerberus_annot_triplets.h5'\n",
    "gtf = '../cerberus.gtf'\n",
    "swan_file = '../swan/swan.p'\n",
    "filt_ab = '../cerberus_filtered_abundance.tsv'\n",
    "h5 = '../cerberus_annot.h5'"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b41a60af-4d73-4960-b5f9-7cef67544457",
   "metadata": {},
   "source": [
    "## Get # isoforms in a sample for a given gene in the major set"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 55,
   "id": "39a5f243-60f4-420a-a03a-b75c89ac2f4d",
   "metadata": {},
   "outputs": [],
   "source": [
    "major_set_df = pd.read_csv(major_set, sep='\\t')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 56,
   "id": "9d614c33-fee6-4474-b166-f90e08cf2d56",
   "metadata": {},
   "outputs": [],
   "source": [
    "df = major_set_df[['tid', 'gname', 'gid', 'sample']]\n",
    "df = df.groupby(['gname', 'gid', 'sample']).count().reset_index()\n",
    "df.rename({'tid': 'n_tid'}, axis=1, inplace=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "e12edb9f-5f89-449a-8731-bb306de54436",
   "metadata": {},
   "outputs": [],
   "source": [
    "# df.loc[df.gname == 'ELN']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "id": "00edaded-c203-4f6d-8e66-a864a45396ea",
   "metadata": {},
   "outputs": [],
   "source": [
    "# df.loc[(df.gname == 'CTCF')&(df.n_tid >= 5)]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "32d0a566-2241-47bf-a1ca-2bb7abd67a6d",
   "metadata": {},
   "source": [
    "## Get # TSSs in a sample for a given gene in the major set"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "id": "b5f3a10e-9b62-433a-953d-4084c902ca20",
   "metadata": {},
   "outputs": [],
   "source": [
    "major_set_df = pd.read_csv(major_set, sep='\\t')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "id": "150a7ca3-0f53-4f46-b65b-9761e584d033",
   "metadata": {},
   "outputs": [],
   "source": [
    "major_set_df = add_feat(major_set_df, kind='tss', col='tid')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "id": "a3f26d90-a8a3-4bcc-bc22-067bd5f61280",
   "metadata": {},
   "outputs": [],
   "source": [
    "df = major_set_df[['tss', 'gname', 'gid', 'sample']]\n",
    "df = df.groupby(['gname', 'gid', 'sample']).nunique().reset_index()\n",
    "df.rename({'tss': 'n_tss'}, axis=1, inplace=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 54,
   "id": "7b6e6952-e2c4-4b92-a675-3d4fdcdc642d",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gname</th>\n",
       "      <th>gid</th>\n",
       "      <th>sample</th>\n",
       "      <th>n_tss</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>152862</th>\n",
       "      <td>CTCF</td>\n",
       "      <td>ENSG00000102974.15</td>\n",
       "      <td>gm12878</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>152870</th>\n",
       "      <td>CTCF</td>\n",
       "      <td>ENSG00000102974.15</td>\n",
       "      <td>h9_panc_beta</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>152880</th>\n",
       "      <td>CTCF</td>\n",
       "      <td>ENSG00000102974.15</td>\n",
       "      <td>hl60_m1_72hr</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>152883</th>\n",
       "      <td>CTCF</td>\n",
       "      <td>ENSG00000102974.15</td>\n",
       "      <td>hl60_m2_72hr</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>152902</th>\n",
       "      <td>CTCF</td>\n",
       "      <td>ENSG00000102974.15</td>\n",
       "      <td>vessels</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       gname                 gid        sample  n_tss\n",
       "152862  CTCF  ENSG00000102974.15       gm12878      4\n",
       "152870  CTCF  ENSG00000102974.15  h9_panc_beta      4\n",
       "152880  CTCF  ENSG00000102974.15  hl60_m1_72hr      4\n",
       "152883  CTCF  ENSG00000102974.15  hl60_m2_72hr      4\n",
       "152902  CTCF  ENSG00000102974.15       vessels      4"
      ]
     },
     "execution_count": 54,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.loc[(df.gname == 'CTCF')&(df.n_tss == 4)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 62,
   "id": "94c0bdca-0b60-4e9e-a607-10a50a3419a2",
   "metadata": {},
   "outputs": [],
   "source": [
    "temp = major_set_df.loc[(major_set_df.gname=='CTCF')&(major_set_df['sample'] == 'vessels')].copy(deep=True)\n",
    "temp = add_feat(temp, kind='tss', col='tid')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 64,
   "id": "46057774-f3ad-4ad7-9279-7306dd7a6cd1",
   "metadata": {},
   "outputs": [],
   "source": [
    "tsss = temp.tss.tolist()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 68,
   "id": "f44b0157-867a-4606-b651-6f81319d56ad",
   "metadata": {},
   "outputs": [],
   "source": [
    "ca = cerberus.read(c_annot)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 76,
   "id": "82a63ae0-78ba-4962-9ae3-293f96248a2b",
   "metadata": {},
   "outputs": [],
   "source": [
    "tss = pr.PyRanges(ca.tss)\n",
    "tss.to_bed('cerberus_tss.bed')\n",
    "\n",
    "tss_map = pr.PyRanges(ca.tss_map)\n",
    "tss_map.to_bed('cerberus_tss_map.bed')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 71,
   "id": "b4fc3eb3-b9fe-47ba-84b0-55e1f278d5dc",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Strand</th>\n",
       "      <th>Name</th>\n",
       "      <th>source</th>\n",
       "      <th>novelty</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>tss</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>112371</th>\n",
       "      <td>chr16</td>\n",
       "      <td>67562416</td>\n",
       "      <td>67562639</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000102974_1</td>\n",
       "      <td>v40,v29,lapa,gtex,encode_cage,fantom_cage,pls</td>\n",
       "      <td>Known</td>\n",
       "      <td>ENSG00000102974</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>112372</th>\n",
       "      <td>chr16</td>\n",
       "      <td>67626458</td>\n",
       "      <td>67626559</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000102974_10</td>\n",
       "      <td>v40,v29,lapa,dels</td>\n",
       "      <td>Known</td>\n",
       "      <td>ENSG00000102974</td>\n",
       "      <td>10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>112376</th>\n",
       "      <td>chr16</td>\n",
       "      <td>67610539</td>\n",
       "      <td>67610640</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000102974_3</td>\n",
       "      <td>v40,v29,lapa</td>\n",
       "      <td>Known</td>\n",
       "      <td>ENSG00000102974</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>112382</th>\n",
       "      <td>chr16</td>\n",
       "      <td>67620836</td>\n",
       "      <td>67620937</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000102974_9</td>\n",
       "      <td>v40,v29,lapa</td>\n",
       "      <td>Known</td>\n",
       "      <td>ENSG00000102974</td>\n",
       "      <td>9</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       Chromosome     Start       End Strand                Name  \\\n",
       "112371      chr16  67562416  67562639      +   ENSG00000102974_1   \n",
       "112372      chr16  67626458  67626559      +  ENSG00000102974_10   \n",
       "112376      chr16  67610539  67610640      +   ENSG00000102974_3   \n",
       "112382      chr16  67620836  67620937      +   ENSG00000102974_9   \n",
       "\n",
       "                                               source novelty  \\\n",
       "112371  v40,v29,lapa,gtex,encode_cage,fantom_cage,pls   Known   \n",
       "112372                              v40,v29,lapa,dels   Known   \n",
       "112376                                   v40,v29,lapa   Known   \n",
       "112382                                   v40,v29,lapa   Known   \n",
       "\n",
       "                gene_id  tss  \n",
       "112371  ENSG00000102974    1  \n",
       "112372  ENSG00000102974   10  \n",
       "112376  ENSG00000102974    3  \n",
       "112382  ENSG00000102974    9  "
      ]
     },
     "execution_count": 71,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ca.tss.loc[ca.tss.Name.isin(tsss)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 75,
   "id": "e78bcc9e-db81-422e-afb4-cc389fe356b1",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Strand</th>\n",
       "      <th>source</th>\n",
       "      <th>Name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1910538</th>\n",
       "      <td>chr16</td>\n",
       "      <td>67562416</td>\n",
       "      <td>67562639</td>\n",
       "      <td>+</td>\n",
       "      <td>v40</td>\n",
       "      <td>ENSG00000102974_1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1932191</th>\n",
       "      <td>chr16</td>\n",
       "      <td>67562595</td>\n",
       "      <td>67562617</td>\n",
       "      <td>+</td>\n",
       "      <td>fantom_cage</td>\n",
       "      <td>ENSG00000102974_1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1932190</th>\n",
       "      <td>chr16</td>\n",
       "      <td>67562514</td>\n",
       "      <td>67562594</td>\n",
       "      <td>+</td>\n",
       "      <td>fantom_cage</td>\n",
       "      <td>ENSG00000102974_1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1932189</th>\n",
       "      <td>chr16</td>\n",
       "      <td>67562481</td>\n",
       "      <td>67562508</td>\n",
       "      <td>+</td>\n",
       "      <td>fantom_cage</td>\n",
       "      <td>ENSG00000102974_1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1920551</th>\n",
       "      <td>chr16</td>\n",
       "      <td>67562356</td>\n",
       "      <td>67562639</td>\n",
       "      <td>+</td>\n",
       "      <td>v29</td>\n",
       "      <td>ENSG00000102974_1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1932188</th>\n",
       "      <td>chr16</td>\n",
       "      <td>67562453</td>\n",
       "      <td>67562464</td>\n",
       "      <td>+</td>\n",
       "      <td>fantom_cage</td>\n",
       "      <td>ENSG00000102974_1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1932187</th>\n",
       "      <td>chr16</td>\n",
       "      <td>67562433</td>\n",
       "      <td>67562450</td>\n",
       "      <td>+</td>\n",
       "      <td>fantom_cage</td>\n",
       "      <td>ENSG00000102974_1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1928129</th>\n",
       "      <td>chr16</td>\n",
       "      <td>67562465</td>\n",
       "      <td>67562721</td>\n",
       "      <td>+</td>\n",
       "      <td>encode_cage</td>\n",
       "      <td>ENSG00000102974_1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1924721</th>\n",
       "      <td>chr16</td>\n",
       "      <td>67562232</td>\n",
       "      <td>67562662</td>\n",
       "      <td>+</td>\n",
       "      <td>lapa</td>\n",
       "      <td>ENSG00000102974_1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1926877</th>\n",
       "      <td>chr16</td>\n",
       "      <td>67562487</td>\n",
       "      <td>67562588</td>\n",
       "      <td>+</td>\n",
       "      <td>gtex</td>\n",
       "      <td>ENSG00000102974_1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1940547</th>\n",
       "      <td>chr16</td>\n",
       "      <td>67562321</td>\n",
       "      <td>67562507</td>\n",
       "      <td>+</td>\n",
       "      <td>pls</td>\n",
       "      <td>ENSG00000102974_1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1953225</th>\n",
       "      <td>chr16</td>\n",
       "      <td>67626443</td>\n",
       "      <td>67626775</td>\n",
       "      <td>+</td>\n",
       "      <td>dels</td>\n",
       "      <td>ENSG00000102974_10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1920562</th>\n",
       "      <td>chr16</td>\n",
       "      <td>67626458</td>\n",
       "      <td>67626559</td>\n",
       "      <td>+</td>\n",
       "      <td>v29</td>\n",
       "      <td>ENSG00000102974_10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1910547</th>\n",
       "      <td>chr16</td>\n",
       "      <td>67626458</td>\n",
       "      <td>67626559</td>\n",
       "      <td>+</td>\n",
       "      <td>v40</td>\n",
       "      <td>ENSG00000102974_10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1924728</th>\n",
       "      <td>chr16</td>\n",
       "      <td>67626458</td>\n",
       "      <td>67626559</td>\n",
       "      <td>+</td>\n",
       "      <td>lapa</td>\n",
       "      <td>ENSG00000102974_10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1924724</th>\n",
       "      <td>chr16</td>\n",
       "      <td>67610539</td>\n",
       "      <td>67610640</td>\n",
       "      <td>+</td>\n",
       "      <td>lapa</td>\n",
       "      <td>ENSG00000102974_3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1920555</th>\n",
       "      <td>chr16</td>\n",
       "      <td>67610539</td>\n",
       "      <td>67610640</td>\n",
       "      <td>+</td>\n",
       "      <td>v29</td>\n",
       "      <td>ENSG00000102974_3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1910541</th>\n",
       "      <td>chr16</td>\n",
       "      <td>67610539</td>\n",
       "      <td>67610640</td>\n",
       "      <td>+</td>\n",
       "      <td>v40</td>\n",
       "      <td>ENSG00000102974_3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1920561</th>\n",
       "      <td>chr16</td>\n",
       "      <td>67620836</td>\n",
       "      <td>67620937</td>\n",
       "      <td>+</td>\n",
       "      <td>v29</td>\n",
       "      <td>ENSG00000102974_9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1910546</th>\n",
       "      <td>chr16</td>\n",
       "      <td>67620836</td>\n",
       "      <td>67620937</td>\n",
       "      <td>+</td>\n",
       "      <td>v40</td>\n",
       "      <td>ENSG00000102974_9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1924727</th>\n",
       "      <td>chr16</td>\n",
       "      <td>67620836</td>\n",
       "      <td>67620937</td>\n",
       "      <td>+</td>\n",
       "      <td>lapa</td>\n",
       "      <td>ENSG00000102974_9</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "        Chromosome     Start       End Strand       source                Name\n",
       "1910538      chr16  67562416  67562639      +          v40   ENSG00000102974_1\n",
       "1932191      chr16  67562595  67562617      +  fantom_cage   ENSG00000102974_1\n",
       "1932190      chr16  67562514  67562594      +  fantom_cage   ENSG00000102974_1\n",
       "1932189      chr16  67562481  67562508      +  fantom_cage   ENSG00000102974_1\n",
       "1920551      chr16  67562356  67562639      +          v29   ENSG00000102974_1\n",
       "1932188      chr16  67562453  67562464      +  fantom_cage   ENSG00000102974_1\n",
       "1932187      chr16  67562433  67562450      +  fantom_cage   ENSG00000102974_1\n",
       "1928129      chr16  67562465  67562721      +  encode_cage   ENSG00000102974_1\n",
       "1924721      chr16  67562232  67562662      +         lapa   ENSG00000102974_1\n",
       "1926877      chr16  67562487  67562588      +         gtex   ENSG00000102974_1\n",
       "1940547      chr16  67562321  67562507      +          pls   ENSG00000102974_1\n",
       "1953225      chr16  67626443  67626775      +         dels  ENSG00000102974_10\n",
       "1920562      chr16  67626458  67626559      +          v29  ENSG00000102974_10\n",
       "1910547      chr16  67626458  67626559      +          v40  ENSG00000102974_10\n",
       "1924728      chr16  67626458  67626559      +         lapa  ENSG00000102974_10\n",
       "1924724      chr16  67610539  67610640      +         lapa   ENSG00000102974_3\n",
       "1920555      chr16  67610539  67610640      +          v29   ENSG00000102974_3\n",
       "1910541      chr16  67610539  67610640      +          v40   ENSG00000102974_3\n",
       "1920561      chr16  67620836  67620937      +          v29   ENSG00000102974_9\n",
       "1910546      chr16  67620836  67620937      +          v40   ENSG00000102974_9\n",
       "1924727      chr16  67620836  67620937      +         lapa   ENSG00000102974_9"
      ]
     },
     "execution_count": 75,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ca.tss_map.loc[ca.tss_map.Name.isin(tsss)].sort_values(by='Name')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 82,
   "id": "aa019ff3-e822-4c77-9dff-b46d394c4e33",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>original_transcript_id</th>\n",
       "      <th>ic</th>\n",
       "      <th>ic_id</th>\n",
       "      <th>tss_id</th>\n",
       "      <th>tss</th>\n",
       "      <th>tes_id</th>\n",
       "      <th>tes</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>original_transcript_name</th>\n",
       "      <th>transcript_triplet</th>\n",
       "      <th>transcript_id</th>\n",
       "      <th>transcript_name</th>\n",
       "      <th>tss_first_sd_issue</th>\n",
       "      <th>tes_last_sa_issue</th>\n",
       "      <th>source</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>81328</th>\n",
       "      <td>ENST00000264010.10</td>\n",
       "      <td>1</td>\n",
       "      <td>ENSG00000102974_1</td>\n",
       "      <td>ENSG00000102974_1</td>\n",
       "      <td>1</td>\n",
       "      <td>ENSG00000102974_1</td>\n",
       "      <td>1</td>\n",
       "      <td>ENSG00000102974</td>\n",
       "      <td>CTCF</td>\n",
       "      <td>CTCF-201</td>\n",
       "      <td>[1,1,1]</td>\n",
       "      <td>ENSG00000102974[1,1,1]</td>\n",
       "      <td>CTCF[1,1,1]</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>v40</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>81886</th>\n",
       "      <td>ENST00000401394.6</td>\n",
       "      <td>7</td>\n",
       "      <td>ENSG00000102974_7</td>\n",
       "      <td>ENSG00000102974_1</td>\n",
       "      <td>1</td>\n",
       "      <td>ENSG00000102974_1</td>\n",
       "      <td>1</td>\n",
       "      <td>ENSG00000102974</td>\n",
       "      <td>CTCF</td>\n",
       "      <td>CTCF-202</td>\n",
       "      <td>[1,7,1]</td>\n",
       "      <td>ENSG00000102974[1,7,1]</td>\n",
       "      <td>CTCF[1,7,1]</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>v40</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>86576</th>\n",
       "      <td>ENST00000644852.1</td>\n",
       "      <td>9</td>\n",
       "      <td>ENSG00000102974_9</td>\n",
       "      <td>ENSG00000102974_1</td>\n",
       "      <td>1</td>\n",
       "      <td>ENSG00000102974_2</td>\n",
       "      <td>2</td>\n",
       "      <td>ENSG00000102974</td>\n",
       "      <td>CTCF</td>\n",
       "      <td>CTCF-210</td>\n",
       "      <td>[1,9,2]</td>\n",
       "      <td>ENSG00000102974[1,9,2]</td>\n",
       "      <td>CTCF[1,9,2]</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>v40</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>86606</th>\n",
       "      <td>ENST00000646076.1</td>\n",
       "      <td>2</td>\n",
       "      <td>ENSG00000102974_2</td>\n",
       "      <td>ENSG00000102974_1</td>\n",
       "      <td>1</td>\n",
       "      <td>ENSG00000102974_1</td>\n",
       "      <td>1</td>\n",
       "      <td>ENSG00000102974</td>\n",
       "      <td>CTCF</td>\n",
       "      <td>CTCF-215</td>\n",
       "      <td>[1,2,1]</td>\n",
       "      <td>ENSG00000102974[1,2,1]</td>\n",
       "      <td>CTCF[1,2,1]</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>v40</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>86630</th>\n",
       "      <td>ENST00000646771.1</td>\n",
       "      <td>6</td>\n",
       "      <td>ENSG00000102974_6</td>\n",
       "      <td>ENSG00000102974_1</td>\n",
       "      <td>1</td>\n",
       "      <td>ENSG00000102974_1</td>\n",
       "      <td>1</td>\n",
       "      <td>ENSG00000102974</td>\n",
       "      <td>CTCF</td>\n",
       "      <td>CTCF-217</td>\n",
       "      <td>[1,6,1]</td>\n",
       "      <td>ENSG00000102974[1,6,1]</td>\n",
       "      <td>CTCF[1,6,1]</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>v40</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>68664</th>\n",
       "      <td>ENST00000264010.9</td>\n",
       "      <td>1</td>\n",
       "      <td>ENSG00000102974_1</td>\n",
       "      <td>ENSG00000102974_1</td>\n",
       "      <td>1</td>\n",
       "      <td>ENSG00000102974_1</td>\n",
       "      <td>1</td>\n",
       "      <td>ENSG00000102974</td>\n",
       "      <td>CTCF</td>\n",
       "      <td>CTCF-201</td>\n",
       "      <td>[1,1,1]</td>\n",
       "      <td>ENSG00000102974[1,1,1]</td>\n",
       "      <td>CTCF[1,1,1]</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>v29</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>69233</th>\n",
       "      <td>ENST00000401394.6</td>\n",
       "      <td>7</td>\n",
       "      <td>ENSG00000102974_7</td>\n",
       "      <td>ENSG00000102974_1</td>\n",
       "      <td>1</td>\n",
       "      <td>ENSG00000102974_1</td>\n",
       "      <td>1</td>\n",
       "      <td>ENSG00000102974</td>\n",
       "      <td>CTCF</td>\n",
       "      <td>CTCF-202</td>\n",
       "      <td>[1,7,1]</td>\n",
       "      <td>ENSG00000102974[1,7,1]</td>\n",
       "      <td>CTCF[1,7,1]</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>v29</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>73976</th>\n",
       "      <td>ENST00000644852.1</td>\n",
       "      <td>9</td>\n",
       "      <td>ENSG00000102974_9</td>\n",
       "      <td>ENSG00000102974_1</td>\n",
       "      <td>1</td>\n",
       "      <td>ENSG00000102974_2</td>\n",
       "      <td>2</td>\n",
       "      <td>ENSG00000102974</td>\n",
       "      <td>CTCF</td>\n",
       "      <td>CTCF-210</td>\n",
       "      <td>[1,9,2]</td>\n",
       "      <td>ENSG00000102974[1,9,2]</td>\n",
       "      <td>CTCF[1,9,2]</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>v29</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>74007</th>\n",
       "      <td>ENST00000646076.1</td>\n",
       "      <td>2</td>\n",
       "      <td>ENSG00000102974_2</td>\n",
       "      <td>ENSG00000102974_1</td>\n",
       "      <td>1</td>\n",
       "      <td>ENSG00000102974_1</td>\n",
       "      <td>1</td>\n",
       "      <td>ENSG00000102974</td>\n",
       "      <td>CTCF</td>\n",
       "      <td>CTCF-215</td>\n",
       "      <td>[1,2,1]</td>\n",
       "      <td>ENSG00000102974[1,2,1]</td>\n",
       "      <td>CTCF[1,2,1]</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>v29</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>74032</th>\n",
       "      <td>ENST00000646771.1</td>\n",
       "      <td>6</td>\n",
       "      <td>ENSG00000102974_6</td>\n",
       "      <td>ENSG00000102974_1</td>\n",
       "      <td>1</td>\n",
       "      <td>ENSG00000102974_1</td>\n",
       "      <td>1</td>\n",
       "      <td>ENSG00000102974</td>\n",
       "      <td>CTCF</td>\n",
       "      <td>CTCF-217</td>\n",
       "      <td>[1,6,1]</td>\n",
       "      <td>ENSG00000102974[1,6,1]</td>\n",
       "      <td>CTCF[1,6,1]</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>v29</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>86502</th>\n",
       "      <td>ENST00000642420.1</td>\n",
       "      <td>17</td>\n",
       "      <td>ENSG00000102974_17</td>\n",
       "      <td>ENSG00000102974_10</td>\n",
       "      <td>10</td>\n",
       "      <td>ENSG00000102974_1</td>\n",
       "      <td>1</td>\n",
       "      <td>ENSG00000102974</td>\n",
       "      <td>CTCF</td>\n",
       "      <td>CTCF-204</td>\n",
       "      <td>[10,17,1]</td>\n",
       "      <td>ENSG00000102974[10,17,1]</td>\n",
       "      <td>CTCF[10,17,1]</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>v40</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>73902</th>\n",
       "      <td>ENST00000642420.1</td>\n",
       "      <td>17</td>\n",
       "      <td>ENSG00000102974_17</td>\n",
       "      <td>ENSG00000102974_10</td>\n",
       "      <td>10</td>\n",
       "      <td>ENSG00000102974_1</td>\n",
       "      <td>1</td>\n",
       "      <td>ENSG00000102974</td>\n",
       "      <td>CTCF</td>\n",
       "      <td>CTCF-204</td>\n",
       "      <td>[10,17,1]</td>\n",
       "      <td>ENSG00000102974[10,17,1]</td>\n",
       "      <td>CTCF[10,17,1]</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>v29</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>86596</th>\n",
       "      <td>ENST00000645699.1</td>\n",
       "      <td>5</td>\n",
       "      <td>ENSG00000102974_5</td>\n",
       "      <td>ENSG00000102974_3</td>\n",
       "      <td>3</td>\n",
       "      <td>ENSG00000102974_1</td>\n",
       "      <td>1</td>\n",
       "      <td>ENSG00000102974</td>\n",
       "      <td>CTCF</td>\n",
       "      <td>CTCF-214</td>\n",
       "      <td>[3,5,1]</td>\n",
       "      <td>ENSG00000102974[3,5,1]</td>\n",
       "      <td>CTCF[3,5,1]</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>v40</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>73997</th>\n",
       "      <td>ENST00000645699.1</td>\n",
       "      <td>5</td>\n",
       "      <td>ENSG00000102974_5</td>\n",
       "      <td>ENSG00000102974_3</td>\n",
       "      <td>3</td>\n",
       "      <td>ENSG00000102974_1</td>\n",
       "      <td>1</td>\n",
       "      <td>ENSG00000102974</td>\n",
       "      <td>CTCF</td>\n",
       "      <td>CTCF-214</td>\n",
       "      <td>[3,5,1]</td>\n",
       "      <td>ENSG00000102974[3,5,1]</td>\n",
       "      <td>CTCF[3,5,1]</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>v29</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>86579</th>\n",
       "      <td>ENST00000644950.1</td>\n",
       "      <td>16</td>\n",
       "      <td>ENSG00000102974_16</td>\n",
       "      <td>ENSG00000102974_9</td>\n",
       "      <td>9</td>\n",
       "      <td>ENSG00000102974_1</td>\n",
       "      <td>1</td>\n",
       "      <td>ENSG00000102974</td>\n",
       "      <td>CTCF</td>\n",
       "      <td>CTCF-211</td>\n",
       "      <td>[9,16,1]</td>\n",
       "      <td>ENSG00000102974[9,16,1]</td>\n",
       "      <td>CTCF[9,16,1]</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>v40</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>73979</th>\n",
       "      <td>ENST00000644950.1</td>\n",
       "      <td>16</td>\n",
       "      <td>ENSG00000102974_16</td>\n",
       "      <td>ENSG00000102974_9</td>\n",
       "      <td>9</td>\n",
       "      <td>ENSG00000102974_1</td>\n",
       "      <td>1</td>\n",
       "      <td>ENSG00000102974</td>\n",
       "      <td>CTCF</td>\n",
       "      <td>CTCF-211</td>\n",
       "      <td>[9,16,1]</td>\n",
       "      <td>ENSG00000102974[9,16,1]</td>\n",
       "      <td>CTCF[9,16,1]</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>v29</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      original_transcript_id  ic               ic_id              tss_id  tss  \\\n",
       "81328     ENST00000264010.10   1   ENSG00000102974_1   ENSG00000102974_1    1   \n",
       "81886      ENST00000401394.6   7   ENSG00000102974_7   ENSG00000102974_1    1   \n",
       "86576      ENST00000644852.1   9   ENSG00000102974_9   ENSG00000102974_1    1   \n",
       "86606      ENST00000646076.1   2   ENSG00000102974_2   ENSG00000102974_1    1   \n",
       "86630      ENST00000646771.1   6   ENSG00000102974_6   ENSG00000102974_1    1   \n",
       "68664      ENST00000264010.9   1   ENSG00000102974_1   ENSG00000102974_1    1   \n",
       "69233      ENST00000401394.6   7   ENSG00000102974_7   ENSG00000102974_1    1   \n",
       "73976      ENST00000644852.1   9   ENSG00000102974_9   ENSG00000102974_1    1   \n",
       "74007      ENST00000646076.1   2   ENSG00000102974_2   ENSG00000102974_1    1   \n",
       "74032      ENST00000646771.1   6   ENSG00000102974_6   ENSG00000102974_1    1   \n",
       "86502      ENST00000642420.1  17  ENSG00000102974_17  ENSG00000102974_10   10   \n",
       "73902      ENST00000642420.1  17  ENSG00000102974_17  ENSG00000102974_10   10   \n",
       "86596      ENST00000645699.1   5   ENSG00000102974_5   ENSG00000102974_3    3   \n",
       "73997      ENST00000645699.1   5   ENSG00000102974_5   ENSG00000102974_3    3   \n",
       "86579      ENST00000644950.1  16  ENSG00000102974_16   ENSG00000102974_9    9   \n",
       "73979      ENST00000644950.1  16  ENSG00000102974_16   ENSG00000102974_9    9   \n",
       "\n",
       "                  tes_id  tes          gene_id gene_name  \\\n",
       "81328  ENSG00000102974_1    1  ENSG00000102974      CTCF   \n",
       "81886  ENSG00000102974_1    1  ENSG00000102974      CTCF   \n",
       "86576  ENSG00000102974_2    2  ENSG00000102974      CTCF   \n",
       "86606  ENSG00000102974_1    1  ENSG00000102974      CTCF   \n",
       "86630  ENSG00000102974_1    1  ENSG00000102974      CTCF   \n",
       "68664  ENSG00000102974_1    1  ENSG00000102974      CTCF   \n",
       "69233  ENSG00000102974_1    1  ENSG00000102974      CTCF   \n",
       "73976  ENSG00000102974_2    2  ENSG00000102974      CTCF   \n",
       "74007  ENSG00000102974_1    1  ENSG00000102974      CTCF   \n",
       "74032  ENSG00000102974_1    1  ENSG00000102974      CTCF   \n",
       "86502  ENSG00000102974_1    1  ENSG00000102974      CTCF   \n",
       "73902  ENSG00000102974_1    1  ENSG00000102974      CTCF   \n",
       "86596  ENSG00000102974_1    1  ENSG00000102974      CTCF   \n",
       "73997  ENSG00000102974_1    1  ENSG00000102974      CTCF   \n",
       "86579  ENSG00000102974_1    1  ENSG00000102974      CTCF   \n",
       "73979  ENSG00000102974_1    1  ENSG00000102974      CTCF   \n",
       "\n",
       "      original_transcript_name transcript_triplet             transcript_id  \\\n",
       "81328                 CTCF-201            [1,1,1]    ENSG00000102974[1,1,1]   \n",
       "81886                 CTCF-202            [1,7,1]    ENSG00000102974[1,7,1]   \n",
       "86576                 CTCF-210            [1,9,2]    ENSG00000102974[1,9,2]   \n",
       "86606                 CTCF-215            [1,2,1]    ENSG00000102974[1,2,1]   \n",
       "86630                 CTCF-217            [1,6,1]    ENSG00000102974[1,6,1]   \n",
       "68664                 CTCF-201            [1,1,1]    ENSG00000102974[1,1,1]   \n",
       "69233                 CTCF-202            [1,7,1]    ENSG00000102974[1,7,1]   \n",
       "73976                 CTCF-210            [1,9,2]    ENSG00000102974[1,9,2]   \n",
       "74007                 CTCF-215            [1,2,1]    ENSG00000102974[1,2,1]   \n",
       "74032                 CTCF-217            [1,6,1]    ENSG00000102974[1,6,1]   \n",
       "86502                 CTCF-204          [10,17,1]  ENSG00000102974[10,17,1]   \n",
       "73902                 CTCF-204          [10,17,1]  ENSG00000102974[10,17,1]   \n",
       "86596                 CTCF-214            [3,5,1]    ENSG00000102974[3,5,1]   \n",
       "73997                 CTCF-214            [3,5,1]    ENSG00000102974[3,5,1]   \n",
       "86579                 CTCF-211           [9,16,1]   ENSG00000102974[9,16,1]   \n",
       "73979                 CTCF-211           [9,16,1]   ENSG00000102974[9,16,1]   \n",
       "\n",
       "      transcript_name tss_first_sd_issue tes_last_sa_issue source  \n",
       "81328     CTCF[1,1,1]              False             False    v40  \n",
       "81886     CTCF[1,7,1]              False             False    v40  \n",
       "86576     CTCF[1,9,2]              False             False    v40  \n",
       "86606     CTCF[1,2,1]              False             False    v40  \n",
       "86630     CTCF[1,6,1]              False             False    v40  \n",
       "68664     CTCF[1,1,1]              False             False    v29  \n",
       "69233     CTCF[1,7,1]              False             False    v29  \n",
       "73976     CTCF[1,9,2]              False             False    v29  \n",
       "74007     CTCF[1,2,1]              False             False    v29  \n",
       "74032     CTCF[1,6,1]              False             False    v29  \n",
       "86502   CTCF[10,17,1]              False             False    v40  \n",
       "73902   CTCF[10,17,1]              False             False    v29  \n",
       "86596     CTCF[3,5,1]              False             False    v40  \n",
       "73997     CTCF[3,5,1]              False             False    v29  \n",
       "86579    CTCF[9,16,1]              False             False    v40  \n",
       "73979    CTCF[9,16,1]              False             False    v29  "
      ]
     },
     "execution_count": 82,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ca.t_map.loc[(ca.t_map.tss_id.isin(tsss))&(ca.t_map.source != 'lapa')].sort_values(by='tss_id')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ba170259-585e-48b1-93af-d24460136cd2",
   "metadata": {},
   "source": [
    "## Get GTF for only the major isoforms for a given gene"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "id": "85c54c88-65b8-4288-bd3e-2d352625d9b0",
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_gene_major_iso_gtf(gene, gtf, major_set, sample=None):\n",
    "    if not sample:\n",
    "        ofile = '{}_major_isos.gtf'.format(gene.lower())\n",
    "    elif sample:\n",
    "        ofile = '{}_{}_major_isos.gtf'.format(gene.lower(), sample)\n",
    "    df = pd.read_csv(major_set, sep='\\t')\n",
    "    df = df.loc[df.gname == gene]\n",
    "    if sample:\n",
    "        df = df.loc[df['sample'] == sample]\n",
    "    tids = df.tid.unique().tolist()\n",
    "    tid_str = \"'\"\n",
    "    print('Found {} isoforms'.format(len(tids)))\n",
    "    for i, tid in enumerate(tids):\n",
    "        if i == 0: \n",
    "            tid_str += tid\n",
    "        else:\n",
    "            tid_str += '|'+tid\n",
    "    tid_str += \"'\"\n",
    "    grep_cmd = 'grep -E {} {} > {}'.format(tid_str, gtf, ofile)\n",
    "    grep_cmd = grep_cmd.replace('[', '\\[')\n",
    "    grep_cmd = grep_cmd.replace(']', '\\]')\n",
    "    # print(grep_cmd)\n",
    "    # grep_cmd = grep_cmd.split()\n",
    "    # print(grep_cmd)\n",
    "    # subprocess.run(grep_cmd, check=True)\n",
    "    print(grep_cmd)\n",
    "    return tids, ofile"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "id": "61db60f2-8f73-49b0-9dc0-1a31ab8f186f",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Found 4 isoforms\n",
      "grep -E 'ENSG00000102974\\[1,1,1\\]|ENSG00000102974\\[3,5,1\\]|ENSG00000102974\\[10,17,1\\]|ENSG00000102974\\[9,16,1\\]' ../cerberus.gtf > ctcf_vessels_major_isos.gtf\n"
     ]
    }
   ],
   "source": [
    "# gene = 'CTCF'\n",
    "# sample = 'vessels'\n",
    "# tids, gtf = get_gene_major_iso_gtf(gene, gtf, major_set, sample=sample)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "id": "82f2212c-29b7-459a-8ffd-ea3865617f9e",
   "metadata": {},
   "outputs": [],
   "source": [
    "# gene = 'ELN'\n",
    "# sample = 'vessels'\n",
    "# get_gene_major_iso_gtf(gene, gtf, major_set, sample=sample)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "id": "6c4dcd54-edd0-4c73-92bf-3feb94b91c7e",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Read in graph from ../swan/swan.p\n"
     ]
    }
   ],
   "source": [
    "sg = swan.read(swan_file)\n",
    "df = swan.calc_tpm(sg.adata, obs_col='sample').sparse.to_dense()\n",
    "df = df.transpose()\n",
    "df = df.loc[tids, sample]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 254,
   "id": "3dccafd4-de3b-4884-8ce0-793b4a1438ce",
   "metadata": {},
   "outputs": [],
   "source": [
    "def gtf_to_bed12(sg, obs_col, obs_condition, tids, gtf, ofile):\n",
    "    \n",
    "    # get tpm df\n",
    "    tpm_df = swan.calc_tpm(sg.adata, obs_col=obs_col).sparse.to_dense()\n",
    "    tpm_df = tpm_df.transpose()\n",
    "    tpm_df = tpm_df.loc[tids, sample]\n",
    "    \n",
    "    # print(tpm_df)\n",
    "    \n",
    "    allids={}\n",
    "    transcript_re = r'transcript_id \\\"([\\w\\.]+\\[[0-9,]*\\])\\\"'\n",
    "    \n",
    "    # colormap definitio\n",
    "    light_shade = get_sector_colors()[0]['mixed']\n",
    "    dark_shade = get_sector_colors()[0]['simple']\n",
    "    cmap = mpl.colors.LinearSegmentedColormap.from_list('', [light_shade, dark_shade])  \n",
    "    \n",
    "    # plot a cheeky colorbar\n",
    "    fig, ax = plt.subplots(figsize=(1,3))\n",
    "    cb1 = mpl.colorbar.ColorbarBase(ax, cmap=cmap,\n",
    "                                    orientation='vertical')\n",
    "    plt.savefig('figures/exp_colorbar.pdf', dpi=300)\n",
    "\n",
    "    with open(ofile, 'w') as out:\n",
    "        def make_entry(estart,eend,field,nline):\n",
    "            try:\n",
    "                estp=estart[0]-1\n",
    "                eedp=eend[-1]\n",
    "                # use regular expression to get transcript_id, gene_id and expression level\n",
    "                geneid=re.findall(r'gene_id \\\"([\\w\\.]+)\\\"',field[8])\n",
    "                transid=re.findall(transcript_re,field[8])\n",
    "\n",
    "                fpkmval=re.findall(r'FPKM \\\"([\\d\\.]+)\\\"',field[8])\n",
    "                if len(geneid)==0:\n",
    "                    print('Warning: no gene_id field ',file=sys.stderr)\n",
    "                else:\n",
    "                    geneid=geneid[0]\n",
    "                if len(transid)==0:\n",
    "                    print(field, file=sys.stderr)\n",
    "                    print('Warning: no transcript_id field',file=sys.stderr)\n",
    "                    transid='Trans_'+str(nline)\n",
    "                else:\n",
    "                    transid=transid[0]\n",
    "                if transid in allids.keys():\n",
    "                    transid2=transid+'_DUP'+str(allids[transid])\n",
    "                    allids[transid]=allids[transid]+1\n",
    "                    transid=transid2\n",
    "                else:\n",
    "                    allids[transid]=1\n",
    "                if len(fpkmval)==0:\n",
    "                    #print('Warning: no FPKM field',file=sys.stderr)\n",
    "                    fpkmval='100'\n",
    "                else:\n",
    "                    fpkmval=fpkmval[0]\n",
    "                fpkmint=round(float(fpkmval))\n",
    "                \n",
    "                # color by TPM\n",
    "                tpm = tpm_df[transid]\n",
    "                norm_val = (tpm-tpm_df.min())/(tpm_df.max()-tpm_df.min())\n",
    "                color = cmap(norm_val)\n",
    "                r = color[0]*255\n",
    "                g = color[1]*255\n",
    "                b = color[2]*255\n",
    "                color = '{},{},{}'.format(r,g,b)\n",
    "                \n",
    "                # get relative positions of exon starts lengths \n",
    "                # of each exon\n",
    "                seglen=[eend[i]-estart[i]+1 for i in range(len(estart))]\n",
    "                segstart=[estart[i]-estart[0] for i in range(len(estart))]\n",
    "                strl=str(seglen[0])\n",
    "                for i in range(1,len(seglen)):\n",
    "                    strl+=','+str(seglen[i])\n",
    "                strs=str(segstart[0])\n",
    "                for i in range(1,len(segstart)):\n",
    "                    strs+=','+str(segstart[i])\n",
    "                \n",
    "                # entry\n",
    "                entry = {'Chromosome': field[0],\n",
    "                         'Start': estp,\n",
    "                         'End': eedp,\n",
    "                         'Name': transid,\n",
    "                         'Score': tpm,\n",
    "                         'Strand': field[6],\n",
    "                         'ThickStart': estp,\n",
    "                         'ThickEnd': eedp,\n",
    "                         'ItemRGB': color,\n",
    "                         'BlockCount': len(estart),\n",
    "                         'BlockSizes': strl,\n",
    "                         'BlockStarts': strs}\n",
    "                \n",
    "                # print(entry['Name'])\n",
    "                # print(entry['Score'])\n",
    "                # print(entry['ItemRGB'])\n",
    "                # print()\n",
    "                entry = pd.DataFrame.from_dict(entry, orient='index').transpose()\n",
    "            except ValueError:\n",
    "                print('Error: non-number fields at line '+str(nline),file=sys.stderr)\n",
    "                \n",
    "            return entry\n",
    "\n",
    "        estart=[]\n",
    "        eend=[]\n",
    "        # read lines one to one\n",
    "        nline=0\n",
    "        prevfield=[]\n",
    "        prevtransid=''\n",
    "        df = pd.DataFrame()\n",
    "        for lines in open(gtf):\n",
    "            field=lines.strip().split('\\t')\n",
    "            nline=nline+1\n",
    "            if len(field)<9:\n",
    "                print('Error: the GTF should has at least 9 fields at line '+str(nline),file=sys.stderr)\n",
    "                continue\n",
    "            if field[1]!='Cufflinks':\n",
    "                pass\n",
    "                #print('Warning: the second field is expected to be \\'Cufflinks\\' at line '+str(nline),file=sys.stderr)\n",
    "            if field[2]!='exon' and field[2] !='transcript':\n",
    "                #print('Error: the third filed is expected to be \\'exon\\' or \\'transcript\\' at line '+str(nline),file=sys.stderr)\n",
    "                continue\n",
    "            transid=re.findall(transcript_re,field[8])\n",
    "            if len(transid)>0:\n",
    "                transid=transid[0]\n",
    "            else:\n",
    "                transid=''\n",
    "            if field[2]=='transcript' or (prevtransid != '' and transid!='' and transid != prevtransid):\n",
    "                #print('prev:'+prevtransid+', current:'+transid)\n",
    "                # A new transcript record, write\n",
    "                if len(estart)!=0:\n",
    "                    entry = make_entry(estart,eend,prevfield,nline)\n",
    "                    df = pd.concat([df, entry], axis=0)\n",
    "                    \n",
    "                estart=[]\n",
    "                eend=[]\n",
    "            prevfield=field\n",
    "            prevtransid=transid\n",
    "            if field[2]=='exon':\n",
    "                try:    \n",
    "                    est=int(field[3])\n",
    "                    eed=int(field[4])\n",
    "                    estart+=[est]\n",
    "                    eend+=[eed]\n",
    "                except ValueError:\n",
    "                    print('Error: non-number fields at line '+str(nline),file=sys.stderr)\n",
    "                    \n",
    "        # the last record\n",
    "        if len(estart)!=0:\n",
    "            entry = make_entry(estart,eend,field,nline)\n",
    "        \n",
    "        # concat entry w/ df \n",
    "        df = pd.concat([df, entry], axis=0)\n",
    "        \n",
    "        # sort by TPM?\n",
    "        df.Score = df.Score.astype(float)\n",
    "        df = df.sort_values(by='Score', ascending=False)\n",
    "        # print()\n",
    "        # print(df)\n",
    "        \n",
    "        # write header line and save\n",
    "        t_name = 'Major isoforms of {}'.format(obs_condition)\n",
    "        header = 'track name=\"{}\" visibility=2 itemRgb=\"On\"\\n'.format(t_name)\n",
    "        with open(ofile, 'w') as o:\n",
    "            o.write(header)\n",
    "        df.to_csv(ofile, mode='a', sep='\\t', index=False, header=False)\n",
    "        \n",
    "        return df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 255,
   "id": "7c03f6ae-7d62-45c6-bd44-4cd38b03d05c",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Found 4 isoforms\n",
      "grep -E 'ENSG00000102974\\[1,1,1\\]|ENSG00000102974\\[3,5,1\\]|ENSG00000102974\\[10,17,1\\]|ENSG00000102974\\[9,16,1\\]' ctcf_vessels_major_isos.gtf > ctcf_vessels_major_isos.gtf\n"
     ]
    }
   ],
   "source": [
    "gene = 'CTCF'\n",
    "obs_condition = 'vessels'\n",
    "tids, gtf = get_gene_major_iso_gtf(gene, gtf, major_set, sample=obs_condition)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 256,
   "id": "15e51656-109d-4907-855a-245ff6441def",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAF0AAAC7CAYAAAAHbJJcAAAABHNCSVQICAgIfAhkiAAAAAlwSFlzAAALEgAACxIB0t1+/AAAADh0RVh0U29mdHdhcmUAbWF0cGxvdGxpYiB2ZXJzaW9uMy4yLjIsIGh0dHA6Ly9tYXRwbG90bGliLm9yZy+WH4yJAAAGk0lEQVR4nO2dX4ilZR3HP991lJSkhbZEtgKJra4STFsvDP9ENTsQSxAURtGiRJBd15VeeKN4o5K6LCLLQuRNUlNsSTehYEsW1NYqxqCwbSvIligoEuP+vHjPynGceZ53nMUvzvv9wIE5533P78x85uH3/J5zfs97VFWE95cd7l9gikS6gUg3EOkGIt1ApBuI9AaSHpH0kqR/bnBcku6XtCLpuKSrxsSN9DaHgcXG8X3AntntB8BDY4JGeoOqegL4X+OU/cCRGjgG7JR0eS9upG+N3cC/5+6fmj3WZKF1UNJ2f4/gBPDG3P1DVXVoE8/XOo91nTWlA0jrxd0eVNUbVXX1FkKcAj45d/8TwOnek6YufashloHbJD0K7AVeqaoXe0/qSt+xY/um/bNnzzaPS/oFcAOwS9Ip4A7gQoCqOggcBZaAFeB14MCY11Xrvy2pFha6/5cPLKurq3/dYnp5T/Qm0m090l3EqIFJ53QXk65eXHRzeqSffzLSDUS6gUykBmLUQBZHBpLTDUS6gUg3kMWRgZSMBmLUQEa6gRg1kMWRgZSMBiLdQKQbyOLIQEpGAzHaQdKipOdmjf8/Xef4RyT9RtLfJZ2Q1O3yykhvIOkC4AHgKwzNok9LWq6qZ+ZO+xHwTFV9XdLHgOck/byq/r9R3OT0Nl8EVqrqeYBZo+h+YF56AZdqEPVhhk0Eq62gU69edkn6y9z9tf3p6zX9710T42cM3bungUuBb1VVszN16tLPdBpIxzT9fw34G3AT8GngD5KerKpXNwo6dek9xjT9HwDuqqH9eUXSC8DngD9vFDTS2zwN7JF0BfAf4NvAzWvOOQl8GXhS0mXAZ4HnW0EzkTaoqlVJtwGPAxcAj1TVCUk/nB0/CNwJHJb0D4Z09JOqOtOKm5KxQ1UdZdhxMf/YwbmfTwNf3UzMpBcDkW4gOd1ARrqBSDcQ6QbywbSBjHQDkW4g0g1EuoEsjgxkpBuIdAN5a9dAcrqBpBcDkW4g0g1EuoFMpAZSMhqIUQMZ6R0kLQL3MTQbPVxVd61zzg3AvQxXJz1TVde3YianNxjTny5pJ/AgsFhVJyV9vBc31UubMf3pNwOPVdVJgKp6qRc00tuM6U//DHChpD8y9KffV1VHWkGnLr23KWBMf/oC8AWGzt2LgT9JOlZV/9roRafeDdDbFDCmP/3ULM5rwGuSngCuBDaUvq2Nngfe7k+XdBFDf/rymnN+DXxJ0oKkSxjSz7OtoCkZG4zpT6+qZyX9HjgOnGUoK9f9XqRzTD2nd+n1p8/u3wPcMzZmpBvI4shARrqBTKQGYtTA1BdHFmLUQHK6gVQvBlKnG8hINxDpBjKRGohRA1kcGYhRA8npBlK9GMjiyEBGuoFIN5CJ1ECMGsjiyEDX6LkKZjvexqDORevnzrtG0puSvtmLmYm0wZhNAXPn3c3Qftcl0tuM2RQA8GPgl8A1Y4JmcdSmuylA0m7gGwzXT9+6dNj2JeP52BRwL8OVpN8cO0C70rc552NTwNXAozPhu4AlSatV9auNgk59pPfoXrS+qq4497Okw8BvW8IhI73JyIvWb5osjjqM2RQw9/j3x8RMyWgg0g1EuoEsjgykZDQQowYy0g0kpxtI9WIg0g1EuoFIN5CJ1EBKRgNJLwYi3UByuoGMdAORbiDSDeSDaQMZ6QYi3UCkG8imgP7f39wUIOk7ko7Pbk9JurIXM4ujBiM3BbwAXF9VL0vaBxzi3ddYfwdJL226mwKq6qm5848xdPY2mbr0Xn/6mG8KmOcW4He9F536W7u9/vQxmwKGE6UbGaRf13vR5PQ2YzYFIOnzwMPAvqr6by/o1NNLj+6mAEmfAh4Dvtv6Hox5Ir3ByE0BtwMfBR6cuVrtpKxI79HbFFBVtwK3biZmpBvIRGpg6iWjhRg1kJFuIDndQKoXA5FuININpBvAQIwaSMloIEYNJKcbSPViININZCI1EKMGMpEaiFEDyekGUr0YyIcYBtKf3v/7e/3pknT/7PhxSVf1Yia9NBjZn74P2DO77QUeYqv96ROfSMdctH4/cKSqCjgmaaeky6vqxY2CTtroCNbrT9/9Hs55B82RvrKy8vjS0tKuTfySHzQ+1NkUMKY/fXQP+zma0qtqsXV8AozpTx/Vwz5P0kubt/vTJV3E0J++vOacZeB7syrmWuCVVj6HXLS+ycj+9KPAErACvA4c6MXVMOmG95OkFwORbiDSDUS6gUg3EOkGIt1ApBt4C1itKaqMuiouAAAAAElFTkSuQmCC\n",
      "text/plain": [
       "<Figure size 72x216 with 1 Axes>"
      ]
     },
     "metadata": {
      "needs_background": "light"
     },
     "output_type": "display_data"
    }
   ],
   "source": [
    "ofile = '{}_{}_major_isos.bed'.format(gene.lower(), obs_condition)\n",
    "df = gtf_to_bed12(sg, 'sample', obs_condition, tids, gtf, ofile)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 258,
   "id": "3d554c78-5904-4dca-a2e3-7545feccc2c4",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>source</th>\n",
       "      <th>gid</th>\n",
       "      <th>n_tss</th>\n",
       "      <th>n_tes</th>\n",
       "      <th>n_ic</th>\n",
       "      <th>n_iso</th>\n",
       "      <th>splicing_ratio</th>\n",
       "      <th>gname</th>\n",
       "      <th>sample</th>\n",
       "      <th>gene_tpm</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>210041</th>\n",
       "      <td>sample_major</td>\n",
       "      <td>ENSG00000102974</td>\n",
       "      <td>4.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>4.0</td>\n",
       "      <td>4.0</td>\n",
       "      <td>1.6</td>\n",
       "      <td>CTCF</td>\n",
       "      <td>vessels</td>\n",
       "      <td>19.757177</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "              source              gid  n_tss  n_tes  n_ic  n_iso  \\\n",
       "210041  sample_major  ENSG00000102974    4.0    1.0   4.0    4.0   \n",
       "\n",
       "        splicing_ratio gname   sample   gene_tpm  \n",
       "210041             1.6  CTCF  vessels  19.757177  "
      ]
     },
     "execution_count": 258,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ca.triplets.loc[(ca.triplets.source=='sample_major')&(ca.triplets['sample']==obs_condition)&(ca.triplets.gname==gene)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 235,
   "id": "9d394383-f825-4c52-91be-4a1a29f64905",
   "metadata": {},
   "outputs": [],
   "source": [
    "# get bed file of all used CTCF promoters\n",
    "df = df[['Name']] \n",
    "df = add_feat(df, kind='tss', col='Name')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 213,
   "id": "5383334a-6518-4edb-ae66-3f46badcdcb2",
   "metadata": {},
   "outputs": [],
   "source": [
    "ca = cerberus.read(c_annot)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 236,
   "id": "96abe88b-a168-47d2-9f7d-3871ac8a8677",
   "metadata": {},
   "outputs": [],
   "source": [
    "tss = ca.tss.loc[ca.tss.Name.isin(df.tss.unique().tolist())]\n",
    "tss = tss[['Chromosome', 'Start', 'End', 'Strand', 'Name']]\n",
    "tss['Score'] = 0\n",
    "tss = tss[['Chromosome', 'Start', 'End', 'Name', 'Score', 'Strand']]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 237,
   "id": "f5e80e57-fdb2-4783-b689-31465e8fd800",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "ctcf_vessels_tss.bed\n"
     ]
    }
   ],
   "source": [
    "ofile = '{}_{}_tss.bed'.format(gene.lower(), obs_condition)\n",
    "print(ofile)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 243,
   "id": "cfe03700-03cb-4734-a0ad-8520499e178e",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "86,180,233\n"
     ]
    }
   ],
   "source": [
    "# write header line and save\n",
    "color = get_feat_colors()[0]['tss']\n",
    "color = color[1:] # remove hash\n",
    "r,g,b = tuple(int(color[i:i+2], 16) for i in (0, 2, 4))\n",
    "color = '{},{},{}'.format(r,g,b)\n",
    "print(color)\n",
    "\n",
    "t_name = 'TSSs from major isoforms of {}'.format(obs_condition)\n",
    "header = 'track name=\"{}\" visibility=2 color={}\\n'.format(t_name, color)\n",
    "with open(ofile, 'w') as o:\n",
    "    o.write(header)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 244,
   "id": "bf21babb-3a81-4e97-b2f6-003351b657c8",
   "metadata": {},
   "outputs": [],
   "source": [
    "tss.to_csv(ofile, sep='\\t', index=False, header=False, mode='a')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 246,
   "id": "2558e044-62da-486c-bf92-b77ad9d8aa85",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>source</th>\n",
       "      <th>gid</th>\n",
       "      <th>n_tss</th>\n",
       "      <th>n_tes</th>\n",
       "      <th>n_ic</th>\n",
       "      <th>n_iso</th>\n",
       "      <th>splicing_ratio</th>\n",
       "      <th>gname</th>\n",
       "      <th>sample</th>\n",
       "      <th>gene_tpm</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>208438</th>\n",
       "      <td>sample_major</td>\n",
       "      <td>ENSG00000049540</td>\n",
       "      <td>3.0</td>\n",
       "      <td>6.0</td>\n",
       "      <td>26.0</td>\n",
       "      <td>37.0</td>\n",
       "      <td>5.777778</td>\n",
       "      <td>ELN</td>\n",
       "      <td>vessels</td>\n",
       "      <td>427.713043</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "              source              gid  n_tss  n_tes  n_ic  n_iso  \\\n",
       "208438  sample_major  ENSG00000049540    3.0    6.0  26.0   37.0   \n",
       "\n",
       "        splicing_ratio gname   sample    gene_tpm  \n",
       "208438        5.777778   ELN  vessels  427.713043  "
      ]
     },
     "execution_count": 246,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ca.triplets.loc[(ca.triplets.gname=='ELN')&(ca.triplets.source=='sample_major')&(ca.triplets['sample']=='vessels')]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 247,
   "id": "4e3f3f2c-ed49-4011-98e4-e22e1a1cb0ca",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>source</th>\n",
       "      <th>gid</th>\n",
       "      <th>n_tss</th>\n",
       "      <th>n_tes</th>\n",
       "      <th>n_ic</th>\n",
       "      <th>n_iso</th>\n",
       "      <th>splicing_ratio</th>\n",
       "      <th>gname</th>\n",
       "      <th>sample</th>\n",
       "      <th>gene_tpm</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>210041</th>\n",
       "      <td>sample_major</td>\n",
       "      <td>ENSG00000102974</td>\n",
       "      <td>4.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>4.0</td>\n",
       "      <td>4.0</td>\n",
       "      <td>1.6</td>\n",
       "      <td>CTCF</td>\n",
       "      <td>vessels</td>\n",
       "      <td>19.757177</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "              source              gid  n_tss  n_tes  n_ic  n_iso  \\\n",
       "210041  sample_major  ENSG00000102974    4.0    1.0   4.0    4.0   \n",
       "\n",
       "        splicing_ratio gname   sample   gene_tpm  \n",
       "210041             1.6  CTCF  vessels  19.757177  "
      ]
     },
     "execution_count": 247,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ca.triplets.loc[(ca.triplets.gname=='CTCF')&(ca.triplets.source=='sample_major')&(ca.triplets['sample']=='vessels')]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "25e79563-6ace-4ee6-86d3-6bba81370fa8",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 127,
   "id": "9b439b0b-3773-448c-ae69-c2808832413b",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Found 37 isoforms\n",
      "grep -E 'ENSG00000049540\\[1,2,3\\]|ENSG00000049540\\[1,76,3\\]|ENSG00000049540\\[1,1,3\\]|ENSG00000049540\\[1,8,3\\]|ENSG00000049540\\[1,76,1\\]|ENSG00000049540\\[1,2,1\\]|ENSG00000049540\\[1,1,1\\]|ENSG00000049540\\[1,8,1\\]|ENSG00000049540\\[1,147,3\\]|ENSG00000049540\\[1,14,3\\]|ENSG00000049540\\[1,149,3\\]|ENSG00000049540\\[1,108,3\\]|ENSG00000049540\\[1,3,3\\]|ENSG00000049540\\[1,98,1\\]|ENSG00000049540\\[1,135,3\\]|ENSG00000049540\\[1,107,3\\]|ENSG00000049540\\[1,112,3\\]|ENSG00000049540\\[6,33,13\\]|ENSG00000049540\\[1,136,1\\]|ENSG00000049540\\[1,147,1\\]|ENSG00000049540\\[1,18,7\\]|ENSG00000049540\\[1,96,3\\]|ENSG00000049540\\[1,111,3\\]|ENSG00000049540\\[1,108,1\\]|ENSG00000049540\\[1,36,3\\]|ENSG00000049540\\[1,127,3\\]|ENSG00000049540\\[1,111,1\\]|ENSG00000049540\\[1,136,3\\]|ENSG00000049540\\[1,103,3\\]|ENSG00000049540\\[1,97,1\\]|ENSG00000049540\\[1,96,1\\]|ENSG00000049540\\[1,77,1\\]|ENSG00000049540\\[1,59,3\\]|ENSG00000049540\\[1,4,3\\]|ENSG00000049540\\[2,2,2\\]|ENSG00000049540\\[1,97,3\\]|ENSG00000049540\\[1,17,6\\]' ctcf_vessels_major_isos.gtf > eln_vessels_major_isos.gtf\n"
     ]
    }
   ],
   "source": [
    "gene = 'ELN'\n",
    "sample = 'vessels'\n",
    "tids, gtf = get_gene_major_iso_gtf(gene, gtf, major_set, sample=sample)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 128,
   "id": "1f21ca21-bf8d-4472-a422-d591351b48d3",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "  Chromosome     Start       End                      Name      Score Strand  \\\n",
      "0       chr7  74028050  74069728    ENSG00000049540[1,2,3]  60.760303      +   \n",
      "0       chr7  74028050  74069728    ENSG00000049540[1,1,3]  43.177643      +   \n",
      "0       chr7  74028050  74069728   ENSG00000049540[1,76,3]  39.778652      +   \n",
      "0       chr7  74028050  74069728    ENSG00000049540[1,8,3]  35.264694      +   \n",
      "0       chr7  74028050  74069957    ENSG00000049540[1,2,1]  23.386642      +   \n",
      "0       chr7  74028050  74069957    ENSG00000049540[1,1,1]  19.494343      +   \n",
      "0       chr7  74028050  74069957   ENSG00000049540[1,76,1]  17.654505      +   \n",
      "0       chr7  74028050  74069957    ENSG00000049540[1,8,1]  11.774277      +   \n",
      "0       chr7  74028050  74069728   ENSG00000049540[1,14,3]   6.237474      +   \n",
      "0       chr7  74028050  74069728  ENSG00000049540[1,108,3]   6.014239      +   \n",
      "0       chr7  74028050  74069728  ENSG00000049540[1,147,3]   4.104070      +   \n",
      "0       chr7  74028050  74069728  ENSG00000049540[1,107,3]   4.072419      +   \n",
      "0       chr7  74028050  74069957   ENSG00000049540[1,98,1]   3.622576      +   \n",
      "0       chr7  74027738  74068813    ENSG00000049540[2,2,2]   3.017969      +   \n",
      "0       chr7  74028050  74069728  ENSG00000049540[1,111,3]   2.904807      +   \n",
      "0       chr7  74028050  74069957  ENSG00000049540[1,147,1]   2.904807      +   \n",
      "0       chr7  74028050  74069728  ENSG00000049540[1,103,3]   2.703271      +   \n",
      "0       chr7  74028050  74069728  ENSG00000049540[1,149,3]   2.656784      +   \n",
      "0       chr7  74028050  74069728   ENSG00000049540[1,96,3]   2.646832      +   \n",
      "0       chr7  74028050  74069957  ENSG00000049540[1,108,1]   2.624850      +   \n",
      "0       chr7  74028050  74069957  ENSG00000049540[1,136,1]   2.423597      +   \n",
      "0       chr7  74028050  74069728  ENSG00000049540[1,135,3]   2.310435      +   \n",
      "0       chr7  74028050  74069728  ENSG00000049540[1,112,3]   2.253712      +   \n",
      "0       chr7  74028050  74069728    ENSG00000049540[1,3,3]   2.232014      +   \n",
      "0       chr7  74028050  74069957  ENSG00000049540[1,111,1]   2.108616      +   \n",
      "0       chr7  74028050  74069728    ENSG00000049540[1,4,3]   2.108616      +   \n",
      "0       chr7  74028050  74069728   ENSG00000049540[1,59,3]   1.850641      +   \n",
      "0       chr7  74028050  74069728   ENSG00000049540[1,36,3]   1.828942      +   \n",
      "0       chr7  74028050  74069957   ENSG00000049540[1,97,1]   1.793918      +   \n",
      "0       chr7  74028050  74043334   ENSG00000049540[1,18,7]   1.772219      +   \n",
      "0       chr7  74046608  74052286  ENSG00000049540[6,33,13]   1.772219      +   \n",
      "0       chr7  74028050  74069728  ENSG00000049540[1,136,3]   1.570683      +   \n",
      "0       chr7  74028050  74069957   ENSG00000049540[1,96,1]   1.570683      +   \n",
      "0       chr7  74028050  74069728   ENSG00000049540[1,97,3]   1.513961      +   \n",
      "0       chr7  74028050  74069728  ENSG00000049540[1,127,3]   1.199263      +   \n",
      "0       chr7  74028050  74069957   ENSG00000049540[1,77,1]   0.604608      +   \n",
      "0       chr7  74028050  74042891   ENSG00000049540[1,17,6]   0.604608      +   \n",
      "\n",
      "  ThickStart  ThickEnd                                            ItemRGB  \\\n",
      "0   74028050  74069728                                        0.0,0.0,0.0   \n",
      "0   74028050  74069728  53.10588235294118,53.10588235294118,53.1058823...   \n",
      "0   74028050  74069728  63.870588235294115,63.870588235294115,63.87058...   \n",
      "0   74028050  74069728  77.50588235294119,77.50588235294119,77.5058823...   \n",
      "0   74028050  74069957  114.10588235294117,114.10588235294117,114.1058...   \n",
      "0   74028050  74069957  125.58823529411765,125.58823529411765,125.5882...   \n",
      "0   74028050  74069957  131.3294117647059,131.3294117647059,131.329411...   \n",
      "0   74028050  74069957  149.27058823529413,149.27058823529413,149.2705...   \n",
      "0   74028050  74069728  166.49411764705883,166.49411764705883,166.4941...   \n",
      "0   74028050  74069728  166.49411764705883,166.49411764705883,166.4941...   \n",
      "0   74028050  74069728  172.9529411764706,172.9529411764706,172.952941...   \n",
      "0   74028050  74069728  172.9529411764706,172.9529411764706,172.952941...   \n",
      "0   74028050  74069957  174.38823529411764,174.38823529411764,174.3882...   \n",
      "0   74027738  74068813  175.8235294117647,175.8235294117647,175.823529...   \n",
      "0   74028050  74069728  176.54117647058823,176.54117647058823,176.5411...   \n",
      "0   74028050  74069957  176.54117647058823,176.54117647058823,176.5411...   \n",
      "0   74028050  74069728  177.25882352941176,177.25882352941176,177.2588...   \n",
      "0   74028050  74069728  177.25882352941176,177.25882352941176,177.2588...   \n",
      "0   74028050  74069728  177.25882352941176,177.25882352941176,177.2588...   \n",
      "0   74028050  74069957  177.25882352941176,177.25882352941176,177.2588...   \n",
      "0   74028050  74069957  177.9764705882353,177.9764705882353,177.976470...   \n",
      "0   74028050  74069728  177.9764705882353,177.9764705882353,177.976470...   \n",
      "0   74028050  74069728  177.9764705882353,177.9764705882353,177.976470...   \n",
      "0   74028050  74069728  178.69411764705885,178.69411764705885,178.6941...   \n",
      "0   74028050  74069957  178.69411764705885,178.69411764705885,178.6941...   \n",
      "0   74028050  74069728  178.69411764705885,178.69411764705885,178.6941...   \n",
      "0   74028050  74069728  179.41176470588238,179.41176470588238,179.4117...   \n",
      "0   74028050  74069728  179.41176470588238,179.41176470588238,179.4117...   \n",
      "0   74028050  74069957  179.41176470588238,179.41176470588238,179.4117...   \n",
      "0   74028050  74043334  180.12941176470588,180.12941176470588,180.1294...   \n",
      "0   74046608  74052286  180.12941176470588,180.12941176470588,180.1294...   \n",
      "0   74028050  74069728  180.12941176470588,180.12941176470588,180.1294...   \n",
      "0   74028050  74069957  180.12941176470588,180.12941176470588,180.1294...   \n",
      "0   74028050  74069728  180.8470588235294,180.8470588235294,180.847058...   \n",
      "0   74028050  74069728  181.56470588235294,181.56470588235294,181.5647...   \n",
      "0   74028050  74069957                                  183.0,183.0,183.0   \n",
      "0   74028050  74042891                                  183.0,183.0,183.0   \n",
      "\n",
      "  BlockCount                                         BlockSizes  \\\n",
      "0         32  219,51,30,33,36,93,51,51,42,72,30,72,42,60,54,...   \n",
      "0         33  219,51,30,33,36,93,51,51,42,72,30,72,42,60,54,...   \n",
      "0         31  219,51,30,33,36,93,51,51,42,72,30,72,42,60,54,...   \n",
      "0         32  219,51,30,33,36,93,51,51,42,72,30,72,42,60,54,...   \n",
      "0         32  219,51,30,33,36,93,51,51,42,72,30,72,42,60,54,...   \n",
      "0         33  219,51,30,33,36,93,51,51,42,72,30,72,42,60,54,...   \n",
      "0         31  219,51,30,33,36,93,51,51,42,72,30,72,42,60,54,...   \n",
      "0         32  219,51,30,33,36,93,51,51,42,72,30,72,42,60,54,...   \n",
      "0         30  219,51,33,36,93,51,51,42,72,30,72,42,60,54,90,...   \n",
      "0         31  219,51,30,33,36,93,51,51,42,72,30,72,60,54,90,...   \n",
      "0         31  219,51,33,36,93,51,51,42,72,30,72,42,60,54,90,...   \n",
      "0         32  219,51,30,33,36,93,51,51,42,72,30,72,60,54,90,...   \n",
      "0         30  219,51,30,33,36,93,51,51,42,72,30,72,42,60,54,...   \n",
      "0         32  531,51,30,33,36,93,51,51,42,72,30,72,42,60,54,...   \n",
      "0         31  219,51,30,33,36,93,51,51,42,72,30,72,60,54,90,...   \n",
      "0         31  219,51,33,36,93,51,51,42,72,30,72,42,60,54,90,...   \n",
      "0         32  219,51,30,33,36,93,51,51,42,72,30,72,42,60,54,...   \n",
      "0         31  219,51,33,36,93,51,51,42,72,30,72,42,60,54,90,...   \n",
      "0         31  219,51,30,33,36,93,51,51,42,72,30,72,42,60,54,...   \n",
      "0         31  219,51,30,33,36,93,51,51,42,72,30,72,60,54,90,...   \n",
      "0         31  219,51,30,33,93,51,51,42,72,30,72,42,60,54,90,...   \n",
      "0         32  219,51,30,33,93,51,51,42,72,30,72,42,60,54,90,...   \n",
      "0         30  219,51,30,33,36,93,51,51,42,72,30,72,60,54,90,...   \n",
      "0         32  219,51,30,33,36,93,51,66,42,72,30,72,42,60,54,...   \n",
      "0         31  219,51,30,33,36,93,51,51,42,72,30,72,60,54,90,...   \n",
      "0         31  219,51,30,33,36,93,51,66,42,72,30,72,42,60,54,...   \n",
      "0         30  219,51,30,33,36,93,51,51,42,72,30,72,42,60,54,...   \n",
      "0         33  219,51,30,33,36,93,51,66,42,72,30,72,42,60,54,...   \n",
      "0         31  219,51,30,33,36,93,51,51,42,72,30,72,42,60,54,...   \n",
      "0          7                             219,51,30,33,36,93,351   \n",
      "0          5                                   159,42,60,54,537   \n",
      "0         31  219,51,30,33,93,51,51,42,72,30,72,42,60,54,90,...   \n",
      "0         31  219,51,30,33,36,93,51,51,42,72,30,72,42,60,54,...   \n",
      "0         31  219,51,30,33,36,93,51,51,42,72,30,72,42,60,54,...   \n",
      "0         31  219,51,30,33,36,93,51,55,42,72,30,72,42,60,54,...   \n",
      "0         30  219,51,30,33,36,93,51,51,42,72,30,72,42,60,54,...   \n",
      "0          6                                219,51,30,33,36,278   \n",
      "\n",
      "                                         BlockStarts  \n",
      "0  0,7313,8504,9656,13165,14563,14933,15067,15828...  \n",
      "0  0,7313,8504,9656,13165,14563,14933,15067,15828...  \n",
      "0  0,7313,8504,9656,13165,14563,14933,15067,15828...  \n",
      "0  0,7313,8504,9656,13165,14563,14933,15067,15828...  \n",
      "0  0,7313,8504,9656,13165,14563,14933,15067,15828...  \n",
      "0  0,7313,8504,9656,13165,14563,14933,15067,15828...  \n",
      "0  0,7313,8504,9656,13165,14563,14933,15067,15828...  \n",
      "0  0,7313,8504,9656,13165,14563,14933,15067,15828...  \n",
      "0  0,7313,9656,13165,14563,14933,15067,15828,1717...  \n",
      "0  0,7313,8504,9656,13165,14563,14933,15067,15828...  \n",
      "0  0,7313,9656,13165,14563,14933,15067,15828,1717...  \n",
      "0  0,7313,8504,9656,13165,14563,14933,15067,15828...  \n",
      "0  0,7313,8504,9656,13165,14563,14933,15067,15828...  \n",
      "0  0,7625,8816,9968,13477,14875,15245,15379,16140...  \n",
      "0  0,7313,8504,9656,13165,14563,14933,15067,15828...  \n",
      "0  0,7313,9656,13165,14563,14933,15067,15828,1717...  \n",
      "0  0,7313,8504,9656,13165,14563,14933,15067,15828...  \n",
      "0  0,7313,9656,13165,14563,14933,15067,15828,1717...  \n",
      "0  0,7313,8504,9656,13165,14563,14933,15067,15828...  \n",
      "0  0,7313,8504,9656,13165,14563,14933,15067,15828...  \n",
      "0  0,7313,8504,9656,14563,14933,15067,15828,17171...  \n",
      "0  0,7313,8504,9656,14563,14933,15067,15828,17171...  \n",
      "0  0,7313,8504,9656,13165,14563,14933,15067,15828...  \n",
      "0  0,7313,8504,9656,13165,14563,14933,15052,15828...  \n",
      "0  0,7313,8504,9656,13165,14563,14933,15067,15828...  \n",
      "0  0,7313,8504,9656,13165,14563,14933,15052,15828...  \n",
      "0  0,7313,8504,9656,13165,14563,14933,15067,15828...  \n",
      "0  0,7313,8504,9656,13165,14563,14933,15052,15828...  \n",
      "0  0,7313,8504,9656,13165,14563,14933,15067,15828...  \n",
      "0                 0,7313,8504,9656,13165,14563,14933  \n",
      "0                              0,1066,1533,1894,5141  \n",
      "0  0,7313,8504,9656,14563,14933,15067,15828,17171...  \n",
      "0  0,7313,8504,9656,13165,14563,14933,15067,15828...  \n",
      "0  0,7313,8504,9656,13165,14563,14933,15067,15828...  \n",
      "0  0,7313,8504,9656,13165,14563,14933,15067,15828...  \n",
      "0  0,7313,8504,9656,13165,14563,14933,15067,15828...  \n",
      "0                       0,7313,8504,9656,13165,14563  \n"
     ]
    }
   ],
   "source": [
    "ofile = '{}_{}_major_isos.bed'.format(gene.lower(), sample)\n",
    "df = gtf_to_bed12(sg, 'sample', sample, tids, gtf, ofile)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a6bc1b3e-0c9d-4acb-b0c9-ecb4bb5539de",
   "metadata": {},
   "source": [
    "## Can I just use swan??"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 194,
   "id": "789ddec0-ab4e-4016-acb4-227ad94c83bd",
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_gene_major_isos(gene, gtf, major_set, sample=None):\n",
    "    if not sample:\n",
    "        ofile = '{}_major_isos.gtf'.format(gene.lower())\n",
    "    elif sample:\n",
    "        ofile = '{}_{}_major_isos.gtf'.format(gene.lower(), sample)\n",
    "    df = pd.read_csv(major_set, sep='\\t')\n",
    "    df = df.loc[df.gname == gene]\n",
    "    if sample:\n",
    "        df = df.loc[df['sample'] == sample]\n",
    "    tids = df.tid.unique().tolist()\n",
    "    return tids"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 199,
   "id": "02f78906-ddf2-4f2f-9c82-62b0e1d8865d",
   "metadata": {},
   "outputs": [],
   "source": [
    "gene = 'ELN'\n",
    "sample = 'vessels'\n",
    "obs_col = 'sample'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 200,
   "id": "13dacd39-71bf-4da9-9520-913516b59975",
   "metadata": {},
   "outputs": [],
   "source": [
    "df = pd.read_csv(major_set, sep='\\t')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 201,
   "id": "42a7b796-9d22-4767-a697-9f0734e21862",
   "metadata": {},
   "outputs": [],
   "source": [
    "df = df.loc[df.gname == gene]\n",
    "df = df.loc[df[obs_col] == sample]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 202,
   "id": "d2bd6767-716d-4ecb-bb20-7e20c19662c1",
   "metadata": {},
   "outputs": [],
   "source": [
    "# get tpm df\n",
    "tpm_df = swan.calc_tpm(sg.adata, obs_col=obs_col).sparse.to_dense()\n",
    "tpm_df = tpm_df.transpose()\n",
    "tpm_df = tpm_df.loc[tids, sample]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e4ec1b1b-451c-4f28-b988-5dadb8d135ea",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 193,
   "id": "c644e8e5-e8e4-47fc-8a49-844b02668587",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAxsAAACmCAYAAACyYpL/AAAABHNCSVQICAgIfAhkiAAAAAlwSFlzAAALEgAACxIB0t1+/AAAADh0RVh0U29mdHdhcmUAbWF0cGxvdGxpYiB2ZXJzaW9uMy4yLjIsIGh0dHA6Ly9tYXRwbG90bGliLm9yZy+WH4yJAAAFtklEQVR4nO3dMW4cZQCG4d8rEAghQQf3QPRA5YLS4ircCfkArkDchpRpEEHCSxOJaLK7s2bm3fU4z9OksP3ltz076xfb4Wa/3w8AAIC17a59AAAA4GUSGwAAQEJsAAAACbEBAAAkxAYAAJAQGwAAQEJsAAAACbEBAAAkxAYAAJAQGwAAQEJsAAAACbEBAAAkxAYAAJAQGwAAQEJsAAAACbEBAAAkxAYAAJAQGwAAQEJsAAAACbEBAAAkxAYAAJAQGwAAQEJsAAAACbEBAAAkxAYAAJAQGwAAQEJsAAAACbEBAAAkxAYAAJAQGwAAQEJsAAAACbEBAAAkxAYAAJAQGwAAQEJsAAAACbEBAAAkxAYAAJAQGwAAQEJsAAAACbEBAAAkxAYAAJAQGwAAQEJsAAAACbEBAAAkxAYAAJAQGwAAQEJsAAAACbEBAAAkxAYAAJAQGwAAQEJsAAAACbEBAAAkxAYAAJAQGwAAQEJsAAAACbEBAAAkxAYAAJAQGwAAQEJsAAAACbEBAAAkxAYAAJAQGwAAQEJsAAAACbEBAAAkxAYAAJAQGwAAQEJsAAAACbEBAAAkxAYAAJAQGwAAQEJsAAAACbEBAAAkxAYAAJAQGwAAQOKjUy/c3d79Mcb4asH+q8eH+68XvD2w0BmPY49TgBXN3Hdn77krfP11rjXP8t7WVp5/tnLOqQPnPnrO+pp6fLi/Ofayue9sLD3UJR4owGlzj0OPU4B1nbqvnnPPvdR9ec2zHHq9rTz/bOWcU9NzLb3uEn6MCgAASIgNAAAgITYAAICE2AAAABIXi43d7d3Pu9u71X6T3569a+4Vm/bs2bNnz95z4HzLfGjXy5yb/X5/9IW727vjLzzf72///GaM8fEY468VNscY49O3f9qzd429YrPa++SM1339hL3n/v7as2fP3rX3vph5+ZuZvbm3X9Oh+/+77+9TzjLdOudtz3n+mVr787v2OS91/R0697FzptfUqX/69uT/Z2Nlf48xPh//feHzZiz7JPwzxvjMnr0r7RWb1d5atvL+2rNnz96Htre26fmem619/F769XJSHhuPD/ff7W7vvh9j/DLG+OHx4f63pZvv7P1oz96l94rNcm+M8evc6z8+3H95rfPZs2fP3kvbO+MnQ/48tbfST5acZXr/n76/TznLga3Zt517/pmKPr+Lnycnexe5/g59fI+d85LX1Ht/94X+nm/HGD+t9YWePXtX3is27dmzZ8/eh7E3gr01Fe/vmorP75q2cP1dVP47G6d+hgvonflfljxOAVYyd9+du+de+Dsbq51lurWV55+tnHPqyHc2Dp6zvqZOfXz807cAAEBCbAAAAAmxAQAAJOZi49XC/aVvDyw39zj0OAVY16n76jn33Evdl9c8y6HX28rzz1bOOTU919LrLnHyF8QBAAD+Lz9GBQAAJMQGAACQEBsAAEBCbAAAAAmxAQAAJMQGAACQEBsAAEBCbAAAAAmxAQAAJMQGAACQEBsAAEBCbAAAAAmxAQAAJMQGAACQEBsAAEBCbAAAAAmxAQAAJMQGAACQEBsAAEBCbAAAAAmxAQAAJMQGAACQEBsAAEBCbAAAAAmxAQAAJMQGAACQEBsAAEBCbAAAAAmxAQAAJMQGAACQEBsAAEBCbAAAAAmxAQAAJMQGAACQEBsAAEBCbAAAAAmxAQAAJMQGAACQEBsAAEBCbAAAAAmxAQAAJMQGAACQEBsAAEBCbAAAAAmxAQAAJMQGAACQEBsAAEBCbAAAAAmxAQAAJMQGAACQEBsAAEBCbAAAAAmxAQAAJMQGAACQEBsAAEBCbAAAAAmxAQAAJMQGAACQEBsAAEBCbAAAAAmxAQAAJMQGAACQEBsAAEBCbAAAAAmxAQAAJMQGAACQEBsAAEBCbAAAAAmxAQAAJMQGAACQEBsAAEBCbAAAAAmxAQAAJMQGAACQ+BdDlHlo06a/EQAAAABJRU5ErkJggg==\n",
      "text/plain": [
       "<Figure size 1008x201.6 with 1 Axes>"
      ]
     },
     "metadata": {
      "needs_background": "light"
     },
     "output_type": "display_data"
    }
   ],
   "source": [
    "sg.plot_transcript_path(tids[0], browser=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 188,
   "id": "422ffbf8-78da-4d16-a445-0cbd4b77fbc5",
   "metadata": {},
   "outputs": [],
   "source": [
    "# df = pd.read_csv(filt_ab, sep='\\t')\n",
    "# df, tids = get_tpm_table(df,\n",
    "#                how='iso',\n",
    "#                min_tpm=1,\n",
    "#                gene_subset='polya',\n",
    "#                h5=h5)\n",
    "\n",
    "# df = df.transpose()\n",
    "# df.reset_index(inplace=True)\n",
    "# df.rename({'index':'dataset'}, axis=1, inplace=True)\n",
    "# df['biosample'] = df.dataset.str.rsplit('_', n=2, expand=True)[0]\n",
    "# meta = get_tissue_metadata()[['tissue', 'biosample']]\n",
    "# df = df.merge(meta, how='left', on='biosample')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 189,
   "id": "e5f7ea58-7144-445d-bfa2-25422d30f50e",
   "metadata": {},
   "outputs": [],
   "source": [
    "# df = df.loc[df.tissue=='vessels']\n",
    "# df.set_index('dataset', inplace=True)\n",
    "# df = df[[c for c in df.columns if 'ENSG00000049540' in c]]\n",
    "# df = df.transpose()\n",
    "# df.head()\n",
    "# df['sumbert'] = df.sum(axis=1)\n",
    "# 2.41843/4 # dividing expression of ENSG00000049540[1,77,1] by averaging\n",
    "# df = df.loc[df.sumbert != 0]\n",
    "# df = df.sort_values(by='sumbert', ascending=False)\n",
    "# df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 190,
   "id": "5e582e52-e314-4299-b599-139864efffbc",
   "metadata": {},
   "outputs": [],
   "source": [
    "# df.sumbert.tolist()"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.4"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
