{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "613bf09a-bd89-4750-8db7-f3c5de7992f5",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "import scipy.stats as st\n",
    "import seaborn as sns\n",
    "import sys\n",
    "import os\n",
    "import gseapy as gp\n",
    "import cerberus\n",
    "\n",
    "p = os.path.dirname(os.path.dirname(os.getcwd()))\n",
    "sys.path.append(p)\n",
    "\n",
    "from scripts.utils import *\n",
    "from scripts.plotting import *"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "3d30c1ba-39b8-40b6-ba6c-77c4a74dad5e",
   "metadata": {},
   "outputs": [],
   "source": [
    "filt_ab = 'cerberus_filtered_abundance.tsv'\n",
    "c_annot = 'cerberus_annot.h5'\n",
    "ver = 'v40_cerberus'"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "170e6040-5de4-4bc2-aef9-2470117d2b27",
   "metadata": {},
   "source": [
    "## PolyA gene transcript detection"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7203cbf6-f653-4b4d-aef3-e8ab1745ad5a",
   "metadata": {},
   "source": [
    "What fraction of polyA genes from the annotation do we detect with at least 1 known isoform >= 1 TPM in any library?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "dfe7b42f-8a42-4d87-9d12-009c009cedca",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Calculating iso TPM values\n",
      "Subsetting for novelty categories ['Known']\n",
      "Subsetting for polya genes\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/fairliereese/Documents/programming/mortazavi_lab/data/rnawg/scripts/utils.py:1297: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead.  To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  df[total_col] = df[d].sum()\n",
      "/Users/fairliereese/Documents/programming/mortazavi_lab/data/rnawg/scripts/utils.py:1298: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead.  To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  df[tpm_col] = (df[d]*1000000)/df[total_col]\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Enforcing minimum TPM\n",
      "Total # isos detected: 245379\n",
      "# isos >= 1 tpm: 223477\n",
      "Applying gene type and novelty subset\n",
      "Number of isos reported: 132274\n",
      "Detected 26446 / 52274 (50.59%) annotated polyA genes w/ at least 1 known transcript >= 1 TPM\n"
     ]
    }
   ],
   "source": [
    "df = pd.read_csv(filt_ab, sep='\\t')\n",
    "df, tids = get_tpm_table(df,\n",
    "               how='iso',\n",
    "               nov=['Known'],\n",
    "               min_tpm=1,\n",
    "               gene_subset='polya')\n",
    "\n",
    "# merge with gene id\n",
    "gene_df = pd.read_csv(filt_ab, sep='\\t')\n",
    "gene_df = gene_df[['annot_gene_id', 'annot_transcript_id']]\n",
    "df = df.merge(gene_df, how='left', left_index=True, right_on='annot_transcript_id')\n",
    "\n",
    "gene_df, _, _ = get_gtf_info(how='gene', ver=ver)\n",
    "\n",
    "polya_biotypes = ['protein_coding', 'pseudogene', 'lncRNA']\n",
    "polya_genes = gene_df.loc[gene_df.biotype_category.isin(polya_biotypes), 'gid'].tolist()\n",
    "n_polya = len(polya_genes)\n",
    "n_det_polya = len(df.annot_gene_id.unique().tolist())\n",
    "\n",
    "print('Detected {} / {} ({:.2f}%) annotated polyA genes w/ at least 1 known transcript >= 1 TPM'.format(n_det_polya, n_polya, (n_det_polya/n_polya)*100))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ca78fffa-5694-42d7-980a-1a37ea233685",
   "metadata": {},
   "source": [
    "## TF gene transcript detection\n",
    "\n",
    "What fraction of TF genes from the annotation do we detect with at least 1 known isoform >= 1 TPM in any library?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "5f2fd8cd-e329-46f8-b6dd-7041bf2ce415",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Calculating iso TPM values\n",
      "Subsetting for novelty categories ['Known']\n",
      "Subsetting for tf genes\n",
      "Enforcing minimum TPM\n",
      "Total # isos detected: 245379\n",
      "# isos >= 1 tpm: 223477\n",
      "Applying gene type and novelty subset\n",
      "Number of isos reported: 9102\n",
      "Detected 1326 / 1419 (93.45%) annotated TF genes w/ at least 1 known transcript >= 1 TPM\n"
     ]
    }
   ],
   "source": [
    "df = pd.read_csv(filt_ab, sep='\\t')\n",
    "df, tids = get_tpm_table(df,\n",
    "                   how='iso',\n",
    "                   nov=['Known'],\n",
    "                   min_tpm=1,\n",
    "                   gene_subset='tf')\n",
    "\n",
    "# merge with gene id\n",
    "gene_df = pd.read_csv(filt_ab, sep='\\t')\n",
    "gene_df = gene_df[['annot_gene_id', 'annot_transcript_id']]\n",
    "df = df.merge(gene_df, how='left', left_index=True, right_on='annot_transcript_id')\n",
    "\n",
    "gene_df, _, _ = get_gtf_info(how='gene', subset='tf', ver=ver)\n",
    "\n",
    "n_tf = len(gene_df.index)\n",
    "n_det_tf = len(df.annot_gene_id.unique().tolist())\n",
    "\n",
    "print('Detected {} / {} ({:.2f}%) annotated TF genes w/ at least 1 known transcript >= 1 TPM'.format(n_det_tf, n_tf, (n_det_tf/n_tf)*100))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "153b85ce-0abc-43e5-9df9-d48da23d316a",
   "metadata": {},
   "source": [
    "Which TFs have the most isoforms?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "c6e1bbf6-51a1-4bc3-beb8-4a8272d81a7d",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Calculating iso TPM values\n",
      "Subsetting for novelty categories ['Known', 'NIC', 'NNC', 'ISM_rescue']\n",
      "Subsetting for tf genes\n",
      "Enforcing minimum TPM\n",
      "Total # isos detected: 245379\n",
      "# isos >= 1 tpm: 223477\n",
      "Applying gene type and novelty subset\n",
      "Number of isos reported: 13122\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>annot_gene_id</th>\n",
       "      <th>all</th>\n",
       "      <th>gid_stable</th>\n",
       "      <th>gname</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>87</th>\n",
       "      <td>ENSG00000081189.15</td>\n",
       "      <td>74</td>\n",
       "      <td>ENSG00000081189</td>\n",
       "      <td>MEF2C</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>ENSG00000005801.17</td>\n",
       "      <td>71</td>\n",
       "      <td>ENSG00000005801</td>\n",
       "      <td>ZNF195</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>68</th>\n",
       "      <td>ENSG00000071564.14</td>\n",
       "      <td>67</td>\n",
       "      <td>ENSG00000071564</td>\n",
       "      <td>TCF3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>120</th>\n",
       "      <td>ENSG00000095794.19</td>\n",
       "      <td>66</td>\n",
       "      <td>ENSG00000095794</td>\n",
       "      <td>CREM</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>44</th>\n",
       "      <td>ENSG00000062194.15</td>\n",
       "      <td>64</td>\n",
       "      <td>ENSG00000062194</td>\n",
       "      <td>GPBP1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>843</th>\n",
       "      <td>ENSG00000175387.15</td>\n",
       "      <td>64</td>\n",
       "      <td>ENSG00000175387</td>\n",
       "      <td>SMAD2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>62</th>\n",
       "      <td>ENSG00000068305.17</td>\n",
       "      <td>59</td>\n",
       "      <td>ENSG00000068305</td>\n",
       "      <td>MEF2A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>302</th>\n",
       "      <td>ENSG00000120837.7</td>\n",
       "      <td>57</td>\n",
       "      <td>ENSG00000120837</td>\n",
       "      <td>NFYB</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>404</th>\n",
       "      <td>ENSG00000130844.17</td>\n",
       "      <td>57</td>\n",
       "      <td>ENSG00000130844</td>\n",
       "      <td>ZNF331</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>100</th>\n",
       "      <td>ENSG00000085274.15</td>\n",
       "      <td>56</td>\n",
       "      <td>ENSG00000085274</td>\n",
       "      <td>MYNN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "          annot_gene_id  all       gid_stable   gname\n",
       "87   ENSG00000081189.15   74  ENSG00000081189   MEF2C\n",
       "5    ENSG00000005801.17   71  ENSG00000005801  ZNF195\n",
       "68   ENSG00000071564.14   67  ENSG00000071564    TCF3\n",
       "120  ENSG00000095794.19   66  ENSG00000095794    CREM\n",
       "44   ENSG00000062194.15   64  ENSG00000062194   GPBP1\n",
       "843  ENSG00000175387.15   64  ENSG00000175387   SMAD2\n",
       "62   ENSG00000068305.17   59  ENSG00000068305   MEF2A\n",
       "302   ENSG00000120837.7   57  ENSG00000120837    NFYB\n",
       "404  ENSG00000130844.17   57  ENSG00000130844  ZNF331\n",
       "100  ENSG00000085274.15   56  ENSG00000085274    MYNN"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df = pd.read_csv(filt_ab, sep='\\t')\n",
    "df = get_isos_per_gene(df,\n",
    "                       min_tpm=1,\n",
    "                       gene_subset='tf',\n",
    "                       groupby='all', \n",
    "                       nov=['Known', 'NIC', 'NNC', 'ISM_rescue'])\n",
    "gene_df, _, _ = get_gtf_info(how='gene', ver=ver, add_stable_gid=True)\n",
    "\n",
    "df.reset_index(inplace=True)\n",
    "df['gid_stable'] = cerberus.get_stable_gid(df, 'annot_gene_id')\n",
    "df.set_index('annot_gene_id')\n",
    "      \n",
    "df = df.merge(gene_df[['gid_stable', 'gname']], \n",
    "              how='left', \n",
    "              on='gid_stable')\n",
    "df = df.sort_values(by='all', ascending=False)\n",
    "df.head(10)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f8008835-d4b3-4ac1-a6d0-9f0eb0951c14",
   "metadata": {},
   "source": [
    "Which TF has the most isoforms expressed in a single sample? "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "0f47243c-28ea-4275-bd92-a0e218d6dc31",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Calculating iso TPM values\n",
      "Subsetting for novelty categories ['Known', 'NIC', 'NNC', 'ISM_rescue']\n",
      "Subsetting for tf genes\n",
      "Enforcing minimum TPM\n",
      "Total # isos detected: 245379\n",
      "# isos >= 1 tpm: 223477\n",
      "Applying gene type and novelty subset\n",
      "Number of isos reported: 13122\n",
      "Found 50 total samples\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>annot_gene_id</th>\n",
       "      <th>max_sample</th>\n",
       "      <th>gid_stable</th>\n",
       "      <th>gname</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>250</th>\n",
       "      <td>ENSG00000115415.18</td>\n",
       "      <td>44.0</td>\n",
       "      <td>ENSG00000115415</td>\n",
       "      <td>STAT1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>120</th>\n",
       "      <td>ENSG00000095794.19</td>\n",
       "      <td>42.0</td>\n",
       "      <td>ENSG00000095794</td>\n",
       "      <td>CREM</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>87</th>\n",
       "      <td>ENSG00000081189.15</td>\n",
       "      <td>41.0</td>\n",
       "      <td>ENSG00000081189</td>\n",
       "      <td>MEF2C</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>62</th>\n",
       "      <td>ENSG00000068305.17</td>\n",
       "      <td>41.0</td>\n",
       "      <td>ENSG00000068305</td>\n",
       "      <td>MEF2A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>44</th>\n",
       "      <td>ENSG00000062194.15</td>\n",
       "      <td>41.0</td>\n",
       "      <td>ENSG00000062194</td>\n",
       "      <td>GPBP1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>471</th>\n",
       "      <td>ENSG00000137504.13</td>\n",
       "      <td>38.0</td>\n",
       "      <td>ENSG00000137504</td>\n",
       "      <td>CREBZF</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>728</th>\n",
       "      <td>ENSG00000168610.14</td>\n",
       "      <td>36.0</td>\n",
       "      <td>ENSG00000168610</td>\n",
       "      <td>STAT3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>ENSG00000005801.17</td>\n",
       "      <td>36.0</td>\n",
       "      <td>ENSG00000005801</td>\n",
       "      <td>ZNF195</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>364</th>\n",
       "      <td>ENSG00000126561.16</td>\n",
       "      <td>35.0</td>\n",
       "      <td>ENSG00000126561</td>\n",
       "      <td>STAT5A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>101</th>\n",
       "      <td>ENSG00000085276.17</td>\n",
       "      <td>34.0</td>\n",
       "      <td>ENSG00000085276</td>\n",
       "      <td>MECOM</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "          annot_gene_id  max_sample       gid_stable   gname\n",
       "250  ENSG00000115415.18        44.0  ENSG00000115415   STAT1\n",
       "120  ENSG00000095794.19        42.0  ENSG00000095794    CREM\n",
       "87   ENSG00000081189.15        41.0  ENSG00000081189   MEF2C\n",
       "62   ENSG00000068305.17        41.0  ENSG00000068305   MEF2A\n",
       "44   ENSG00000062194.15        41.0  ENSG00000062194   GPBP1\n",
       "471  ENSG00000137504.13        38.0  ENSG00000137504  CREBZF\n",
       "728  ENSG00000168610.14        36.0  ENSG00000168610   STAT3\n",
       "5    ENSG00000005801.17        36.0  ENSG00000005801  ZNF195\n",
       "364  ENSG00000126561.16        35.0  ENSG00000126561  STAT5A\n",
       "101  ENSG00000085276.17        34.0  ENSG00000085276   MECOM"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df = pd.read_csv(filt_ab, sep='\\t')\n",
    "df = get_isos_per_gene(df,\n",
    "                       min_tpm=1,\n",
    "                       gene_subset='tf',\n",
    "                       groupby='sample', \n",
    "                       nov=['Known', 'NIC', 'NNC', 'ISM_rescue'])\n",
    "df['max_sample'] = df.max(1)\n",
    "df = df['max_sample'].to_frame()\n",
    "\n",
    "gene_df, _, _ = get_gtf_info(how='gene', ver=ver, add_stable_gid=True)\n",
    "df.reset_index(inplace=True)\n",
    "df['gid_stable'] = cerberus.get_stable_gid(df, 'annot_gene_id')\n",
    "df.set_index('annot_gene_id')\n",
    "df = df.merge(gene_df[['gid_stable', 'gname']], \n",
    "              how='left', \n",
    "              on='gid_stable')\n",
    "\n",
    "df = df.sort_values(by='max_sample', ascending=False)\n",
    "df.head(10)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c5446cca-2207-44a8-bf45-484e2e8ea3b5",
   "metadata": {},
   "source": [
    "## Protein coding gene transcript detection\n",
    "\n",
    "What fraction of protein-coding genes do we detect with at least 1 known isoform >= 1 TPM in any library?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "53e19663-1f06-46c2-a281-217712288532",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Calculating iso TPM values\n",
      "Subsetting for novelty categories ['Known']\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/fairliereese/Documents/programming/mortazavi_lab/data/rnawg/scripts/utils.py:1294: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead.  To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  df[total_col] = df[d].sum()\n",
      "/Users/fairliereese/Documents/programming/mortazavi_lab/data/rnawg/scripts/utils.py:1295: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead.  To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  df[tpm_col] = (df[d]*1000000)/df[total_col]\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Enforcing minimum TPM\n",
      "Total # isos detected: 245379\n",
      "# isos >= 1 tpm: 223477\n",
      "Applying gene type and novelty subset\n",
      "Number of isos reported: 135125\n",
      "gene_df\n",
      "                  gid     gname  length         biotype biotype_category  \\\n",
      "0  ENSG00000000460.17  C1orf112  192174  protein_coding   protein_coding   \n",
      "1  ENSG00000000971.17       CFH  100823  protein_coding   protein_coding   \n",
      "2  ENSG00000001461.17    NIPAL3   57275  protein_coding   protein_coding   \n",
      "3  ENSG00000004487.18     KDM1A   64347  protein_coding   protein_coding   \n",
      "4  ENSG00000007933.13      FMO3   27019  protein_coding   protein_coding   \n",
      "\n",
      "      tf       gid_stable  \n",
      "0  False  ENSG00000000460  \n",
      "1  False  ENSG00000000971  \n",
      "2  False  ENSG00000001461  \n",
      "3  False  ENSG00000004487  \n",
      "4  False  ENSG00000007933  \n",
      "df\n",
      "    mcf7_1_1  k562_1_1  heart_right_ventricle_2_1  gm12878_3_1  \\\n",
      "8        0.0       0.0                        0.0          0.0   \n",
      "9        0.0       0.0                        0.0          0.0   \n",
      "10       0.0       0.0                        0.0          0.0   \n",
      "11       0.0       0.0                        0.0          0.0   \n",
      "12       0.0       0.0                        0.0          0.0   \n",
      "\n",
      "    heart_right_ventricle_3_1  pc3_1_1  lower_lobe_of_right_lung_1_1  \\\n",
      "8                         0.0      0.0                           0.0   \n",
      "9                         0.0      0.0                           0.0   \n",
      "10                        0.0      0.0                           0.0   \n",
      "11                        0.0      0.0                           0.0   \n",
      "12                        0.0      0.0                           0.0   \n",
      "\n",
      "    hl60_m1_24hr_1_1  hl60_m1_24hr_1_2  hl60_m2_24hr_1_1  ...  hl60_1_1  \\\n",
      "8                0.0               0.0          0.000000  ...       0.0   \n",
      "9                0.0               0.0          0.000000  ...       0.0   \n",
      "10               0.0               0.0          0.000000  ...       0.0   \n",
      "11               0.0               0.0          0.000000  ...       0.0   \n",
      "12               0.0               0.0          1.632088  ...       0.0   \n",
      "\n",
      "    hl60_1_2  pgp1_endo_1_1  pgp1_endo_1_2  left_cardiac_atrium_1_1  \\\n",
      "8        0.0            0.0            0.0                      0.0   \n",
      "9        0.0            0.0            0.0                      0.0   \n",
      "10       0.0            0.0            0.0                      0.0   \n",
      "11       0.0            0.0            0.0                      0.0   \n",
      "12       0.0            0.0            0.0                      0.0   \n",
      "\n",
      "    h9_neural_crest_1_1  h9_neural_crest_1_2  left_lung_2_1  \\\n",
      "8                   0.0                  0.0            0.0   \n",
      "9                   0.0                  0.0            0.0   \n",
      "10                  0.0                  0.0            0.0   \n",
      "11                  0.0                  0.0            0.0   \n",
      "12                  0.0                  0.0            0.0   \n",
      "\n",
      "        annot_gene_id     annot_transcript_id  \n",
      "8   ENSG00000278267.1  ENSG00000278267[1,1,1]  \n",
      "9   ENSG00000243485.5  ENSG00000243485[2,2,1]  \n",
      "10  ENSG00000237613.2  ENSG00000237613[1,1,1]  \n",
      "11  ENSG00000237613.2  ENSG00000237613[1,2,2]  \n",
      "12  ENSG00000238009.6  ENSG00000238009[4,5,5]  \n",
      "\n",
      "[5 rows x 140 columns]\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>biotype_category</th>\n",
       "      <th>detected_counts</th>\n",
       "      <th>gencode_counts</th>\n",
       "      <th>prop_detected</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>other</td>\n",
       "      <td>2182</td>\n",
       "      <td>7391</td>\n",
       "      <td>0.295224</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>miRNA</td>\n",
       "      <td>466</td>\n",
       "      <td>1879</td>\n",
       "      <td>0.248004</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>pseudogene</td>\n",
       "      <td>3430</td>\n",
       "      <td>14538</td>\n",
       "      <td>0.235933</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>lncRNA</td>\n",
       "      <td>3389</td>\n",
       "      <td>17748</td>\n",
       "      <td>0.190951</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>protein_coding</td>\n",
       "      <td>787</td>\n",
       "      <td>19988</td>\n",
       "      <td>0.039374</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  biotype_category  detected_counts  gencode_counts  prop_detected\n",
       "2            other             2182            7391       0.295224\n",
       "1            miRNA              466            1879       0.248004\n",
       "4       pseudogene             3430           14538       0.235933\n",
       "0           lncRNA             3389           17748       0.190951\n",
       "3   protein_coding              787           19988       0.039374"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df = pd.read_csv(filt_ab, sep='\\t')\n",
    "# no gene subset here so we can look at all other genes\n",
    "df, tids = get_tpm_table(df,\n",
    "                   how='iso',\n",
    "                   nov=['Known'],\n",
    "                   min_tpm=1)\n",
    "\n",
    "# merge with gene id\n",
    "gene_df = pd.read_csv(filt_ab, sep='\\t')\n",
    "gene_df = gene_df[['annot_gene_id', 'annot_transcript_id']]\n",
    "df = df.merge(gene_df, how='left', left_index=True, right_on='annot_transcript_id')\n",
    "\n",
    "gene_df, b_counts, b_cat_counts = get_gtf_info(how='gene', ver=ver, add_stable_gid=True)\n",
    "df['gid_stable'] = cerberus.get_stable_gid(df, 'annot_gene_id')\n",
    "print('gene_df')\n",
    "print(gene_df.head())\n",
    "print('df')\n",
    "print(df.head())\n",
    "df = df.merge(gene_df, how='left', left_on='annot_gene_id', right_on='gid')\n",
    "\n",
    "temp = df[['gid', 'biotype_category']].drop_duplicates().groupby('biotype_category').count()\n",
    "temp.reset_index(inplace=True)\n",
    "temp.rename({'gid': 'detected_counts'}, axis=1, inplace=True)\n",
    "\n",
    "temp = temp.merge(b_cat_counts, on='biotype_category', how='outer')\n",
    "temp['prop_detected'] = temp.detected_counts/temp.gencode_counts\n",
    "temp.sort_values(by='prop_detected', ascending=False, inplace=True)\n",
    "temp"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "39e0ab7e-1200-4540-8698-2eacee85911b",
   "metadata": {},
   "source": [
    "## What fraction of protein coding gene transcripts do we detect?\n",
    "\n",
    "Of the transcripts associated with protein coding genes GENCODE, how many do we detect >=1 TPM?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "cb257623-1c8f-49f5-93e4-beb409d351b4",
   "metadata": {},
   "outputs": [],
   "source": [
    "ca = cerberus.read(c_annot)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "aedb575c-e5a7-4d31-a05b-371e31b6cc6e",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>original_transcript_id</th>\n",
       "      <th>ic</th>\n",
       "      <th>ic_id</th>\n",
       "      <th>tss_id</th>\n",
       "      <th>tss</th>\n",
       "      <th>tes_id</th>\n",
       "      <th>tes</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>original_transcript_name</th>\n",
       "      <th>transcript_triplet</th>\n",
       "      <th>transcript_id</th>\n",
       "      <th>transcript_name</th>\n",
       "      <th>tss_first_sd_issue</th>\n",
       "      <th>tes_last_sa_issue</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENCODEHT000206942</td>\n",
       "      <td>1</td>\n",
       "      <td>ENCODEHG000058846_1</td>\n",
       "      <td>ENCODEHG000058846_1</td>\n",
       "      <td>1</td>\n",
       "      <td>ENCODEHG000058846_1</td>\n",
       "      <td>1</td>\n",
       "      <td>ENCODEHG000058846</td>\n",
       "      <td>ENCODEHG000058846</td>\n",
       "      <td>ENCODEHT000206942</td>\n",
       "      <td>[1,1,1]</td>\n",
       "      <td>ENCODEHG000058846[1,1,1]</td>\n",
       "      <td>ENCODEHG000058846[1,1,1]</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENCODEHT000206942#0</td>\n",
       "      <td>1</td>\n",
       "      <td>ENCODEHG000058846_1</td>\n",
       "      <td>ENCODEHG000058846_1</td>\n",
       "      <td>1</td>\n",
       "      <td>ENCODEHG000058846_1</td>\n",
       "      <td>1</td>\n",
       "      <td>ENCODEHG000058846</td>\n",
       "      <td>ENCODEHG000058846</td>\n",
       "      <td>ENCODEHT000206942</td>\n",
       "      <td>[1,1,1]</td>\n",
       "      <td>ENCODEHG000058846[1,1,1]</td>\n",
       "      <td>ENCODEHG000058846[1,1,1]</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENCODEHT000206867</td>\n",
       "      <td>4</td>\n",
       "      <td>ENCODEHG000058837_4</td>\n",
       "      <td>ENCODEHG000058837_2</td>\n",
       "      <td>2</td>\n",
       "      <td>ENCODEHG000058837_1</td>\n",
       "      <td>1</td>\n",
       "      <td>ENCODEHG000058837</td>\n",
       "      <td>ENCODEHG000058837</td>\n",
       "      <td>ENCODEHT000206867</td>\n",
       "      <td>[2,4,1]</td>\n",
       "      <td>ENCODEHG000058837[2,4,1]</td>\n",
       "      <td>ENCODEHG000058837[2,4,1]</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENCODEHT000206867#0</td>\n",
       "      <td>4</td>\n",
       "      <td>ENCODEHG000058837_4</td>\n",
       "      <td>ENCODEHG000058837_2</td>\n",
       "      <td>2</td>\n",
       "      <td>ENCODEHG000058837_1</td>\n",
       "      <td>1</td>\n",
       "      <td>ENCODEHG000058837</td>\n",
       "      <td>ENCODEHG000058837</td>\n",
       "      <td>ENCODEHT000206867</td>\n",
       "      <td>[2,4,1]</td>\n",
       "      <td>ENCODEHG000058837[2,4,1]</td>\n",
       "      <td>ENCODEHG000058837[2,4,1]</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENCODEHT000206868</td>\n",
       "      <td>2</td>\n",
       "      <td>ENCODEHG000058837_2</td>\n",
       "      <td>ENCODEHG000058837_2</td>\n",
       "      <td>2</td>\n",
       "      <td>ENCODEHG000058837_1</td>\n",
       "      <td>1</td>\n",
       "      <td>ENCODEHG000058837</td>\n",
       "      <td>ENCODEHG000058837</td>\n",
       "      <td>ENCODEHT000206868</td>\n",
       "      <td>[2,2,1]</td>\n",
       "      <td>ENCODEHG000058837[2,2,1]</td>\n",
       "      <td>ENCODEHG000058837[2,2,1]</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  original_transcript_id  ic                ic_id               tss_id  tss  \\\n",
       "0      ENCODEHT000206942   1  ENCODEHG000058846_1  ENCODEHG000058846_1    1   \n",
       "1    ENCODEHT000206942#0   1  ENCODEHG000058846_1  ENCODEHG000058846_1    1   \n",
       "2      ENCODEHT000206867   4  ENCODEHG000058837_4  ENCODEHG000058837_2    2   \n",
       "3    ENCODEHT000206867#0   4  ENCODEHG000058837_4  ENCODEHG000058837_2    2   \n",
       "4      ENCODEHT000206868   2  ENCODEHG000058837_2  ENCODEHG000058837_2    2   \n",
       "\n",
       "                tes_id  tes            gene_id          gene_name  \\\n",
       "0  ENCODEHG000058846_1    1  ENCODEHG000058846  ENCODEHG000058846   \n",
       "1  ENCODEHG000058846_1    1  ENCODEHG000058846  ENCODEHG000058846   \n",
       "2  ENCODEHG000058837_1    1  ENCODEHG000058837  ENCODEHG000058837   \n",
       "3  ENCODEHG000058837_1    1  ENCODEHG000058837  ENCODEHG000058837   \n",
       "4  ENCODEHG000058837_1    1  ENCODEHG000058837  ENCODEHG000058837   \n",
       "\n",
       "  original_transcript_name transcript_triplet             transcript_id  \\\n",
       "0        ENCODEHT000206942            [1,1,1]  ENCODEHG000058846[1,1,1]   \n",
       "1        ENCODEHT000206942            [1,1,1]  ENCODEHG000058846[1,1,1]   \n",
       "2        ENCODEHT000206867            [2,4,1]  ENCODEHG000058837[2,4,1]   \n",
       "3        ENCODEHT000206867            [2,4,1]  ENCODEHG000058837[2,4,1]   \n",
       "4        ENCODEHT000206868            [2,2,1]  ENCODEHG000058837[2,2,1]   \n",
       "\n",
       "            transcript_name  tss_first_sd_issue  tes_last_sa_issue  \n",
       "0  ENCODEHG000058846[1,1,1]               False              False  \n",
       "1  ENCODEHG000058846[1,1,1]               False              False  \n",
       "2  ENCODEHG000058837[2,4,1]               False              False  \n",
       "3  ENCODEHG000058837[2,4,1]               False              False  \n",
       "4  ENCODEHG000058837[2,2,1]               False              False  "
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "temp = ca.t_map[['original_transcript_id', 'original_transcript_name'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "204c7c4d-da5e-4d69-93fa-f4fe08f2fd62",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.4"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
