{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "1d6323ee-6fc4-459d-b992-5efefa5bef71",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "import scipy.stats as st\n",
    "import seaborn as sns\n",
    "import sys\n",
    "import os\n",
    "import gseapy as gp\n",
    "import pdb\n",
    "import copy\n",
    "from scipy import sparse\n",
    "import anndata\n",
    "import swan_vis as swan\n",
    "\n",
    "\n",
    "p = os.path.dirname(os.path.dirname(os.getcwd()))\n",
    "sys.path.append(p)\n",
    "\n",
    "from scripts.utils import *\n",
    "from scripts.plotting import *"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "id": "a0067de5-b765-436c-a838-b0a961f0962a",
   "metadata": {},
   "outputs": [],
   "source": [
    "# load expression data and metadata into swan\n",
    "annot = 'v29_cerberus.gtf'\n",
    "ab = 'cerberus_filtered_abundance.tsv'\n",
    "gene_ab = '/Users/fairliereese/mortazavi_lab/data/rnawg/lr_bulk/talon/human_talon_abundance.tsv'\n",
    "gtf = 'cerberus.gtf'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "id": "f4695ba5-c976-414a-ae0e-9df0d1a7fc32",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "Adding annotation to the SwanGraph\n",
      "\n",
      "Adding transcriptome to the SwanGraph\n",
      "Saving graph as swan_no_ab.p\n"
     ]
    }
   ],
   "source": [
    "sg = swan.SwanGraph()\n",
    "sg.add_annotation(annot)\n",
    "sg.add_transcriptome(gtf, include_isms=True)\n",
    "sg.save_graph('swan_no_ab')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "id": "61d1fbde-a552-4e60-a4a9-2e6694e9fb56",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Read in graph from swan_no_ab.p\n",
      "\n",
      "Adding abundance for datasets hl60_m2_24hr_1_1, mucosa_of_descending_colon_1_1, hl60_1_2, posterior_vena_cava_2_1, aorta_2_1... (and 133 more) to SwanGraph\n",
      "Calculating TPM...\n",
      "Calculating PI...\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/fairliereese/miniconda3/lib/python3.7/site-packages/swan_vis/utils.py:410: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead.  To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  df.reset_index(inplace=True)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Calculating edge usage...\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/fairliereese/miniconda3/lib/python3.7/site-packages/anndata/_core/anndata.py:120: ImplicitModificationWarning: Transforming to str index.\n",
      "  warnings.warn(\"Transforming to str index.\", ImplicitModificationWarning)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Calculating TSS usage...\n",
      "Using cerberus IDs to calculate\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/fairliereese/miniconda3/lib/python3.7/site-packages/swan_vis/utils.py:410: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead.  To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  df.reset_index(inplace=True)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Calculating TES usage...\n",
      "Using cerberus IDs to calculate\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/fairliereese/miniconda3/lib/python3.7/site-packages/swan_vis/utils.py:410: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead.  To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  df.reset_index(inplace=True)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Calculating IC usage...\n",
      "Using cerberus IDs to calculate\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/fairliereese/miniconda3/lib/python3.7/site-packages/swan_vis/utils.py:410: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead.  To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  df.reset_index(inplace=True)\n"
     ]
    }
   ],
   "source": [
    "sg = swan.read('swan_no_ab.p')\n",
    "sg.add_abundance(ab)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "id": "2b0b0c4a-0cb0-42ed-853c-d3143b91f736",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Saving graph as swan.p\n"
     ]
    }
   ],
   "source": [
    "sg.save_graph('swan')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "id": "66c4c895-636e-4192-a229-6f91c610901c",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Read in graph from swan.p\n",
      "\n",
      "Adding abundance for datasets gm12878_1_1, gm12878_1_2, gm12878_1_3, gm12878_1_4, gm12878_3_1... (and 133 more) to SwanGraph\n",
      "Calculating TPM...\n"
     ]
    }
   ],
   "source": [
    "sg = swan.read('swan.p')\n",
    "sg.add_abundance(gene_ab, how='gene')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "id": "e0947285-bebb-4b13-800a-ff1c1345cc7b",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Saving graph as swan.p\n"
     ]
    }
   ],
   "source": [
    "sg.save_graph('swan')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "id": "99b6df35-0247-44f8-88ae-9154f0066050",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array(['Undefined', 'Known', 'NIC', 'ISM', 'NNC'], dtype=object)"
      ]
     },
     "execution_count": 33,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sg.t_df.novelty.unique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "id": "177d480c-5e5f-4d38-a8cf-b5d4ba8c519a",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>tid</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>novelty</th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>ISM</th>\n",
       "      <td>27635</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Known</th>\n",
       "      <td>262367</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NIC</th>\n",
       "      <td>48526</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NNC</th>\n",
       "      <td>14539</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Undefined</th>\n",
       "      <td>1282</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "              tid\n",
       "novelty          \n",
       "ISM         27635\n",
       "Known      262367\n",
       "NIC         48526\n",
       "NNC         14539\n",
       "Undefined    1282"
      ]
     },
     "execution_count": 34,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "temp = sg.t_df[['novelty', 'tid']]\n",
    "temp.reset_index(drop=True, inplace=True)\n",
    "temp[['novelty', 'tid']].groupby('novelty').count()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "id": "6f47e30b-2ddf-4f0d-a714-06d71281ce46",
   "metadata": {
    "collapsed": true,
    "jupyter": {
     "outputs_hidden": true
    },
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>tname</th>\n",
       "      <th>gid</th>\n",
       "      <th>gname</th>\n",
       "      <th>path</th>\n",
       "      <th>tid</th>\n",
       "      <th>tss_id</th>\n",
       "      <th>ic_id</th>\n",
       "      <th>tes_id</th>\n",
       "      <th>loc_path</th>\n",
       "      <th>annotation</th>\n",
       "      <th>novelty</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>tid</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>ENCODEHG000058784[1,1,2]</th>\n",
       "      <td>ENCODEHG000058784[1,1,2]</td>\n",
       "      <td>ENCODEHG000058784</td>\n",
       "      <td>ENCODEHG000058784</td>\n",
       "      <td>[869635, 869617, 869636, 869621, 869637]</td>\n",
       "      <td>ENCODEHG000058784[1,1,2]</td>\n",
       "      <td>ENCODEHG000058784_1</td>\n",
       "      <td>ENCODEHG000058784_1</td>\n",
       "      <td>ENCODEHG000058784_2</td>\n",
       "      <td>[43, 41, 40, 37, 36, 34]</td>\n",
       "      <td>False</td>\n",
       "      <td>Undefined</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENCODEHG000058784[2,2,2]</th>\n",
       "      <td>ENCODEHG000058784[2,2,2]</td>\n",
       "      <td>ENCODEHG000058784</td>\n",
       "      <td>ENCODEHG000058784</td>\n",
       "      <td>[869638, 869639, 869616, 869617, 869618, 86961...</td>\n",
       "      <td>ENCODEHG000058784[2,2,2]</td>\n",
       "      <td>ENCODEHG000058784_2</td>\n",
       "      <td>ENCODEHG000058784_2</td>\n",
       "      <td>ENCODEHG000058784_2</td>\n",
       "      <td>[46, 44, 42, 41, 40, 39, 38, 37, 36, 34]</td>\n",
       "      <td>False</td>\n",
       "      <td>Undefined</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENCODEHG000058784[2,3,1]</th>\n",
       "      <td>ENCODEHG000058784[2,3,1]</td>\n",
       "      <td>ENCODEHG000058784</td>\n",
       "      <td>ENCODEHG000058784</td>\n",
       "      <td>[869614, 869615, 869616, 869617, 869618, 86961...</td>\n",
       "      <td>ENCODEHG000058784[2,3,1]</td>\n",
       "      <td>ENCODEHG000058784_2</td>\n",
       "      <td>ENCODEHG000058784_3</td>\n",
       "      <td>ENCODEHG000058784_1</td>\n",
       "      <td>[46, 45, 42, 41, 40, 39, 38, 37, 36, 35, 33, 3...</td>\n",
       "      <td>False</td>\n",
       "      <td>Undefined</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENCODEHG000058837[1,1,1]</th>\n",
       "      <td>ENCODEHG000058837[1,1,1]</td>\n",
       "      <td>ENCODEHG000058837</td>\n",
       "      <td>ENCODEHG000058837</td>\n",
       "      <td>[869612, 869598, 869599, 869600, 869613, 86960...</td>\n",
       "      <td>ENCODEHG000058837[1,1,1]</td>\n",
       "      <td>ENCODEHG000058837_1</td>\n",
       "      <td>ENCODEHG000058837_1</td>\n",
       "      <td>ENCODEHG000058837_1</td>\n",
       "      <td>[11, 9, 8, 6, 5, 2, 1, 0]</td>\n",
       "      <td>False</td>\n",
       "      <td>Undefined</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENCODEHG000058837[2,2,1]</th>\n",
       "      <td>ENCODEHG000058837[2,2,1]</td>\n",
       "      <td>ENCODEHG000058837</td>\n",
       "      <td>ENCODEHG000058837</td>\n",
       "      <td>[869606, 869607, 869597, 869598, 869599, 86960...</td>\n",
       "      <td>ENCODEHG000058837[2,2,1]</td>\n",
       "      <td>ENCODEHG000058837_2</td>\n",
       "      <td>ENCODEHG000058837_2</td>\n",
       "      <td>ENCODEHG000058837_1</td>\n",
       "      <td>[17, 12, 10, 9, 8, 6, 5, 4, 3, 2, 1, 0]</td>\n",
       "      <td>False</td>\n",
       "      <td>Undefined</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENCODEHG000570158[1,1,1]</th>\n",
       "      <td>ENCODEHG000570158[1,1,1]</td>\n",
       "      <td>ENCODEHG000570158</td>\n",
       "      <td>ENCODEHG000570158</td>\n",
       "      <td>[928099, 928100, 928101]</td>\n",
       "      <td>ENCODEHG000570158[1,1,1]</td>\n",
       "      <td>ENCODEHG000570158_1</td>\n",
       "      <td>ENCODEHG000570158_1</td>\n",
       "      <td>ENCODEHG000570158_1</td>\n",
       "      <td>[600964, 600962, 600957, 600956]</td>\n",
       "      <td>False</td>\n",
       "      <td>Undefined</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENCODEHG000570465[1,1,1]</th>\n",
       "      <td>ENCODEHG000570465[1,1,1]</td>\n",
       "      <td>ENCODEHG000570465</td>\n",
       "      <td>ENCODEHG000570465</td>\n",
       "      <td>[956853, 956854, 956855]</td>\n",
       "      <td>ENCODEHG000570465[1,1,1]</td>\n",
       "      <td>ENCODEHG000570465_1</td>\n",
       "      <td>ENCODEHG000570465_1</td>\n",
       "      <td>ENCODEHG000570465_1</td>\n",
       "      <td>[898398, 898399, 898400, 898401]</td>\n",
       "      <td>False</td>\n",
       "      <td>Undefined</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENCODEHG000573849[1,1,1]</th>\n",
       "      <td>ENCODEHG000573849[1,1,1]</td>\n",
       "      <td>ENCODEHG000573849</td>\n",
       "      <td>ENCODEHG000573849</td>\n",
       "      <td>[947086, 947087, 947088, 947089, 947090, 94709...</td>\n",
       "      <td>ENCODEHG000573849[1,1,1]</td>\n",
       "      <td>ENCODEHG000573849_1</td>\n",
       "      <td>ENCODEHG000573849_1</td>\n",
       "      <td>ENCODEHG000573849_1</td>\n",
       "      <td>[829227, 829228, 829229, 829230, 829231, 82923...</td>\n",
       "      <td>False</td>\n",
       "      <td>Undefined</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENCODEHG000576557[1,1,1]</th>\n",
       "      <td>ENCODEHG000576557[1,1,1]</td>\n",
       "      <td>ENCODEHG000576557</td>\n",
       "      <td>ENCODEHG000576557</td>\n",
       "      <td>[924605, 924606, 924607]</td>\n",
       "      <td>ENCODEHG000576557[1,1,1]</td>\n",
       "      <td>ENCODEHG000576557_1</td>\n",
       "      <td>ENCODEHG000576557_1</td>\n",
       "      <td>ENCODEHG000576557_1</td>\n",
       "      <td>[565062, 565061, 565060, 565059]</td>\n",
       "      <td>False</td>\n",
       "      <td>Undefined</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENCODEHG000600414[1,1,1]</th>\n",
       "      <td>ENCODEHG000600414[1,1,1]</td>\n",
       "      <td>ENCODEHG000600414</td>\n",
       "      <td>ENCODEHG000600414</td>\n",
       "      <td>[938051, 938052, 938053]</td>\n",
       "      <td>ENCODEHG000600414[1,1,1]</td>\n",
       "      <td>ENCODEHG000600414_1</td>\n",
       "      <td>ENCODEHG000600414_1</td>\n",
       "      <td>ENCODEHG000600414_1</td>\n",
       "      <td>[715620, 715621, 715622, 715623]</td>\n",
       "      <td>False</td>\n",
       "      <td>Undefined</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1282 rows Ã— 11 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                             tname                gid  \\\n",
       "tid                                                                     \n",
       "ENCODEHG000058784[1,1,2]  ENCODEHG000058784[1,1,2]  ENCODEHG000058784   \n",
       "ENCODEHG000058784[2,2,2]  ENCODEHG000058784[2,2,2]  ENCODEHG000058784   \n",
       "ENCODEHG000058784[2,3,1]  ENCODEHG000058784[2,3,1]  ENCODEHG000058784   \n",
       "ENCODEHG000058837[1,1,1]  ENCODEHG000058837[1,1,1]  ENCODEHG000058837   \n",
       "ENCODEHG000058837[2,2,1]  ENCODEHG000058837[2,2,1]  ENCODEHG000058837   \n",
       "...                                            ...                ...   \n",
       "ENCODEHG000570158[1,1,1]  ENCODEHG000570158[1,1,1]  ENCODEHG000570158   \n",
       "ENCODEHG000570465[1,1,1]  ENCODEHG000570465[1,1,1]  ENCODEHG000570465   \n",
       "ENCODEHG000573849[1,1,1]  ENCODEHG000573849[1,1,1]  ENCODEHG000573849   \n",
       "ENCODEHG000576557[1,1,1]  ENCODEHG000576557[1,1,1]  ENCODEHG000576557   \n",
       "ENCODEHG000600414[1,1,1]  ENCODEHG000600414[1,1,1]  ENCODEHG000600414   \n",
       "\n",
       "                                      gname  \\\n",
       "tid                                           \n",
       "ENCODEHG000058784[1,1,2]  ENCODEHG000058784   \n",
       "ENCODEHG000058784[2,2,2]  ENCODEHG000058784   \n",
       "ENCODEHG000058784[2,3,1]  ENCODEHG000058784   \n",
       "ENCODEHG000058837[1,1,1]  ENCODEHG000058837   \n",
       "ENCODEHG000058837[2,2,1]  ENCODEHG000058837   \n",
       "...                                     ...   \n",
       "ENCODEHG000570158[1,1,1]  ENCODEHG000570158   \n",
       "ENCODEHG000570465[1,1,1]  ENCODEHG000570465   \n",
       "ENCODEHG000573849[1,1,1]  ENCODEHG000573849   \n",
       "ENCODEHG000576557[1,1,1]  ENCODEHG000576557   \n",
       "ENCODEHG000600414[1,1,1]  ENCODEHG000600414   \n",
       "\n",
       "                                                                       path  \\\n",
       "tid                                                                           \n",
       "ENCODEHG000058784[1,1,2]           [869635, 869617, 869636, 869621, 869637]   \n",
       "ENCODEHG000058784[2,2,2]  [869638, 869639, 869616, 869617, 869618, 86961...   \n",
       "ENCODEHG000058784[2,3,1]  [869614, 869615, 869616, 869617, 869618, 86961...   \n",
       "ENCODEHG000058837[1,1,1]  [869612, 869598, 869599, 869600, 869613, 86960...   \n",
       "ENCODEHG000058837[2,2,1]  [869606, 869607, 869597, 869598, 869599, 86960...   \n",
       "...                                                                     ...   \n",
       "ENCODEHG000570158[1,1,1]                           [928099, 928100, 928101]   \n",
       "ENCODEHG000570465[1,1,1]                           [956853, 956854, 956855]   \n",
       "ENCODEHG000573849[1,1,1]  [947086, 947087, 947088, 947089, 947090, 94709...   \n",
       "ENCODEHG000576557[1,1,1]                           [924605, 924606, 924607]   \n",
       "ENCODEHG000600414[1,1,1]                           [938051, 938052, 938053]   \n",
       "\n",
       "                                               tid               tss_id  \\\n",
       "tid                                                                       \n",
       "ENCODEHG000058784[1,1,2]  ENCODEHG000058784[1,1,2]  ENCODEHG000058784_1   \n",
       "ENCODEHG000058784[2,2,2]  ENCODEHG000058784[2,2,2]  ENCODEHG000058784_2   \n",
       "ENCODEHG000058784[2,3,1]  ENCODEHG000058784[2,3,1]  ENCODEHG000058784_2   \n",
       "ENCODEHG000058837[1,1,1]  ENCODEHG000058837[1,1,1]  ENCODEHG000058837_1   \n",
       "ENCODEHG000058837[2,2,1]  ENCODEHG000058837[2,2,1]  ENCODEHG000058837_2   \n",
       "...                                            ...                  ...   \n",
       "ENCODEHG000570158[1,1,1]  ENCODEHG000570158[1,1,1]  ENCODEHG000570158_1   \n",
       "ENCODEHG000570465[1,1,1]  ENCODEHG000570465[1,1,1]  ENCODEHG000570465_1   \n",
       "ENCODEHG000573849[1,1,1]  ENCODEHG000573849[1,1,1]  ENCODEHG000573849_1   \n",
       "ENCODEHG000576557[1,1,1]  ENCODEHG000576557[1,1,1]  ENCODEHG000576557_1   \n",
       "ENCODEHG000600414[1,1,1]  ENCODEHG000600414[1,1,1]  ENCODEHG000600414_1   \n",
       "\n",
       "                                        ic_id               tes_id  \\\n",
       "tid                                                                  \n",
       "ENCODEHG000058784[1,1,2]  ENCODEHG000058784_1  ENCODEHG000058784_2   \n",
       "ENCODEHG000058784[2,2,2]  ENCODEHG000058784_2  ENCODEHG000058784_2   \n",
       "ENCODEHG000058784[2,3,1]  ENCODEHG000058784_3  ENCODEHG000058784_1   \n",
       "ENCODEHG000058837[1,1,1]  ENCODEHG000058837_1  ENCODEHG000058837_1   \n",
       "ENCODEHG000058837[2,2,1]  ENCODEHG000058837_2  ENCODEHG000058837_1   \n",
       "...                                       ...                  ...   \n",
       "ENCODEHG000570158[1,1,1]  ENCODEHG000570158_1  ENCODEHG000570158_1   \n",
       "ENCODEHG000570465[1,1,1]  ENCODEHG000570465_1  ENCODEHG000570465_1   \n",
       "ENCODEHG000573849[1,1,1]  ENCODEHG000573849_1  ENCODEHG000573849_1   \n",
       "ENCODEHG000576557[1,1,1]  ENCODEHG000576557_1  ENCODEHG000576557_1   \n",
       "ENCODEHG000600414[1,1,1]  ENCODEHG000600414_1  ENCODEHG000600414_1   \n",
       "\n",
       "                                                                   loc_path  \\\n",
       "tid                                                                           \n",
       "ENCODEHG000058784[1,1,2]                           [43, 41, 40, 37, 36, 34]   \n",
       "ENCODEHG000058784[2,2,2]           [46, 44, 42, 41, 40, 39, 38, 37, 36, 34]   \n",
       "ENCODEHG000058784[2,3,1]  [46, 45, 42, 41, 40, 39, 38, 37, 36, 35, 33, 3...   \n",
       "ENCODEHG000058837[1,1,1]                          [11, 9, 8, 6, 5, 2, 1, 0]   \n",
       "ENCODEHG000058837[2,2,1]            [17, 12, 10, 9, 8, 6, 5, 4, 3, 2, 1, 0]   \n",
       "...                                                                     ...   \n",
       "ENCODEHG000570158[1,1,1]                   [600964, 600962, 600957, 600956]   \n",
       "ENCODEHG000570465[1,1,1]                   [898398, 898399, 898400, 898401]   \n",
       "ENCODEHG000573849[1,1,1]  [829227, 829228, 829229, 829230, 829231, 82923...   \n",
       "ENCODEHG000576557[1,1,1]                   [565062, 565061, 565060, 565059]   \n",
       "ENCODEHG000600414[1,1,1]                   [715620, 715621, 715622, 715623]   \n",
       "\n",
       "                          annotation    novelty  \n",
       "tid                                              \n",
       "ENCODEHG000058784[1,1,2]       False  Undefined  \n",
       "ENCODEHG000058784[2,2,2]       False  Undefined  \n",
       "ENCODEHG000058784[2,3,1]       False  Undefined  \n",
       "ENCODEHG000058837[1,1,1]       False  Undefined  \n",
       "ENCODEHG000058837[2,2,1]       False  Undefined  \n",
       "...                              ...        ...  \n",
       "ENCODEHG000570158[1,1,1]       False  Undefined  \n",
       "ENCODEHG000570465[1,1,1]       False  Undefined  \n",
       "ENCODEHG000573849[1,1,1]       False  Undefined  \n",
       "ENCODEHG000576557[1,1,1]       False  Undefined  \n",
       "ENCODEHG000600414[1,1,1]       False  Undefined  \n",
       "\n",
       "[1282 rows x 11 columns]"
      ]
     },
     "execution_count": 37,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sg.t_df.loc[sg.t_df.novelty == 'Undefined']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "id": "1c053645-748b-4152-938a-5bfa226f7ffa",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Saving graph as swan.p\n"
     ]
    }
   ],
   "source": [
    "sg.save_graph('swan')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8dd4b48c-1575-4141-9e3d-8419b82a480a",
   "metadata": {},
   "source": [
    "## Add metadata"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "id": "185a5c7f-a653-4e04-b002-00689f540d9b",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Read in graph from swan.p\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/fairliereese/miniconda3/lib/python3.7/site-packages/anndata/_core/anndata.py:794: UserWarning: \n",
      "AnnData expects .obs.index to contain strings, but got values like:\n",
      "    [0, 1, 2, 3, 4]\n",
      "\n",
      "    Inferred to be: integer\n",
      "\n",
      "  value_idx = self._prep_dim_index(value.index, attr)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Saving graph as swan.p\n"
     ]
    }
   ],
   "source": [
    "sg = swan.read('swan.p')\n",
    "meta = '../swan/swan_metadata.tsv'\n",
    "sg.add_metadata(meta)\n",
    "\n",
    "# colors\n",
    "c_dict, order = get_biosample_colors()\n",
    "sg.set_metadata_colors('sample', c_dict)\n",
    "\n",
    "c_dict, order = get_ad_colors()\n",
    "sg.set_metadata_colors('health_status', c_dict)\n",
    "\n",
    "sg.save_graph('swan')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2b32fb3b-6709-4a55-b49d-c6ddd7819e1f",
   "metadata": {},
   "source": [
    "## Swan reports"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 80,
   "id": "46e883a4-46a5-430a-a124-5cb45d3ac5c3",
   "metadata": {
    "collapsed": true,
    "jupyter": {
     "outputs_hidden": true
    },
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Read in graph from swan.p\n",
      "['h9_chondro', 'h9_panc_beta', 'h9_de', 'h9_neural_crest', 'h9', 'h9_osteocyte', 'h9_panc_progen', 'pgp1_excite_neuron', 'pgp1_astro', 'pgp1', 'pgp1_endo', 'caco2']\n",
      "\n",
      "Plotting transcripts for ENSG00000081189.15\n",
      "Saving transcript path graph for ENSG00000081189[1,10,5] as figures/mef2c_h9_pgp1_caco2_novel_ENSG00000081189[1,10,5]_path.png\n",
      "Saving transcript path graph for ENSG00000081189[1,13,8] as figures/mef2c_h9_pgp1_caco2_novel_ENSG00000081189[1,13,8]_path.png\n",
      "Saving transcript path graph for ENSG00000081189[1,64,4] as figures/mef2c_h9_pgp1_caco2_novel_ENSG00000081189[1,64,4]_path.png\n",
      "Saving transcript path graph for ENSG00000081189[1,62,4] as figures/mef2c_h9_pgp1_caco2_novel_ENSG00000081189[1,62,4]_path.png\n",
      "Saving transcript path graph for ENSG00000081189[1,43,20] as figures/mef2c_h9_pgp1_caco2_novel_ENSG00000081189[1,43,20]_path.png\n",
      "Saving transcript path graph for ENSG00000081189[1,64,5] as figures/mef2c_h9_pgp1_caco2_novel_ENSG00000081189[1,64,5]_path.png\n",
      "Saving transcript path graph for ENSG00000081189[1,62,28] as figures/mef2c_h9_pgp1_caco2_novel_ENSG00000081189[1,62,28]_path.png\n",
      "Saving transcript path graph for ENSG00000081189[1,64,28] as figures/mef2c_h9_pgp1_caco2_novel_ENSG00000081189[1,64,28]_path.png\n",
      "Saving transcript path graph for ENSG00000081189[1,13,28] as figures/mef2c_h9_pgp1_caco2_novel_ENSG00000081189[1,13,28]_path.png\n",
      "Saving transcript path graph for ENSG00000081189[1,10,28] as figures/mef2c_h9_pgp1_caco2_novel_ENSG00000081189[1,10,28]_path.png\n",
      "Saving transcript path graph for ENSG00000081189[1,62,5] as figures/mef2c_h9_pgp1_caco2_novel_ENSG00000081189[1,62,5]_path.png\n",
      "Saving transcript path graph for ENSG00000081189[1,13,5] as figures/mef2c_h9_pgp1_caco2_novel_ENSG00000081189[1,13,5]_path.png\n",
      "Saving transcript path graph for ENSG00000081189[1,64,7] as figures/mef2c_h9_pgp1_caco2_novel_ENSG00000081189[1,64,7]_path.png\n",
      "Saving transcript path graph for ENSG00000081189[1,10,7] as figures/mef2c_h9_pgp1_caco2_novel_ENSG00000081189[1,10,7]_path.png\n",
      "Saving transcript path graph for ENSG00000081189[1,62,7] as figures/mef2c_h9_pgp1_caco2_novel_ENSG00000081189[1,62,7]_path.png\n",
      "Saving transcript path graph for ENSG00000081189[1,13,7] as figures/mef2c_h9_pgp1_caco2_novel_ENSG00000081189[1,13,7]_path.png\n",
      "Saving transcript path graph for ENSG00000081189[1,12,2] as figures/mef2c_h9_pgp1_caco2_novel_ENSG00000081189[1,12,2]_path.png\n",
      "Saving transcript path graph for ENSG00000081189[1,13,4] as figures/mef2c_h9_pgp1_caco2_novel_ENSG00000081189[1,13,4]_path.png\n",
      "Saving transcript path graph for ENSG00000081189[1,10,4] as figures/mef2c_h9_pgp1_caco2_novel_ENSG00000081189[1,10,4]_path.png\n",
      "Saving transcript path graph for ENSG00000081189[1,64,2] as figures/mef2c_h9_pgp1_caco2_novel_ENSG00000081189[1,64,2]_path.png\n",
      "Saving transcript path graph for ENSG00000081189[1,10,2] as figures/mef2c_h9_pgp1_caco2_novel_ENSG00000081189[1,10,2]_path.png\n",
      "Saving transcript path graph for ENSG00000081189[8,11,4] as figures/mef2c_h9_pgp1_caco2_novel_ENSG00000081189[8,11,4]_path.png\n",
      "Saving transcript path graph for ENSG00000081189[1,1,5] as figures/mef2c_h9_pgp1_caco2_novel_ENSG00000081189[1,1,5]_path.png\n",
      "Saving transcript path graph for ENSG00000081189[1,12,28] as figures/mef2c_h9_pgp1_caco2_novel_ENSG00000081189[1,12,28]_path.png\n",
      "Saving transcript path graph for ENSG00000081189[1,62,2] as figures/mef2c_h9_pgp1_caco2_novel_ENSG00000081189[1,62,2]_path.png\n",
      "Saving transcript path graph for ENSG00000081189[10,22,9] as figures/mef2c_h9_pgp1_caco2_novel_ENSG00000081189[10,22,9]_path.png\n",
      "Saving transcript path graph for ENSG00000081189[1,13,2] as figures/mef2c_h9_pgp1_caco2_novel_ENSG00000081189[1,13,2]_path.png\n",
      "Generating report for ENSG00000081189.15\n"
     ]
    }
   ],
   "source": [
    "sg = swan.read('swan.p')\n",
    "datasets = []\n",
    "datasets += sg.adata.obs.loc[sg.adata.obs['sample'].str.contains('h9'), 'sample'].unique().tolist()\n",
    "datasets += sg.adata.obs.loc[sg.adata.obs['sample'].str.contains('pgp1'), 'sample'].unique().tolist()\n",
    "datasets.append('caco2')\n",
    "print(datasets)\n",
    "\n",
    "# sg.gen_report('MEF2C',\n",
    "#               prefix='figures/mef2c_h9_pgp1_caco2',\n",
    "#               layer='pi',\n",
    "#               cmap='magma',\n",
    "#               novelty=True,\n",
    "#               groupby='sample',\n",
    "#               transcript_col='tname',\n",
    "#               metadata_cols=['sample'],\n",
    "#               datasets={'sample': datasets},\n",
    "#               browser=True)\n",
    "\n",
    "sg.gen_report('MEF2C',\n",
    "              prefix='figures/mef2c_h9_pgp1_caco2',\n",
    "              layer='tpm',\n",
    "              cmap='viridis',\n",
    "              novelty=True,\n",
    "              groupby='sample',\n",
    "              transcript_col='tname',\n",
    "              metadata_cols=['sample'],\n",
    "              indicate_novel=True,\n",
    "              datasets={'sample': datasets})\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "3c44b73d-f8e0-4bd9-899f-8b688b4d85f8",
   "metadata": {
    "collapsed": true,
    "jupyter": {
     "outputs_hidden": true
    },
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Read in graph from swan.p\n",
      "\n",
      "Plotting transcripts for ENSG00000136717.14\n",
      "Saving transcript path graph for ENSG00000136717[1,2,1] as figures/BIN1_browser_ENSG00000136717[1,2,1]_path.png\n",
      "Saving transcript path graph for ENSG00000136717[2,2,1] as figures/BIN1_browser_ENSG00000136717[2,2,1]_path.png\n",
      "Saving transcript path graph for ENSG00000136717[1,4,1] as figures/BIN1_browser_ENSG00000136717[1,4,1]_path.png\n",
      "Saving transcript path graph for ENSG00000136717[2,1,1] as figures/BIN1_browser_ENSG00000136717[2,1,1]_path.png\n",
      "Saving transcript path graph for ENSG00000136717[1,1,1] as figures/BIN1_browser_ENSG00000136717[1,1,1]_path.png\n",
      "Saving transcript path graph for ENSG00000136717[1,20,1] as figures/BIN1_browser_ENSG00000136717[1,20,1]_path.png\n",
      "Saving transcript path graph for ENSG00000136717[4,15,1] as figures/BIN1_browser_ENSG00000136717[4,15,1]_path.png\n",
      "Saving transcript path graph for ENSG00000136717[1,18,1] as figures/BIN1_browser_ENSG00000136717[1,18,1]_path.png\n",
      "Saving transcript path graph for ENSG00000136717[2,21,1] as figures/BIN1_browser_ENSG00000136717[2,21,1]_path.png\n",
      "Saving transcript path graph for ENSG00000136717[1,21,1] as figures/BIN1_browser_ENSG00000136717[1,21,1]_path.png\n",
      "Saving transcript path graph for ENSG00000136717[2,19,1] as figures/BIN1_browser_ENSG00000136717[2,19,1]_path.png\n",
      "Saving transcript path graph for ENSG00000136717[7,2,1] as figures/BIN1_browser_ENSG00000136717[7,2,1]_path.png\n",
      "Saving transcript path graph for ENSG00000136717[1,3,1] as figures/BIN1_browser_ENSG00000136717[1,3,1]_path.png\n",
      "Saving transcript path graph for ENSG00000136717[1,19,1] as figures/BIN1_browser_ENSG00000136717[1,19,1]_path.png\n",
      "Saving transcript path graph for ENSG00000136717[2,20,1] as figures/BIN1_browser_ENSG00000136717[2,20,1]_path.png\n",
      "Saving transcript path graph for ENSG00000136717[2,3,1] as figures/BIN1_browser_ENSG00000136717[2,3,1]_path.png\n",
      "Saving transcript path graph for ENSG00000136717[2,18,1] as figures/BIN1_browser_ENSG00000136717[2,18,1]_path.png\n",
      "Saving transcript path graph for ENSG00000136717[7,4,1] as figures/BIN1_browser_ENSG00000136717[7,4,1]_path.png\n",
      "Saving transcript path graph for ENSG00000136717[2,4,1] as figures/BIN1_browser_ENSG00000136717[2,4,1]_path.png\n",
      "Saving transcript path graph for ENSG00000136717[7,18,1] as figures/BIN1_browser_ENSG00000136717[7,18,1]_path.png\n",
      "Saving transcript path graph for ENSG00000136717[7,3,1] as figures/BIN1_browser_ENSG00000136717[7,3,1]_path.png\n",
      "Saving transcript path graph for ENSG00000136717[2,6,1] as figures/BIN1_browser_ENSG00000136717[2,6,1]_path.png\n",
      "Saving transcript path graph for ENSG00000136717[1,6,1] as figures/BIN1_browser_ENSG00000136717[1,6,1]_path.png\n",
      "Saving transcript path graph for ENSG00000136717[7,20,1] as figures/BIN1_browser_ENSG00000136717[7,20,1]_path.png\n",
      "Saving transcript path graph for ENSG00000136717[3,12,1] as figures/BIN1_browser_ENSG00000136717[3,12,1]_path.png\n",
      "Generating report for ENSG00000136717.14\n"
     ]
    }
   ],
   "source": [
    "# bin1\n",
    "sg = swan.read('swan.p')\n",
    "\n",
    "gene = 'BIN1'\n",
    "opref = 'figures/{}'.format(gene)\n",
    "conds = sg.adata.obs.loc[sg.adata.obs['sample'].str.contains('pgp1'), 'sample'].unique().tolist()\n",
    "conds = ['pgp1', 'pgp1_astro', 'pgp1_endo', 'pgp1_excite_neuron']\n",
    "sg.gen_report(gene,\n",
    "          prefix=opref,\n",
    "          layer='pi',\n",
    "          order='pi',\n",
    "          cmap='magma',\n",
    "          novelty=True,\n",
    "          metadata_cols=['sample'],\n",
    "          groupby='sample',\n",
    "          datasets={'sample': conds},\n",
    "          transcript_col='tname',\n",
    "          display_numbers=True,\n",
    "          browser=True)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2833238d-8656-47ef-adeb-85020edc4c17",
   "metadata": {},
   "source": [
    "### How many isoforms does it take to get 90% expression / gene / sample?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "567c4539-29f9-48c5-acdb-ac4d09ea1c5d",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Read in graph from swan.p\n"
     ]
    }
   ],
   "source": [
    "sg = swan.read('swan.p')\n",
    "t_df = pd.read_csv(ab, sep='\\t')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "eb860c47-8e76-47c4-bf12-d2f2293042a7",
   "metadata": {},
   "outputs": [],
   "source": [
    "t_df = t_df[['annot_gene_name', 'annot_transcript_id', 'annot_gene_id']]\n",
    "t_df.rename({'annot_gene_name': 'gname',\n",
    "             'annot_gene_id': 'gid',\n",
    "             'annot_transcript_id': 'tid'}, \n",
    "             axis=1, \n",
    "             inplace=True)\n",
    "\n",
    "df, _ = swan.calc_pi(sg.adata, sg.t_df, obs_col='sample')\n",
    "df = df.sparse.to_dense()\n",
    "df = df.transpose()\n",
    "df = df.merge(t_df[['tid', 'gname', 'gid']], how='inner', on='tid')\n",
    "\n",
    "df.set_index(['tid', 'gname', 'gid'], inplace=True)\n",
    "df = df.melt(ignore_index=False, value_name='pi', var_name='sample')\n",
    "df = df.dropna(subset=['pi'])\n",
    "df.reset_index(inplace=True)\n",
    "\n",
    "# determine the rank of each pi value for each sample / gene combo\n",
    "df = df.sort_values(by='pi', ascending=False)\n",
    "df['pi_rank'] = df.sort_values(by='pi', ascending=False).groupby(['gname', 'gid', 'sample']).cumcount()+1\n",
    "\n",
    "# add a column that we can check for convergence with \n",
    "df['gname_gid_biosamp'] = df.gname+'_'+df.gid+'_'+df['sample']\n",
    "\n",
    "iso_df = pd.DataFrame()\n",
    "max_pi_rank = df.pi_rank.max()\n",
    "for max_pi in range(1, max_pi_rank+1):\n",
    "    pi_ranks = [i for i in range(1, max_pi+1)]\n",
    "    # for the first iteration, we don't have to limit which genes we look at\n",
    "    if max_pi == 1:\n",
    "        temp = df.loc[df.pi_rank.isin(pi_ranks)].groupby(['gname_gid_biosamp']).sum().reset_index()\n",
    "    else:\n",
    "        ids = iso_df.gname_gid_biosamp.tolist()\n",
    "        temp = df.loc[(~df.gname_gid_biosamp.isin(ids))&(df.pi_rank.isin(pi_ranks))].groupby(['gname_gid_biosamp']).sum().reset_index()\n",
    "    \n",
    "    # converged if no more entries to analyze\n",
    "    if len(temp.index) == 0:\n",
    "        break\n",
    "    \n",
    "    # get isoforms that have >90% isoform exp accounted for\n",
    "    temp = temp.loc[temp.pi >= 90]\n",
    "    temp.drop(['pi_rank'], axis=1, inplace=True)\n",
    "    temp['n_isos'] = max_pi\n",
    "    \n",
    "    iso_df = pd.concat([iso_df, temp])\n",
    "\n",
    "# get list of isoforms required for each sample / gene combination as well\n",
    "df = df.merge(iso_df, how='left', on='gname_gid_biosamp')\n",
    "df['in_90_set'] = df.pi_rank <= df.n_isos\n",
    "df = df.loc[df.in_90_set]\n",
    "df[['gname', 'gid', 'sample']] = df.gname_gid_biosamp.str.split('_', n=2, expand=True)    \n",
    "df.rename({'pi_x': 'pi'}, axis=1, inplace=True)\n",
    "df.drop(['gname_gid_biosamp', \n",
    "        'pi_y', 'n_isos', 'in_90_set'], axis=1, inplace=True)\n",
    "\n",
    "# get the sample / gene vs. n isoforms required for 90%    \n",
    "iso_df[['gname', 'gid', 'sample']] = iso_df.gname_gid_biosamp.str.split('_', n=2, expand=True)    \n",
    "iso_df.drop('gname_gid_biosamp', axis=1, inplace=True)\n",
    "iso_df = iso_df.sort_values('n_isos', ascending=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "c539ea98-e8b9-4e19-a7b0-2a9a4a74330c",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>tid</th>\n",
       "      <th>gname</th>\n",
       "      <th>gid</th>\n",
       "      <th>sample</th>\n",
       "      <th>pi</th>\n",
       "      <th>pi_rank</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENSG00000238009[4,5,5]</td>\n",
       "      <td>AL627309.1</td>\n",
       "      <td>ENSG00000238009.6</td>\n",
       "      <td>hl60_m2_24hr</td>\n",
       "      <td>100.0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENSG00000259335[1,1,1]</td>\n",
       "      <td>HNRNPMP1</td>\n",
       "      <td>ENSG00000259335.1</td>\n",
       "      <td>h9_de</td>\n",
       "      <td>100.0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENSG00000133111[1,1,1]</td>\n",
       "      <td>RFXAP</td>\n",
       "      <td>ENSG00000133111.3</td>\n",
       "      <td>heart</td>\n",
       "      <td>100.0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENSG00000227973[1,1,1]</td>\n",
       "      <td>PIN4P1</td>\n",
       "      <td>ENSG00000227973.1</td>\n",
       "      <td>h9_de</td>\n",
       "      <td>100.0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENSG00000230283[1,1,1]</td>\n",
       "      <td>RPS12P24</td>\n",
       "      <td>ENSG00000230283.1</td>\n",
       "      <td>heart</td>\n",
       "      <td>100.0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                      tid       gname                gid        sample     pi  \\\n",
       "0  ENSG00000238009[4,5,5]  AL627309.1  ENSG00000238009.6  hl60_m2_24hr  100.0   \n",
       "1  ENSG00000259335[1,1,1]    HNRNPMP1  ENSG00000259335.1         h9_de  100.0   \n",
       "2  ENSG00000133111[1,1,1]       RFXAP  ENSG00000133111.3         heart  100.0   \n",
       "3  ENSG00000227973[1,1,1]      PIN4P1  ENSG00000227973.1         h9_de  100.0   \n",
       "4  ENSG00000230283[1,1,1]    RPS12P24  ENSG00000230283.1         heart  100.0   \n",
       "\n",
       "   pi_rank  \n",
       "0        1  \n",
       "1        1  \n",
       "2        1  \n",
       "3        1  \n",
       "4        1  "
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "7970ce54-5bb7-4bdf-b030-4fb0a6d84ddd",
   "metadata": {},
   "outputs": [],
   "source": [
    "df.to_csv('isos_sample_gene_90.tsv', sep='\\t', index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7dad5a29-453c-43f5-b350-855826512af2",
   "metadata": {},
   "source": [
    "## Which isoforms are the major isoform in any sample?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 156,
   "id": "e59b58bd-f153-4884-a3f1-f7a5038755b8",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "import scipy.stats as st\n",
    "import seaborn as sns\n",
    "import sys\n",
    "import os\n",
    "import gseapy as gp\n",
    "import pdb\n",
    "import copy\n",
    "from scipy import sparse\n",
    "import anndata\n",
    "import swan_vis as swan\n",
    "\n",
    "\n",
    "p = os.path.dirname(os.path.dirname(os.getcwd()))\n",
    "sys.path.append(p)\n",
    "\n",
    "from scripts.utils import *\n",
    "from scripts.plotting import *\n",
    "\n",
    "# load expression data and metadata into swan\n",
    "annot = 'v29_cerberus.gtf'\n",
    "ab = 'cerberus_filtered_abundance.tsv'\n",
    "gene_ab = '/Users/fairliereese/mortazavi_lab/data/rnawg/lr_bulk/talon/human_talon_abundance.tsv'\n",
    "gtf = 'cerberus.gtf'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 157,
   "id": "b8ac0b82-610d-48ac-b249-f2b632f84b03",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Read in graph from swan.p\n"
     ]
    }
   ],
   "source": [
    "inds = pd.read_csv('temp_gids_1_tpm.tsv', sep='\\t')\n",
    "inds = inds.gids.tolist()\n",
    "obs_col = 'sample'\n",
    "sg = swan.read('swan.p')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "c02d2e0a-c6fe-429a-9526-8dc225f4483f",
   "metadata": {},
   "outputs": [],
   "source": [
    "# df = pd.read_csv(gene_ab, sep='\\t')\n",
    "# df, inds = get_tpm_table(df,\n",
    "#                    how='gene',\n",
    "#                    min_tpm=1,\n",
    "#                    gene_subset='polya')\n",
    "\n",
    "# # need stable gid\n",
    "# df.head()\n",
    "# df['temp_tid'] = df.index.tolist()\n",
    "# inds = cerberus.get_stable_gid(df, col='temp_tid')\n",
    "# inds[:5]\n",
    "\n",
    "# temp = pd.DataFrame()\n",
    "# temp['gids'] = inds\n",
    "# temp.to_csv('temp_gids_1_tpm.tsv', sep='\\t', index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 154,
   "id": "26fe2b14-172c-4cbd-8f62-28d8900b04b2",
   "metadata": {},
   "outputs": [],
   "source": [
    "def uses_principal_feat(x):\n",
    "    \"\"\"\n",
    "    Agg function to determine if individual feats\n",
    "    (tss, ic, tes) use the principal version of that \n",
    "    feature\n",
    "    \"\"\"\n",
    "    return '1' in x.unique()\n",
    "\n",
    "def uses_principal_iso(x):\n",
    "    \"\"\"\n",
    "    Agg function to determine if the isoform triplet\n",
    "    is the principal version of that gene\n",
    "    \"\"\"\n",
    "    return '1,1,1' in x.unique()\n",
    "\n",
    "def get_major_iso(sg, obs_col='sample', gene_subset=None):\n",
    "    \"\"\"\n",
    "    Determine what the major isoform is in each grouping\n",
    "    for each gene \n",
    "    \n",
    "    Parameters:\n",
    "        sg (swan_vis SwanGraph): SwanGraph with transcript\n",
    "            abundance data added\n",
    "        obs_col (str): Column in `sg.adata.obs` to group\n",
    "            datasets by \n",
    "        gene_subset (list of str or None): If not None, \n",
    "            list of stable gene ids to subset on \n",
    "            \n",
    "    Returns:\n",
    "        df (pandas DataFrame): DF where each row is the major \n",
    "            isoform in each expressed gene / sample combination\n",
    "    \"\"\"\n",
    "    df, _ = swan.calc_pi(sg.adata, sg.t_df, obs_col=obs_col)\n",
    "    df = df.sparse.to_dense()\n",
    "    df = df.transpose()\n",
    "\n",
    "    t_df = sg.t_df.copy(deep=True)\n",
    "    t_df.reset_index(drop=True, inplace=True)\n",
    "    t_df = t_df[['tid', 'gname', 'gid']]\n",
    "    df = df.merge(t_df, how='inner', on='tid')\n",
    "    \n",
    "    df.set_index(['tid', 'gname', 'gid'], inplace=True)\n",
    "    df = df.melt(ignore_index=False, value_name='pi', var_name=obs_col)\n",
    "    df = df.dropna(subset=['pi'])\n",
    "    df.reset_index(inplace=True)\n",
    "    \n",
    "    # remove unexpressed isoforms\n",
    "    df = df.loc[df.pi > 0]\n",
    "    \n",
    "    # limit to detected genes\n",
    "    df['gid_stable'] = cerberus.get_stable_gid(df, 'gid')\n",
    "    if gene_subset:\n",
    "        df = df.loc[df.gid_stable.isin(gene_subset)]\n",
    "        \n",
    "    # sort by gene, sample, and pi value\n",
    "    # dedupe across the gene and sample cols; take the top-expressed isoform\n",
    "    df = df.sort_values(by=['gid', obs_col, 'pi'],\n",
    "                        ascending=[False, False, False])\n",
    "\n",
    "    df = df.drop_duplicates(subset=['gname', 'gid', obs_col], keep='first')       \n",
    "    \n",
    "    return df\n",
    "\n",
    "def get_major_principal_feats(sg, **kwargs):\n",
    "    \"\"\"\n",
    "    Determine how many genes have at least one sample where\n",
    "    the features and full-length isoform is the principal isoform\n",
    "    \n",
    "    Parameters:\n",
    "        sg (swan_vis SwanGraph): SwanGraph with transcript\n",
    "                    abundance data added   \n",
    "                    \n",
    "    Returns:\n",
    "        df (pandas DataFrame): DF w/ boolean columns for \n",
    "            whether or not there's at least one sample\n",
    "            where the major feat / isoform is the principal one\n",
    "    \"\"\"\n",
    "    df = get_major_iso(sg, **kwargs)\n",
    "    \n",
    "    # add features\n",
    "    for feat in ['ic', 'tss', 'tes']:\n",
    "        df = add_feat(df, kind=feat,\n",
    "                      col='tid',\n",
    "                      as_number=True,\n",
    "                      drop_triplet=False)\n",
    "\n",
    "    # remove unnecessary columns and determine which genes\n",
    "    # have at least one sample who's major iso uses the principal\n",
    "    # of each feature\n",
    "    temp = df.copy(deep=True)\n",
    "    temp.drop([obs_col, 'tid', 'pi', 'gid'], axis=1, inplace=True)\n",
    "\n",
    "    temp = temp.groupby(['gname',\n",
    "                         'gid_stable']).agg({'tss': uses_principal_feat,\n",
    "                                             'ic': uses_principal_feat,\n",
    "                                             'tes': uses_principal_feat,\n",
    "                                             'triplet': uses_principal_iso}).reset_index()\n",
    "\n",
    "    # output\n",
    "    for feat in ['ic', 'tss', 'tes', 'triplet']:\n",
    "        n_genes = len(temp.gid_stable.unique().tolist())\n",
    "        n_major = len(temp.loc[temp[feat] == True].index)\n",
    "        print(n_genes)\n",
    "        print(n_major)\n",
    "        print(feat)\n",
    "        print('{:.2f}% of genes have >=1 sample where major isoform contains principal {}'.format((n_major/n_genes)*100, feat))\n",
    "        print()\n",
    "        \n",
    "    return temp\n",
    "\n",
    "def count_major_principal_feats(sg, **kwargs):\n",
    "    \"\"\"\n",
    "    Count the number of samples that the major isoform\n",
    "    is the principal isoform and vice versa. Do the same\n",
    "    for individual features as well.\n",
    "    \n",
    "    Parameters:\n",
    "        sg (swan_vis SwanGraph): SwanGraph with transcript\n",
    "            abundance data added\n",
    "            \n",
    "    Returns:\n",
    "        df (pandas DataFrame): DF with # samples per gene \n",
    "            that use the principal feature as their major\n",
    "            feature / iso\n",
    "    \"\"\"\n",
    "    \n",
    "    df = get_major_iso(sg, **kwargs)\n",
    "    \n",
    "    # add features\n",
    "    for feat in ['ic', 'tss', 'tes']:\n",
    "        df = add_feat(df, kind=feat,\n",
    "                      col='tid',\n",
    "                      as_number=True,\n",
    "                      drop_triplet=False)\n",
    "        \n",
    "    temp = df.copy(deep=True)\n",
    "    temp.drop(['tid', 'pi', 'gid'], axis=1, inplace=True)\n",
    "\n",
    "    temp = temp.groupby([obs_col, 'gname',\n",
    "                        'gid_stable'],\n",
    "                        observed=True).agg({'tss': uses_principal_feat,\n",
    "                                            'ic': uses_principal_feat,\n",
    "                                            'tes': uses_principal_feat,\n",
    "                                            'triplet': uses_principal_iso}).reset_index()\n",
    "    \n",
    "    # convert from detection bool to int so we can sum up\n",
    "    # also add the opposiite number\n",
    "    for feat in ['tss', 'ic', 'tes', 'triplet']:\n",
    "        col = 'not_{}'.format(feat)\n",
    "        temp[feat] = temp[feat].astype(int)\n",
    "        temp[col] = (temp[feat] == False).astype(int)\n",
    "\n",
    "    # count n samples where major != principal and vice versa\n",
    "    temp = temp.groupby(['gname', 'gid_stable']).sum().reset_index()\n",
    "    \n",
    "    # total number of samples that this gene is expressed / \n",
    "    # has a complete isoform in \n",
    "    temp['n_samples'] = temp.triplet+temp.not_triplet\n",
    "    \n",
    "    for feat in ['triplet', 'tss', 'ic', 'tes']:\n",
    "        col = 'not_{}'.format(feat)\n",
    "        n_genes = len(temp.gid_stable.unique())\n",
    "        n_num = len(temp.loc[temp[col] >= 1].index)\n",
    "        print()\n",
    "        print('{:.2f}% ({}/{}) genes with >= 1 sample where the major {} is not the principal one'.format((n_num/n_genes)*100, n_num, n_genes, feat)) \n",
    "        \n",
    "    for feat in ['triplet', 'tss', 'ic', 'tes']:\n",
    "        col = 'perc_{}'.format(feat)\n",
    "        temp[col] = (temp[feat]/temp.n_samples)*100\n",
    "        \n",
    "    return temp\n",
    "\n",
    "def plot_major_principal_feat_counts(sg, opref='figures/', **kwargs):\n",
    "    \"\"\"\n",
    "    Plot a histogram of the number of datasets where the major isoform\n",
    "    is the principal isoform.\n",
    "    \n",
    "    Parameters:\n",
    "        sg (swan_vis SwanGraph): SwanGraph with transcript\n",
    "                abundance data added\n",
    "    \n",
    "    Returns:\n",
    "        temp (pandas DataFrame): Output from `count_major_principal_feats`.\n",
    "    \"\"\"\n",
    "    temp = count_major_principal_feats(sg, **kwargs)\n",
    "    c_dict, order = get_feat_colors()    \n",
    "    for feat in ['triplet', 'tss', 'ic', 'tes']:\n",
    "        col = 'perc_{}'.format(feat)\n",
    "        ax = sns.displot(data=temp,\n",
    "                         x=col,\n",
    "                         linewidth=0, \n",
    "                         color=c_dict[feat])\n",
    "        ylabel = '# genes'\n",
    "        xlabel = '% of samples where MANE is major {}'.format(feat)\n",
    "        ax.set(ylabel=ylabel, xlabel=xlabel)\n",
    "\n",
    "        fname = '{}/MANE_vs_major_{}_hist.png'.format(opref, feat)\n",
    "        plt.savefig(fname, dpi=800, bbox_inches='tight')\n",
    "        \n",
    "    return temp"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 158,
   "id": "ba40653b-c5f8-4602-af4c-700326ac7107",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "26603\n",
      "22235\n",
      "ic\n",
      "83.58% of genes have >=1 sample where major isoform contains principal ic\n",
      "\n",
      "26603\n",
      "23734\n",
      "tss\n",
      "89.22% of genes have >=1 sample where major isoform contains principal tss\n",
      "\n",
      "26603\n",
      "22160\n",
      "tes\n",
      "83.30% of genes have >=1 sample where major isoform contains principal tes\n",
      "\n",
      "26603\n",
      "18295\n",
      "triplet\n",
      "68.77% of genes have >=1 sample where major isoform contains principal triplet\n",
      "\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gname</th>\n",
       "      <th>gid_stable</th>\n",
       "      <th>tss</th>\n",
       "      <th>ic</th>\n",
       "      <th>tes</th>\n",
       "      <th>triplet</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>A1BG</td>\n",
       "      <td>ENSG00000121410</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>A1BG-AS1</td>\n",
       "      <td>ENSG00000268895</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>A1CF</td>\n",
       "      <td>ENSG00000148584</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>A2M</td>\n",
       "      <td>ENSG00000175899</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>A2M-AS1</td>\n",
       "      <td>ENSG00000245105</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26598</th>\n",
       "      <td>ZXDC</td>\n",
       "      <td>ENSG00000070476</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26599</th>\n",
       "      <td>ZYG11A</td>\n",
       "      <td>ENSG00000203995</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26600</th>\n",
       "      <td>ZYG11B</td>\n",
       "      <td>ENSG00000162378</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26601</th>\n",
       "      <td>ZYX</td>\n",
       "      <td>ENSG00000159840</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26602</th>\n",
       "      <td>ZZEF1</td>\n",
       "      <td>ENSG00000074755</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>26603 rows Ã— 6 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "          gname       gid_stable   tss     ic   tes  triplet\n",
       "0          A1BG  ENSG00000121410  True   True  True    False\n",
       "1      A1BG-AS1  ENSG00000268895  True   True  True     True\n",
       "2          A1CF  ENSG00000148584  True  False  True    False\n",
       "3           A2M  ENSG00000175899  True   True  True     True\n",
       "4       A2M-AS1  ENSG00000245105  True   True  True     True\n",
       "...         ...              ...   ...    ...   ...      ...\n",
       "26598      ZXDC  ENSG00000070476  True   True  True     True\n",
       "26599    ZYG11A  ENSG00000203995  True   True  True     True\n",
       "26600    ZYG11B  ENSG00000162378  True   True  True     True\n",
       "26601       ZYX  ENSG00000159840  True   True  True     True\n",
       "26602     ZZEF1  ENSG00000074755  True   True  True     True\n",
       "\n",
       "[26603 rows x 6 columns]"
      ]
     },
     "execution_count": 158,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "obs_col = 'sample'\n",
    "\n",
    "get_major_principal_feats(sg, obs_col=obs_col, gene_subset=inds)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 161,
   "id": "fc404ad4-460b-40e0-a3c9-bacaec7a1b35",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "69.00% (18355/26603) genes with >= 1 sample where the major triplet is not the principal one\n",
      "\n",
      "44.49% (11835/26603) genes with >= 1 sample where the major tss is not the principal one\n",
      "\n",
      "54.96% (14622/26603) genes with >= 1 sample where the major ic is not the principal one\n",
      "\n",
      "51.78% (13774/26603) genes with >= 1 sample where the major tes is not the principal one\n"
     ]
    }
   ],
   "source": [
    "df = count_major_principal_feats(sg, obs_col=obs_col, gene_subset=inds)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 162,
   "id": "7426e0c7-31b5-4ddc-ab85-e8352ce1d977",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "50"
      ]
     },
     "execution_count": 162,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# do we have the correct number of samples? Should be 50\n",
    "max(df.tss)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 142,
   "id": "d01272cc-1dac-47ed-8286-0ce07bd4a94b",
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_feat_colors(cats=None):\n",
    "    tss = '#56B4E9'\n",
    "    tes = '#E69F00'\n",
    "    splicing = '#CC79A7'\n",
    "    triplet = '#009E73'\n",
    "    c_dict = {'tss': tss,\n",
    "              'ic': splicing,\n",
    "              'tes': tes,\n",
    "              'triplet': triplet}\n",
    "    order = ['triplet', 'tss', 'ic', 'tes']\n",
    "    \n",
    "    c_dict, order = rm_color_cats(c_dict, order, cats)\n",
    "    return c_dict, order\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 143,
   "id": "fe435b88-d27d-4844-8767-a27b842241bb",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 166,
   "id": "76e78fa5-fcc9-4af9-8390-81dd3473f09f",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a10b1afa-b176-47a4-af7c-9dd8d21e17cb",
   "metadata": {},
   "outputs": [],
   "source": [
    "df = plot_major_principal_feat_counts(sg, opref='figures/', obs_col=obs_col, gene_subset=inds)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 116,
   "id": "73c402fa-e941-48db-9e63-29ebbb509d18",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gname</th>\n",
       "      <th>gid_stable</th>\n",
       "      <th>n_sample_major_iso</th>\n",
       "      <th>n_sample_not_major_iso</th>\n",
       "      <th>n_expressed_samples</th>\n",
       "      <th>prop_major_samples</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>8842</th>\n",
       "      <td>COL1A1</td>\n",
       "      <td>ENSG00000108821</td>\n",
       "      <td>0</td>\n",
       "      <td>29</td>\n",
       "      <td>29</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16140</th>\n",
       "      <td>MEF2C</td>\n",
       "      <td>ENSG00000081189</td>\n",
       "      <td>0</td>\n",
       "      <td>42</td>\n",
       "      <td>42</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "        gname       gid_stable  n_sample_major_iso  n_sample_not_major_iso  \\\n",
       "8842   COL1A1  ENSG00000108821                   0                      29   \n",
       "16140   MEF2C  ENSG00000081189                   0                      42   \n",
       "\n",
       "       n_expressed_samples  prop_major_samples  \n",
       "8842                    29                 0.0  \n",
       "16140                   42                 0.0  "
      ]
     },
     "execution_count": 116,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gnames = ['MEF2C', 'COL1A1']\n",
    "temp.loc[temp.gname.isin(gnames)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d7fe8fb6-8cfa-4bb1-a083-1f2b5d4cd65b",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.4"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
