{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "252e7a34-6a87-45d7-8517-dbf280adace8",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "import scipy.stats as st\n",
    "import seaborn as sns\n",
    "import sys\n",
    "import os\n",
    "import gseapy as gp\n",
    "import pdb\n",
    "import copy\n",
    "from scipy import sparse\n",
    "import anndata\n",
    "import swan_vis as swan\n",
    "\n",
    "\n",
    "p = os.path.dirname(os.path.dirname(os.path.dirname(os.getcwd())))\n",
    "sys.path.append(p)\n",
    "\n",
    "from scripts.utils import *\n",
    "from scripts.plotting import *"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "709fabe5-2698-4116-a6d6-fdd78be7dee1",
   "metadata": {},
   "outputs": [],
   "source": [
    "h5 = '../cerberus_annot.h5'\n",
    "ab = '../../talon/human_talon_abundance.tsv'\n",
    "filt_ab = '../cerberus_filtered_abundance.tsv'\n",
    "obs_col = 'sample'\n",
    "min_tpm = 1\n",
    "major_set = 'isos_sample_gene_90.tsv'\n",
    "swan_file = 'swan.p'\n",
    "mane_file = o = '/Users/fairliereese/mortazavi_lab/data/rnawg/refs/v40_gene_metadata.tsv'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "d89d0ef3-35bf-4d64-994d-170ce3fe1cca",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Read in graph from swan.p\n"
     ]
    }
   ],
   "source": [
    "sg = swan.read(swan_file)\n",
    "ca = cerberus.read(h5)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "509ccd34-ea09-4abb-b8aa-b15a22975ab0",
   "metadata": {},
   "outputs": [],
   "source": [
    "# plot\n",
    "# if Mane is principal\n",
    "# - exp of mane vs. exp of secondary\n",
    "# if mane != principal\n",
    "# - exp of mane vs. exp of principal"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "b8fb81b9-4a74-411d-8ae0-9c7c2d3499f6",
   "metadata": {},
   "outputs": [],
   "source": [
    "def uses_principal_feat(x):\n",
    "    \"\"\"\n",
    "    Agg function to determine if individual feats\n",
    "    (tss, ic, tes) use the principal version of that \n",
    "    feature\n",
    "    \"\"\"\n",
    "    return '1' in x.unique()\n",
    "\n",
    "def uses_principal_iso(x):\n",
    "    \"\"\"\n",
    "    Agg function to determine if the isoform triplet\n",
    "    is the principal version of that gene\n",
    "    \"\"\"\n",
    "    return '1,1,1' in x.unique()\n",
    "\n",
    "def count_major_principal_feats(df, **kwargs):\n",
    "    \"\"\"\n",
    "    Count the number of samples that the major isoform\n",
    "    is the principal isoform and vice versa. Do the same\n",
    "    for individual features as well.\n",
    "    \n",
    "    Parameters:\n",
    "        sg (swan_vis SwanGraph): SwanGraph with transcript\n",
    "            abundance data added\n",
    "            \n",
    "    Returns:\n",
    "        df (pandas DataFrame): DF with # samples per gene \n",
    "            that use the principal feature as their major\n",
    "            feature / iso\n",
    "    \"\"\"\n",
    "\n",
    "    # df = get_major_feats(sg, **kwargs)\n",
    "    # df = major_df.copy(deep=True)\n",
    "\n",
    "    # drop unnecessary columns\n",
    "    temp = df.copy(deep=True)\n",
    "    drop_cols = [c for c in temp.columns if 'pi' in c]\n",
    "    drop_cols += [c for c in temp.columns if '_id' in c]\n",
    "    drop_cols += ['tid', 'gid']\n",
    "    temp.drop(drop_cols, axis=1, inplace=True)\n",
    "\n",
    "    temp = temp.groupby([obs_col, 'gname',\n",
    "                        'gid_stable'],\n",
    "                        observed=True).agg({'tss': uses_principal_feat,\n",
    "                                            'ic': uses_principal_feat,\n",
    "                                            'tes': uses_principal_feat,\n",
    "                                            'triplet': uses_principal_iso}).reset_index()\n",
    "\n",
    "    # convert from detection bool to int so we can sum up\n",
    "    # also add the opposiite number\n",
    "    for feat in ['tss', 'ic', 'tes', 'triplet']:\n",
    "        col = 'not_{}'.format(feat)\n",
    "        temp[feat] = temp[feat].astype(int)\n",
    "        temp[col] = (temp[feat] == False).astype(int)\n",
    "\n",
    "    # count n samples where major != principal and vice versa\n",
    "    temp = temp.groupby(['gname', 'gid_stable']).sum().reset_index()\n",
    "\n",
    "    # total number of samples that this gene is expressed / \n",
    "    # has a complete isoform in \n",
    "    temp['n_samples'] = temp.triplet+temp.not_triplet\n",
    "\n",
    "    for feat in ['triplet', 'tss', 'ic', 'tes']:\n",
    "        col = 'not_{}'.format(feat)\n",
    "        n_genes = len(temp.gid_stable.unique())\n",
    "        n_num = len(temp.loc[temp[col] >= 1].index)\n",
    "        print()\n",
    "        print('{:.2f}% ({}/{}) genes with >= 1 sample where the major {} is not MANE'.format((n_num/n_genes)*100, n_num, n_genes, feat)) \n",
    "\n",
    "    for feat in ['triplet', 'tss', 'ic', 'tes']:\n",
    "        col = 'perc_{}'.format(feat)\n",
    "        temp[col] = (temp[feat]/temp.n_samples)*100\n",
    "\n",
    "    return temp\n",
    "\n",
    "def get_major_principal_feats(df, **kwargs):\n",
    "    \"\"\"\n",
    "    Determine how many genes have at least one sample where\n",
    "    the features and full-length isoform is the principal isoform\n",
    "    \n",
    "    Parameters:\n",
    "        sg (swan_vis SwanGraph): SwanGraph with transcript\n",
    "                    abundance data added   \n",
    "                    \n",
    "    Returns:\n",
    "        df (pandas DataFrame): DF w/ boolean columns for \n",
    "            whether or not there's at least one sample\n",
    "            where the major feat / isoform is the principal one\n",
    "    \"\"\"\n",
    "    # df = get_major_feats(sg, **kwargs)\n",
    "    # df = major_df.copy(deep=True)\n",
    "\n",
    "    # add feature numbers\n",
    "    for feat in ['tss', 'ic', 'tes']:\n",
    "        id_col = '{}_id'.format(feat)\n",
    "        df[feat] = df[id_col].str.split('_', expand=True)[1]\n",
    "\n",
    "    # add triplet\n",
    "    df['triplet'] = df.tid.str.split('[', expand=True)[1].str.split(']', expand=True)[0]\n",
    "\n",
    "\n",
    "    # remove unnecessary columns and determine which genes\n",
    "    # have at least one sample who's major iso uses the principal\n",
    "    # of each feature\n",
    "    temp = df.copy(deep=True)\n",
    "    drop_cols = [c for c in temp.columns if 'pi' in c]\n",
    "    drop_cols += [c for c in temp.columns if '_id' in c]\n",
    "    drop_cols += ['tid', 'gid']\n",
    "    temp.drop(drop_cols, axis=1, inplace=True)\n",
    "\n",
    "    temp = temp.groupby(['gname',\n",
    "                         'gid_stable']).agg({'tss': uses_principal_feat,\n",
    "                                             'ic': uses_principal_feat,\n",
    "                                             'tes': uses_principal_feat,\n",
    "                                             'triplet': uses_principal_iso}).reset_index()\n",
    "\n",
    "    # output\n",
    "    for feat in ['ic', 'tss', 'tes', 'triplet']:\n",
    "        n_genes = len(temp.gid_stable.unique().tolist())\n",
    "        n_major = len(temp.loc[temp[feat] == True].index)\n",
    "        print(n_genes)\n",
    "        print(n_major)\n",
    "        print(feat)\n",
    "        print('{:.2f}% of genes have >=1 sample where major {} is MANE'.format((n_major/n_genes)*100, feat))\n",
    "        print()\n",
    "\n",
    "    return temp\n",
    "\n",
    "def plot_major_principal_feat_counts(df, opref='figures/', **kwargs):\n",
    "    \"\"\"\n",
    "    Plot a histogram of the number of datasets where the major isoform\n",
    "    is the principal isoform.\n",
    "    \n",
    "    Parameters:\n",
    "        sg (swan_vis SwanGraph): SwanGraph with transcript\n",
    "                abundance data added\n",
    "    \n",
    "    Returns:\n",
    "        temp (pandas DataFrame): Output from `count_major_principal_feats`.\n",
    "    \"\"\"\n",
    "    temp = count_major_principal_feats(df, **kwargs)\n",
    "    mpl.rcParams['font.family'] = 'Arial'\n",
    "    mpl.rcParams['pdf.fonttype'] = 42\n",
    "    sns.set_context('paper', font_scale=1.8)\n",
    "    c_dict, order = get_feat_triplet_colors()  \n",
    "    for feat in ['triplet', 'tss', 'ic', 'tes']:\n",
    "        col = 'perc_{}'.format(feat)\n",
    "        ax = sns.displot(data=temp,\n",
    "                         x=col,\n",
    "                         linewidth=0, \n",
    "                         color=c_dict[feat],\n",
    "                         binwidth=5,\n",
    "                         alpha=1)\n",
    "        ylabel = '# genes'\n",
    "        if feat in ['tss', 'ic', 'tes']: \n",
    "            xlabel = '% of samples where major {} is from MANE'.format(feat.upper())\n",
    "        else:\n",
    "            xlabel = '% of samples where major isoform is MANE'\n",
    "            \n",
    "        ax.set(ylabel=ylabel, xlabel=xlabel)\n",
    "\n",
    "        fname = '{}/MANE_vs_major_{}_hist.pdf'.format(opref, feat)\n",
    "        plt.savefig(fname, dpi=800, bbox_inches='tight')\n",
    "        \n",
    "    return temp\n",
    "\n",
    "\n",
    "def compute_feat_tpm(adata, obs_col, feat, how, min_tpm=None):\n",
    "    tpm_df = swan.calc_tpm(adata, obs_col=obs_col, how=how)\n",
    "    tpm_df = tpm_df.sparse.to_dense()\n",
    "    tpm_df = tpm_df.T\n",
    "    tpm_df = tpm_df.melt(var_name=obs_col, value_name='tpm', ignore_index=False)\n",
    "    tpm_df.reset_index(inplace=True)\n",
    "    tpm_df.rename({'index': id_col}, axis=1, inplace=True)\n",
    "    if min_tpm:\n",
    "        tpm_df = tpm_df.loc[tpm_df.tpm >= min_tpm]\n",
    "    tpm_df.rename({'tpm': '{}_tpm'.format(feat)}, axis=1, inplace=True)\n",
    "    return tpm_df\n",
    "\n",
    "def get_major_feats(sg, obs_col='sample', gene_subset=None, min_tpm=None):\n",
    "    \"\"\"\n",
    "    Determine what the major isoform, tss, tes, ic is in each grouping\n",
    "    for each gene \n",
    "    \n",
    "    Parameters:\n",
    "        sg (swan_vis SwanGraph): SwanGraph with transcript\n",
    "            abundance data added\n",
    "        obs_col (str): Column in `sg.adata.obs` to group\n",
    "            datasets by \n",
    "        gene_subset (list of str or None): If not None, \n",
    "            list of stable gene ids to subset on \n",
    "            \n",
    "    Returns:\n",
    "        df (pandas DataFrame): DF where each row is the major \n",
    "            isoform in each expressed gene / sample combination\n",
    "    \"\"\"\n",
    "    # get the major tss, ic, tes, and isoform from each sample\n",
    "    adatas = [sg.adata, sg.tss_adata, sg.ic_adata, sg.tes_adata]\n",
    "    t_dfs = [sg.t_df, sg.tss_adata.var, sg.ic_adata.var, sg.tes_adata.var]\n",
    "    feats = ['triplet', 'tss', 'ic', 'tes']\n",
    "    id_cols = {'triplet': 'tid', 'tss': 'tss_id', 'ic': 'ic_id', 'tes': 'tes_id'}\n",
    "    major_df = pd.DataFrame()\n",
    "\n",
    "    for feat, adata, t_df in zip(feats, adatas, t_dfs):\n",
    "\n",
    "        # get pi value for each feature in the specified gb category\n",
    "        df, _ = swan.calc_pi(adata, t_df, obs_col=obs_col)\n",
    "        df = df.sparse.to_dense()\n",
    "        df = df.transpose()\n",
    "\n",
    "        # merge with gene info\n",
    "        id_col = id_cols[feat]\n",
    "        t_df = t_df.copy(deep=True)\n",
    "        if feat == 'triplet':\n",
    "            drop = True\n",
    "        else:\n",
    "            drop = False\n",
    "        t_df.reset_index(drop=drop,inplace=True)\n",
    "        t_df = t_df[[id_col, 'gid']]\n",
    "        df = df.merge(t_df, how='inner', on=id_col)\n",
    "        df.set_index([id_col, 'gid'], inplace=True)\n",
    "\n",
    "        # pivot and coerce into lengthwise format\n",
    "        df = df.melt(ignore_index=False, value_name='pi', var_name=obs_col)\n",
    "        df = df.dropna(subset=['pi'])\n",
    "        df.reset_index(inplace=True)\n",
    "\n",
    "        # enforce min tpm for isoforms\n",
    "        if min_tpm and feat == 'triplet':\n",
    "            tpm_df = compute_feat_tpm(adata, obs_col, feat, how='max', min_tpm=min_tpm)\n",
    "            df = df.merge(tpm_df, how='inner', on=[id_col, 'sample'])\n",
    "            df.drop('{}_tpm'.format(feat), axis=1, inplace=True)\n",
    "\n",
    "        # merge in avg. tpm for each feat and enforce min\n",
    "        tpm_df = compute_feat_tpm(adata, obs_col, feat, how='mean')\n",
    "        df = df.merge(tpm_df, how='left', on=[id_col, 'sample'])\n",
    "\n",
    "        # remove unexpressed isoforms\n",
    "        df = df.loc[df.pi > 0]\n",
    "\n",
    "        # limit to detected genes\n",
    "        df['gid_stable'] = cerberus.get_stable_gid(df, 'gid')\n",
    "        if gene_subset:\n",
    "            df = df.loc[df.gid_stable.isin(gene_subset)]\n",
    "\n",
    "        # sort by gene, sample, and pi value\n",
    "        # dedupe across the gene and sample cols; take the top-expressed isoform\n",
    "        df = df.sort_values(by=['gid', obs_col, 'pi', '{}_tpm'.format(feat)],\n",
    "                            ascending=[False, False, False, False])\n",
    "\n",
    "        df = df.drop_duplicates(subset=['gid', obs_col], keep='first')\n",
    "\n",
    "        # rename some columns\n",
    "        if feat == 'triplet': \n",
    "            df.rename({'pi': 'tid_pi'}, axis=1, inplace=True)\n",
    "        else:\n",
    "            pi_col = '{}_pi'.format(feat)\n",
    "            df.rename({'pi': pi_col}, axis=1, inplace=True)\n",
    "\n",
    "        # first entry\n",
    "        if major_df.empty:\n",
    "            major_df = df.copy(deep=True)\n",
    "        else:\n",
    "            major_df = major_df.merge(df, how='outer', on=['gid', 'sample', 'gid_stable'])        \n",
    "\n",
    "    # add gene name\n",
    "    g_df = sg.t_df[['gid', 'gname']].reset_index(drop=True).drop_duplicates()\n",
    "    major_df = major_df.merge(g_df, how='left', on='gid')\n",
    "    \n",
    "    # add feature numbers\n",
    "    for feat in ['tss', 'ic', 'tes']:\n",
    "        id_col = '{}_id'.format(feat)\n",
    "        major_df[feat] = major_df[id_col].str.split('_', expand=True)[1]\n",
    "\n",
    "    # add triplet\n",
    "    major_df['triplet'] = major_df.tid.str.split('[', expand=True)[1].str.split(']', expand=True)[0]\n",
    "\n",
    "    return major_df\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "00dab5ae-00d8-48a5-a093-660b5af65ed3",
   "metadata": {},
   "outputs": [],
   "source": [
    "def plot_mane_v_princ_tpm(temp, feat):\n",
    "    temp['log_princ'] = np.log2(temp['{}_tpm_principal'.format(feat)]+1)\n",
    "    temp['log_mane'] = np.log2(temp['{}_tpm_mane'.format(feat)]+1)\n",
    "\n",
    "    lim_max = max(temp.log_princ.max(), temp.log_mane.max())\n",
    "    xlim = ylim = (-2, lim_max)\n",
    "\n",
    "    c_dict, order = get_feat_triplet_colors()  \n",
    "    color = c_dict[feat]\n",
    "\n",
    "    ax = sns.jointplot(data=temp, x='log_mane', y='log_princ', color=color, \n",
    "                       xlim=xlim, ylim=ylim)\n",
    "\n",
    "    if feat == 'triplet':\n",
    "        label = 'isoform'\n",
    "    else:\n",
    "        label = feat.upper()\n",
    "    ylabel = 'log2(Principal {} TPM+1)'.format(label)\n",
    "    xlabel = 'log2(MANE {} TPM+1)'.format(label)\n",
    "\n",
    "    tick_range = range(0, int(lim_max+1), 5)\n",
    "    ax.ax_joint.set_xticks(tick_range)\n",
    "    ax.ax_joint.set_yticks(tick_range)\n",
    "    ax.ax_joint.set(ylabel=ylabel, xlabel=xlabel)\n",
    "    \n",
    "    fname = 'figures/mane_vs_principal_{}_tpm.pdf'.format(feat)\n",
    "    plt.savefig(fname, dpi=800, bbox_inches='tight')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "dd7b7a73-031b-4148-a92b-07858032ed6d",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "triplet\n",
      "> \u001b[0;32m<ipython-input-13-1737c85d370d>\u001b[0m(17)\u001b[0;36m<module>\u001b[0;34m()\u001b[0m\n",
      "\u001b[0;32m     15 \u001b[0;31m    \u001b[0;31m# get mane tids -- look at v40 metadata and cross ref. w/ cerberus\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0m\u001b[0;32m     16 \u001b[0;31m    \u001b[0;31m# also grab their TPMs\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0m\u001b[0;32m---> 17 \u001b[0;31m    \u001b[0mmeta_df\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mpd\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mread_csv\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m'../../../refs/v40_transcript_metadata.tsv'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0msep\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;34m'\\t'\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0m\u001b[0;32m     18 \u001b[0;31m    \u001b[0mmane_tids\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mmeta_df\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mloc\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0mmeta_df\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mMANE_Select\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mtid\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mtolist\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0m\u001b[0;32m     19 \u001b[0;31m    \u001b[0mmane_df\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mca\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mt_map\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mcopy\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mdeep\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;32mTrue\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0m\n"
     ]
    },
    {
     "name": "stdin",
     "output_type": "stream",
     "text": [
      "ipdb>  tpm_df.head()\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "                      tid sample  triplet_tpm\n",
      "0  ENSG00000227232[1,2,1]  caco2     1.435214\n",
      "1  ENSG00000227232[1,3,1]  caco2     1.799986\n",
      "2  ENSG00000227232[1,4,1]  caco2     0.432952\n",
      "3  ENSG00000227232[1,5,1]  caco2     0.501131\n",
      "4  ENSG00000227232[2,2,1]  caco2     0.000000\n"
     ]
    }
   ],
   "source": [
    "adatas = [sg.adata, sg.tss_adata, sg.ic_adata, sg.tes_adata]\n",
    "t_dfs = [sg.t_df, sg.tss_adata.var, sg.ic_adata.var, sg.tes_adata.var]\n",
    "feats = ['triplet', 'tss', 'ic', 'tes']\n",
    "id_cols = {'triplet': 'tid', 'tss': 'tss_id', 'ic': 'ic_id', 'tes': 'tes_id'}\n",
    "id_cols_2 = {'triplet': 'transcript_id', 'tss': 'tss_id', 'ic': 'ic_id', 'tes': 'tes_id'}\n",
    "how = 'mean'\n",
    "\n",
    "for feat, adata, t_df in zip(feats, adatas, t_dfs):\n",
    "    print(feat)\n",
    "    id_col = id_cols[feat]\n",
    "    id_col_2 = id_cols_2[feat]\n",
    "    tpm_df = compute_feat_tpm(adata, obs_col, feat, how, min_tpm=None)\n",
    "    pdb.set_trace()\n",
    "    \n",
    "    # get mane tids -- look at v40 metadata and cross ref. w/ cerberus\n",
    "    # also grab their TPMs\n",
    "    meta_df = pd.read_csv('../../../refs/v40_transcript_metadata.tsv', sep='\\t')\n",
    "    mane_tids = meta_df.loc[meta_df.MANE_Select].tid.tolist()\n",
    "    mane_df = ca.t_map.copy(deep=True)\n",
    "    mane_df = mane_df.loc[mane_df.source=='v40']\n",
    "    mane_df = mane_df.loc[mane_df.original_transcript_id.isin(mane_tids)]\n",
    "    mane_feats = mane_df[id_col_2].unique().tolist()\n",
    "    mane_feats[:5]\n",
    "\n",
    "    # limit tpm_df to just the mane features\n",
    "    print(len(tpm_df.index))\n",
    "    tpm_df = tpm_df.loc[tpm_df[id_col].isin(mane_feats)]\n",
    "    print(len(tpm_df.index))\n",
    "\n",
    "    # add the gene id \n",
    "    if feat == 'triplet':\n",
    "        pat = '['\n",
    "    else:\n",
    "        pat = '_'\n",
    "    tpm_df['gid'] = tpm_df[id_col].str.split(pat, expand=True)[0]\n",
    "    tpm_df.head()\n",
    "\n",
    "    # merge in TPM of MANE features\n",
    "    df = df.merge(tpm_df, how='left', on=['gid', obs_col], suffixes=('_principal', '_mane'))\n",
    "    # fill nans for tpm w/ 0\n",
    "    df['{}_tpm_mane'.format(feat)].fillna(0, inplace=True)\n",
    "\n",
    "    prin_mane_col = '{}_principal_is_mane'.format(feat)\n",
    "    df[prin_mane_col] = df[id_col+'_principal']==df[id_col+'_mane']\n",
    "\n",
    "    # when is the mane isoform not the principal isoform, but mane is still expressed?\n",
    "    temp = df.loc[df[prin_mane_col] == False]\n",
    "    n = len(temp.index)\n",
    "    temp = temp.loc[temp['{}_tpm_mane'.format(feat)]>0]\n",
    "    n_num = len(temp.index)\n",
    "    print('{:.2f}% ({}/{}) of gene / sample combos have MANE expression where principal {} is not MANE '.format((n_num/n)*100, n_num, n, feat))\n",
    "    plot_mane_v_princ_tpm(temp, feat)\n",
    "    \n",
    "    # when is the mane isoform not the principal isoform, and mane is NOT expressed?\n",
    "    temp = df.loc[df[prin_mane_col] == False]\n",
    "    n = len(temp.index)\n",
    "    temp = temp.loc[temp['{}_tpm_mane'.format(feat)]==0]\n",
    "    n_num = len(temp.index)\n",
    "    print('{:.2f}% ({}/{}) of gene / sample combos have NO MANE expression where principal {} is not MANE '.format((n_num/n)*100, n_num, n, feat))\n",
    "\n",
    "    print()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "09ffff57-86d4-4696-99d5-0fc8b9872664",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.4"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
