{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "6e146929-84bd-47ac-b816-f51ef3bf2d4c",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import pyranges as pr"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "d199639b-7d96-4470-8c75-66ff31eb0df2",
   "metadata": {},
   "outputs": [],
   "source": [
    "# fname = '/Users/fairliereese/mortazavi_lab/ref/gencode.v39/gencode.v39.annotation.gtf'\n",
    "# v39 = pr.read_gtf(fname, duplicate_attr=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "b1db1335-7be2-4661-ba63-dcc6b264bf80",
   "metadata": {},
   "outputs": [],
   "source": [
    "# v39.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "8143157e-c064-40ea-8885-3f8961734491",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "this block is running\n"
     ]
    }
   ],
   "source": [
    "fname = '/Users/fairliereese/mortazavi_lab/data/rnawg/lr_bulk/talon/human_known_nic_nnc_talon.gtf'\n",
    "print('this block is running')\n",
    "# fname = '/Users/fairliereese/mortazavi_lab/data/rnawg/lr_bulk/talon/test.gtf'\n",
    "talon = pr.read_gtf(fname, duplicate_attr=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "b5c55811-78c4-4a6e-96e8-23d6d4e694db",
   "metadata": {},
   "outputs": [],
   "source": [
    "talon = talon.df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "5a71cade-c1f3-44da-9ce4-e14577262235",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Source</th>\n",
       "      <th>Feature</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Score</th>\n",
       "      <th>Strand</th>\n",
       "      <th>Frame</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>...</th>\n",
       "      <th>tag</th>\n",
       "      <th>transcript_type</th>\n",
       "      <th>transcript_support_level</th>\n",
       "      <th>havana_transcript</th>\n",
       "      <th>NNC_transcript</th>\n",
       "      <th>protein_id</th>\n",
       "      <th>ccdsid</th>\n",
       "      <th>intergenic_novel</th>\n",
       "      <th>antisense_gene</th>\n",
       "      <th>gene_antisense_to_IDs</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>SIRV1</td>\n",
       "      <td>TALON</td>\n",
       "      <td>gene</td>\n",
       "      <td>10647</td>\n",
       "      <td>11606</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ENCODEHG000058846</td>\n",
       "      <td>ENCODEHG000058846</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>TRUE</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>SIRV1</td>\n",
       "      <td>TALON</td>\n",
       "      <td>transcript</td>\n",
       "      <td>10647</td>\n",
       "      <td>11606</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ENCODEHG000058846</td>\n",
       "      <td>ENCODEHG000058846</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>TRUE</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>SIRV1</td>\n",
       "      <td>TALON</td>\n",
       "      <td>exon</td>\n",
       "      <td>10647</td>\n",
       "      <td>10791</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ENCODEHG000058846</td>\n",
       "      <td>ENCODEHG000058846</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>SIRV1</td>\n",
       "      <td>TALON</td>\n",
       "      <td>exon</td>\n",
       "      <td>10882</td>\n",
       "      <td>11057</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ENCODEHG000058846</td>\n",
       "      <td>ENCODEHG000058846</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>SIRV1</td>\n",
       "      <td>TALON</td>\n",
       "      <td>exon</td>\n",
       "      <td>11434</td>\n",
       "      <td>11606</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ENCODEHG000058846</td>\n",
       "      <td>ENCODEHG000058846</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 36 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "  Chromosome Source     Feature  Start    End Score Strand Frame  \\\n",
       "0      SIRV1  TALON        gene  10647  11606     .      +     .   \n",
       "1      SIRV1  TALON  transcript  10647  11606     .      +     .   \n",
       "2      SIRV1  TALON        exon  10647  10791     .      +     .   \n",
       "3      SIRV1  TALON        exon  10882  11057     .      +     .   \n",
       "4      SIRV1  TALON        exon  11434  11606     .      +     .   \n",
       "\n",
       "             gene_id          gene_name  ...  tag transcript_type  \\\n",
       "0  ENCODEHG000058846  ENCODEHG000058846  ...  NaN             NaN   \n",
       "1  ENCODEHG000058846  ENCODEHG000058846  ...  NaN             NaN   \n",
       "2  ENCODEHG000058846  ENCODEHG000058846  ...  NaN             NaN   \n",
       "3  ENCODEHG000058846  ENCODEHG000058846  ...  NaN             NaN   \n",
       "4  ENCODEHG000058846  ENCODEHG000058846  ...  NaN             NaN   \n",
       "\n",
       "  transcript_support_level havana_transcript NNC_transcript protein_id ccdsid  \\\n",
       "0                      NaN               NaN            NaN        NaN    NaN   \n",
       "1                      NaN               NaN           TRUE        NaN    NaN   \n",
       "2                      NaN               NaN            NaN        NaN    NaN   \n",
       "3                      NaN               NaN            NaN        NaN    NaN   \n",
       "4                      NaN               NaN            NaN        NaN    NaN   \n",
       "\n",
       "  intergenic_novel antisense_gene gene_antisense_to_IDs  \n",
       "0             TRUE            NaN                   NaN  \n",
       "1              NaN            NaN                   NaN  \n",
       "2              NaN            NaN                   NaN  \n",
       "3              NaN            NaN                   NaN  \n",
       "4              NaN            NaN                   NaN  \n",
       "\n",
       "[5 rows x 36 columns]"
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "talon.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "af207c7c-0801-4c1a-9eb7-43a280557d2f",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Source</th>\n",
       "      <th>Feature</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Score</th>\n",
       "      <th>Strand</th>\n",
       "      <th>Frame</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>...</th>\n",
       "      <th>tag</th>\n",
       "      <th>transcript_type</th>\n",
       "      <th>transcript_support_level</th>\n",
       "      <th>havana_transcript</th>\n",
       "      <th>NNC_transcript</th>\n",
       "      <th>protein_id</th>\n",
       "      <th>ccdsid</th>\n",
       "      <th>intergenic_novel</th>\n",
       "      <th>antisense_gene</th>\n",
       "      <th>gene_antisense_to_IDs</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>30</th>\n",
       "      <td>SIRV2</td>\n",
       "      <td>TALON</td>\n",
       "      <td>transcript</td>\n",
       "      <td>995</td>\n",
       "      <td>5906</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ENCODEHG000058784</td>\n",
       "      <td>ENCODEHG000058784</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>TRUE</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31</th>\n",
       "      <td>SIRV2</td>\n",
       "      <td>TALON</td>\n",
       "      <td>exon</td>\n",
       "      <td>5788</td>\n",
       "      <td>5906</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ENCODEHG000058784</td>\n",
       "      <td>ENCODEHG000058784</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32</th>\n",
       "      <td>SIRV2</td>\n",
       "      <td>TALON</td>\n",
       "      <td>exon</td>\n",
       "      <td>4687</td>\n",
       "      <td>4800</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ENCODEHG000058784</td>\n",
       "      <td>ENCODEHG000058784</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>33</th>\n",
       "      <td>SIRV2</td>\n",
       "      <td>TALON</td>\n",
       "      <td>exon</td>\n",
       "      <td>4338</td>\n",
       "      <td>4479</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ENCODEHG000058784</td>\n",
       "      <td>ENCODEHG000058784</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>34</th>\n",
       "      <td>SIRV2</td>\n",
       "      <td>TALON</td>\n",
       "      <td>exon</td>\n",
       "      <td>3966</td>\n",
       "      <td>4094</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ENCODEHG000058784</td>\n",
       "      <td>ENCODEHG000058784</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>35</th>\n",
       "      <td>SIRV2</td>\n",
       "      <td>TALON</td>\n",
       "      <td>exon</td>\n",
       "      <td>3665</td>\n",
       "      <td>3825</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ENCODEHG000058784</td>\n",
       "      <td>ENCODEHG000058784</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>36</th>\n",
       "      <td>SIRV2</td>\n",
       "      <td>TALON</td>\n",
       "      <td>exon</td>\n",
       "      <td>3105</td>\n",
       "      <td>3325</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ENCODEHG000058784</td>\n",
       "      <td>ENCODEHG000058784</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>37</th>\n",
       "      <td>SIRV2</td>\n",
       "      <td>TALON</td>\n",
       "      <td>exon</td>\n",
       "      <td>2881</td>\n",
       "      <td>3010</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ENCODEHG000058784</td>\n",
       "      <td>ENCODEHG000058784</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>38</th>\n",
       "      <td>SIRV2</td>\n",
       "      <td>TALON</td>\n",
       "      <td>exon</td>\n",
       "      <td>2674</td>\n",
       "      <td>2802</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ENCODEHG000058784</td>\n",
       "      <td>ENCODEHG000058784</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>39</th>\n",
       "      <td>SIRV2</td>\n",
       "      <td>TALON</td>\n",
       "      <td>exon</td>\n",
       "      <td>1973</td>\n",
       "      <td>2064</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ENCODEHG000058784</td>\n",
       "      <td>ENCODEHG000058784</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>40</th>\n",
       "      <td>SIRV2</td>\n",
       "      <td>TALON</td>\n",
       "      <td>exon</td>\n",
       "      <td>1741</td>\n",
       "      <td>1853</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ENCODEHG000058784</td>\n",
       "      <td>ENCODEHG000058784</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>41</th>\n",
       "      <td>SIRV2</td>\n",
       "      <td>TALON</td>\n",
       "      <td>exon</td>\n",
       "      <td>995</td>\n",
       "      <td>1661</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ENCODEHG000058784</td>\n",
       "      <td>ENCODEHG000058784</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>12 rows × 36 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "   Chromosome Source     Feature  Start   End Score Strand Frame  \\\n",
       "30      SIRV2  TALON  transcript    995  5906     .      -     .   \n",
       "31      SIRV2  TALON        exon   5788  5906     .      -     .   \n",
       "32      SIRV2  TALON        exon   4687  4800     .      -     .   \n",
       "33      SIRV2  TALON        exon   4338  4479     .      -     .   \n",
       "34      SIRV2  TALON        exon   3966  4094     .      -     .   \n",
       "35      SIRV2  TALON        exon   3665  3825     .      -     .   \n",
       "36      SIRV2  TALON        exon   3105  3325     .      -     .   \n",
       "37      SIRV2  TALON        exon   2881  3010     .      -     .   \n",
       "38      SIRV2  TALON        exon   2674  2802     .      -     .   \n",
       "39      SIRV2  TALON        exon   1973  2064     .      -     .   \n",
       "40      SIRV2  TALON        exon   1741  1853     .      -     .   \n",
       "41      SIRV2  TALON        exon    995  1661     .      -     .   \n",
       "\n",
       "              gene_id          gene_name  ...  tag transcript_type  \\\n",
       "30  ENCODEHG000058784  ENCODEHG000058784  ...  NaN             NaN   \n",
       "31  ENCODEHG000058784  ENCODEHG000058784  ...  NaN             NaN   \n",
       "32  ENCODEHG000058784  ENCODEHG000058784  ...  NaN             NaN   \n",
       "33  ENCODEHG000058784  ENCODEHG000058784  ...  NaN             NaN   \n",
       "34  ENCODEHG000058784  ENCODEHG000058784  ...  NaN             NaN   \n",
       "35  ENCODEHG000058784  ENCODEHG000058784  ...  NaN             NaN   \n",
       "36  ENCODEHG000058784  ENCODEHG000058784  ...  NaN             NaN   \n",
       "37  ENCODEHG000058784  ENCODEHG000058784  ...  NaN             NaN   \n",
       "38  ENCODEHG000058784  ENCODEHG000058784  ...  NaN             NaN   \n",
       "39  ENCODEHG000058784  ENCODEHG000058784  ...  NaN             NaN   \n",
       "40  ENCODEHG000058784  ENCODEHG000058784  ...  NaN             NaN   \n",
       "41  ENCODEHG000058784  ENCODEHG000058784  ...  NaN             NaN   \n",
       "\n",
       "   transcript_support_level havana_transcript NNC_transcript protein_id  \\\n",
       "30                      NaN               NaN           TRUE        NaN   \n",
       "31                      NaN               NaN            NaN        NaN   \n",
       "32                      NaN               NaN            NaN        NaN   \n",
       "33                      NaN               NaN            NaN        NaN   \n",
       "34                      NaN               NaN            NaN        NaN   \n",
       "35                      NaN               NaN            NaN        NaN   \n",
       "36                      NaN               NaN            NaN        NaN   \n",
       "37                      NaN               NaN            NaN        NaN   \n",
       "38                      NaN               NaN            NaN        NaN   \n",
       "39                      NaN               NaN            NaN        NaN   \n",
       "40                      NaN               NaN            NaN        NaN   \n",
       "41                      NaN               NaN            NaN        NaN   \n",
       "\n",
       "   ccdsid intergenic_novel antisense_gene gene_antisense_to_IDs  \n",
       "30    NaN              NaN            NaN                   NaN  \n",
       "31    NaN              NaN            NaN                   NaN  \n",
       "32    NaN              NaN            NaN                   NaN  \n",
       "33    NaN              NaN            NaN                   NaN  \n",
       "34    NaN              NaN            NaN                   NaN  \n",
       "35    NaN              NaN            NaN                   NaN  \n",
       "36    NaN              NaN            NaN                   NaN  \n",
       "37    NaN              NaN            NaN                   NaN  \n",
       "38    NaN              NaN            NaN                   NaN  \n",
       "39    NaN              NaN            NaN                   NaN  \n",
       "40    NaN              NaN            NaN                   NaN  \n",
       "41    NaN              NaN            NaN                   NaN  \n",
       "\n",
       "[12 rows x 36 columns]"
      ]
     },
     "execution_count": 22,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "talon.loc[talon.transcript_id == 'ENCODEHT000206785']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "5cf5bc26-be04-4a0c-8383-219e0c94a47b",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "31    ENCODEHT000206785\n",
       "Name: transcript_id, dtype: object"
      ]
     },
     "execution_count": 23,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "talon.loc[(talon.gene_id == 'ENCODEHG000058784')&(talon.Start == 5788)].transcript_id"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "id": "775726d6-576c-42ea-9d7f-a7687fb13970",
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_ic(gtf_pr):\n",
    "    \"\"\"\n",
    "    Get a hyphen-separated representation of each transcript's intron chain\n",
    "    from a PyRanges GTF\n",
    "\n",
    "    Parameters:\n",
    "        gtf_pr (pyranges PyRanges): GTF PyRanges object\n",
    "\n",
    "    Returns:\n",
    "        df (pandas DataFrame): DataFrame detailing intron chain, gene, strand,\n",
    "            chromosome, and transcript that intron chain was seen in\n",
    "    \"\"\"\n",
    "    df = gtf_pr.df.copy(deep=True)\n",
    "\n",
    "    # restrict to exon entries\n",
    "    df = df.loc[df.Feature == 'exon']\n",
    "    cols = ['Chromosome', 'Strand', 'Start', 'End', 'transcript_id', 'gene_id']\n",
    "    df = df[cols]\n",
    "\n",
    "    # melt to isolate individual coordinates\n",
    "    df = pd.melt(df, id_vars=['Chromosome', 'Strand', 'transcript_id', 'gene_id'],\n",
    "                value_vars=['Start', 'End'],\n",
    "                value_name='Coord')\n",
    "    df.drop('variable', axis=1, inplace=True)\n",
    "\n",
    "    # sort to order coordinates correctly\n",
    "    df.Coord = df.Coord.astype(int)\n",
    "    fwd = df.loc[df.Strand == '+'].copy(deep=True)\n",
    "    rev = df.loc[df.Strand == '-'].copy(deep=True)\n",
    "\n",
    "    fwd.sort_values(by=['Chromosome', 'transcript_id', 'Coord'],\n",
    "                    ascending=[True, True, True], inplace=True)\n",
    "    rev.sort_values(by=['Chromosome', 'transcript_id', 'Coord'],\n",
    "                    ascending=[True, True, False], inplace=True)\n",
    "    df = pd.concat([fwd, rev])\n",
    "\n",
    "    # create intron chain strings\n",
    "    df.Coord = df.Coord.astype(str)\n",
    "    df = df.groupby(['Chromosome', 'Strand',\n",
    "                     'transcript_id', 'gene_id'], observed=True)['Coord'].apply('-'.join).reset_index()\n",
    "\n",
    "    # remove tss and tes from intron chain\n",
    "    df['temp'] = df.Coord.str.split('-', n=1, expand=True)[1]\n",
    "    df['ic'] = df.temp.str.rsplit('-', n=1, expand=True)[0]\n",
    "\n",
    "    return df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "id": "44fee4f2-4fab-4be5-bfe4-436c2dd74fb4",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Strand</th>\n",
       "      <th>transcript_id</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>Coord</th>\n",
       "      <th>temp</th>\n",
       "      <th>ic</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>127212</th>\n",
       "      <td>SIRV2</td>\n",
       "      <td>-</td>\n",
       "      <td>ENCODEHT000206785</td>\n",
       "      <td>ENCODEHG000058784</td>\n",
       "      <td>5906-5788-4800-4687-4479-4338-4094-3966-3825-3...</td>\n",
       "      <td>5788-4800-4687-4479-4338-4094-3966-3825-3665-3...</td>\n",
       "      <td>5788-4800-4687-4479-4338-4094-3966-3825-3665-3...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       Chromosome Strand      transcript_id            gene_id  \\\n",
       "127212      SIRV2      -  ENCODEHT000206785  ENCODEHG000058784   \n",
       "\n",
       "                                                    Coord  \\\n",
       "127212  5906-5788-4800-4687-4479-4338-4094-3966-3825-3...   \n",
       "\n",
       "                                                     temp  \\\n",
       "127212  5788-4800-4687-4479-4338-4094-3966-3825-3665-3...   \n",
       "\n",
       "                                                       ic  \n",
       "127212  5788-4800-4687-4479-4338-4094-3966-3825-3665-3...  "
      ]
     },
     "execution_count": 25,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.loc[df.transcript_id == 'ENCODEHT000206785']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "id": "61fd0353-15c1-4e73-bc0e-6784be846f70",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Strand</th>\n",
       "      <th>transcript_id</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>Coord</th>\n",
       "      <th>temp</th>\n",
       "      <th>ic</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: [Chromosome, Strand, transcript_id, gene_id, Coord, temp, ic]\n",
       "Index: []"
      ]
     },
     "execution_count": 29,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.loc[df.Coord.isnull()]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "id": "6f21baee-4d8f-4133-b50b-9897890b09a8",
   "metadata": {},
   "outputs": [],
   "source": [
    "df = talon.copy(deep=True)\n",
    "df = get_ic(df)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "id": "9e2c7662-50dc-4d98-a254-4faf2583abc9",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Strand</th>\n",
       "      <th>transcript_id</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>Coord</th>\n",
       "      <th>temp</th>\n",
       "      <th>ic</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>SIRV1</td>\n",
       "      <td>+</td>\n",
       "      <td>ENCODEHT000206942</td>\n",
       "      <td>ENCODEHG000058846</td>\n",
       "      <td>10647-10791-10882-11057-11434-11606</td>\n",
       "      <td>10791-10882-11057-11434-11606</td>\n",
       "      <td>10791-10882-11057-11434</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>SIRV1</td>\n",
       "      <td>-</td>\n",
       "      <td>ENCODEHT000206867</td>\n",
       "      <td>ENCODEHG000058837</td>\n",
       "      <td>10790-10647-10366-10282-7814-7552-6813-6560-64...</td>\n",
       "      <td>10647-10366-10282-7814-7552-6813-6560-6473-633...</td>\n",
       "      <td>10647-10366-10282-7814-7552-6813-6560-6473-633...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>SIRV1</td>\n",
       "      <td>-</td>\n",
       "      <td>ENCODEHT000206868</td>\n",
       "      <td>ENCODEHG000058837</td>\n",
       "      <td>10790-10444-10366-10282-7814-7552-6813-6560-64...</td>\n",
       "      <td>10444-10366-10282-7814-7552-6813-6560-6473-633...</td>\n",
       "      <td>10444-10366-10282-7814-7552-6813-6560-6473-633...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>SIRV1</td>\n",
       "      <td>-</td>\n",
       "      <td>ENCODEHT000206870</td>\n",
       "      <td>ENCODEHG000058837</td>\n",
       "      <td>10790-10553-7808-7552-1484-999</td>\n",
       "      <td>10553-7808-7552-1484-999</td>\n",
       "      <td>10553-7808-7552-1484</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>SIRV1</td>\n",
       "      <td>-</td>\n",
       "      <td>ENCODEHT000206886</td>\n",
       "      <td>ENCODEHG000058837</td>\n",
       "      <td>10366-10282-7814-7552-6813-6337-1484-999</td>\n",
       "      <td>10282-7814-7552-6813-6337-1484-999</td>\n",
       "      <td>10282-7814-7552-6813-6337-1484</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  Chromosome Strand      transcript_id            gene_id  \\\n",
       "0      SIRV1      +  ENCODEHT000206942  ENCODEHG000058846   \n",
       "1      SIRV1      -  ENCODEHT000206867  ENCODEHG000058837   \n",
       "2      SIRV1      -  ENCODEHT000206868  ENCODEHG000058837   \n",
       "3      SIRV1      -  ENCODEHT000206870  ENCODEHG000058837   \n",
       "4      SIRV1      -  ENCODEHT000206886  ENCODEHG000058837   \n",
       "\n",
       "                                               Coord  \\\n",
       "0                10647-10791-10882-11057-11434-11606   \n",
       "1  10790-10647-10366-10282-7814-7552-6813-6560-64...   \n",
       "2  10790-10444-10366-10282-7814-7552-6813-6560-64...   \n",
       "3                     10790-10553-7808-7552-1484-999   \n",
       "4           10366-10282-7814-7552-6813-6337-1484-999   \n",
       "\n",
       "                                                temp  \\\n",
       "0                      10791-10882-11057-11434-11606   \n",
       "1  10647-10366-10282-7814-7552-6813-6560-6473-633...   \n",
       "2  10444-10366-10282-7814-7552-6813-6560-6473-633...   \n",
       "3                           10553-7808-7552-1484-999   \n",
       "4                 10282-7814-7552-6813-6337-1484-999   \n",
       "\n",
       "                                                  ic  \n",
       "0                            10791-10882-11057-11434  \n",
       "1  10647-10366-10282-7814-7552-6813-6560-6473-633...  \n",
       "2  10444-10366-10282-7814-7552-6813-6560-6473-633...  \n",
       "3                               10553-7808-7552-1484  \n",
       "4                     10282-7814-7552-6813-6337-1484  "
      ]
     },
     "execution_count": 35,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "12266695-fd77-4e66-ab74-5aa16d6c81e4",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Source</th>\n",
       "      <th>Feature</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Score</th>\n",
       "      <th>Strand</th>\n",
       "      <th>Frame</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>...</th>\n",
       "      <th>tag</th>\n",
       "      <th>transcript_type</th>\n",
       "      <th>transcript_support_level</th>\n",
       "      <th>havana_transcript</th>\n",
       "      <th>NNC_transcript</th>\n",
       "      <th>protein_id</th>\n",
       "      <th>ccdsid</th>\n",
       "      <th>intergenic_novel</th>\n",
       "      <th>antisense_gene</th>\n",
       "      <th>gene_antisense_to_IDs</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>SIRV1</td>\n",
       "      <td>TALON</td>\n",
       "      <td>gene</td>\n",
       "      <td>10647</td>\n",
       "      <td>11606</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ENCODEHG000058846</td>\n",
       "      <td>ENCODEHG000058846</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>TRUE</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>SIRV1</td>\n",
       "      <td>TALON</td>\n",
       "      <td>transcript</td>\n",
       "      <td>10647</td>\n",
       "      <td>11606</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ENCODEHG000058846</td>\n",
       "      <td>ENCODEHG000058846</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>TRUE</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>SIRV1</td>\n",
       "      <td>TALON</td>\n",
       "      <td>exon</td>\n",
       "      <td>10647</td>\n",
       "      <td>10791</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ENCODEHG000058846</td>\n",
       "      <td>ENCODEHG000058846</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>SIRV1</td>\n",
       "      <td>TALON</td>\n",
       "      <td>exon</td>\n",
       "      <td>10882</td>\n",
       "      <td>11057</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ENCODEHG000058846</td>\n",
       "      <td>ENCODEHG000058846</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>SIRV1</td>\n",
       "      <td>TALON</td>\n",
       "      <td>exon</td>\n",
       "      <td>11434</td>\n",
       "      <td>11606</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ENCODEHG000058846</td>\n",
       "      <td>ENCODEHG000058846</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>SIRV1</td>\n",
       "      <td>TALON</td>\n",
       "      <td>gene</td>\n",
       "      <td>999</td>\n",
       "      <td>10366</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ENCODEHG000058837</td>\n",
       "      <td>ENCODEHG000058837</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>TRUE</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>SIRV1</td>\n",
       "      <td>TALON</td>\n",
       "      <td>transcript</td>\n",
       "      <td>999</td>\n",
       "      <td>10790</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ENCODEHG000058837</td>\n",
       "      <td>ENCODEHG000058837</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>TRUE</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>SIRV1</td>\n",
       "      <td>TALON</td>\n",
       "      <td>exon</td>\n",
       "      <td>10647</td>\n",
       "      <td>10790</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ENCODEHG000058837</td>\n",
       "      <td>ENCODEHG000058837</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>8 rows × 36 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "+--------------+------------+--------------+-----------+-----------+-------+\n",
       "| Chromosome   | Source     | Feature      |     Start |       End | +31   |\n",
       "| (category)   | (object)   | (category)   |   (int32) |   (int32) | ...   |\n",
       "|--------------+------------+--------------+-----------+-----------+-------|\n",
       "| SIRV1        | TALON      | gene         |     10647 |     11606 | ...   |\n",
       "| SIRV1        | TALON      | transcript   |     10647 |     11606 | ...   |\n",
       "| SIRV1        | TALON      | exon         |     10647 |     10791 | ...   |\n",
       "| SIRV1        | TALON      | exon         |     10882 |     11057 | ...   |\n",
       "| SIRV1        | TALON      | exon         |     11434 |     11606 | ...   |\n",
       "| SIRV1        | TALON      | gene         |       999 |     10366 | ...   |\n",
       "| SIRV1        | TALON      | transcript   |       999 |     10790 | ...   |\n",
       "| SIRV1        | TALON      | exon         |     10647 |     10790 | ...   |\n",
       "+--------------+------------+--------------+-----------+-----------+-------+\n",
       "Stranded PyRanges object has 8 rows and 36 columns from 1 chromosomes.\n",
       "For printing, the PyRanges was sorted on Chromosome and Strand.\n",
       "31 hidden columns: Score, Strand, Frame, gene_id, gene_name, gene_status, gene_type, ... (+ 24 more.)"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "talon.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "d3bfa7b7-c185-420b-bccf-b44b10418a41",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Source</th>\n",
       "      <th>Feature</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Score</th>\n",
       "      <th>Strand</th>\n",
       "      <th>Frame</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>...</th>\n",
       "      <th>tag</th>\n",
       "      <th>transcript_type</th>\n",
       "      <th>transcript_support_level</th>\n",
       "      <th>havana_transcript</th>\n",
       "      <th>NNC_transcript</th>\n",
       "      <th>protein_id</th>\n",
       "      <th>ccdsid</th>\n",
       "      <th>intergenic_novel</th>\n",
       "      <th>antisense_gene</th>\n",
       "      <th>gene_antisense_to_IDs</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>30</th>\n",
       "      <td>SIRV2</td>\n",
       "      <td>TALON</td>\n",
       "      <td>transcript</td>\n",
       "      <td>995</td>\n",
       "      <td>5906</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ENCODEHG000058784</td>\n",
       "      <td>ENCODEHG000058784</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>TRUE</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31</th>\n",
       "      <td>SIRV2</td>\n",
       "      <td>TALON</td>\n",
       "      <td>exon</td>\n",
       "      <td>5788</td>\n",
       "      <td>5906</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ENCODEHG000058784</td>\n",
       "      <td>ENCODEHG000058784</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32</th>\n",
       "      <td>SIRV2</td>\n",
       "      <td>TALON</td>\n",
       "      <td>exon</td>\n",
       "      <td>4687</td>\n",
       "      <td>4800</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ENCODEHG000058784</td>\n",
       "      <td>ENCODEHG000058784</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>33</th>\n",
       "      <td>SIRV2</td>\n",
       "      <td>TALON</td>\n",
       "      <td>exon</td>\n",
       "      <td>4338</td>\n",
       "      <td>4479</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ENCODEHG000058784</td>\n",
       "      <td>ENCODEHG000058784</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>34</th>\n",
       "      <td>SIRV2</td>\n",
       "      <td>TALON</td>\n",
       "      <td>exon</td>\n",
       "      <td>3966</td>\n",
       "      <td>4094</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ENCODEHG000058784</td>\n",
       "      <td>ENCODEHG000058784</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>35</th>\n",
       "      <td>SIRV2</td>\n",
       "      <td>TALON</td>\n",
       "      <td>exon</td>\n",
       "      <td>3665</td>\n",
       "      <td>3825</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ENCODEHG000058784</td>\n",
       "      <td>ENCODEHG000058784</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>36</th>\n",
       "      <td>SIRV2</td>\n",
       "      <td>TALON</td>\n",
       "      <td>exon</td>\n",
       "      <td>3105</td>\n",
       "      <td>3325</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ENCODEHG000058784</td>\n",
       "      <td>ENCODEHG000058784</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>37</th>\n",
       "      <td>SIRV2</td>\n",
       "      <td>TALON</td>\n",
       "      <td>exon</td>\n",
       "      <td>2881</td>\n",
       "      <td>3010</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ENCODEHG000058784</td>\n",
       "      <td>ENCODEHG000058784</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>38</th>\n",
       "      <td>SIRV2</td>\n",
       "      <td>TALON</td>\n",
       "      <td>exon</td>\n",
       "      <td>2674</td>\n",
       "      <td>2802</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ENCODEHG000058784</td>\n",
       "      <td>ENCODEHG000058784</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>39</th>\n",
       "      <td>SIRV2</td>\n",
       "      <td>TALON</td>\n",
       "      <td>exon</td>\n",
       "      <td>1973</td>\n",
       "      <td>2064</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ENCODEHG000058784</td>\n",
       "      <td>ENCODEHG000058784</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>40</th>\n",
       "      <td>SIRV2</td>\n",
       "      <td>TALON</td>\n",
       "      <td>exon</td>\n",
       "      <td>1741</td>\n",
       "      <td>1853</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ENCODEHG000058784</td>\n",
       "      <td>ENCODEHG000058784</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>41</th>\n",
       "      <td>SIRV2</td>\n",
       "      <td>TALON</td>\n",
       "      <td>exon</td>\n",
       "      <td>995</td>\n",
       "      <td>1661</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ENCODEHG000058784</td>\n",
       "      <td>ENCODEHG000058784</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>12 rows × 36 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "   Chromosome Source     Feature  Start   End Score Strand Frame  \\\n",
       "30      SIRV2  TALON  transcript    995  5906     .      -     .   \n",
       "31      SIRV2  TALON        exon   5788  5906     .      -     .   \n",
       "32      SIRV2  TALON        exon   4687  4800     .      -     .   \n",
       "33      SIRV2  TALON        exon   4338  4479     .      -     .   \n",
       "34      SIRV2  TALON        exon   3966  4094     .      -     .   \n",
       "35      SIRV2  TALON        exon   3665  3825     .      -     .   \n",
       "36      SIRV2  TALON        exon   3105  3325     .      -     .   \n",
       "37      SIRV2  TALON        exon   2881  3010     .      -     .   \n",
       "38      SIRV2  TALON        exon   2674  2802     .      -     .   \n",
       "39      SIRV2  TALON        exon   1973  2064     .      -     .   \n",
       "40      SIRV2  TALON        exon   1741  1853     .      -     .   \n",
       "41      SIRV2  TALON        exon    995  1661     .      -     .   \n",
       "\n",
       "              gene_id          gene_name  ...  tag transcript_type  \\\n",
       "30  ENCODEHG000058784  ENCODEHG000058784  ...  NaN             NaN   \n",
       "31  ENCODEHG000058784  ENCODEHG000058784  ...  NaN             NaN   \n",
       "32  ENCODEHG000058784  ENCODEHG000058784  ...  NaN             NaN   \n",
       "33  ENCODEHG000058784  ENCODEHG000058784  ...  NaN             NaN   \n",
       "34  ENCODEHG000058784  ENCODEHG000058784  ...  NaN             NaN   \n",
       "35  ENCODEHG000058784  ENCODEHG000058784  ...  NaN             NaN   \n",
       "36  ENCODEHG000058784  ENCODEHG000058784  ...  NaN             NaN   \n",
       "37  ENCODEHG000058784  ENCODEHG000058784  ...  NaN             NaN   \n",
       "38  ENCODEHG000058784  ENCODEHG000058784  ...  NaN             NaN   \n",
       "39  ENCODEHG000058784  ENCODEHG000058784  ...  NaN             NaN   \n",
       "40  ENCODEHG000058784  ENCODEHG000058784  ...  NaN             NaN   \n",
       "41  ENCODEHG000058784  ENCODEHG000058784  ...  NaN             NaN   \n",
       "\n",
       "   transcript_support_level havana_transcript NNC_transcript protein_id  \\\n",
       "30                      NaN               NaN           TRUE        NaN   \n",
       "31                      NaN               NaN            NaN        NaN   \n",
       "32                      NaN               NaN            NaN        NaN   \n",
       "33                      NaN               NaN            NaN        NaN   \n",
       "34                      NaN               NaN            NaN        NaN   \n",
       "35                      NaN               NaN            NaN        NaN   \n",
       "36                      NaN               NaN            NaN        NaN   \n",
       "37                      NaN               NaN            NaN        NaN   \n",
       "38                      NaN               NaN            NaN        NaN   \n",
       "39                      NaN               NaN            NaN        NaN   \n",
       "40                      NaN               NaN            NaN        NaN   \n",
       "41                      NaN               NaN            NaN        NaN   \n",
       "\n",
       "   ccdsid intergenic_novel antisense_gene gene_antisense_to_IDs  \n",
       "30    NaN              NaN            NaN                   NaN  \n",
       "31    NaN              NaN            NaN                   NaN  \n",
       "32    NaN              NaN            NaN                   NaN  \n",
       "33    NaN              NaN            NaN                   NaN  \n",
       "34    NaN              NaN            NaN                   NaN  \n",
       "35    NaN              NaN            NaN                   NaN  \n",
       "36    NaN              NaN            NaN                   NaN  \n",
       "37    NaN              NaN            NaN                   NaN  \n",
       "38    NaN              NaN            NaN                   NaN  \n",
       "39    NaN              NaN            NaN                   NaN  \n",
       "40    NaN              NaN            NaN                   NaN  \n",
       "41    NaN              NaN            NaN                   NaN  \n",
       "\n",
       "[12 rows x 36 columns]"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "talon = talon.df\n",
    "talon.loc[talon.transcript_id == 'ENCODEHT000206785']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a1d83943-8ace-4cae-b3e8-5e501aee5ff1",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.4"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
